---
title: "02-DESeq2-etc-Pcod"
author: "Laura Spencer"
date: '2022-10-25'
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
      number_sections: true
---

```{r include = FALSE}
knitr::opts_chunk$set(echo=FALSE, warning = F, message = F)
```

```{r, message=FALSE, warning=FALSE, results=FALSE}
#### Load libraries and source scripts 

source("../references/biostats.R")

# Add all required libraries that are installed with install.packages() here
list.of.packages <- c("RCurl", "tidyverse", "vegan", "pheatmap", "pastecs", "factoextra", "FactoMineR", "RColorBrewer", "tibble", "reshape2", "plotly", "cowplot", "clipr", "janitor", "ggpubr", "forcats", "apeglm", "car", "vsn", "devtools", "grid", "gridGraphics", "Rfast", "dendextend", "RColorBrewer", "scales", "VennDiagram")

# Add all libraries that are installed using BiocManager here
bioconductor.packages <- c("DESeq2", "WGCNA")

# # Install BiocManager if needed
# if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
# 
# # Get names of all required packages that aren't installed
# new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
# new.bioc.packages <- bioconductor.packages[!(bioconductor.packages %in% installed.packages()[, "Package"])]
# # Install all new packages
# if(length(new.packages)) install.packages(new.packages)
# if(length(new.bioc.packages)) BiocManager::install(new.bioc.packages)

# Load all required libraries
all.packages <- c(list.of.packages, bioconductor.packages)
lapply(all.packages, FUN = function(X) {
  do.call("require", list(X))
})


# Github packages 
install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
library(pairwiseAdonis)

```

```{r}
#### Load counts and sample info 

load(file="../data/sample.info")
load(file = "../results/counts")
load(file = "../results/counts.t")
load(file = "../results/counts.ts")
load( file = "../references/gadMor.blast")
load(file = "../references/gadMor.blast.GO")
```

## Library summary stats

#### No. of fragments per sample? 

```{r}
data.frame(rowSums(counts.ts)) %>% 
                  dplyr::rename(read.total = 1) %>% 
                  rownames_to_column(var="sample") %>% #summary() 
dplyr::summarise(mean=round(mean(read.total, na.rm=T)), 
          sd=round(sd(read.total, na.rm=T)), 
          se=round(sd/sqrt(length(sample))),
          median=median(read.total), 
          min=min(read.total), 
          max=max(read.total)) %>% t() %>% as.data.frame() %>% 
  dplyr::mutate(V1=prettyNum(V1, big.mark = ",")) %>% 
  dplyr::rename("fragments.per.sample"="V1") #use this to average across all samples 
```

#### Did the no. of fragments per treatment differ? 

##### Boxplots of the # of fragments in each sample per treatment 

```{r}
fragments <- data.frame(rowSums(counts.ts)) %>% 
                  dplyr::rename(read.total = 1) %>% 
                  rownames_to_column(var="sample") %>%
  left_join(sample.info %>% dplyr::select(vial_label, treatment, temperature, ph, tank), by=c("sample"="vial_label")) %>%
  mutate(treatment=factor(treatment, ordered = T, levels=c("3_amb", "3_low", "6_amb", "6_low", "10_amb", "10_low")))


ggplotly(
ggplot(data = fragments) +
           geom_bar(aes(x=sample, y=read.total, fill=treatment), stat = "identity") + ggtitle("Total # fragments by sample") + 
             theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())) 
```

##### Statistically test whether no. of fragments / treatment differs 

```{r}
hist(log(fragments$read.total))
shapiro.test(log(fragments$read.total))
anova(lm(log(read.total) ~ treatment, fragments))


ggplot(fragments, aes(x=temperature, y=read.total)) + geom_boxplot() + theme_minimal()
ggplot(fragments, aes(x=ph, y=read.total)) + geom_boxplot() + theme_minimal()
ggplot(fragments, aes(x=treatment, y=read.total)) + geom_boxplot() + theme_minimal()
```

#### How many samples per treatment, tank, etc?

```{r}
fragments %>% group_by(treatment) %>% tally()
fragments %>% group_by(temperature) %>% tally()
fragments %>% group_by(ph) %>% tally()
fragments %>% group_by(treatment, tank) %>% tally() %>% arrange(treatment, tank)
```

#### How many genes total in the filtered count matrix?

```{r}
prettyNum(ncol(counts.ts),big.mark = ",")
```

#### What's the average no. of genes per sample, etc.? 

```{r}
data.frame(rowSums(counts.ts != 0)) %>% 
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample") %>% 
  summarise(mean=round(mean(count.total, na.rm=T)), 
            sd=round(sd(count.total, na.rm=T)), 
            se=round(sd/sqrt(length(sample))),
            median=median(count.total, na.rm=T),
            min=min(count.total, na.rm=T), 
            max=max(count.total, na.rm=T)) %>% t() %>% as.data.frame() %>% 
  mutate(V1=prettyNum(V1, big.mark = ",")) %>% 
  dplyr::rename("genes.per.sample"="V1")
```

#### Total no. of genes identified in each sample

```{r}
ggplotly(
ggplot(data = data.frame(rowSums(counts.t != 0, na.rm=TRUE)) %>% 
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample")) +
           geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total # genes by sample") + 
             theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())) 
```
```{r}
#### Merge sample key info to count data, then sort, and generate heat map for initial inspection by treatment 

# IMPORTANT NOTE: to use data for ALL genes, use object "counts.t". To remove low-frequency genes, use object "counts.ts" 

# merge count data with sample key, reset row names as sample names, and arrange by infection, then temperature, then day 
counts.tk <- merge(x=sample.info[,c("vial_label", "temperature", "ph", "treatment", "tank", "treatment_tank")], by.x="vial_label", y=counts.ts, by.y="row.names") %>% 
  arrange(treatment, tank)  %>% column_to_rownames(var="vial_label") %>% droplevels()

head(counts.tk[1:24]) #check out results of merge/arrange
```

```{r}
#### Reformat for DESeq, ensure correct sample order for 

# NOTE: It is absolutely critical that the **columns of the count matrix** and the **rows of the column data (information about samples)** are in the same order. DESeq2 will not make guesses as to which column of the count matrix belongs to which row of the column data, these must be provided to DESeq2 already in consistent order.

all(rownames(counts.tk) == counts.tk[-1:-5] %>% t() %>% colnames()) #check that rownames of untransformed matrix match column names of transformed matrix. Should print 'TRUE' 
```

```{r}
##### If using counts which included multimapped counted fractionally, need to round to the nearest integer for DESeq2. 
#_Not needed for gene counts produced by STAR aligner_

#counts.tk <- counts.tk %>% mutate_if(is.numeric, round)
```

```{r}
#### Generate DESeq datasets with various treatment comparisons  

dds.ph <- DESeqDataSetFromMatrix(countData = counts.tk[-1:-5] %>% t(),
                              colData = counts.tk[,"ph", drop=FALSE] ,
                              design = ~ ph)

dds.temperature <- DESeqDataSetFromMatrix(countData = counts.tk[-1:-5] %>% t(),
                              colData = counts.tk[,"temperature", drop=FALSE] ,
                              design = ~ temperature)

dds.treatment <- DESeqDataSetFromMatrix(countData = counts.tk[-1:-5] %>% t(),
                              colData = counts.tk[,"treatment", drop=FALSE] ,
                              design = ~ treatment)

dds.tank <- DESeqDataSetFromMatrix(countData = counts.tk[-1:-5] %>% t(),
                              colData = counts.tk[,"tank", drop=FALSE] ,
                              design = ~ tank)

dds.treatment.tank <- DESeqDataSetFromMatrix(countData = counts.tk[-1:-5] %>% t(),
                              colData = counts.tk[,"treatment_tank", drop=FALSE] ,
                              design = ~ treatment_tank)
```


```{r}
#### Transform data for visualization 

# - Here we transform counts using a variance stabilizing transformation (VST), since we have many samples. 
# - Here we use `blind=FALSE` b/c we are interested in differences explained by experimental design, and may wish to use this transformed data in downstream analyses. 
# 
vsd.pH <- varianceStabilizingTransformation(dds.ph, blind=FALSE)
vsd.temperature <- varianceStabilizingTransformation(dds.temperature, blind=FALSE)
vsd.treatment <- varianceStabilizingTransformation(dds.treatment, blind=FALSE)
vsd.tank <- varianceStabilizingTransformation(dds.tank, blind=FALSE)
vsd.treatment.tank <- varianceStabilizingTransformation(dds.treatment.tank, blind=FALSE)

# Save vsd-transformed count dataframes for later ruse 
save(vsd.pH, file = "../results/deseq2/vsd.pH")
save(vsd.temperature, file = "../results/deseq2/vsd.temperature")
save(vsd.treatment, file = "../results/deseq2/vsd.treatment")
save(vsd.tank, file = "../results/deseq2/vsd.tank")
save(vsd.treatment.tank, file = "../results/deseq2/vsd.treatment.tank")
```

## Unsupervised analyses 

### Heat maps

```{r, include=F}
#NOTE: scale="column" b/c range of counts is so huge, so counts have been scaled 

# # from raw counts, but scaled by column (aka gene)
# pheatmap(data.matrix(counts.tk[-1:-5]), Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, 
#                      show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE,
#                   annotation_colors = list(treatment=
#           c(`3_amb`= "navyblue", `3_low`="royalblue1",
#             `6_amb`="darkgreen", `6_low`="yellow3",
#             `10_amb`="firebrick4", `10_low`="sienna2")),
#          scale="column", #color=c("dodgerblue3", "goldenrod1"), 
#          main = "gene counts - not transformed", annotation_row=counts.tk[,c("treatment"), drop=FALSE])
```

#### Heatmap from all gene counts, variance-stabilization transformed

```{r}
pheatmap(t(assay(vsd.treatment)), Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
                     show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE,
         annotation_colors = list(treatment=
          c(`3_amb`= "navyblue", `3_low`="royalblue1",
            `6_amb`="darkgreen", `6_low`="yellow3",
            `10_amb`="firebrick4", `10_low`="sienna2")),
         annotation_row=as.data.frame(colData(vsd.treatment)[,c("treatment"), drop=FALSE]),
         main = "gene counts - VSD-transformed")
```

#### Heat map with top 10% most variable genes, VSD-transformed counts

```{r}
# calculate CV for each gene across all samples 
most.variable <- assay(vsd.treatment) %>% as.data.frame() %>%
  rownames_to_column(var = "gene") %>%
  rowwise() %>% 
  mutate(
     sd = sd(c_across(-gene)),
     mean = mean(c_across(-gene)),
     cv = sd/mean) %>%
  #column_to_rownames("gene") %>%
  arrange(desc(cv)) %>% 
# subset top 10% of most variable genes 
  head(n=round(0.1*nrow(assay(vsd.treatment)))) %>%
  select(gene) %>% unlist() %>% as.vector()


# Generate heat map / cluster those top 10% most variable genes
cluster.most.variable <- pheatmap(t(assay(vsd.treatment))[,most.variable], 
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
                     show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE,
         annotation_colors = list(treatment=
          c(`3_amb`= "navyblue", `3_low`="royalblue1",
            `6_amb`="darkgreen", `6_low`="yellow3",
            `10_amb`="firebrick4", `10_low`="sienna2")),
         annotation_row=as.data.frame(colData(vsd.treatment)[,c("treatment"), drop=FALSE]),
         main = "gene counts - top 10% most variable\n(VSD-transformed)")
```



### Hierarchical clustering

NOTE: this was generated by the pheatmap function in the previous code chunk

```{r}
# Just plot hierarchical cluster, color-code sample by treatment, or by tank number
cluster.dendo <- as.dendrogram(cluster.most.variable$tree_row) 
#plot(cluster.dendo) # bare bones dendogram
#order.dendrogram(cluster.dendo) # to find out the order of the sample labels 

# Create a dataframe ordered the same as the dendogram for annotation purposes 
dendo.df <- sample.info[match(cluster.most.variable$tree_row$labels, sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")]
all(dendo.df$vial_label == cluster.most.variable$tree_row$labels) # Should == TRUE

# Create color coding vectors 
colorCodes.treat <- c(`3_amb`= "navyblue", `3_low`="royalblue1",
            `6_amb`="darkgreen", `6_low`="yellow3",
            `10_amb`="firebrick4", `10_low`="sienna2")
colorCodes.tank <- colorRampPalette(brewer.pal(12, "Paired"))(length(levels(sample.info$tank)))

# # Color code sample names by treatment
#labels_colors(cluster.dendo) <- colorCodes.treat[as.character(dendo.df$treatment[order.dendrogram(cluster.dendo)])]
#plot(cluster.dendo, xlab="Sample (color=treatment)") # plot with sample names color coded by treatment 

# Plot dendogram color-code by treatment, but labels are tank number
labels(cluster.dendo) <- dendo.df$tank[order.dendrogram(cluster.dendo)]
labels_colors(cluster.dendo) <- colorCodes.treat[as.character(dendo.df$treatment[order.dendrogram(cluster.dendo)])]
plot(cluster.dendo, xlab="Tank Number", cex.lab=0.9) # plot with sample names color coded by tank
legend("topright", title = "Treatment", bty="n", legend =
         c("3_amb","3_low","6_amb","6_low","10_amb", "10_low"), 
       fill = c("navyblue","royalblue1","darkgreen","yellow3","firebrick4","sienna2"))


# # Color code sample names by tank
# labels_colors(cluster.dendo) <- colorCodes.tank[dendo.df$tank][order.dendrogram(cluster.dendo)]
# plot(cluster.dendo, xlab="Sample (color=tank)") # plot with sample names color coded by tank

# # A sub tree - looking at branches
# par(cex = 1)
# plot(cluster.dendo[[1]], horiz = TRUE)
```

#### **NOTE: Samples 194 & 137 look like outliers!**

```{r}

#### Save variance-stabilization normalized gene counts x sample matrix, annotated for later use 

####**NOTE: Here forward I primarily work with the vsd.treatment object, so I can do pairwise treatment contrasts** 

counts.trans <- assay(vsd.treatment) %>% t() %>% as.matrix() %>% as.data.frame() %>% 
  rownames_to_column(var = "sample") %>% 
  left_join(sample.info[,c("vial_label", "treatment", "tank", "temperature", "ph")], by = c("sample"="vial_label")) %>%
  column_to_rownames(var = "sample") %>% 
  dplyr::select(treatment, temperature, ph, tank, everything()) %>% 
  mutate(Treatment=factor(treatment, levels=c("3_low", "6_amb", "6_low", "10_amb", "10_low"))) %>%
  mutate(temperature=factor(temperature, levels=c("3", "6", "10"))) %>%
  mutate(ph=factor(ph, levels=c("amb", "low")))
save(counts.trans, file = "../results/star/counts.trans")
```

### PCA's

```{r}
# To view the the guts of the PlotPCA function in DESeq2, execute this code chunk. It reveals that, by default, it only uses the top 500 most influential genes to plot the PCA! 

#getMethod("plotPCA","DESeqTransform")
```

#### PCAs using DESeq2

NOTE: Hover over points to see the sample numbers

```{r}
# Generate PCA with points color coded by pH Treatment with various number of top genes to use for principal components, selected by highest row variance  

ggplotly(
  plotPCA(vsd.treatment, intgroup="treatment", ntop=500) +
           ggtitle("PCA by Treatment (var-stabilizing transformed)\ntop 500 genes") +
    geom_point(size=3, aes(text=colnames(vsd.treatment))) +
      scale_color_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen",
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
      theme_minimal()+ stat_ellipse() , tooltip = "text")
# 
# ggplotly(
#   plotPCA(vsd.treatment, intgroup="treatment", ntop=5000) + 
#            ggtitle("PCA by Treatment (var-stabilizing transformed)\ntop 5k genes") + 
#     geom_point(size=3, aes(text=colnames(vsd.treatment))) + 
#       scale_color_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
#                               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
#     theme_minimal()+ stat_ellipse() , tooltip = "text")
# 
# ggplotly(
#   plotPCA(vsd.treatment, intgroup="treatment", ntop=50000) + 
#            ggtitle("PCA by Treatment (var-stabilizing transformed)\ntop 50k genes") + 
#     geom_point(size=3, aes(text=colnames(vsd.treatment))) + 
#       scale_color_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
#                               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
#     theme_minimal()+ stat_ellipse() , tooltip = "text")

ggplotly(
  plotPCA(vsd.treatment, intgroup="treatment", ntop=nrow(assay(vsd.treatment))) + 
           ggtitle("PCA by Treatment (var-stabilizing transformed)\nall genes") + 
    geom_point(size=3, aes(text=colnames(vsd.treatment))) + 
      scale_color_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
    theme_minimal()+ stat_ellipse() , tooltip = "text")


# PCA with points color coded by tank 
ggplotly(
  plotPCA(vsd.tank, intgroup="tank") + 
           ggtitle("PCA by Tank (var-stabilizing transformed)") + 
    geom_point(size=3, aes(text=colnames(vsd.tank))) + 
   theme_minimal(), tooltip = "text")

# PCA.data.treatment <- plotPCA(vsd.treatment, intgroup=c("treatment"), returnData=TRUE)
# save(PCA.data.treatment, file="../results/PCA.data.treatment")
```

This PCA also suggests that two samples - 137 and 194 - are outliers. Consider removing. 

### PCAs using prcomp

This enables screeplot to ID significant PCs, calculates variance, etc. 

```{r, warning=F, message=F}
#pca.princomp <- prcomp(cov(assay(vsd.pH)), scale=F) #scale=F for variance-covariance matrix
pca.princomp <- prcomp(t(assay(vsd.treatment))) # using the same code that's under the hood of PlotPCA but using all genes 
pca.eigenval(pca.princomp) #The Proporation of Variance = %variance 
pc.percent <- pca.eigenval(pca.princomp)[2,1:5]*100
pc.percent[1:2] %>% sum()
screeplot(pca.princomp, bstick=TRUE) #looks like PC 1-4 are significant 
```

```{r}
#### Generate dataframe with prcomp results 

tab.expr <- data.frame(sample.id = colnames(assay(vsd.treatment)),
    EV1.all = pca.princomp$x[,1],    # the first eigenvector
    EV2.all = pca.princomp$x[,2],    # the second eigenvector
    EV3.all = pca.princomp$x[,3],    # the third eigenvector
    EV4.all = pca.princomp$x[,4],    # the fourth eigenvector
    stringsAsFactors = FALSE)
#shapiro.test(pca.princomp$x) #sample size too large for shapiro test which is weird 
#hist(pca.princomp$x) #normal? hard to say, maybe
tab.expr.annot <- left_join(tab.expr, sample.info, by=c("sample.id"="vial_label")) %>% droplevels()
```

```{r}
#### PCA biplots from prcomp analysis

# GGSCATTER with ellipses and stars - PC1 PC2
ggscatter(tab.expr.annot #%>% 
            # mutate(treatment=str_replace(Treatment, "Ambient", "Ambient (pH 8.0)"),
            #        treatment=str_replace(Treatment, "Moderate", "Moderate (pH 7.8)"),
            #        treatment=str_replace(Treatment, "Low", "Severe (pH 7.6)"))
          , 
          x="EV1.all", y="EV2.all", col="treatment", size=2.5, alpha=0.85, ellipse = TRUE, star.plot = TRUE, 
          # palette = c(`Ambient (pH 8.0)`="#2c7bb6", `Moderate (pH 7.8)`="#fdae61", `Severe (pH 7.6)`="#d7191c")
          ) +
  theme_minimal() + ggtitle("Global gene expression, PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
  labs(color="OA Treatment") + 
  labs(fill="OA Treatment") + 
  theme(legend.position = "right") + 
  scale_color_manual(name="Treatment", values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
  scale_fill_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2"), guide = "none") +
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank")))

# GGSCATTER with ellipses and stars - PC1 PC3
ggscatter(tab.expr.annot, x="EV1.all", y="EV3.all",
          col="treatment", size=2.5, alpha=0.85, ellipse = TRUE, star.plot = TRUE, 
           #palette = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")
          ) +
  theme_minimal() + ggtitle("Global gene expression, PC1xPC3") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC3 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_manual(name="Treatment", values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
  scale_fill_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2"), guide = "none") +
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"))) 

# GGSCATTER with ellipses and stars - PC1 PC4
ggscatter(tab.expr.annot, x="EV1.all", y="EV4.all",
          col="treatment", size=2.5, alpha=0.85, ellipse = TRUE, star.plot = TRUE, 
          #palette = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")
          ) +
  theme_minimal() + ggtitle("Global gene expression, PC1xPC4") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC4 (", round(pc.percent[4], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_manual(name="Treatment", values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
  scale_fill_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2"), guide = "none") +
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"))) 


# GGSCATTER with ellipses and stars - PC3 PC4
ggscatter(tab.expr.annot, x="EV3.all", y="EV4.all",
          col="treatment", size=2.5, alpha=0.85, ellipse = TRUE, star.plot = TRUE, 
          #palette = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")
          ) +
  theme_minimal() + ggtitle("Global gene expression, PC3xPC4") + 
  xlab(paste("PC3 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
  ylab(paste("PC4 (", round(pc.percent[4], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_manual(name="Treatment", values=c("3_amb"= "navyblue",
                                                "3_low"="royalblue1","6_amb"="darkgreen", 
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
  scale_fill_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2"), guide = "none") +
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"))) 
```

```{r}
#### Examine potential tank effects for each treatment

#_NOT WORKING - FIX OR REMOVE_

# pc.axes <- names(tab.expr.annot[2:5])
# x <- c(1,1,1,1,2,2,2,3,3,4)
# y <- c(2,3,4,5,3,4,5,4,5,5)
# 
# pc.plots <- vector("list", length(x))
# for (i in 1:length(x)) {
# pc.plots[[i]] <-
#   #ggplotly(
#     ggplot(
#   tab.expr.annot %>%
#   mutate(Tank_n=as.numeric(as.character(Tank))) %>%
#   mutate_at("Tank_n", function(x) {
#     case_when(
#       x == 1 ~ "A",
#       x == 5 ~ "A",
#       x == 3 ~ "A",
#       x == 2 ~ "B",
#       x == 13 ~ "B",
#       x == 4 ~ "B",
#       x == 7 ~ "C",
#       x == 15 ~ "C",
#       x == 9 ~ "C",
#       x == 10 ~ "D",
#       x == 16 ~ "D",
#       x == 20 ~ "D",
#       x == 11 ~ "E",
#       x == 18 ~ "E")}),
#   aes(x=!!as.symbol(pc.axes[x[i]]),y=!!as.symbol(pc.axes[y[i]]))) +
#   geom_point(aes(col=Treatment, shape=Tank_n, fill=Treatment), size=3.5, alpha=0.85) +
#   theme_minimal() + ggtitle(paste("PC", x[i], " x ", "PC", y[i], sep = "")) +
#   xlab(paste("PC", x[i], " (", round(pc.percent[x[i]], digits = 1), "%)", sep="")) +
#   ylab(paste("PC", y[i], " (", round(pc.percent[y[i]], digits = 1), "%)", sep="")) +
#   theme(legend.position = "right") + scale_shape_manual(values=seq(0,5))#)
# }
# pc.plots
```

### Heatmap of the sample-to-sample distances

Another use of the transformed data is sample clustering. Here, we apply the dist function to the transpose of the transformed count matrix to get sample-to-sample distances.

A heatmap of this distance matrix gives us an overview over similarities and dissimilarities between samples. We have to provide a hierarchical clustering hc to the heatmap function based on the sample distances, or else the heatmap function would calculate a clustering based on the distances between the rows/columns of the distance matrix.

```{r}
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

# Here we show treatment and tank number
sampleDistss <- dist(t(assay(vsd.treatment.tank)))
sampleDistMatrixs <- as.matrix(sampleDistss)
rownames(sampleDistMatrixs) <- vsd.treatment.tank$treatment_tank   #set row names
colnames(sampleDistMatrixs) <- colnames(vsd.treatment)

all(colnames(sampleDistMatrixs) == rownames(counts.tk))

pheatmap(sampleDistMatrixs,
         clustering_distance_rows=sampleDistss,
         clustering_distance_cols=sampleDistss,
         col=rev(colors), fontsize = 6.5)
```


```{r}
### perMANOVA - multivariate analysis of variance 

#### _not working, probably b/c it's a 2-way design_ 

# # Effect of treatment 
# vsd.t <- t(assay(vsd.treatment)) %>% as.data.frame()
# perm <- adonis(vsd.t ~ treatment, data=sample.info %>% remove_rownames() %>% column_to_rownames("vial_label"), permutations = 1000, method="bray")
# 
# # Pairwise comparisons 
# vsd.t <- sample.info %>% dplyr::select(Sample, Treatment, Tank) %>% left_join(t(assay(vsd.pH)) %>% as.data.frame() %>% rownames_to_column("Sample"))
# pairwise.adonis(vsd.t[,-1:-3], vsd.t$Treatment, perm = 1000)

```

```{r}
#### How many unique genes were measured per treatment. This takes the mean of each gene by group, and counts up those that have mean >10. 

#Warning: this takes forever 
# # Counts before filtering 
# sample.info %>% select(Sample, Treatment) %>% left_join(counts.t %>% rownames_to_column("Sample")) %>% 
#   group_by(Treatment) %>% summarise_if(is.numeric, "mean") %>% mutate(genes=rowSums(select_if(., is.numeric)>10)) %>% select(Treatment, genes)

# # Counts after filtering
# sample.info %>% select(Sample, Treatment) %>% left_join(counts.tk %>% rownames_to_column("Sample")) %>% 
#   select(-Sample, -Tank, -Treatment_Tank) %>% group_by(Treatment) %>% summarise_all("mean") %>% mutate(genes=rowSums(is.numeric(.)>10)) %>% select(Treatment, genes)
```

### Differential Expression Analysis

_This is the core DESeq2 analysis!_ 

```{r, message=F, warning=F, include=F}
##### Run the function `DESeq` to assess differential expression 
dds.DESeq.treatment <- DESeq(dds.treatment) 
dds.DESeq.temperature <- DESeq(dds.temperature) 
dds.DESeq.ph <- DESeq(dds.ph) 
```

### Identify differentially expressed genes among contrasts 

#### pH contrasts 

```{r}
print("Comparison: Ambient pH vs. Low pH, 3C")
summary(res.pH.3 <- results(dds.DESeq.treatment, contrast=c("treatment", "3_amb", "3_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pH, 3C:",  sum(res.pH.3$padj < 0.05, na.rm=TRUE))

print("Comparison: Ambient pH vs. Low pH, 6C")
summary(res.pH.6 <- results(dds.DESeq.treatment, contrast=c("treatment", "6_amb", "6_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pH, 6C:",  sum(res.pH.6$padj < 0.05, na.rm=TRUE))

print("Comparison: Ambient pH vs. Low pH, 10C")
summary(res.pH.10 <- results(dds.DESeq.treatment, contrast=c("treatment", "10_amb", "10_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pH, 10C:",  sum(res.pH.10$padj < 0.05, na.rm=TRUE))
```
#### Temperature contrasts 

##### Comparing temperatures within ambient pH 

```{r}
print("Comparison: 3C vs. 6C, ambient pH")
summary(res.temp.3.6.amb <- results(dds.DESeq.treatment, contrast=c("treatment", "3_amb", "6_amb"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 6C & 10C, ambient pH:",  sum(res.temp.3.6.amb$padj < 0.05, na.rm=TRUE))

print("Comparison: 3C vs. 10C, ambient pH")
summary(res.temp.3.10.amb <- results(dds.DESeq.treatment, contrast=c("treatment", "3_amb", "10_amb"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 3C & 10C, ambient pH:",  sum(res.temp.3.10.amb$padj < 0.05, na.rm=TRUE))

print("Comparison: 6C vs. 10C, ambient pH")
summary(res.temp.6.10.amb <- results(dds.DESeq.treatment, contrast=c("treatment", "6_amb", "10_amb"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 6C & 10C, ambient pH:",  sum(res.temp.6.10.amb$padj < 0.05, na.rm=TRUE))
```

##### Comparing temperatures within low pH 

```{r}
print("Comparison: 3C vs. 6C, low pH")
summary(res.temp.3.6.low <- results(dds.DESeq.treatment, contrast=c("treatment", "3_low", "6_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 6C & 10C, low pH:",  sum(res.temp.3.6.low$padj < 0.05, na.rm=TRUE))

print("Comparison: 3C vs. 10C, low pH")
summary(res.temp.3.10.low <- results(dds.DESeq.treatment, contrast=c("treatment", "3_low", "10_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 3C & 10C, low pH:",  sum(res.temp.3.10.low$padj < 0.05, na.rm=TRUE))

print("Comparison: 6C vs. 10C, low pH")
summary(res.temp.6.10.low <- results(dds.DESeq.treatment, contrast=c("treatment", "6_low", "10_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 6C & 10C, low pH:",  sum(res.temp.6.10.low$padj < 0.05, na.rm=TRUE))
```

```{r, include=F, echo=F, message=F}
#### Just looking at temperature and pH 

# print("Comparison: 6C vs 10C, any pH")
# summary(res.temp.3.6 <- results(dds.DESeq.temperature, contrast=c("temperature", "3", "6"), alpha=0.05))
# paste("No. of genes differentially expressed (padj<0.05) between 3C & 6C, any pH:",  sum(res.temp.3.6$padj < 0.05, na.rm=TRUE))
# 
# print("Comparison: 6C vs 10C, any pH")
# summary(res.temp.3.10 <- results(dds.DESeq.temperature, contrast=c("temperature", "3", "10"), alpha=0.05))
# paste("No. of genes differentially expressed (padj<0.05) between 3C & 10C, any pH:",  sum(res.temp.3.10$padj < 0.05, na.rm=TRUE))
# 
# print("Comparison: 6C vs 10C, any pH")
# summary(res.temp.6.10 <- results(dds.DESeq.temperature, contrast=c("temperature", "6", "10"), alpha=0.05))
# paste("No. of genes differentially expressed (padj<0.05) between 6C & 10C, any pH:",  sum(res.temp.6.10$padj < 0.05, na.rm=TRUE))
# 
# print("Comparison: amb and low pH, any temp")
# summary(res.ph <- results(dds.DESeq.ph, contrast=c("ph", "amb", "low"), alpha=0.05))
# paste("No. of genes differentially expressed (padj<0.05) between ambient and low pH, any temperature",  sum(res.ph$padj < 0.05, na.rm=TRUE))
```

#### Number of genes analyzed 

```{r}
nrow(res.pH.6)
```

```{r}
#### Save differential expression results to objects 

#_These are the most straight-forward pairwise contrasts_

save(res.pH.3, file = "../results/deseq2/res.pH.3")
save(res.pH.6, file = "../results/deseq2/res.pH.6")
save(res.pH.10, file = "../results/deseq2/res.pH.10")

save(res.temp.3.6.amb, file = "../results/deseq2/res.temp.3.6.amb")
save(res.temp.3.10.amb, file = "../results/deseq2/res.temp.3.10.amb")
save(res.temp.6.10.amb, file = "../results/deseq2/res.temp.6.10.amb")

save(res.temp.3.6.low, file = "../results/deseq2/res.temp.3.6.low")
save(res.temp.3.10.low, file = "../results/deseq2/res.temp.3.10.low")
save(res.temp.6.10.low, file = "../results/deseq2/res.temp.6.10.low")
```


```{r}
#### Subset the DESeq results for only those statistically sign. ones 

diffex.pH.3 <- subset(res.pH.3, padj < 0.05)
diffex.pH.6 <- subset(res.pH.6, padj < 0.05)
diffex.pH.10 <- subset(res.pH.10, padj < 0.05)
diffex.temp.3.6.amb <- subset(res.temp.3.6.amb, padj < 0.05)
diffex.temp.3.10.amb <- subset(res.temp.3.10.amb, padj < 0.05)
diffex.temp.6.10.amb <- subset(res.temp.6.10.amb, padj < 0.05)
diffex.temp.3.6.low <- subset(res.temp.3.6.low, padj < 0.05)
diffex.temp.3.10.low <- subset(res.temp.3.10.low, padj < 0.05)
diffex.temp.6.10.low <- subset(res.temp.6.10.low, padj < 0.05)

save(diffex.pH.3, file = "../results/deseq2/diffex.pH.3")
save(diffex.pH.6, file = "../results/deseq2/diffex.pH.6")
save(diffex.pH.10, file = "../results/deseq2/diffex.pH.10")
save(diffex.temp.3.6.amb, file = "../results/deseq2/diffex.temp.3.6.amb")
save(diffex.temp.3.10.amb, file = "../results/deseq2/diffex.temp.3.10.amb")
save(diffex.temp.6.10.amb, file = "../results/deseq2/diffex.temp.6.10.amb")
save(diffex.temp.3.6.low, file = "../results/deseq2/diffex.temp.3.6.low")
save(diffex.temp.3.10.low, file = "../results/deseq2/diffex.temp.3.10.low")
save(diffex.temp.6.10.low, file = "../results/deseq2/diffex.temp.6.10.low")
```

### Summary stats for the number of Differentially Expressed Genes (DEGs)

#### Bar plot of the number of DEGs identified in each contrast 

```{r}
# write simple function to return the name of an object as a string

degs <- list(
diffex.pH.3, diffex.pH.6, diffex.pH.10,
diffex.temp.3.6.amb, diffex.temp.3.10.amb, diffex.temp.6.10.amb,
diffex.temp.3.6.low, diffex.temp.3.10.low, diffex.temp.6.10.low)

names(degs) <- c("Amb vs. Low pH @3C", "Amb vs. Low pH @6C", "Amb vs. Low pH @ 10C", 
                 "3 vs. 6C @Amb pH", "3 vs. 10C @Amb pH", "6 vs. 10C @Amb pH", 
                 "3 vs. 6C @Low pH", "3 vs. 10C @Low pH", "6 vs. 10C @Low pH")
grp <- c(rep("pH contrast", times=3), rep("temperature contrast, @ amb pH", times=3), rep("temperature contrast, @ low pH", times=3))


n.degs <- data.frame(matrix(NA, nrow = length(degs), ncol = 4))
names(n.degs) <- c("contrast", "n.degs", "perc", "grp")
for (i in (1:length(degs))) {
  n.degs[i,1] <- names(degs[i])
  n.degs[i,2] <- nrow(degs[[i]])
  n.degs[i,3] <- 100*round(nrow(degs[[i]])/ncol(counts.ts), digits = 2)
  n.degs[i,4] <- grp[i]
}

n.degs$contrast <- factor(n.degs$contrast, levels=n.degs$contrast, ordered = T)

ggplot(data=n.degs, aes(x=contrast, y=n.degs, fill=grp)) + geom_bar(stat = "identity") +
  geom_text(hjust=-0.1, size=3.5, aes(label=paste(comma(n.degs), " (", perc, "%)", sep=""))) + 
#  geom_text(hjust=-1, aes(label=perc)) + 
  xlab("Contrast") + ylab("Number Differentially Expressed Genes (% of all genes)") + coord_flip() + 
  theme_minimal() + ylim(c(0,12500)) + theme(legend.position = "bottom") +
  scale_fill_discrete(name=NULL) +
  guides(fill=guide_legend(nrow=3,byrow=TRUE))
```


#### Number of unique DEGs

##### Among pH, for each temperature?

```{r}
length(unique(c(diffex.pH.3 %>% rownames(), diffex.pH.6 %>% rownames(), diffex.pH.10 %>% rownames())))
```

##### Among temperatures, ambient pH?
```{r}
length(unique(c(diffex.temp.3.6.amb %>% rownames(), diffex.temp.3.10.amb %>% rownames(), diffex.temp.6.10.amb %>% rownames())))
```
##### Among temperatures, low pH?
```{r}
length(unique(c(diffex.temp.3.6.low %>% rownames(), diffex.temp.3.10.low %>% rownames(), diffex.temp.6.10.low %>% rownames())))
```
##### What % of all examined genes are DEGs among pH treatments? 
```{r}
round(100*(length(unique(c(diffex.pH.3 %>% rownames(), diffex.pH.6 %>% rownames(), diffex.pH.10 %>% rownames()))))/(res.pH.3 %>% nrow()), digits = 1)
```
##### What % of all examined genes are DEGs among temperature treatments (ambient pH only)? 
```{r}
round(100*(length(unique(c(diffex.temp.3.6.amb %>% rownames(), diffex.temp.3.10.amb %>% rownames(), diffex.temp.6.10.amb %>% rownames()))))/(res.temp.3.6.amb %>% nrow()), digits = 1) 
```
##### What % of all examined genes are DEGs among temperature treatments (low pH only)? 
```{r}
round(100*(length(unique(c(diffex.temp.3.6.low %>% rownames(), diffex.temp.3.10.low %>% rownames(), diffex.temp.6.10.low %>% rownames()))))/(res.temp.3.6.low %>% nrow()), digits = 1) 
```

### Heat maps of DEGs 

#### Single heatmap with all DEGs among any pH treatment 
NOTE: There are 4,525 DEGs among pH treatment in the 6C group, but only 2 and 58 among the 3C and 10C groups, respectively. This is visible in the heat map- there are clear differences in expression in the 6_amb (green) and 6_low (yellow) treatments. 

```{r}
# Generate counts matrix With DEGs among any pH treatment
diffex.pH.counts <- subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% unique(c(rownames(diffex.pH.3),rownames(diffex.pH.6),rownames(diffex.pH.10))))

# Create a dataframe to annotate heatmap with treatments 
dds.pH.df <- sample.info[match(colnames(diffex.pH.counts), sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")] %>%
  remove_rownames() %>% column_to_rownames(var = "vial_label")

# Make sure treatment order is correct
all(colnames(diffex.pH.counts) == rownames(dds.pH.df)) #double check that samples are still in same order 

# All DEGs among any pH treatment (within temperatures)
pheatmap(diffex.pH.counts, cluster_rows=TRUE, cluster_cols=FALSE,
         show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=
             c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen",
               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")),
         scale="row", main = "Differentially expressed genes (DEGs) among any pH treatment", 
         annotation_col=dds.pH.df[,"treatment", drop=FALSE], fontsize = 8,
         cutree_rows = 2,  border_color=F)

# # another way to do a heatmap 
# heatmap(diffex.pH.counts, Colv = NA, scale="row", ColSideColors = brewer.pal(3, "Set1")[dds.pH.df$Treatment])
```

#### Single heatmap with all DEGs among any temperature treatment, ambient pH only

```{r}
ph.amb <- sample.info[sample.info$ph=="amb",]$vial_label

# Generate counts matrix With DEGs among any tempreature treatment, ambient pH only
diffex.temp.counts.amb <-  
  (subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% 
            unique(c(rownames(diffex.temp.3.6.amb),
                     rownames(diffex.temp.3.10.amb),
                     rownames(diffex.temp.6.10.amb)))) %>% 
     as.data.frame())[ph.amb] %>% as.matrix()

# Create a dataframe to annotate heatmap with treatments 
dds.temp.df <- sample.info[match(colnames(diffex.temp.counts.amb), sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")] %>%
  remove_rownames() %>% column_to_rownames(var = "vial_label")

# Make sure treatment order is correct
all(colnames(diffex.temp.counts.amb) == rownames(dds.temp.df)) #double check that samples are still in same order 

# All DEGs among any temp treatment (within temperatures)
pheatmap(diffex.temp.counts.amb, cluster_rows=TRUE, cluster_cols=FALSE,
         show_rownames=FALSE, na.rm=TRUE, 
         annotation_colors = list(temperature = c("3"= "navyblue", "6"="darkgreen","10"="firebrick4")),
         scale="row", 
         main = "Differentially expressed genes (DEGs) among any temperature treatment\nAmbient pH only", 
         annotation_col=dds.temp.df[,"temperature", drop=FALSE], fontsize = 8,
         cutree_rows = 2,  border_color=F)
```

#### Single heatmap with all DEGs among any temperature treatment, low pH only

```{r}
ph.low <- sample.info[sample.info$ph=="low",]$vial_label

# Generate counts matrix With DEGs among any tempreature treatment, low pH only
diffex.temp.counts.low <-  
  (subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% 
            unique(c(rownames(diffex.temp.3.6.low),
                     rownames(diffex.temp.3.10.low),
                     rownames(diffex.temp.6.10.low)))) %>% 
     as.data.frame())[ph.low] %>% as.matrix()

# Create a dataframe to annotate heatmap with treatments 
dds.temp.df <- sample.info[match(colnames(diffex.temp.counts.low), sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")] %>%
  remove_rownames() %>% column_to_rownames(var = "vial_label")

# Make sure treatment order is correct
all(colnames(diffex.temp.counts.low) == rownames(dds.temp.df)) #double check that samples are still in same order 

# All DEGs among any temp treatment (within temperatures)
pheatmap(diffex.temp.counts.low, cluster_rows=TRUE, cluster_cols=FALSE,
         show_rownames=FALSE, na.rm=TRUE, 
         annotation_colors = list(temperature = c("3"="royalblue1","6"="yellow3","10"="sienna2")),
         scale="row", 
         main = "Differentially expressed genes (DEGs) among any temperature treatment\nLow pH only", 
         annotation_col=dds.temp.df[,"temperature", drop=FALSE], fontsize = 8,
         cutree_rows = 2,  border_color=F)
```

### Venn Diagrams of DEGs by pairwise contrast

```{r}
names(degs)

display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}
```
#### pH contrasts 

```{r}
x <- list(
  AvsL.3 = row.names(degs$`Amb vs. Low pH @3C`), 
  AvsL.6 = row.names(degs$`Amb vs. Low pH @6C`), 
  AvsL.10 = row.names(degs$`Amb vs. Low pH @ 10C`)
  )

#display_venn(x) #bare bones venn diagram

display_venn(
        x,
        category.names = c("Ambient vs.\nLow pH, 3C" , "Ambient vs.\nLow pH, 6C" , "Ambient vs.\nLow pH, 10C"),
        # Circles
        lwd = .5,
        lty = 'blank',
        fill = c("gray25", "gray50", "gray75"),
        # Numbers
        cex = 1.2,
        fontface = "italic",
        # Set names
        cat.cex = 0.7,
        cat.dist = c(0.01, 0.01, 0.01),
        cat.fontface = "bold",
        cat.default.pos = "outer" 
)
```

```{r, include=F, echo=F, message=F}
#### Temperature contrasts, Ambient pH

x <- list(
  amb.3vs6 = row.names(degs$`3 vs. 6C @Amb pH`), 
  amb.3vs10 = row.names(degs$`3 vs. 10C @Amb pH`), 
  amb.6vs10 = row.names(degs$`6 vs. 10C @Amb pH`) #,
  # low.3vs6 = row.names(degs$`3 vs. 6C @Low pH`),
  # low.3vs10 = row.names(degs$`3 vs. 10C @Low pH`),
  # low.6vs10 = row.names(degs$`6 vs. 10C @Low pH`)
)

#display_venn(x) #bare bones venn diagram

display_venn(
        x,
        category.names = c(
          "3C vs. 6C\nAmb pH" , "3C vs. 10C\nAmb pH" , "6C vs. 10C\nAmb pH" #,
          # "3C vs. 6C\nLow pH" , "3C vs. 10C\nLow pH" , "6C vs. 10C\nLow pH
          ),
        # Circles
        lwd = .5,
        lty = 'blank',
        fill = c("gray25", "gray50", "gray75"),
        # Numbers
        cex = 1,
        fontface = "italic",
        # Set names
        cat.cex = 1,
        cat.dist = c(0.02, 0.02, 0.02),
        cat.fontface = "bold",
        cat.default.pos = "outer" 
)
```

```{r, include=F, echo=F, message=F}
#### Temperature contrasts, Low pH

x <- list(
  # amb.3vs6 = row.names(degs$`3 vs. 6C @Amb pH`), 
  # amb.3vs10 = row.names(degs$`3 vs. 10C @Amb pH`), 
  # amb.6vs10 = row.names(degs$`6 vs. 10C @Amb pH`) #,
  low.3vs6 = row.names(degs$`3 vs. 6C @Low pH`),
  low.3vs10 = row.names(degs$`3 vs. 10C @Low pH`),
  low.6vs10 = row.names(degs$`6 vs. 10C @Low pH`)
)

#display_venn(x) #bare bones venn diagram

display_venn(
        x,
        category.names = c(
          #"3C vs. 6C\nAmb pH" , "3C vs. 10C\nAmb pH" , "6C vs. 10C\nAmb pH" #,
          "3C vs. 6C\nLow pH" , "3C vs. 10C\nLow pH" , "6C vs. 10C\nLow pH"
        ),
        # Circles
        lwd = .5,
        lty = 'blank',
        fill = c("gray25", "gray50", "gray75"),
        # Numbers
        cex = 1,
        fontface = "italic",
        # Set names
        cat.cex = 1,
        cat.dist = c(0.02, 0.02, 0.02),
        cat.fontface = "bold",
        cat.default.pos = "outer" 
)
```
#### Same temperature contrasts, compare low vs. ambient pH conditions

```{r}
x <- list(
  `3C vs. 6C\nAmb pH` = row.names(degs$`3 vs. 6C @Amb pH`), 
  `3C vs. 6C\nLow pH` = row.names(degs$`3 vs. 6C @Low pH`))
y <- list(
  `3C vs. 10C\nAmb pH` = row.names(degs$`3 vs. 10C @Amb pH`), 
  `3C vs. 10C\nLow pH` = row.names(degs$`3 vs. 10C @Low pH`))
z <- list(
  `6C vs. 10C\nAmb pH` = row.names(degs$`6 vs. 10C @Amb pH`),
  `6C vs. 10C\nLow pH` = row.names(degs$`6 vs. 10C @Low pH`))

display_venn(x,cat.dist = c(0.02, 0.01),lty = 'blank', fill = c("gray25", "gray50")) 
display_venn(y,cat.dist = c(0.02, 0.01),lty = 'blank', fill = c("gray25", "gray50")) 
display_venn(z,cat.dist = c(0.02, 0.01),lty = 'blank', fill = c("gray25", "gray50")) 
```

###  DAVID enrichment analysis

It is a little clunky, but the DAVID functional analysis tool is one of the best ways to perform enrichment analysis (that I have found to date). To use, you copy a list of Uniprot IDs (or other gene identifier) to clipboard, then paste it into their tool. You then do the same for all genes in the analysis, providing an accurate background list of genes.  Here is code to copy all genes, then upregulated and downregulated DEGs identified by each contrast.

```{r, include=F, echo=F, message=F}
names(degs)

# BACKGROUND - all genes submitted to DESeq2 analysis 
res.pH.6 %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>% 
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names, protein_names), by = "gene_gadmor") %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)


# # ------- NO DEGS
# print(paste("DEGs, ", names(degs[1])))
# degs[1] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
#   left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor") %>%
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# -------
print(paste("DEGs, ", names(degs[2])))
degs[2] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor") %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# -------
print(paste("DEGs, ", names(degs[3])))
degs[3] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor") %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# -------
print(paste("DEGs, ", names(degs[4])))
degs[4] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor") %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# -------
print(paste("DEGs, ", names(degs[5])))
degs[5] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor") %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# -------
print(paste("DEGs, ", names(degs[6])))
degs[6] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor") %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# -------
print(paste("DEGs, ", names(degs[7])))
degs[7] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor") %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# -------
print(paste("DEGs, ", names(degs[8])))
degs[8] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor") %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# -------
print(paste("DEGs, ", names(degs[9])))
degs[9] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor") %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
```

<!-- ```{bash} -->
<!-- ##**_THIS CODE ISN'T WORKING_** -->
<!-- ## Remove files which list GO_MWU results filenames if they already exist -->
<!-- cd ../results/GO-enrichment/ -->
<!-- pwd -->
<!-- rm GO_filenames.txt -->

<!-- for file in GO_enrich_*.txt -->
<!-- do -->
<!-- filename="$(echo $file)" -->
<!-- sample="$(basename -a $filename | cut -d "_" -f 2  | cut -d "." -f 1)" -->
<!-- printf "%s\t%s\n" "$filename" "$sample" >> GO_filenames.txt -->
<!-- done -->
<!-- ``` -->

```{r}
#### Read in enriched biological functions for all contrasts

filenames <- read_delim(file="../results/GO-enrichment/GO_filenames.txt", delim = "\t", col_names = c("filename", "gene_set"))
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=paste("../results/GO-enrichment/", filenames[i,"filename"], sep=""), delim = "\t"))}

enriched.bp.degs <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("go", "process"), sep = "~") %>%
    mutate(gene_set=factor(gene_set, 
                           levels=c("Amb-vs-Low-3C", "Amb-vs-Low-6C", "Amb-vs-Low-10C", 
                                    "3Cvs6C-Amb-pH", "3Cvs10C-Amb-pH", "6Cvs10C-Amb-pH",
                                    "3Cvs6C-Low-pH", "3Cvs10C-Low-pH", "6Cvs10C-Low-pH"))) %>% #Convert gene_set column to factor, define all gene_sets levels
    tidyr::complete(gene_set, category) # add rows for gene_sets that are missing from dataframe

#enriched.bp.degs %>% write_clip() #copy to clipboard to paste into google spreadsheet 
```

### Bubble plot of enriched Biological Processes 

#### pH contrasts for each temperature

```{r}
#pdf(file="../results/GO-enrichment/BP-pH-enriched-bubble.pdf", height = 7, width = 7)

enriched.bp.degs %>% 
  group_by(gene_set, genes, count) %>%
  summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.degs %>% select(gene_set, category, go, process, count, percent, p_value, fold_enrichment, fdr, genes), by = c("gene_set", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(gene_set, p_value) %>% group_by(gene_set) %>% dplyr::slice(1:25) %>%
  mutate(gene_set=factor(gene_set, ordered = TRUE,
                           levels=c("Amb-vs-Low-3C", "Amb-vs-Low-6C", "Amb-vs-Low-10C", 
                                    "3Cvs6C-Amb-pH", "3Cvs6C-Low-pH",
                                    "3Cvs10C-Amb-pH", "3Cvs10C-Low-pH", 
                                    "6Cvs10C-Amb-pH", "6Cvs10C-Low-pH"))) %>%
  mutate(group=gene_set) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "Amb-vs-Low-3C" ~ "pH",
      x == "Amb-vs-Low-6C" ~ "pH",
      x == "Amb-vs-Low-10C" ~ "pH",
      x == "3Cvs6C-Amb-pH" ~ "Temp, Amb pH",
      x == "3Cvs10C-Amb-pH" ~ "Temp, Amb pH",
      x == "6Cvs10C-Amb-pH" ~ "Temp, Amb pH",
      x == "3Cvs6C-Low-pH" ~ "Temp, Low pH",
      x == "3Cvs10C-Low-pH" ~ "Temp, Low pH",
      x == "6Cvs10C-Low-pH" ~ "Temp, Low pH")}) %>%
  mutate(group=factor(group, ordered=TRUE, levels=c("pH", "Temp, Amb pH", "Temp, Low pH"))) %>% 

  filter(!grepl('Amb-pH|Low-pH', gene_set)) %>% # Filter for only the pH contrasts
  ungroup() %>% arrange(group, gene_set, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=gene_set, col=gene_set)) + 
  geom_point(aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2, 
            labeller = as_labeller(c(`GOTERM_BP_DIRECT`="Enriched Biological Processes\nAmbient vs. Low pH\nby temperature"))) +
  # scale_color_manual(name="Upregulated", values=c(Ambient="#2c7bb6", Moderate="#fdae61", Severe="#d7191c"),
  #            guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) +
  coord_cartesian(clip = "off", ylim = c(-.7, 35)) +
  ggtitle(NULL) +
  annotate("text", x = 1, y = -.15, label = "3C", size=3.5, col="gray15") +
  annotate("text", x = 2, y = -.15, label = "6C", size=3.5, col="gray15") +
  annotate("text", x = 3, y = -.15, label = "10C", size=3.5, col="gray15") +
  annotate("text", x = 2, y = 35, label = "Enriched Biological Processes", size=3, col="gray15") + 
  annotate("text", x = 2, y = 33.5, label = "Ambient vs. Low pH contrasts", size=3.25, col="gray15", fontface = 'bold') + 
  annotate("text", x = 2, y = 32, label = "by temperature", size=3, col="gray15", fontface = 'italic')
#dev.off()
```

#### Temperature contrasts for each pH 

```{r}
#pdf(file="../results/GO-enrichment/BP-temperature-enriched-bubble.pdf", height = 11, width = 8)

enriched.bp.degs %>% 
  group_by(gene_set, genes, count) %>%
  summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.degs %>% select(gene_set, category, go, process, count, percent, p_value, fold_enrichment, fdr, genes), by = c("gene_set", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(gene_set, p_value) %>% group_by(gene_set) %>% dplyr::slice(1:20) %>%
  mutate(gene_set=factor(gene_set, ordered = TRUE,
                           levels=c("Amb-vs-Low-3C", "Amb-vs-Low-6C", "Amb-vs-Low-10C", 
                                    "3Cvs6C-Amb-pH", "3Cvs6C-Low-pH",
                                    "3Cvs10C-Amb-pH", "3Cvs10C-Low-pH", 
                                    "6Cvs10C-Amb-pH", "6Cvs10C-Low-pH"))) %>%
  mutate(group=gene_set) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "Amb-vs-Low-3C" ~ "pH",
      x == "Amb-vs-Low-6C" ~ "pH",
      x == "Amb-vs-Low-10C" ~ "pH",
      x == "3Cvs6C-Amb-pH" ~ "Temp, Amb pH",
      x == "3Cvs10C-Amb-pH" ~ "Temp, Amb pH",
      x == "6Cvs10C-Amb-pH" ~ "Temp, Amb pH",
      x == "3Cvs6C-Low-pH" ~ "Temp, Low pH",
      x == "3Cvs10C-Low-pH" ~ "Temp, Low pH",
      x == "6Cvs10C-Low-pH" ~ "Temp, Low pH")}) %>%
  mutate(group=factor(group, ordered=TRUE, levels=c("pH", "Temp, Amb pH", "Temp, Low pH"))) %>% 

  filter(grepl('Amb-pH|Low-pH', gene_set)) %>% # Filter for only the pH contrasts
  ungroup() %>% arrange(group, gene_set, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=gene_set, col=gene_set)) + 
  geom_point(aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2, labeller = as_labeller(c(`GOTERM_BP_DIRECT`="Enriched Biological Processes\nAmbient vs. Low pH\nby temperature"))) +
  # scale_color_manual(name="Upregulated", values=c(Ambient="#2c7bb6", Moderate="#fdae61", Severe="#d7191c"),
  #            guide = guide_legend(override.aes = list(size=4))) +
  #scale_color_manual(guide="none", values=c("#ffffb3","#8dd3c7","#bebada")) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) +
  coord_cartesian(clip = "off", ylim = c(-6, 80)) +
  ggtitle(NULL) +
  annotate("text", x = 1, y = -1.5, label = "Amb", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 2, y = -1.5, label = "Low", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 3, y = -1.5, label = "Amb", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 4, y = -1.5, label = "Low", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 5, y = -1.5, label = "Amb", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 6, y = -1.5, label = "Low", size=2.75, col="gray15", angle=90) +

  annotate("text", x = 1.5, y = -5, label = "3C vs. 6C", size=2.75, col="gray15", fontface = 'bold') +
  annotate("text", x = 3.5, y = -5, label = "3C vs. 10C", size=2.75, col="gray15", fontface = 'bold') +
  annotate("text", x = 5.5, y = -5, label = "6C vs. 10C", size=2.75, col="gray15", fontface = 'bold') +  
  
  annotate("text", x = 3.5, y = 79.5, label = "Enriched Biological Processes", size=3, col="gray15") +
  annotate("text", x = 3.5, y = 78, label = "Temperature contrasts", size=3.25, col="gray15", fontface = 'bold') +
  annotate("text", x = 3.5, y = 76.5, label = "by pH", size=3, col="gray15", fontface = 'italic') +
  geom_segment(aes(x = 2.5, y = -6, xend = 2.5, yend = 73), linetype="dashed", color = "gray65", size=0.75) +
  geom_segment(aes(x = 4.5, y = -6, xend = 4.5, yend = 73), linetype="dashed", color = "gray65", size=0.75)
#dev.off()
```

