---
title: "02-DESeq2-etc-Pcod"
author: "Laura Spencer"
date: '2022-10-25'
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
      number_sections: true
---

```{r include = FALSE}
knitr::opts_chunk$set(echo=FALSE, warning = F, message = F)
```

```{r, message=FALSE, warning=FALSE, results=FALSE}
#### Load libraries and source scripts 

source("../references/biostats.R")
source("../references/iLOO.R")
`%!in%` = Negate(`%in%`)

# Add all required libraries that are installed with install.packages() here
list.of.packages <- c("RCurl", "tidyverse", "vegan", "pheatmap", "pastecs", "factoextra", "FactoMineR", "RColorBrewer", "tibble", "reshape2", "plotly", "cowplot", "clipr", "janitor", "ggpubr", "forcats", "apeglm", "car", "vsn", "devtools", "grid", "gridGraphics", "Rfast", "dendextend", "RColorBrewer", "scales", "VennDiagram", "colorspace", "edgeR", "ggpattern")

# Add all libraries that are installed using BiocManager here
bioconductor.packages <- c("DESeq2", "WGCNA")

# # Install BiocManager if needed
# if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
# 
# Get names of all required packages that aren't installed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
new.bioc.packages <- bioconductor.packages[!(bioconductor.packages %in% installed.packages()[, "Package"])]
# Install all new packages
if(length(new.packages)) install.packages(new.packages)
if(length(new.bioc.packages)) BiocManager::install(new.bioc.packages)

# Load all required libraries
all.packages <- c(list.of.packages, bioconductor.packages)
lapply(all.packages, FUN = function(X) {
  do.call("require", list(X))
})

# Github packages 
#install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
library(pairwiseAdonis)

`%!in%` = Negate(`%in%`)
```

```{r}
#### Load counts and sample info 

load(file="../data/sample.info")
load(file = "../results/counts")
load(file = "../results/counts.t")
load(file = "../results/counts.ts")
load( file = "../references/gadMor.blast")
load(file = "../references/gadMor.blast.GO")
```


## Library summary stats

#### No. of fragments per sample? 

```{r}
data.frame(rowSums(counts.ts)) %>% 
                  dplyr::rename(read.total = 1) %>% 
                  rownames_to_column(var="sample") %>% #summary() 
dplyr::summarise(mean=round(mean(read.total, na.rm=T)), 
          sd=round(sd(read.total, na.rm=T)), 
          se=round(sd/sqrt(length(sample))),
          median=median(read.total), 
          min=min(read.total), 
          max=max(read.total)) %>% t() %>% as.data.frame() %>% 
  dplyr::mutate(V1=prettyNum(V1, big.mark = ",")) %>% 
  dplyr::rename("fragments.per.sample"="V1") #use this to average across all samples 
```

#### Did the no. of fragments per treatment differ? 

##### Boxplots of the # of fragments in each sample per treatment 

```{r}
fragments <- data.frame(rowSums(counts.ts)) %>% 
                  dplyr::rename(read.total = 1) %>% 
                  rownames_to_column(var="sample") %>%
  left_join(sample.info %>% dplyr::select(vial_label, treatment, temperature, ph, tank), by=c("sample"="vial_label")) %>%
  mutate(treatment=factor(treatment, ordered = T, levels=c("3_amb", "3_low", "6_amb", "6_low", "10_amb", "10_low")))


#ggplotly(
ggplot(data = fragments) +
           geom_bar(aes(x=sample, y=read.total, fill=treatment), stat = "identity") + ggtitle("Total # fragments by sample") + 
             theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())#) 
```

##### Statistically test whether no. of fragments / treatment differs 

```{r}
hist(log(fragments$read.total))
shapiro.test(log(fragments$read.total))
anova(lm(log(read.total) ~ treatment, fragments))

anova(lm(log(read.total) ~ temperature, fragments))
anova(lm(log(read.total) ~ ph, fragments))
anova(lm(log(read.total) ~ temperature*ph, fragments))

ggplot(fragments, aes(x=temperature, y=read.total)) + geom_boxplot() + theme_minimal()
ggplot(fragments, aes(x=ph, y=read.total)) + geom_boxplot() + theme_minimal()
ggplot(fragments, aes(x=treatment, y=read.total)) + geom_boxplot() + theme_minimal()
```

#### How many samples per treatment, tank, etc?

```{r}
fragments %>% group_by(treatment) %>% tally()
fragments %>% group_by(temperature) %>% tally()
fragments %>% group_by(ph) %>% tally()
fragments %>% group_by(treatment, tank) %>% tally() %>% arrange(treatment, tank)
```

#### How many genes total in the filtered count matrix?

```{r}
prettyNum(ncol(counts.ts),big.mark = ",")
```

#### What's the average no. of genes per sample, etc.? 

```{r}
data.frame(rowSums(counts.ts != 0)) %>% 
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample") %>% 
  summarise(mean=round(mean(count.total, na.rm=T)), 
            sd=round(sd(count.total, na.rm=T)), 
            se=round(sd/sqrt(length(sample))),
            median=median(count.total, na.rm=T),
            min=min(count.total, na.rm=T), 
            max=max(count.total, na.rm=T)) %>% t() %>% as.data.frame() %>% 
  mutate(V1=prettyNum(V1, big.mark = ",")) %>% 
  dplyr::rename("genes.per.sample"="V1")
```

#### Total no. of genes identified in each sample

```{r}
#ggplotly(
ggplot(data = data.frame(rowSums(counts.t != 0, na.rm=TRUE)) %>% 
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample")) +
           geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total # genes by sample") + 
             theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())#) 
```

```{r}
#### Merge sample key info to count data, then sort, and generate heat map for initial inspection by treatment 

# IMPORTANT NOTE: to use data for ALL genes, use object "counts.t". To remove low-frequency genes, use object "counts.ts" 

# merge count data with sample key, reset row names as sample names, and arrange by infection, then temperature, then day 
counts.tk <- merge(x=sample.info[,c("vial_label", "temperature", "ph", "treatment", "tank", "treatment_tank", "length")], by.x="vial_label", y=counts.ts, by.y="row.names") %>% 
  arrange(treatment, tank)  %>% column_to_rownames(var="vial_label") %>% droplevels()

head(counts.tk[1:24]) #check out results of merge/arrange
save(counts.tk, file = "../results/deseq2/counts.tk")
```

```{r}
#### Reformat for DESeq, ensure correct sample order for 

# NOTE: It is absolutely critical that the **columns of the count matrix** and the **rows of the column data (information about samples)** are in the same order. DESeq2 will not make guesses as to which column of the count matrix belongs to which row of the column data, these must be provided to DESeq2 already in consistent order.

all(rownames(counts.tk) == counts.tk[-1:-6] %>% t() %>% colnames()) #check that rownames of untransformed matrix match column names of transformed matrix. Should print 'TRUE' 
```

```{r}
##### If using counts which included multimapped counted fractionally, need to round to the nearest integer for DESeq2. 
#_Not needed for gene counts produced by STAR aligner_

#counts.tk <- counts.tk %>% mutate_if(is.numeric, round)
```

# Controlling for larval fish length 

## Examine length distributions by treatment

```{r}
## read in and plot sizes of fish that were selected for RNAseq 
length.rnaseq <- read_csv("../data/length_data_rnaseq.csv", na = c("NA", ""))

require(colorspace)

#ggplotly(
length.rnaseq %>%
  clean_names() %>%
  mutate_at(c("temperature", "p_h", "tank"), factor) %>% 
  ggplot(aes(x=value_mm, fill=interaction(temperature:p_h))) + 
  geom_histogram(color="gray50", bins = 25, alpha=0.5, position="dodge") + 
  xlab("Length (mm)") + ylab(NULL) +
  facet_wrap(~interaction(temperature:p_h), nrow=6) + theme_minimal() + 
  ggtitle("Larval cod samples used in RNAseq\n    size by treatment") +
  scale_y_continuous(breaks=c(0,2)) +
    theme(panel.grid = element_blank(),
        strip.text.x = element_blank(),
        strip.background = element_blank(),
        axis.text.y = element_blank(),
        legend.text=element_text(size=7.5)) +
    scale_fill_manual(name="Treatment", 
                     values=c("6:amb"="gray65",
                              "6:low"=lighten("darkorange2", 0.25), 
                              "10:amb"=lighten("firebrick4", 0.45), 
                              "10:low"="firebrick4", 
                              "3:amb"=lighten("navyblue", 0.45),
                              "3:low"="navyblue"),
                     labels=c("6:amb"="Control",
                              "6:low"="High pCO2", 
                              "10:amb"="Warm Temperature", 
                              "10:low"="Warm Temperature + High pCO2", 
                              "3:amb"="Cold Temperature",
                              "3:low"="Cold Temperature + High pCO2"))
#)
```

## Remove genes that are strongly associated with fish length

1. Use DESeq2 to identify genes with expression dependent on larval length. 
2. Use edgeR to test for effects of length as a continuous explanatory variable using all 3 treatments.
3. Use edgeR to test for effects of length as a continuous explanatory variable without the 10C treatment.

### 1. Genes associated with length using DESeq2 and all three treatments 

```{r}
# Determine size-associated genes using DESeq2
dds.length <- DESeqDataSetFromMatrix(countData = counts.tk[-1:-6] %>% t(),
                              colData = counts.tk[,"length", drop=FALSE] %>% scale() %>% as.data.frame(),
                              design = ~ length)
vsd.length <- varianceStabilizingTransformation(dds.length, blind=FALSE)
dds.DESeq.length <- DESeq(dds.length)

summary(res.length <- results(dds.DESeq.length, alpha=0.05)) 
paste("No. of genes associated with length:",  sum(res.length$padj < 0.05, na.rm=TRUE))
diffex.length <- subset(res.length, padj < 0.05) # & abs(log2FoldChange)>0.5
```

### 2. Genes associated with length using edgeR and all three treatments 

```{r}
# Re-organize counts dataframe to use in edgeR
counts.treat <- counts.ts %>% rownames_to_column("sample") %>%
  left_join(sample.info[c("tank", "temperature", "ph", "vial_label", "length")], by=c("sample"="vial_label")) %>%
  mutate(treatment=as.factor(paste(ph, temperature, sep="."))) %>%
  mutate(ph=gsub("amb", "370", ph)) %>% #estimate ambient = pH 8.0 and pco2 370
  mutate(ph=gsub("low", "950", ph)) %>% #estimate low pH of 7.6 as pco2 900
  mutate(ph=as.numeric(as.character(ph)),
         temperature=as.numeric(as.character(temperature)),
         treatment=factor(treatment, 
                          levels=c("amb.6", "low.6", "amb.3", "low.3", "amb.10", "low.10"))) %>% 
  select(sample, tank, temperature, ph, treatment, length, everything())

# Save counts.treat to use in other analyses
save(counts.treat, file = "../data/counts.treat")

counts.length <- counts.treat %>% column_to_rownames("sample") %>%
  dplyr::select(-tank, -temperature, -ph, -treatment, -length) %>%
  t() %>% as.data.frame()

# double check samples (columns) are in the same order as the samples in the annotated count dataframe. Should = TRUE
all(colnames(counts.length) == counts.treat$sample)

# Extract vectors of larval size (length) and treatment
temperature.length <- counts.treat$temperature
length <- counts.treat$length
treatment.length <- counts.treat$treatment

# Create a DGEList object 
y.length <- DGEList(counts.length, group=treatment.length)

#NOTE: counts have already been filtered for low-frequency genes, see "01-Importing-data" notebook

# TMM normalization
y.length <- calcNormFactors(y.length)
y.length$samples #check out the normalization factor column 

# check out distribution of normalized, filtered read counts

# Calculate logCPM
df_log.length <- cpm(y.length, log = TRUE, prior.count = 2)

# Create tabulated dataframe of mean expression across each temperature level with metadata for transcript ID and factors
tab_exp_df.length <- as.data.frame(df_log.length) %>% rownames_to_column("geneid") %>%
  pivot_longer(cols = -geneid) %>% 
  left_join(counts.treat[c("sample", "temperature", "ph", "treatment", "length")], 
            by=c("name"="sample"))

# Fit design matrix
design.length <-
  model.matrix(
    ~ 1 + 
      length +   #linear effect of larval length
      temperature.length:length      #interaction between temperature and length 
    ) # Generate multivariate edgeR glm

#design <- model.matrix(~ 0 + treatment) 
#colnames(design.length) <- levels(treatment)

# Check out coefficients 
head(design.length)

# Estimate dispersion coefficients
y.length <- estimateDisp(y.length, design.length, robust=TRUE)
y.length$common.dispersion
plotBCV(y.length)

# Fit quasi-likelihood, neg binom linear regression with multifactorial design matrix
fit.length <- glmQLFit(y.length, design.length, robust=TRUE)
head(fit.length$coefficients)
#plotQLDisp(fit)

# Linear effect of length
# estimate sign. degs 
lin.length <-
    glmQLFTest(fit.length,
    coef = 2,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs

# Make contrasts
lin.length.res <-
  decideTestsDGE(lin.length, adjust.method = "fdr", p.value = 0.05)

# Summary of genes linearly related to size using edgeR
summary(lin.length.res)
plotMD(lin.length, cex=0.6)

DEGs.lin.length <-
  topTags(
  lin.length,
  n = sum(abs(lin.length.res)),
  adjust.method = "BH",
  p.value = 0.05
  )
```

### 2. Genes associated with length using edgeR using only 3C and 6C groups 

```{r}
# Identify genes that are linearly related to size when considering only 3C & 6C

counts.length.no10 <- counts.treat %>% column_to_rownames("sample") %>%
  filter(temperature!=10) %>%
  dplyr::select(-tank, -temperature, -ph, -treatment, -length) %>%
  t() %>% as.data.frame()

# double check samples (columns) are in the same order as the samples in the annotated count dataframe. Should = TRUE
all(colnames(counts.length.no10) == counts.treat[counts.treat$temperature!=10,]$sample)

# Extract vectors of larval size (length) and treatment
temperature.length.no10 <- counts.treat[counts.treat$temperature!=10,]$temperature
length.no10 <- counts.treat[counts.treat$temperature!=10,]$length
treatment.length.no10 <- counts.treat[counts.treat$temperature!=10,]$treatment

# Create a DGEList object 
y.length.no10 <- DGEList(counts.length.no10, group=treatment.length.no10)

#NOTE: counts have already been filtered for low-frequency genes, see "01-Importing-data" notebook

# TMM normalization
y.length.no10.no10 <- calcNormFactors(y.length.no10)
y.length.no10.no10$samples #check out the normalization factor column 

# check out distribution of normalized, filtered read counts

# Calculate logCPM
df_log.length.no10 <- cpm(y.length.no10, log = TRUE, prior.count = 2)

# Create tabulated dataframe of mean expression across each temperature level with metadata for transcript ID and factors
tab_exp_df.length.no10 <- as.data.frame(df_log.length.no10) %>% rownames_to_column("geneid") %>%
  pivot_longer(cols = -geneid) %>% 
  left_join(counts.treat[c("sample", "temperature", "ph", "treatment", "length")], 
            by=c("name"="sample"))

# # Fit design matrix
# design.length.no10 <-
#   model.matrix(
#     ~ 1 + length.no10) # Generate edgeR glm

# Fit design matrix
design.length.no10 <-
  model.matrix(
    ~ 1 + 
      length.no10 +   #linear effect of larval length
      temperature.length.no10:length.no10      #interaction between temperature and length 
    ) # Generate multivariate edgeR glm

# Check out coefficients 
head(design.length.no10)

# Estimate dispersion coefficients
y.length.no10 <- estimateDisp(y.length.no10, design.length.no10, robust=TRUE)
y.length.no10$common.dispersion
plotBCV(y.length.no10)

# Fit quasi-likelihood, neg binom linear regression with multifactorial design matrix
fit.length.no10 <- glmQLFit(y.length.no10, design.length.no10, robust=TRUE)
head(fit.length.no10$coefficients)
#plotQLDisp(fit)

# effect of size
# estimate sign. degs 
lin.length.no10 <-
    glmQLFTest(fit.length.no10,
    coef = 2,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs

# Make contrasts
lin.length.res.no10 <-
  decideTestsDGE(lin.length.no10, adjust.method = "fdr", p.value = 0.05)

# Summary of genes linearly related to size using edgeR
summary(lin.length.res.no10)
plotMD(lin.length.no10, cex=0.6)

DEGs.length.no10 <-
  topTags(
  lin.length.no10,
  n = sum(abs(lin.length.res.no10)),
  adjust.method = "BH",
  p.value = 0.05
  )
```

Consensus set of genes that are associated with length (across all treatments) using DESeq2 and both edgeR analyses (all three treatments, no 10C) 

```{r}
# How many genes are associated with length according to both edgeR & DESeq2? 
# Here I look for the overlap among genes that were deemed linearly associated with length, while removing those that have temperature-dependent length dependent. I want to make sure I'm only removing genes where length certainly predicts gene expression, not just across one or two temperatures. 

genes.length <- Reduce(intersect, list(
          rownames(diffex.length), #From DESeq2
          rownames(DEGs.lin.length$table),  #From edgeR all 3 treatments
          rownames(DEGs.length.no10$table))) # From edgeR no 10C
length(genes.length) #how many are linearly associated with length?
```

How many genes in the final dataset after removing length-associated ones?

```{r}
counts.tk[, !(colnames(counts.tk) %in% genes.length)][-1:-6] %>% ncol()
```

### Here is a DESeq2 object containing ALL genes (not filtered to remove length-associated ones) 

```{r}
dds.treatment.all <- DESeqDataSetFromMatrix(countData = counts.tk[-1:-6] %>% t(),
                              colData = counts.tk[,"treatment", drop=FALSE] ,
                              design = ~ treatment)
dds.DESeq.treatment.all <- DESeq(dds.treatment.all)

vsd.treatment.all <- varianceStabilizingTransformation(dds.treatment.all, blind=FALSE)
```

#### Back to DESeq analysis

Generate DESeq datasets with various treatment comparisons  - remove genes associated with length from all analyses 

```{r}
dds.ph <- DESeqDataSetFromMatrix(countData = counts.tk[, !(colnames(counts.tk) %in% genes.length)][-1:-6] %>% t(),
                              colData = counts.tk[,"ph", drop=FALSE] ,
                              design = ~ ph)

dds.temperature <- DESeqDataSetFromMatrix(countData = counts.tk[, !(colnames(counts.tk) %in% genes.length)][-1:-6] %>% t(),
                              colData = counts.tk[,"temperature", drop=FALSE] ,
                              design = ~ temperature)

dds.treatment <- DESeqDataSetFromMatrix(countData = counts.tk[, !(colnames(counts.tk) %in% genes.length)][-1:-6] %>% t(),
                              colData = counts.tk[,"treatment", drop=FALSE] ,
                              design = ~ treatment)

dds.tank <- DESeqDataSetFromMatrix(countData = counts.tk[, !(colnames(counts.tk) %in% genes.length)][-1:-6] %>% t(),
                              colData = counts.tk[,"tank", drop=FALSE] ,
                              design = ~ tank)

dds.treatment.tank <- DESeqDataSetFromMatrix(countData = counts.tk[, !(colnames(counts.tk) %in% genes.length)][-1:-6] %>% t(),
                              colData = counts.tk[,"treatment_tank", drop=FALSE] ,
                              design = ~ treatment_tank)
```


```{r}
#### Transform data for visualization 

# - Here we transform counts using a variance stabilizing transformation (VST), since we have many samples. 
# - Here we use `blind=FALSE` b/c we are interested in differences explained by experimental design, and may wish to use this transformed data in downstream analyses. 
# 
# vsd.pH <- varianceStabilizingTransformation(dds.ph, blind=FALSE)
# vsd.temperature <- varianceStabilizingTransformation(dds.temperature, blind=FALSE)
vsd.treatment <- varianceStabilizingTransformation(dds.treatment, blind=FALSE)
vsd.tank <- varianceStabilizingTransformation(dds.tank, blind=FALSE)
vsd.treatment.tank <- varianceStabilizingTransformation(dds.treatment.tank, blind=FALSE)

# Save vsd-transformed count dataframes for later use 
# save(vsd.pH, file = "../results/deseq2/vsd.pH")
# save(vsd.temperature, file = "../results/deseq2/vsd.temperature")
save(vsd.treatment, file = "../results/deseq2/vsd.treatment")
# save(vsd.tank, file = "../results/deseq2/vsd.tank")
# save(vsd.treatment.tank, file = "../results/deseq2/vsd.treatment.tank")
```

## Unsupervised analyses 

### Heat maps

```{r, include=F}
#NOTE: scale="column" b/c range of counts is so huge, so counts have been scaled 

# # from raw counts, but scaled by column (aka gene)
# pheatmap(data.matrix(counts.tk[-1:-5]), Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, 
#                      show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE,
#                   annotation_colors = list(treatment=
#           c(`3_amb`= "navyblue", `3_low`="royalblue1",
#             `6_amb`="darkgreen", `6_low`="yellow3",
#             `10_amb`="firebrick4", `10_low`="sienna2")),
#          scale="column", #color=c("dodgerblue3", "goldenrod1"), 
#          main = "gene counts - not transformed", annotation_row=counts.tk[,c("treatment"), drop=FALSE])
```

#### Heatmap from all gene counts, variance-stabilization transformed

```{r}
# pheatmap(t(assay(vsd.treatment)), Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
#                      show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE,
#          annotation_colors = list(treatment=
#           c(`3_amb`= "navyblue", `3_low`="royalblue1",
#             `6_amb`="darkgreen", `6_low`="yellow3",
#             `10_amb`="firebrick4", `10_low`="sienna2")),
#          annotation_row=as.data.frame(colData(vsd.treatment)[,c("treatment"), drop=FALSE]),
#          main = "gene counts - VSD-transformed")
```

#### Heat map with top 10% most variable genes, VSD-transformed counts

```{r}
# calculate CV for each gene across all samples 
most.variable <- assay(vsd.treatment) %>% as.data.frame() %>%
  rownames_to_column(var = "gene") %>%
  rowwise() %>% 
  dplyr::mutate(
     sd = sd(c_across(-gene)),
     mean = mean(c_across(-gene)),
     cv = sd/mean) %>%
  #column_to_rownames("gene") %>%
  arrange(desc(cv)) %>% 
# subset top 10% of most variable genes 
  head(n=round(0.1*nrow(assay(vsd.treatment)))) %>%
  select(gene) %>% unlist() %>% as.vector()

length(most.variable)

# Generate heat map / cluster those top 10% most variable genes
cluster.most.variable <- pheatmap(t(assay(vsd.treatment))[,most.variable], 
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
                     show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE,
         annotation_colors = list(treatment=
          c(`3_amb`= "navyblue", `3_low`="royalblue1",
            `6_amb`="darkgreen", `6_low`="yellow3",
            `10_amb`="firebrick4", `10_low`="sienna2")),
         annotation_row=as.data.frame(colData(vsd.treatment)[,c("treatment"), drop=FALSE]),
         main = "gene counts - top 10% most variable\n(VSD-transformed)")
```

### Hierarchical clustering

NOTE: this was generated by the pheatmap function in the previous code chunk

```{r}
# Just plot hierarchical cluster, color-code sample by treatment, or by tank number
cluster.dendo <- as.dendrogram(cluster.most.variable$tree_row) 
#plot(cluster.dendo) # bare bones dendogram
#order.dendrogram(cluster.dendo) # to find out the order of the sample labels 

# Create a dataframe ordered the same as the dendogram for annotation purposes 
dendo.df <- sample.info[match(cluster.most.variable$tree_row$labels, sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")]
all(dendo.df$vial_label == cluster.most.variable$tree_row$labels) # Should == TRUE

# Create color coding vectors 
colorCodes.treat <- c(`3_amb`= "navyblue", `3_low`="royalblue1",
            `6_amb`="darkgreen", `6_low`="yellow3",
            `10_amb`="firebrick4", `10_low`="sienna2")
colorCodes.tank <- colorRampPalette(brewer.pal(12, "Paired"))(length(levels(sample.info$tank)))

# # Color code sample names by treatment
#labels_colors(cluster.dendo) <- colorCodes.treat[as.character(dendo.df$treatment[order.dendrogram(cluster.dendo)])]
#plot(cluster.dendo, xlab="Sample (color=treatment)") # plot with sample names color coded by treatment 

# Plot dendogram color-code by treatment, but labels are tank number
labels(cluster.dendo) <- dendo.df$tank[order.dendrogram(cluster.dendo)]
labels_colors(cluster.dendo) <- colorCodes.treat[as.character(dendo.df$treatment[order.dendrogram(cluster.dendo)])]
plot(cluster.dendo, xlab="Tank Number", cex.lab=0.9) # plot with sample names color coded by tank
legend("topright", title = "Treatment", bty="n", legend =
         c("3_amb","3_low","6_amb","6_low","10_amb", "10_low"), 
       fill = c("navyblue","royalblue1","darkgreen","yellow3","firebrick4","sienna2"))


# # Color code sample names by tank
# labels_colors(cluster.dendo) <- colorCodes.tank[dendo.df$tank][order.dendrogram(cluster.dendo)]
# plot(cluster.dendo, xlab="Sample (color=tank)") # plot with sample names color coded by tank

# # A sub tree - looking at branches
# par(cex = 1)
# plot(cluster.dendo[[1]], horiz = TRUE)
```

#### **NOTE: Samples 194 & 137 look like outliers!** - therefore they were removed 

```{r}

#### Save variance-stabilization normalized gene counts x sample matrix, annotated for later use 

####**NOTE: Here forward I primarily work with the vsd.treatment object, so I can do pairwise treatment contrasts** 

counts.trans <- assay(vsd.treatment) %>% t() %>% as.matrix() %>% as.data.frame() %>% 
  rownames_to_column(var = "sample") %>% 
  left_join(sample.info[,c("vial_label", "treatment", "tank", "temperature", "ph")], by = c("sample"="vial_label")) %>%
  column_to_rownames(var = "sample") %>% 
  dplyr::select(treatment, temperature, ph, tank, everything()) %>% 
  mutate(Treatment=factor(treatment, levels=c("3_low", "6_amb", "6_low", "10_amb", "10_low"))) %>%
  mutate(temperature=factor(temperature, levels=c("3", "6", "10"))) %>%
  mutate(ph=factor(ph, levels=c("amb", "low")))
save(counts.trans, file = "../results/star/counts.trans")
```

### PCA's

```{r}
# To view the the guts of the PlotPCA function in DESeq2, execute this code chunk. It reveals that, by default, it only uses the top 500 most influential genes to plot the PCA! 

#getMethod("plotPCA","DESeqTransform")
```

#### PCAs using DESeq2

NOTE: Hover over points to see the sample numbers

```{r}
# Generate PCA with points color coded by pH Treatment with various number of top genes to use for principal components, selected by highest row variance  

#ggplotly(
  plotPCA(vsd.treatment, intgroup="treatment", ntop=500) +
           ggtitle("PCA by Treatment (var-stabilizing transformed)\ntop 500 genes") +
    geom_point(size=2, aes(text=colnames(vsd.treatment))) +
      scale_color_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen",
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
      theme_minimal()+ stat_ellipse() #, tooltip = "text")
# 
# ggplotly(
#   plotPCA(vsd.treatment, intgroup="treatment", ntop=5000) + 
#            ggtitle("PCA by Treatment (var-stabilizing transformed)\ntop 5k genes") + 
#     geom_point(size=3, aes(text=colnames(vsd.treatment))) + 
#       scale_color_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
#                               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
#     theme_minimal()+ stat_ellipse() , tooltip = "text")
# 
# ggplotly(
#   plotPCA(vsd.treatment, intgroup="treatment", ntop=50000) + 
#            ggtitle("PCA by Treatment (var-stabilizing transformed)\ntop 50k genes") + 
#     geom_point(size=3, aes(text=colnames(vsd.treatment))) + 
#       scale_color_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
#                               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
#     theme_minimal()+ stat_ellipse() , tooltip = "text")

#ggplotly(
  plotPCA(vsd.treatment, intgroup="treatment", ntop=nrow(assay(vsd.treatment))) + 
           ggtitle("PCA by Treatment (var-stabilizing transformed)\nall genes") + 
    geom_point(size=3, aes(text=colnames(vsd.treatment))) + 
      scale_color_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
                              "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
    theme_minimal()+ stat_ellipse() #, tooltip = "text")


# PCA with points color coded by tank
colData(dds.treatment.tank)
  #ggplotly(
  plotPCA(vsd.treatment.tank, intgroup="treatment_tank", ntop=14400) +
           ggtitle("PCA by Tank (var-stabilizing transformed)") +
    geom_point(size=3, aes(text=colnames(vsd.tank))) +
   theme_minimal()#, tooltip = "text")

# PCA.data.treatment <- plotPCA(vsd.treatment, intgroup=c("treatment"), returnData=TRUE)
# save(PCA.data.treatment, file="../results/PCA.data.treatment")
```

This PCA also suggests that two samples - 137 and 194 - are outliers. Consider removing. 

### PCAs using prcomp

This enables screeplot to ID significant PCs, calculates variance, etc. 

```{r, warning=F, message=F}
#pca.princomp <- prcomp(cov(assay(vsd.treatment)), scale=F) #scale=F for variance-covariance matrix
pca.princomp <- prcomp(t(assay(vsd.treatment))) # using the same code that's under the hood of PlotPCA but using all genes 
pca.eigenval(pca.princomp) #The Proporation of Variance = %variance 
pc.percent <- pca.eigenval(pca.princomp)[2,1:6]*100
screeplot(pca.princomp, bstick=TRUE) #looks like PC 1-34 are significant 
pc.percent[1:2] %>% sum()
```

```{r}
#### Generate dataframe with prcomp results 
tab.expr <- data.frame(sample.id = colnames(assay(vsd.treatment)),
    EV1.all = pca.princomp$x[,1],    # the first eigenvector
    EV2.all = pca.princomp$x[,2],    # the second eigenvector
    EV3.all = pca.princomp$x[,3],    # the third eigenvector
    EV4.all = pca.princomp$x[,4],    # the fourth eigenvector
    EV5.all = pca.princomp$x[,5],    # the fourth eigenvector
    EV6.all = pca.princomp$x[,6],    # the fourth eigenvector
    stringsAsFactors = FALSE)
#shapiro.test(pca.princomp$x) #sample size too large for shapiro test which is weird 
#hist(pca.princomp$x) #normal? hard to say, maybe
tab.expr.annot <- left_join(tab.expr, sample.info, by=c("sample.id"="vial_label")) %>% droplevels()
```

```{r}
# #### Old PCA biplots figures from prcomp analysis
# 
# # GGSCATTER with ellipses and stars - PC1 PC2
# ggscatter(tab.expr.annot #%>% 
#             # mutate(treatment=str_replace(Treatment, "Ambient", "Ambient (pH 8.0)"),
#             #        treatment=str_replace(Treatment, "Moderate", "Moderate (pH 7.8)"),
#             #        treatment=str_replace(Treatment, "Low", "Severe (pH 7.6)"))
#           , 
#           x="EV1.all", y="EV2.all", col="treatment", size=2.5, alpha=0.85, ellipse = TRUE, star.plot = TRUE, 
#           # palette = c(`Ambient (pH 8.0)`="#2c7bb6", `Moderate (pH 7.8)`="#fdae61", `Severe (pH 7.6)`="#d7191c")
#           ) +
#   theme_minimal() + ggtitle("Global gene expression, PC1xPC2") + 
#   xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
#   ylab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
#   labs(color="OA Treatment") + 
#   labs(fill="OA Treatment") + 
#   theme(legend.position = "right") + 
#   scale_color_manual(name="Treatment", values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
#                               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
#   scale_fill_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
#                               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2"), guide = "none") +
#   guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank")))
# 
# 
# # GGSCATTER with ellipses and stars - PC1 PC3
# ggscatter(tab.expr.annot, x="EV1.all", y="EV3.all",
#           col="treatment", size=2.5, alpha=0.85, ellipse = TRUE, star.plot = TRUE, 
#            #palette = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")
#           ) +
#   theme_minimal() + ggtitle("Global gene expression, PC1xPC3") + 
#   xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
#   ylab(paste("PC3 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
#   theme(legend.position = "right") + 
#   scale_color_manual(name="Treatment", values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
#                               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
#   scale_fill_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
#                               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2"), guide = "none") +
#   guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"))) 
# 
# # GGSCATTER with ellipses and stars - PC2 PC3
# ggscatter(tab.expr.annot, x="EV2.all", y="EV3.all",
#           col="treatment", size=2.5, alpha=0.85, ellipse = TRUE, star.plot = TRUE, 
#           #palette = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")
#           ) +
#   theme_minimal() + ggtitle("Global gene expression, PC2xPC3") + 
#   xlab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
#   ylab(paste("PC3 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
#   theme(legend.position = "right") + 
#   scale_color_manual(name="Treatment", values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
#                               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")) +
#   scale_fill_manual(values=c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen", 
#                               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2"), guide = "none") +
#   guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"))) 
```

## GGSCATTER - nicer formatting

```{r}
# ggscatter(tab.expr.annot, group=c("temperature", "ph"),
#           x="EV1.all", y="EV2.all", col="temperature", shape="ph", size=3.5, alpha=0.85, 
#           ellipse = FALSE, star.plot = FALSE) +
#   theme_minimal() + ggtitle("Gene Expression PC1xPC2\nall genes") + 
#   xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
#   ylab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
#   labs(color="OA Treatment") + 
#   theme(legend.position = "right") + 
#   scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"),
#                      labels=c("3"="3 deg C", "6"="6 deg C", "10"="10 deg C")) +
#   scale_shape_manual(name="pCO2", values=c(16, 1), labels=c("amb"="Ambient pCO2", "low"="High pCO2")) + 
#   guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"), order=1))
# 
# ggscatter(tab.expr.annot, group=c("temperature", "ph"),
#           x="EV1.all", y="EV3.all", col="temperature", shape="ph", size=3.5, alpha=0.85, 
#           ellipse = FALSE, star.plot = FALSE) +
#   theme_minimal() + ggtitle("Global gene expression PC1xPC3") + 
#   xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
#   ylab(paste("PC3 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
#   labs(color="OA Treatment") + 
#   theme(legend.position = "right") + 
#   scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"),
#                      labels=c("3"="3 deg C", "6"="6 deg C", "10"="10 deg C")) +
#   scale_shape_manual(name="pCO2", values=c(16, 1), labels=c("amb"="Ambient pCO2", "low"="High pCO2")) + 
#   guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"), order=1))
# 
# ggscatter(tab.expr.annot, group=c("temperature", "ph"),
#           x="EV1.all", y="EV4.all", col="temperature", shape="ph", size=3.5, alpha=0.85, 
#           ellipse = FALSE, star.plot = FALSE) +
#   theme_minimal() + ggtitle("Global gene expression PC1xPC4") + 
#   xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
#   ylab(paste("PC4 (", round(pc.percent[4], digits = 1), "%)", sep="")) + 
#   labs(color="OA Treatment") + 
#   theme(legend.position = "right") + 
#   scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"),
#                      labels=c("3"="3 deg C", "6"="6 deg C", "10"="10 deg C")) +
#   scale_shape_manual(name="pCO2", values=c(16, 1), labels=c("amb"="Ambient pCO2", "low"="High pCO2")) + 
#   guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"), order=1))
# 
# ggscatter(tab.expr.annot, group=c("temperature", "ph"),
#           x="EV2.all", y="EV3.all", col="temperature", shape="ph", size=3.5, alpha=0.85, 
#           ellipse = FALSE, star.plot = FALSE) +
#   theme_minimal() + ggtitle("Gene Expression PC2xPC3\nall genes") + 
#   xlab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
#   ylab(paste("PC3 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
#   labs(color="OA Treatment") + 
#   theme(legend.position = "right") + 
#   scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"),
#                      labels=c("3"="3 deg C", "6"="6 deg C", "10"="10 deg C")) +
#   scale_shape_manual(name="pCO2", values=c(16, 1), labels=c("amb"="Ambient pCO2", "low"="High pCO2")) + 
#   guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"), order=1))
# 
# ggscatter(tab.expr.annot, group=c("temperature", "ph"),
#           x="EV3.all", y="EV4.all", col="temperature", shape="ph", size=3.5, alpha=0.85, 
#           ellipse = FALSE, star.plot = FALSE) +
#   theme_minimal() + ggtitle("Global gene expression PC3xPC4") + 
#   xlab(paste("PC3 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
#   ylab(paste("PC4 (", round(pc.percent[4], digits = 1), "%)", sep="")) + 
#   labs(color="OA Treatment") + 
#   theme(legend.position = "right") + 
#   scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) +
#   scale_shape_manual(name="pCO2", values=c("amb"=19, "low"=1), labels=c("amb"="ambient", "low"="high pCO2")) + 
#   guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"), order=1))
```

## Figure 1 - PCA 1x2

```{r}
#pdf(file="../results/deseq2/Figure1_PCA.pdf", height = 5.25, width = 7.5)
pdf(file="../manuscript/CJFAS-submission/Fig1.pdf", height = 3.5, width = 5)
ggscatter(tab.expr.annot %>%
            mutate(treatment=factor(treatment, 
                             levels=c("6_amb", "6_low", "10_amb", 
                                      "10_low", "3_amb", "3_low"), ordered = T)), 
          group=c("temperature", "ph"),
          x="EV1.all", y="EV2.all", col="treatment", shape="ph", size=2.25, alpha=0.8, 
          ellipse = FALSE, star.plot = TRUE, star.size=0.1) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
  theme(axis.text = element_text(size=8), title = element_text(size=8),
        legend.position = "right", legend.text=element_text(size=7), legend.title=element_text(size=7)) + 
  scale_color_manual(name=NULL, 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4", 
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="Acidified", 
                              "10_amb"="Warm", 
                              "10_low"="Warm + Acidified", 
                              "3_amb"="Cold",
                              "3_low"="Cold + Acidified")) +
  scale_shape_manual(guide="none", values=c("amb"=19, "low"=17)) + 
  guides(colour = guide_legend(override.aes = 
                                 list(size=2.5, linetype="blank", shape=c(19, 17, 19, 17, 19, 17))))
dev.off()
```

EXPLORE TANK EFFECTS FOR REVISION

This PCA also suggests that two samples - 137 and 194 - are outliers. Consider removing. 

### PCAs using prcomp

This enables screeplot to ID significant PCs, calculates variance, etc. 

```{r, warning=F, message=F}
#pca.princomp <- prcomp(cov(assay(vsd.treatment)), scale=F) #scale=F for variance-covariance matrix
pca.tank <- prcomp(t(assay(vsd.treatment.tank))) # using the same code that's under the hood of PlotPCA but using all genes 
pca.eigenval(pca.tank) #The Proporation of Variance = %variance 
perc.tank <- pca.eigenval(pca.tank)[2,1:7]*100
screeplot(pca.tank, bstick=TRUE) #looks like PC 1-34 are significant 
perc.tank[1:2] %>% sum()

#### Generate dataframe with prcomp results 
tab.tank <- data.frame(sample.id = colnames(assay(vsd.treatment.tank)),
    EV1.all = pca.tank$x[,1],    # the first eigenvector
    EV2.all = pca.tank$x[,2],    # the second eigenvector
    EV3.all = pca.tank$x[,3],    # the third eigenvector
    EV4.all = pca.tank$x[,4],    # the fourth eigenvector
    EV5.all = pca.tank$x[,5],    # the fourth eigenvector
    EV6.all = pca.tank$x[,6],    # the fourth eigenvector
    EV7.all = pca.tank$x[,7],
    stringsAsFactors = FALSE)
tab.tank.annot <- left_join(tab.tank, sample.info, by=c("sample.id"="vial_label")) %>% droplevels()

#pdf(file="../manuscript/CJFAS-submission/Fig1.pdf", height = 3.5, width = 5)
tab.tank.annot %>% mutate(tank_rep=case_when(
  tank %in% c(1,11,19,13,3,7)~"1",
  tank %in% c(2,12,14,4,8)~"2",
  tank %in% c(9,23,21,17,15)~"3",
  tank %in% c(10,24,22,18,16)~"4")) %>% 
  ggplot(aes(x=treatment, y=EV1.all, color=treatment, shape=tank_rep)) + 
  geom_boxplot(outlier.shape = NULL) +
  geom_jitter(width=0.25)


ggscatter(tab.tank.annot %>% 
            mutate(tank_rep=case_when(
  tank %in% c(1,11,19,13,3,7)~"1",
  tank %in% c(2,12,14,4,8)~"2",
  tank %in% c(9,23,21,17,15)~"3",
  tank %in% c(10,24,22,18,16)~"4")) %>%
            mutate(treatment=factor(treatment, 
                             levels=c("6_amb", "6_low", "10_amb", 
                                      "10_low", "3_amb", "3_low"), ordered = T)), 
          group=c("temperature", "tank"),
          x="EV1.all", y="EV2.all", col="treatment", shape="tank_rep", size=3.5, alpha=0.6, 
          ellipse = FALSE, star.plot = T, star.size=0.1) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
  theme(axis.text = element_text(size=8), title = element_text(size=8),
        legend.position = "right", legend.text=element_text(size=7), legend.title=element_text(size=7)) + 
  scale_color_manual(name=NULL,
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25),
                              "10_amb"=lighten("firebrick4", 0.45),
                              "10_low"="firebrick4",
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="Acidified",
                              "10_amb"="Warm",
                              "10_low"="Warm + Acidified",
                              "3_amb"="Cold",
                              "3_low"="Cold + Acidified")) +
  scale_shape_manual(values=c(15,16,17,18)) + #guide="none", 
  guides(colour = guide_legend(override.aes =
         list(size=2.5, linetype="blank", shape=c(19, 17, 19, 17, 19, 17))))

centroids <- tab.tank.annot %>%
  group_by(treatment) %>%
  summarise(across(c(EV1.all, EV2.all), mean), .groups = "drop")

# Compute Euclidean distance to centroid
tab.tank.annot <- tab.tank.annot %>% mutate(tank_rep=case_when(
  tank %in% c(1,11,19,13,3,7)~"1",
  tank %in% c(2,12,14,4,8)~"2",
  tank %in% c(9,23,21,17,15)~"3",
  tank %in% c(10,24,22,18,16)~"4")) %>%
  left_join(centroids, by = "treatment", suffix = c("", "_centroid")) %>%
  mutate(euclidean_dist = sqrt((EV1.all - EV1.all_centroid)^2 + (EV2.all - EV2.all_centroid)^2)) %>% 
  mutate(treatment=factor(treatment, 
                 levels=c("6_amb", "6_low", "10_amb", 
                          "10_low", "3_amb", "3_low"), ordered = T))

hist(log(tab.tank.annot$euclidean_dist))
summary(aov(log(euclidean_dist)~treatment, tab.tank.annot))
TukeyHSD(aov(log(euclidean_dist)~treatment, tab.tank.annot))

plot(TukeyHSD(aov(log(euclidean_dist)~treat, tab.tank.annot %>% 
  mutate(treat=case_when(
    treatment=="3_amb"~"Cold",
    treatment=="3_low"~"Cold+OA",
    treatment=="10_amb"~"Warm",
    treatment=="10_low"~"Warm+OA",
    treatment=="6_low"~"OA",
    treatment=="6_amb"~"Control",
  )))), las=2)

tab.tank.annot %>%
  ggplot(aes(x=treatment, color=treatment, y=euclidean_dist)) + 
  geom_boxplot(outlier.shape = NA) + geom_jitter(aes(shape=tank_rep), 
                                                 alpha=0.75, size=2, width = 0.25) + 
  theme_minimal() +
    scale_color_manual(name=NULL,
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25),
                              "10_amb"=lighten("firebrick4", 0.45),
                              "10_low"="firebrick4",
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="Acidified",
                              "10_amb"="Warm",
                              "10_low"="Warm + Acidified",
                              "3_amb"="Cold",
                              "3_low"="Cold + Acidified")) +
  ggtitle("Distance to treatment centroid, PC1xPC2") + 
  ylab("Euclidean distance") + xlab("Treatment") + 
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())

#dev.off()
```
## Explore CV 

```{r}
# Calculate per-gene variance within each tank
gene_var_by_tank <- assay(vst(dds.DESeq.treatment)) %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(-gene, names_to = "sample", values_to = "expression") %>%
  left_join(sample.info %>% select(treatment, temperature, ph, tank, vial_label), by = c("sample"="vial_label")) %>%
  group_by(gene, treatment, tank) %>%
  summarise(variance = var(expression), .groups = "drop")%>% 
  mutate(treat=case_when(
    treatment=="3_amb"~"Cold",
    treatment=="3_low"~"Cold+OA",
    treatment=="10_amb"~"Warm",
    treatment=="10_low"~"Warm+OA",
    treatment=="6_low"~"OA",
    treatment=="6_amb"~"Control",
  ))

# Calculate per-gene variance within each treatment
gene_var_by_treatment <- assay(vst(dds.DESeq.treatment)) %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(-gene, names_to = "sample", values_to = "expression") %>%
  left_join(sample.info %>% select(treatment, temperature, ph, tank, vial_label), by = c("sample"="vial_label")) %>%
  group_by(gene, treatment) %>%
  summarise(variance = var(expression), .groups = "drop") %>% 
  mutate(treat=case_when(
    treatment=="3_amb"~"Cold",
    treatment=="3_low"~"Cold+OA",
    treatment=="10_amb"~"Warm",
    treatment=="10_low"~"Warm+OA",
    treatment=="6_low"~"OA",
    treatment=="6_amb"~"Control",
  ))

# Boxplot of variance per treatment
gene_var_by_treatment %>% 
ggplot(aes(x = treat, y = variance)) +
  geom_boxplot() + 
  scale_y_log10() +  # Log scale to handle high variability
  theme_minimal() +
  labs(title = "Variance of Gene Expression by Treatment", y = "Gene Expression Variance")

# Mean by treatment 
gene_var_by_treatment %>% group_by(treat) %>% summarize(mean=round(mean(variance), digits = 3), sd=round(sd(variance), digits=3)) 

# Mean of mean tank variance 
gene_var_by_tank %>% group_by(treat, tank) %>% summarize(mean=round(mean(variance), digits = 3), sd=round(sd(variance), digits=3)) %>% ungroup() %>% 
  summarize(mean_var_tank=mean(mean, na.rm=T),sd_var_tank=sd(mean, na.rm=T))

# normal distribution after log transformation? 
hist(log(gene_var_by_treatment$variance))
leveneTest(expression ~ treatment, data=expr.testing)
summary(aov(variance~treat,gene_var_by_treatment))
par(mar = c(5, 9, 4, 2))
plot(TukeyHSD(aov(variance~treat,gene_var_by_treatment)), las=2)
```
```{r}
?pairwise.wilcox.test
```

FOR PACIFIC COD WORKSHOP, NOVEMBER 2023. MAKE PCA SHOWING ONLY EFFECTS OF OA AND WARMING 

```{r}
ggscatter(tab.expr.annot %>%
            filter(temperature != 3) %>%
            mutate(treatment=factor(treatment, 
                             levels=c("6_amb", "6_low", "10_amb", 
                                      "10_low"), ordered = T)), 
#          group=c("temperature", "ph"),
          x="EV1.all", y="EV2.all", col="treatment", size=3.5, alpha=0.85, 
          ellipse = F, star.plot = F, ellipse.alpha=0, star.plot.lwd=.02) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment", 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4"), 
                     labels=c("6_amb"="Control",
                              "6_low"="Acidified", 
                              "10_amb"="Warm",
                              "10_low"="Warm + Acidified")) +
  scale_shape_manual(guide="none", values=c("amb"=19, "low"=17)) + 
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank")))

```
Supplemental PCA's with PC3+

```{r}
pdf(file="../results/deseq2/Supp_PCA-1x3.pdf", height = 5.25, width = 7.5)
ggscatter(tab.expr.annot %>%
            mutate(treatment=factor(treatment, 
                             levels=c("6_amb", "6_low", "10_amb", 
                                      "10_low", "3_amb", "3_low"), ordered = T)), 
          group=c("temperature", "ph"),
          x="EV1.all", y="EV3.all", col="treatment", shape="ph", size=3.5, alpha=0.85, 
          ellipse = FALSE, star.plot = FALSE) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC3") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC3 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment", 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4", 
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="High pCO2", 
                              "10_amb"="Warm Temperature", 
                              "10_low"="Warm Temperature + High pCO2", 
                              "3_amb"="Cold Temperature",
                              "3_low"="Cold Temperature + High pCO2")) +
  scale_shape_manual(guide="none", values=c("amb"=19, "low"=17)) + 
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank", shape=c(19, 17, 19, 17, 19, 17))))
dev.off()


pdf(file="../results/deseq2/Supp_PCA-1x4.pdf", height = 5.25, width = 7.5)
ggscatter(tab.expr.annot %>%
            mutate(treatment=factor(treatment, 
                             levels=c("6_amb", "6_low", "10_amb", 
                                      "10_low", "3_amb", "3_low"), ordered = T)), 
          group=c("temperature", "ph"),
          x="EV1.all", y="EV4.all", col="treatment", shape="ph", size=3.5, alpha=0.85, 
          ellipse = FALSE, star.plot = FALSE) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC4") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC4 (", round(pc.percent[4], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment", 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4", 
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="High pCO2", 
                              "10_amb"="Warm Temperature", 
                              "10_low"="Warm Temperature + High pCO2", 
                              "3_amb"="Cold Temperature",
                              "3_low"="Cold Temperature + High pCO2")) +
  scale_shape_manual(guide="none", values=c("amb"=19, "low"=17)) + 
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank", shape=c(19, 17, 19, 17, 19, 17))))
dev.off()

pdf(file="../results/deseq2/Supp_PCA-1x5.pdf", height = 5.25, width = 7.5)
ggscatter(tab.expr.annot %>%
            mutate(treatment=factor(treatment, 
                             levels=c("6_amb", "6_low", "10_amb", 
                                      "10_low", "3_amb", "3_low"), ordered = T)), 
          group=c("temperature", "ph"),
          x="EV1.all", y="EV5.all", col="treatment", shape="ph", size=3.5, alpha=0.85, 
          ellipse = FALSE, star.plot = FALSE) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC5") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC5 (", round(pc.percent[5], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment", 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4", 
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="High pCO2", 
                              "10_amb"="Warm Temperature", 
                              "10_low"="Warm Temperature + High pCO2", 
                              "3_amb"="Cold Temperature",
                              "3_low"="Cold Temperature + High pCO2")) +
  scale_shape_manual(guide="none", values=c("amb"=19, "low"=17)) + 
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank", shape=c(19, 17, 19, 17, 19, 17))))
dev.off()

pdf(file="../results/deseq2/Supp_PCA-1x6.pdf", height = 5.25, width = 7.5)
ggscatter(tab.expr.annot %>%
            mutate(treatment=factor(treatment, 
                             levels=c("6_amb", "6_low", "10_amb", 
                                      "10_low", "3_amb", "3_low"), ordered = T)), 
          group=c("temperature", "ph"),
          x="EV1.all", y="EV6.all", col="treatment", shape="ph", size=3.5, alpha=0.85, 
          ellipse = FALSE, star.plot = FALSE) +
  theme_minimal() + ggtitle("Global gene expression PC1xPC6") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC6 (", round(pc.percent[6], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment", 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4", 
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="High pCO2", 
                              "10_amb"="Warm Temperature", 
                              "10_low"="Warm Temperature + High pCO2", 
                              "3_amb"="Cold Temperature",
                              "3_low"="Cold Temperature + High pCO2")) +
  scale_shape_manual(guide="none", values=c("amb"=19, "low"=17)) + 
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank", shape=c(19, 17, 19, 17, 19, 17))))
dev.off()
```

# Correlation plots, PC scores (from all genes) ~ temperature

```{r}
ggplot() + 
  geom_jitter(data= tab.expr.annot, width = 0.5, size=2.5,
              aes(x=as.numeric(as.character((temperature))), y=EV1.all, color=temperature, shape=ph,
                  ellipse = FALSE, star.plot = FALSE)) + 
  theme_minimal() + xlab("Temperature") + ylab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  scale_x_continuous(breaks = c(3,6,10)) + 
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"),
                     labels=c("3"="3 deg C", "6"="6 deg C", "10"="10 deg C")) +
  scale_shape_manual(name="pCO2", values=c(16, 1), labels=c("amb"="Ambient pCO2", "low"="High pCO2")) + 
  geom_point(data=tab.expr.annot %>% group_by(temperature, ph) %>% dplyr::summarise(stat = mean(EV1.all)), 
      aes(x=as.numeric(as.character((temperature))), y=stat, color=temperature, shape=ph), size = 6) +
  guides(colour = guide_legend(override.aes = list(size=4, linetype="blank"), order=1),
         shape = guide_legend(override.aes = list(size=4, linetype="blank")))

ggplot() + 
  geom_jitter(data= tab.expr.annot, width = 0.5, size=2.5,
              aes(x=as.numeric(as.character((temperature))), y=EV2.all, color=temperature, shape=ph)) + 
  theme_minimal() + xlab("Temperature") + ylab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
  scale_x_continuous(breaks = c(3,6,10)) + 
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"),
                     labels=c("3"="3 deg C", "6"="6 deg C", "10"="10 deg C")) +
  scale_shape_manual(name="pCO2", values=c(16, 1), labels=c("amb"="Ambient pCO2", "low"="High pCO2")) + 
  geom_point(data=tab.expr.annot %>% group_by(temperature, ph) %>% dplyr::summarise(stat = mean(EV2.all)), 
      aes(x=as.numeric(as.character((temperature))), y=stat, color=temperature, shape=ph), size = 6) +
  guides(colour = guide_legend(override.aes = list(size=4, linetype="blank"), order=1),
         shape = guide_legend(override.aes = list(size=4, linetype="blank")))

ggplot() + 
  geom_jitter(data= tab.expr.annot, width = 0.5, size=2.5,
              aes(x=as.numeric(as.character((temperature))), y=EV3.all, color=temperature, shape=ph)) + 
  theme_minimal() + xlab("Temperature") + ylab(paste("PC3 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
  scale_x_continuous(breaks = c(3,6,10)) + 
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"),
                     labels=c("3"="3 deg C", "6"="6 deg C", "10"="10 deg C")) +
  scale_shape_manual(name="pCO2", values=c(16, 1), labels=c("amb"="Ambient pCO2", "low"="High pCO2")) + 
  geom_point(data=tab.expr.annot %>% group_by(temperature, ph) %>% dplyr::summarise(stat = mean(EV3.all)), 
      aes(x=as.numeric(as.character((temperature))), y=stat, color=temperature, shape=ph), size = 6) +
  guides(colour = guide_legend(override.aes = list(size=4, linetype="blank"), order=1),
         shape = guide_legend(override.aes = list(size=4, linetype="blank")))

ggplot() + 
  geom_jitter(data= tab.expr.annot, width = 0.5, size=2.5,
              aes(x=as.numeric(as.character((temperature))), y=EV4.all, color=temperature, shape=ph)) + 
  theme_minimal() + xlab("Temperature") + ylab(paste("PC4 (", round(pc.percent[4], digits = 1), "%)", sep="")) + 
  scale_x_continuous(breaks = c(3,6,10)) + 
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"),
                     labels=c("3"="3 deg C", "6"="6 deg C", "10"="10 deg C")) +
  scale_shape_manual(name="pCO2", values=c(16, 1), labels=c("amb"="Ambient pCO2", "low"="High pCO2")) + 
  geom_point(data=tab.expr.annot %>% group_by(temperature, ph) %>% dplyr::summarise(stat = mean(EV3.all)), 
      aes(x=as.numeric(as.character((temperature))), y=stat, color=temperature, shape=ph), size = 6) +
  guides(colour = guide_legend(override.aes = list(size=4, linetype="blank"), order=1),
         shape = guide_legend(override.aes = list(size=4, linetype="blank")))
```

## Generate PCA using 10% most variable genes 

```{r}
pca.princomp.var <- prcomp(t(assay(vsd.treatment)[most.variable,])) # using the same code that's under the hood of PlotPCA 
pca.eigenval(pca.princomp.var) #The Proporation of Variance = %variance 
pc.percent.var <- pca.eigenval(pca.princomp.var)[2,1:5]*100
pc.percent.var[1:2] %>% sum()
screeplot(pca.princomp.var, bstick=TRUE) #looks like PC 1-3 are significant 

tab.expr.var <- data.frame(sample.id = colnames(assay(vsd.treatment[most.variable,])),
    EV1.all = pca.princomp.var$x[,1],    # the first eigenvector
    EV2.all = pca.princomp.var$x[,2],    # the second eigenvector
    EV3.all = pca.princomp.var$x[,3],    # the third eigenvector
    stringsAsFactors = FALSE)
#shapiro.test(pca.princomp$x) #sample size too large for shapiro test which is weird 
#hist(pca.princomp$x) #normal? hard to say, maybe
tab.expr.annot.var <- left_join(tab.expr.var, sample.info, by=c("sample.id"="vial_label")) %>% droplevels()

# GGSCATTER with ellipses and stars - PC1 PC2
ggscatter(tab.expr.annot.var, group=c("temperature", "ph"),
          x="EV1.all", y="EV2.all", col="temperature", shape="ph", size=3.5, alpha=0.85, 
          ellipse = FALSE, star.plot = FALSE) +
  theme_minimal() + ggtitle("Gene Expression PC1xPC2\n10% most variable genes") + 
  xlab(paste("PC1 (", round(pc.percent.var[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.var[2], digits = 1), "%)", sep="")) + 
  labs(color="OA Treatment") + 
  theme(legend.position = "right") + 
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) +
  scale_shape_manual(name="pCO2", values=c("amb"=19, "low"=1), labels=c("amb"="ambient", "low"="high pCO2")) + 
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"), order=1))

# GGSCATTER with ellipses and stars - PC1 PC3
ggscatter(tab.expr.annot.var, 
          x="EV1.all", y="EV3.all", col="temperature", shape="ph", size=3.5, alpha=0.85, 
          ellipse = FALSE, star.plot = FALSE) +
  theme_minimal() + ggtitle("Gene Expression PC1xPC3\n10% most variable genes") + 
  xlab(paste("PC1 (", round(pc.percent.var[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC3 (", round(pc.percent.var[3], digits = 1), "%)", sep="")) + 
  labs(color="OA Treatment") + 
  theme(legend.position = "left") + 
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) +
  scale_shape_manual(name="pCO2", values=c("amb"=19, "low"=1), labels=c("amb"="ambient", "low"="high pCO2")) + 
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"), order=1))
```

```{r}
# ggplot(tab.expr.annot.var, 
#        aes(x=as.numeric(as.character((temperature))), y=EV1.all, color=ph)) + 
#   geom_point() + 
#   geom_smooth(method="lm", se=F, level=0.95, linetype="solid", cex=0.4, alpha=0.15) +
#   theme_minimal() + xlab("Temperature") + ylab("PC1 score") + 
#   scale_x_continuous(breaks = c(3,6,10)) + 
#   scale_color_manual(name="pCO2", values=c("amb"="blue", "low"="red"), 
#                      labels=c("amb"="Ambient", "low"="High")) 

ggplot() + 
  geom_jitter(data= tab.expr.annot.var, width = 0.5, size=2.5,
              aes(x=as.numeric(as.character((temperature))), y=EV1.all, color=temperature, shape=ph)) + 
  theme_minimal() + xlab("Temperature") + ylab("PC1 score") + 
  scale_x_continuous(breaks = c(3,6,10)) + 
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) +
  scale_shape_manual(name="pCO2", values=c(16, 1)) + 
  geom_point(data=tab.expr.annot.var %>% group_by(temperature, ph) %>% dplyr::summarise(stat = mean(EV1.all)), 
      aes(x=as.numeric(as.character((temperature))), y=stat, color=temperature, shape=ph), size = 6) +
  guides(colour = guide_legend(override.aes = list(size=4, linetype="blank"), order=1),
         shape = guide_legend(override.aes = list(size=4, linetype="blank")))

ggplot() + 
  geom_jitter(data= tab.expr.annot.var, width = 0.5, size=2.5,
              aes(x=as.numeric(as.character((temperature))), y=EV2.all, color=temperature, shape=ph)) + 
  theme_minimal() + xlab("Temperature") + ylab("PC2 score") + 
  scale_x_continuous(breaks = c(3,6,10)) + 
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) +
  scale_shape_manual(name="pCO2", values=c(16, 1)) + 
  geom_point(data=tab.expr.annot.var %>% group_by(temperature, ph) %>% dplyr::summarise(stat = mean(EV2.all)), 
      aes(x=as.numeric(as.character((temperature))), y=stat, color=temperature, shape=ph), size = 6) +
  guides(colour = guide_legend(override.aes = list(size=4, linetype="blank"), order=1),
         shape = guide_legend(override.aes = list(size=4, linetype="blank")))

ggplot() + 
  geom_jitter(data= tab.expr.annot.var, width = 0.5, size=2.5,
              aes(x=as.numeric(as.character((temperature))), y=EV3.all, color=temperature, shape=ph)) + 
  theme_minimal() + xlab("Temperature") + ylab("PC3 score") + 
  scale_x_continuous(breaks = c(3,6,10)) + 
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) +
  scale_shape_manual(name="pCO2", values=c(16, 1)) + 
  geom_point(data=tab.expr.annot.var %>% group_by(temperature, ph) %>% dplyr::summarise(stat = mean(EV3.all)), 
      aes(x=as.numeric(as.character((temperature))), y=stat, color=temperature, shape=ph), size = 6) +
  guides(colour = guide_legend(override.aes = list(size=4, linetype="blank"), order=1),
         shape = guide_legend(override.aes = list(size=4, linetype="blank")))
```

### Heatmap of the sample-to-sample distances

Another use of the transformed data is sample clustering. Here, we apply the dist function to the transpose of the transformed count matrix to get sample-to-sample distances.

A heatmap of this distance matrix gives us an overview over similarities and dissimilarities between samples. We have to provide a hierarchical clustering hc to the heatmap function based on the sample distances, or else the heatmap function would calculate a clustering based on the distances between the rows/columns of the distance matrix.

```{r}
# colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)
# 
# # Here we show treatment and tank number
# sampleDistss <- dist(t(assay(vsd.treatment.tank)))
# sampleDistMatrixs <- as.matrix(sampleDistss)
# rownames(sampleDistMatrixs) <- vsd.treatment.tank$treatment_tank   #set row names
# colnames(sampleDistMatrixs) <- colnames(vsd.treatment)
# 
# all(colnames(sampleDistMatrixs) == rownames(counts.tk))
# 
# pheatmap(sampleDistMatrixs,
#          clustering_distance_rows=sampleDistss,
#          clustering_distance_cols=sampleDistss,
#          col=rev(colors), fontsize = 6.5)
```


```{r}
### perMANOVA - multivariate analysis of variance

# Effect of treatment
vsd.t <- t(assay(vsd.treatment)) %>% as.data.frame()
perm <- adonis(vsd.t ~ treatment, data=sample.info %>% remove_rownames() %>% column_to_rownames("vial_label"), permutations = 1000, method="bray")

# Pairwise comparisons
vsd.t <- sample.info %>% dplyr::select(vial_label, temperature, ph, treatment, tank) %>% left_join(t(assay(vsd.treatment)) %>% as.data.frame() %>% rownames_to_column("vial_label"))

# Clustering by temperature? Yes! all temps differ in multivariate space 
pairwise.adonis(vsd.t[,-1:-5], vsd.t$temperature, perm = 1000)

# Clustering by ph? No! no effect of ph
pairwise.adonis(vsd.t[,-1:-5], vsd.t$ph, perm = 1000)

# Clustering by temp x ph treatment? Yes! Treatments that differ are below. 
pairwise.adonis(vsd.t[,-1:-5], vsd.t$treatment, perm = 1000) %>% arrange(p.adjusted) %>%
  filter(p.adjusted<0.05) %>% select(pairs, p.adjusted)

# 10_amb vs 6_amb	0.01498501			
# 10_amb vs 3_low	0.01498501			
# 10_amb vs 3_amb	0.01498501			
# 10_low vs 3_low	0.01498501			
# 10_low vs 3_amb	0.01498501			
# 6_low vs 3_low	0.01498501			
# 6_low vs 3_amb	0.01498501			
# 10_amb vs 6_low	0.02997003	
```

```{r}
#### How many unique genes were measured per treatment. This takes the mean of each gene by group, and counts up those that have mean >10. 

#Warning: this takes forever 
# # Counts before filtering 
# sample.info %>% select(Sample, Treatment) %>% left_join(counts.t %>% rownames_to_column("Sample")) %>% 
#   group_by(Treatment) %>% summarise_if(is.numeric, "mean") %>% mutate(genes=rowSums(select_if(., is.numeric)>10)) %>% select(Treatment, genes)

# # Counts after filtering
# sample.info %>% select(Sample, Treatment) %>% left_join(counts.tk %>% rownames_to_column("Sample")) %>% 
#   select(-Sample, -Tank, -Treatment_Tank) %>% group_by(Treatment) %>% summarise_all("mean") %>% mutate(genes=rowSums(is.numeric(.)>10)) %>% select(Treatment, genes)
```

### Differential Expression Analysis

_This is the core DESeq2 analysis!_ 

```{r, message=F, warning=F, include=F}
##### Run the function `DESeq` to assess differential expression 
# NOTE, could consider including 'minReplicatesForReplace=Inf' to keep DESeq2 from replacing outlier values with trimmed means. 
## The default is for this to happen when there are more than 7 replicates. 

dds.DESeq.treatment <- DESeq(dds.treatment)
# dds.DESeq.temperature <- DESeq(dds.temperature)
# dds.DESeq.ph <- DESeq(dds.ph)
dds.DESeq.tank <- DESeq(dds.tank)
```

### Identify differentially expressed genes among contrasts 

#### pH contrasts 

```{r}
print("Comparison: Ambient pH vs. Low pH, 3C")
summary(res.pH.3 <- results(dds.DESeq.treatment, contrast=c("treatment", "3_amb", "3_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pH, 3C:",  sum(res.pH.3$padj < 0.05, na.rm=TRUE))

print("Comparison: Ambient pH vs. Low pH, 6C")
summary(res.pH.6 <- results(dds.DESeq.treatment, contrast=c("treatment", "6_amb", "6_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pH, 6C:",  sum(res.pH.6$padj < 0.05, na.rm=TRUE))

print("Comparison: Ambient pH vs. Low pH, 10C")
summary(res.pH.10 <- results(dds.DESeq.treatment, contrast=c("treatment", "10_amb", "10_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pH, 10C:",  sum(res.pH.10$padj < 0.05, na.rm=TRUE))
```

#### Temperature contrasts 

##### Comparing temperatures within ambient pH 

```{r}
print("Comparison: 3C vs. 6C, ambient pH")
summary(res.temp.3.6.amb <- results(dds.DESeq.treatment, contrast=c("treatment", "3_amb", "6_amb"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 3C & 6C, ambient pH:",  sum(res.temp.3.6.amb$padj < 0.05, na.rm=TRUE))

# This contrast isn't interesting 
# print("Comparison: 3C vs. 10C, ambient pH")
# summary(res.temp.3.10.amb <- results(dds.DESeq.treatment, contrast=c("treatment", "3_amb", "10_amb"), alpha=0.05))
# paste("No. of genes differentially expressed (padj<0.05) between 3C & 10C, ambient pH:",  sum(res.temp.3.10.amb$padj < 0.05, na.rm=TRUE))

print("Comparison: 6C vs. 10C, ambient pH")
summary(res.temp.6.10.amb <- results(dds.DESeq.treatment, contrast=c("treatment", "6_amb", "10_amb"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 6C & 10C, ambient pH:",  sum(res.temp.6.10.amb$padj < 0.05, na.rm=TRUE))
```

##### Comparing temperatures within low pH 

```{r}
print("Comparison: 3C vs. 6C, low pH")
summary(res.temp.3.6.low <- results(dds.DESeq.treatment, contrast=c("treatment", "3_low", "6_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 3C & 6C, low pH:",  sum(res.temp.3.6.low$padj < 0.05, na.rm=TRUE))

# This contrast isn't interesting 
# print("Comparison: 3C vs. 10C, low pH")
# summary(res.temp.3.10.low <- results(dds.DESeq.treatment, contrast=c("treatment", "3_low", "10_low"), alpha=0.05))
# paste("No. of genes differentially expressed (padj<0.05) between 3C & 10C, low pH:",  sum(res.temp.3.10.low$padj < 0.05, na.rm=TRUE))

print("Comparison: 6C vs. 10C, low pH")
summary(res.temp.6.10.low <- results(dds.DESeq.treatment, contrast=c("treatment", "6_low", "10_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 6C & 10C, low pH:",  sum(res.temp.6.10.low$padj < 0.05, na.rm=TRUE))
```

##### Comparing optimal conditions (6-amb) to combined stressors (3-low, 10-low)

```{r}
print("Comparison: 6C-ambient vs. 3C")
summary(res.both.3low.6amb <- results(dds.DESeq.treatment, contrast=c("treatment", "3_low", "6_amb"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 6C-amb & 3-low:",  sum(res.both.3low.6amb$padj < 0.05, na.rm=TRUE))

print("Comparison: 6C-ambient vs. 10C-low")
summary(res.both.10low.6amb <- results(dds.DESeq.treatment, contrast=c("treatment", "6_amb", "10_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) between 6C-amb & 10-low:",  sum(res.both.10low.6amb$padj < 0.05, na.rm=TRUE))
```

```{r, include=F, echo=F, message=F}
#### Just looking at temperature and pH 

# print("Comparison: 6C vs 10C, any pH")
# summary(res.temp.3.6 <- results(dds.DESeq.temperature, contrast=c("temperature", "3", "6"), alpha=0.05))
# paste("No. of genes differentially expressed (padj<0.05) between 3C & 6C, any pH:",  sum(res.temp.3.6$padj < 0.05, na.rm=TRUE))
# 
# print("Comparison: 6C vs 10C, any pH")
# summary(res.temp.3.10 <- results(dds.DESeq.temperature, contrast=c("temperature", "3", "10"), alpha=0.05))
# paste("No. of genes differentially expressed (padj<0.05) between 3C & 10C, any pH:",  sum(res.temp.3.10$padj < 0.05, na.rm=TRUE))
# 
# print("Comparison: 6C vs 10C, any pH")
# summary(res.temp.6.10 <- results(dds.DESeq.temperature, contrast=c("temperature", "6", "10"), alpha=0.05))
# paste("No. of genes differentially expressed (padj<0.05) between 6C & 10C, any pH:",  sum(res.temp.6.10$padj < 0.05, na.rm=TRUE))
# 
# print("Comparison: amb and low pH, any temp")
# summary(res.ph <- results(dds.DESeq.ph, contrast=c("ph", "amb", "low"), alpha=0.05))
# paste("No. of genes differentially expressed (padj<0.05) between ambient and low pH, any temperature",  sum(res.ph$padj < 0.05, na.rm=TRUE))
```

#### Total number of genes analyzed 

```{r}
nrow(res.pH.6)
```

```{r}
#### Save differential expression results to objects 

save(res.pH.3, file = "../results/deseq2/res.pH.3")
save(res.pH.6, file = "../results/deseq2/res.pH.6")
save(res.pH.10, file = "../results/deseq2/res.pH.10")

save(res.temp.3.6.amb, file = "../results/deseq2/res.temp.3.6.amb")
#save(res.temp.3.10.amb, file = "../results/deseq2/res.temp.3.10.amb") #don't use these anymore
save(res.temp.6.10.amb, file = "../results/deseq2/res.temp.6.10.amb")

save(res.temp.3.6.low, file = "../results/deseq2/res.temp.3.6.low")
#save(res.temp.3.10.low, file = "../results/deseq2/res.temp.3.10.low") #don't use these anymore
save(res.temp.6.10.low, file = "../results/deseq2/res.temp.6.10.low")

save(res.both.3low.6amb, file = "../results/deseq2/res.both.3low.6amb")
save(res.both.10low.6amb, file = "../results/deseq2/res.both.10low.6amb")

# load(file = "../results/deseq2/res.pH.3")
# load(file = "../results/deseq2/res.pH.6")
# load(file = "../results/deseq2/res.pH.10")
# load(file = "../results/deseq2/res.temp.3.6.amb")
# load(file = "../results/deseq2/res.temp.6.10.amb")
# load(file = "../results/deseq2/res.temp.3.6.low")
# load(file = "../results/deseq2/res.temp.6.10.low")
# load(file = "../results/deseq2/res.both.3low.6amb")
# load(file = "../results/deseq2/res.both.10low.6amb")
```

```{r}
#### Subset the DESeq results for only those statistically sign. ones
# Could also filter to only include those with abs(L2FC)>0.5, but that might filter out interesting genes - 

diffex.pH.3 <- subset(res.pH.3, padj < 0.05) # & abs(log2FoldChange)>0.5
diffex.pH.6 <- subset(res.pH.6, padj < 0.05)
diffex.pH.10 <- subset(res.pH.10, padj < 0.0)
diffex.temp.3.6.amb <- subset(res.temp.3.6.amb, padj < 0.05)
#diffex.temp.3.10.amb <- subset(res.temp.3.10.amb, padj < 0.05) #not interesting
diffex.temp.6.10.amb <- subset(res.temp.6.10.amb, padj < 0.05)
diffex.temp.3.6.low <- subset(res.temp.3.6.low, padj < 0.05)
#diffex.temp.3.10.low <- subset(res.temp.3.10.low, padj < 0.05) #not interesting
diffex.temp.6.10.low <- subset(res.temp.6.10.low, padj < 0.05)

diffex.both.3low.6amb <- subset(res.both.3low.6amb, padj < 0.05)
diffex.both.10low.6amb <- subset(res.both.10low.6amb, padj < 0.05)
```

### Examine some differentially expressed genes

Plot counts
NOTE: using the DeSeq2 function `plotCounts` plots original counts, i.e. those that are normalized, but outlier values are NOT replaced with trimmed means.  I'm really interested to see how the counts look that DESeq2 is analyzing, i.e. I want to see the outlier-replaced counts. So, I must include the argument `replaced=TRUE` in the plotCounts function.

```{r, warning=FALSE, message=F}
#b <- c("3_amb", "3_low")
#b <- c("6_amb", "6_low")
#b <- c("10_amb", "10_low")

#test <- 
#  diffex.pH.3 %>% as.data.frame() %>% rownames()
#  diffex.pH.6 %>% as.data.frame() %>% rownames()
#  diffex.pH.10 %>% as.data.frame() %>% rownames()


# Plot genes using Uniprot IDs
install.packages("rSHAPE")
require(rSHAPE)
test <- counts.annot.gadmor %>% 
#  filter(spid %in% c("P49285"))
#  filter(gene_gadmor %in% c("pcdh8"))
  filter(gene_id %in% as.character(quote(c(115533477, 115552173, 115552184, 115552186, 115552451, 115552198, 115552771, 115552177, 115552200, 115552179, 115552774, 115552456, 115552765, 115552205, 115552766, 115547299, 115561650, 115552768, 115549085, 115552170, 115552181, 115552160))))
  
for (i in 1:nrow(test)) {
a <-  plotCounts(dds.DESeq.treatment, gene=test$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #filter(treatment %in% b) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))

print(
  ggplot(a,
       aes(x=temperature, y=count, color=temperature, label=sample)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(aes(shape=ph), size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle(str_wrap(test$protein_names[i])) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) + 
  facet_wrap(~ph))
}
```

There definitely seem to be some outliers influencing DEGs. For example, in 10C-ambient Tank 10 (sample #61, there's only one sample from that tank). 

Inspect Cook's distances for all genes: 

```{r}
par(mar=c(8,5,2,2))
boxplot(log10(assays(dds.DESeq.treatment)[["cooks"]]), range=0, las=2)
```
'
Inspect Cook's distances for genes with potential outliers:

Here are some genes with potential outliers betwen 10-amb and 10-low: 
- phlda3 - PCG061
- LOC115542103 - PCG061
- hamp - PCG035
- LOC115551232 - PCG061
- paip2 - PCG061
- map3k8 - pcg059
- LOC115537409 - PCG061
- LOC115535354 - PCG061
- mid1ip1 - PCG061
- LOC115532810 - PCG061
- LOC115559061 - PCG061

And for 6-amb vs 6-low: c("tomm6", "snrpe", "slc16a", "rab5if", "LOC115553606", "LOC115552062", "LOC115548877")0

```{r}
outs.10ph <- c("phlda3", "LOC115542103", "hamp", "LOC115551232", "paip2", "map3k8", "LOC115537409", "LOC115535354", "mid1ip1", "LOC115532810", "LOC115559061")
outs.6ph <- c("tomm6", "snrpe", "slc16a", "rab5if", "LOC115553606", "LOC115552062", "LOC115548877")

#ggplotly(
assays(dds.DESeq.treatment)[["cooks"]] %>% as.data.frame() %>% rownames_to_column("gene") %>%
#  filter(gene %in% outs.10ph) %>%
  filter(gene %in% outs.6ph) %>%
  pivot_longer(cols = -gene) %>%
  left_join(sample.info[,c("vial_label", "treatment")], by=c("name"="vial_label")) %>%
#  filter(treatment %in% c("10_amb", "10_low")) %>%
  filter(treatment %in% c("6_amb", "6_low")) %>%
  mutate(gene=as.factor(gene)) %>% 
  ggplot(aes(x=gene, y=value, label=name)) #+ geom_text())
```

Oddly, Cooks Distances for the potential outlier samples aren't low. Doesn't indicate that it influences diff. expression.

What if I remove potential outlier samples? What are the DEG results? 

10degC pH contrast: 
Crazy- when I remove PCG061 from the dataset there are NO DEGs between 10_amb and 10_low!!  That seems weird, perhaps it's largely due to reduced power? Could try to remove another random sample to see how that affects things. 

```{r}
# find index number for row containing sample PCG061
ind <- grep("PCG061", rownames(counts.tk))

dds.test <- DESeqDataSetFromMatrix(countData = counts.tk[-ind, -1:-6] %>% t(),
                              colData = counts.tk[-ind,"treatment", drop=FALSE] ,
                              design = ~ treatment)
dds.test.DESeq <- DESeq(dds.test)
summary(res.test.out <- results(dds.test.DESeq, contrast=c("treatment", "10_amb", "10_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pH, 10C:",  sum(res.test.out$padj < 0.05, na.rm=TRUE))
```

6degC pH contrast: 
What if I remove PCG093, will those genes still be DEGs?
Way fewer DEGs - 2,922 vs 4,657

```{r}
ind <- grep("PCG193", rownames(counts.tk)) # find index number for row containing sample PCG061
dds.test <- DESeqDataSetFromMatrix(countData = counts.tk[-ind, -1:-6] %>% t(),
                              colData = counts.tk[-ind,"treatment", drop=FALSE] ,
                              design = ~ treatment)
dds.test.DESeq <- DESeq(dds.test)
summary(res.test.out <- results(dds.test.DESeq, contrast=c("treatment", "6_amb", "6_low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pH, 10C:",  sum(res.test.out$padj < 0.05, na.rm=TRUE))

b <- c("6_amb", "6_low")
test <- subset(res.test.out, padj < 0.05 & abs(log2FoldChange)>0.5) %>% as.data.frame() %>% rownames()

for (i in 1:length(test[1:25])) {
a <-  plotCounts(dds.test.DESeq, gene=test[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  filter(treatment %in% b)

print(ggplot(a,
       aes(x=treatment, y=count, color=treatment, label=sample)) +
    geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle(test[i]) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none"))
}
```

Were outliers actually replaced with trimmed means? 

```{r}
assays(dds.DESeq.treatment)[["replaceCounts"]] #replaced counts 
assays(dds.DESeq.treatment)[["counts"]] #original counts 

counts.replace.check <- left_join(
  assays(dds.DESeq.treatment)[["replaceCounts"]] %>% as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(-gene, values_to = "replacedCounts", names_to = "sample"),
  
  assays(dds.DESeq.treatment)[["counts"]] %>% as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(-gene, values_to = "originalCounts", names_to = "sample")) %>%
  
  mutate(sample=as.factor(sample),
         ratio=originalCounts/replacedCounts)

# How many data points were considered outliers and replaced?  Answer: 48 
counts.replace.check %>% filter(ratio!=1) 
  
# How many samples had outliers? Answer: 16 samples. Most outliers were detected in samples 59, 35, and 163. It's suprising that none were detected in sample 61. 
counts.replace.check %>% filter(ratio!=1) %>% group_by(sample) %>% tally() %>% arrange(desc(n))
```


### Investigate other methods for outlier detection. 

Outlier detection using the iterative Leave-One-Out method (iLOO) and script from [George et al. 2015](https://doi.org/10.1371/journal.pone.0125224), see [iLOO.R script](../references/iLOO.R)

Note: The function iLOO requires an input  matrix, which represents a matrix of read counts for a treatment or homogeneous group containing   samples and  features. iLOO returns a  matrix, where only the outlier read counts are present (all non-outlier values have been NAed). 

```{r}
# Create vectors of sample IDs per treatment
samples.3amb <- (sample.info[c("vial_label", "treatment")] %>% filter(treatment=="3_amb"))$vial_label
samples.3low <- (sample.info[c("vial_label", "treatment")] %>% filter(treatment=="3_low"))$vial_label
samples.6amb <- (sample.info[c("vial_label", "treatment")] %>% filter(treatment=="6_amb"))$vial_label
samples.6low <- (sample.info[c("vial_label", "treatment")] %>% filter(treatment=="6_low"))$vial_label
samples.10amb <- (sample.info[c("vial_label", "treatment")] %>% filter(treatment=="10_amb"))$vial_label
samples.10low <- (sample.info[c("vial_label", "treatment")] %>% filter(treatment=="10_low"))$vial_label

# Create vector of all DEGs identified among any contrast 
degs <- unique(c(rownames(diffex.pH.3),rownames(diffex.pH.6),rownames(diffex.pH.10),
#          rownames(diffex.temp.3.10.amb),rownames(diffex.temp.3.10.low),
          rownames(diffex.temp.3.6.amb), rownames(diffex.temp.3.6.low),
          rownames(diffex.temp.6.10.amb), rownames(diffex.temp.6.10.low),
          diffex.both.3low.6amb, diffex.both.10low.6amb))

# Extract gene counts that have outliers replaced with trimmed means, and filter for DEGs
degs.counts <- assays(dds.DESeq.treatment)[["replaceCounts"]] %>% as.data.frame() %>% rownames_to_column("gene") %>% filter(gene %in% degs)

# Extract counts for 
counts.3amb <- degs.counts %>% select(gene, all_of(samples.3amb)) %>% column_to_rownames("gene")
counts.3low <- degs.counts %>% select(gene, all_of(samples.3low)) %>% column_to_rownames("gene")
counts.6amb <- degs.counts %>% select(gene, all_of(samples.6amb)) %>% column_to_rownames("gene")
counts.6low <- degs.counts %>% select(gene, all_of(samples.6low)) %>% column_to_rownames("gene")
counts.10amb <- degs.counts %>% select(gene, all_of(samples.10amb)) %>% column_to_rownames("gene")
counts.10low <- degs.counts %>% select(gene, all_of(samples.10low)) %>% column_to_rownames("gene")

# Run the iterative Leave-One-Out function on gene counts for each treatment 
iLOO.3amb <- iLOO(counts.3amb)
iLOO.3low <- iLOO(counts.3low)
iLOO.6amb <- iLOO(counts.6amb)
iLOO.6low <- iLOO(counts.6low)
iLOO.10amb <- iLOO(counts.10amb)
iLOO.10low <- iLOO(counts.10low)

# find samples with lots of outliers in DEG lists 
iLOO.3amb %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  pivot_longer(-gene) %>% filter(!is.na(value)) %>%
  mutate(name=as.factor(name)) %>%
  filter(gene %in% c(rownames(diffex.pH.3), rownames(diffex.temp.3.6.amb))) %>% #, rownames(diffex.temp.3.10.amb)
  group_by(name) %>% tally() %>% arrange(desc(n))

iLOO.3low %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  pivot_longer(-gene) %>% filter(!is.na(value)) %>%
  mutate(name=as.factor(name)) %>%
  filter(gene %in% c(rownames(diffex.pH.3), rownames(diffex.temp.3.6.low))) %>% #, rownames(diffex.temp.3.10.low)
  group_by(name) %>% tally() %>% arrange(desc(n))

iLOO.6amb %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  pivot_longer(-gene) %>% filter(!is.na(value)) %>%
  mutate(name=as.factor(name)) %>%
  filter(gene %in% c(rownames(diffex.pH.6), rownames(diffex.temp.3.6.amb), rownames(diffex.temp.6.10.amb))) %>%
  group_by(name) %>% tally() %>% arrange(desc(n))

iLOO.6low %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  pivot_longer(-gene) %>% filter(!is.na(value)) %>%
  mutate(name=as.factor(name)) %>%
  filter(gene %in% c(rownames(diffex.pH.6), rownames(diffex.temp.3.6.low), rownames(diffex.temp.6.10.low))) %>%
  group_by(name) %>% tally() %>% arrange(desc(n))

iLOO.10amb %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  pivot_longer(-gene) %>% filter(!is.na(value)) %>%
  mutate(name=as.factor(name)) %>%
  filter(gene %in% c(rownames(diffex.pH.10), rownames(diffex.temp.6.10.amb))) %>% #, rownames(diffex.temp.3.10.amb)
  group_by(name) %>% tally() %>% arrange(desc(n))

iLOO.10low %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  pivot_longer(-gene) %>% filter(!is.na(value)) %>%
  mutate(name=as.factor(name)) %>%
  filter(gene %in% c(rownames(diffex.pH.10), rownames(diffex.temp.6.10.low))) %>% #, rownames(diffex.temp.3.10.low)
  group_by(name) %>% tally() %>% arrange(desc(n))
```

### Generate lists of degs that have an outlier in at least one sample

```{r}
### pH contrasts within temperature treatments 

# 3C ph ambient vs. low 
(outs.diffex.3.ph <- full_join(
  iLOO.3amb[rowSums(is.na(iLOO.3amb)) != ncol(iLOO.3amb), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.pH.3)),

iLOO.3low[rowSums(is.na(iLOO.3low)) != ncol(iLOO.3low), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.pH.3)),
by="gene"))

# 6C ph ambient vs. low 
(outs.diffex.6.ph <- full_join(
  iLOO.6amb[rowSums(is.na(iLOO.6amb)) != ncol(iLOO.6amb), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.pH.6)),

iLOO.6low[rowSums(is.na(iLOO.6low)) != ncol(iLOO.6low), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.pH.6)),
by="gene"))

# 10C ph ambient vs. low 
(outs.diffex.10.ph <- full_join(
  iLOO.10amb[rowSums(is.na(iLOO.10amb)) != ncol(iLOO.10amb), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.pH.10)),

iLOO.10low[rowSums(is.na(iLOO.10low)) != ncol(iLOO.10low), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.pH.10)),
by="gene"))

### Temperature contrasts within ambient pH 

# 3C vs 6C at ambient pH
(outs.diffex.3.6.amb <- full_join(
  iLOO.3amb[rowSums(is.na(iLOO.3amb)) != ncol(iLOO.3amb), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.temp.3.6.amb)),

iLOO.6amb[rowSums(is.na(iLOO.6amb)) != ncol(iLOO.6amb), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.temp.3.6.amb)),
by="gene"))

# Don't use this contrast anymore
# # 3C vs 10C at ambient pH
# (outs.diffex.3.10.amb <- full_join(
#   iLOO.3amb[rowSums(is.na(iLOO.3amb)) != ncol(iLOO.3amb), ] %>% 
#   as.data.frame() %>% rownames_to_column("gene") %>%
#   filter(gene %in% rownames(diffex.temp.3.10.amb)),
# 
# iLOO.10amb[rowSums(is.na(iLOO.10amb)) != ncol(iLOO.10amb), ] %>% 
#   as.data.frame() %>% rownames_to_column("gene") %>%
#   filter(gene %in% rownames(diffex.temp.3.10.amb)),
# by="gene"))

# 6C vs 10C at ambient pH
(outs.diffex.6.10.amb <- full_join(
  iLOO.6amb[rowSums(is.na(iLOO.6amb)) != ncol(iLOO.6amb), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.temp.6.10.amb)),

iLOO.10amb[rowSums(is.na(iLOO.10amb)) != ncol(iLOO.10amb), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.temp.6.10.amb)),
by="gene"))

### Temperature contrasts within low pH 

# 3C vs 6C at low pH
(outs.diffex.3.6.low <- full_join(
  iLOO.3low[rowSums(is.na(iLOO.3low)) != ncol(iLOO.3low), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.temp.3.6.low)),

iLOO.6low[rowSums(is.na(iLOO.6low)) != ncol(iLOO.6low), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.temp.3.6.low)),
by="gene"))

# Don't use this contrast anymore
# # 3C vs 10C at low pH
# (outs.diffex.3.10.low <- full_join(
#   iLOO.3low[rowSums(is.na(iLOO.3low)) != ncol(iLOO.3low), ] %>% 
#   as.data.frame() %>% rownames_to_column("gene") %>%
#   filter(gene %in% rownames(diffex.temp.3.10.low)),
# 
# iLOO.10low[rowSums(is.na(iLOO.10low)) != ncol(iLOO.10low), ] %>% 
#   as.data.frame() %>% rownames_to_column("gene") %>%
#   filter(gene %in% rownames(diffex.temp.3.10.low)),
# by="gene"))

# 6C vs 10C at low pH
(outs.diffex.6.10.low <- full_join(
  iLOO.6low[rowSums(is.na(iLOO.6low)) != ncol(iLOO.6low), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.temp.6.10.low)),

iLOO.10low[rowSums(is.na(iLOO.10low)) != ncol(iLOO.10low), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.temp.6.10.low)),
by="gene"))


### Combined stressor contrasts 

# 6C-ambient pH vs. 3C-low pH 
(outs.diffex.both.3low.6amb <- full_join(
  iLOO.6amb[rowSums(is.na(iLOO.6amb)) != ncol(iLOO.6amb), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.both.3low.6amb)),

iLOO.3low[rowSums(is.na(iLOO.3low)) != ncol(iLOO.3low), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.both.3low.6amb)),
by="gene"))

# 6C-ambient pH vs. 10C-low pH 
(outs.diffex.both.10low.6amb <- full_join(
  iLOO.6amb[rowSums(is.na(iLOO.6amb)) != ncol(iLOO.6amb), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.both.3low.6amb)),

iLOO.10low[rowSums(is.na(iLOO.10low)) != ncol(iLOO.10low), ] %>% 
  as.data.frame() %>% rownames_to_column("gene") %>%
  filter(gene %in% rownames(diffex.both.10low.6amb)),
by="gene"))
```

For all genes with outliers identified in at least one sample, visually inspect the counts (with trimmed means replaced for DESeq2-identified outliers). Convincing? 

Decision 12/7/2022: Use count plots to identify samples that are outliers in lots of genes- then remove the subset of outlier genes

3 Ambient pH:  359 (330 outliers), 366 (173 outliers)
3 Low pH: 349 kinda (73 outliers)
6 Ambient pH: 189 kinda (52 outliers)
6 Low pH: 168 kinda (60 outliers)
10 Ambient pH: 61 (312 genes), 59 (115)
10 Low pH: 62 (186), 20 kinda (83)

To look at genes with outliers identified in DEG sets, comment out the different "b" and "test" lines of code, then run for loop to plot genes 

```{r}
# b <- c("3_amb", "3_low")
# test <- outs.diffex.3.ph$gene   # no outliers in this set

b <- c("6_amb", "6_low")
test <- outs.diffex.6.ph$gene

# b <- c("10_amb", "10_low")
# test <- outs.diffex.10.ph$gene
# 
# b <- c("3_amb", "6_amb")
# test <- outs.diffex.3.6.amb$gene
# 
# b <- c("3_amb", "10_amb")
# test <- outs.diffex.3.10.amb$gene
# 
# b <- c("6_amb", "10_amb")
# test <- outs.diffex.6.10.amb$gene
# 
# b <- c("3_low", "6_low")
# test <- outs.diffex.3.6.low$gene
# 
# b <- c("3_low", "10_low")
# test <- outs.diffex.3.10.low$gene
# 
# b <- c("6_low", "10_low")
# test <- outs.diffex.6.10.low$gene


for (i in 1:length(test)) {
a <-  plotCounts(dds.DESeq.treatment, gene=test[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  filter(treatment %in% b)

print(ggplot(a,
       aes(x=treatment, y=count, color=treatment, label=sample)) +
    geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle(test[i]) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none"))
}
```


### Control for tank! 

Generate lists of DEGs that differ among tanks

```{r}
# here are the tanks by treatment
sample.info %>% select(tank, treatment) %>% distinct() %>% arrange(treatment)
```

#### Generate list of DEGs among tanks within each treatment. Here I set alpha=0.01, to ensure that tank-effect DEGs are highly significant. 

```{r}
# 3C ambient - no DEGs among tank, since only 1 tank in that treatment
# & abs(log2FoldChange)>0.5))
# 3C low pH 
(diffex.3_low.1 <- subset(results(dds.DESeq.tank, contrast=c("tank", "22", "13"), alpha=0.01), padj < 0.01))
(diffex.3_low.2 <- subset(results(dds.DESeq.tank, contrast=c("tank", "22", "21"), alpha=0.01), padj < 0.01))
(diffex.3_low.3 <- subset(results(dds.DESeq.tank, contrast=c("tank", "22", "14"), alpha=0.01), padj < 0.01))
(diffex.3_low.4 <- subset(results(dds.DESeq.tank, contrast=c("tank", "13", "21"), alpha=0.01), padj < 0.01))
(diffex.3_low.5 <- subset(results(dds.DESeq.tank, contrast=c("tank", "13", "14"), alpha=0.01), padj < 0.01))
(diffex.3_low.6 <- subset(results(dds.DESeq.tank, contrast=c("tank", "21", "14"), alpha=0.01), padj < 0.01))
diffex.tank.3_low <- unique(c(rownames(diffex.3_low.1), rownames(diffex.3_low.2), rownames(diffex.3_low.3), 
                     rownames(diffex.3_low.4), rownames(diffex.3_low.5), rownames(diffex.3_low.6)))

# 6C ambient pH
(diffex.6_amb.1 <- subset(results(dds.DESeq.tank, contrast=c("tank", "7", "8"), alpha=0.01), padj < 0.01))
(diffex.6_amb.2 <- subset(results(dds.DESeq.tank, contrast=c("tank", "7", "15"), alpha=0.01), padj < 0.01))
(diffex.6_amb.3 <- subset(results(dds.DESeq.tank, contrast=c("tank", "7", "16"), alpha=0.01), padj < 0.01))
(diffex.6_amb.4 <- subset(results(dds.DESeq.tank, contrast=c("tank", "8", "15"), alpha=0.01), padj < 0.01))
(diffex.6_amb.5 <- subset(results(dds.DESeq.tank, contrast=c("tank", "8", "16"), alpha=0.01), padj < 0.01))
(diffex.6_amb.6 <- subset(results(dds.DESeq.tank, contrast=c("tank", "15", "16"), alpha=0.01), padj < 0.01))
diffex.tank.6_amb <- unique(c(rownames(diffex.6_amb.1), rownames(diffex.6_amb.2), rownames(diffex.6_amb.3), 
                     rownames(diffex.6_amb.4), rownames(diffex.6_amb.5), rownames(diffex.6_amb.6)))

# 6C low pH
(diffex.6_low.1 <- subset(results(dds.DESeq.tank, contrast=c("tank", "11", "12"), alpha=0.01), padj < 0.01))
(diffex.6_low.2 <- subset(results(dds.DESeq.tank, contrast=c("tank", "11", "23"), alpha=0.01), padj < 0.01))
(diffex.6_low.3 <- subset(results(dds.DESeq.tank, contrast=c("tank", "11", "24"), alpha=0.01), padj < 0.01))
(diffex.6_low.4 <- subset(results(dds.DESeq.tank, contrast=c("tank", "12", "23"), alpha=0.01), padj < 0.01))
(diffex.6_low.5 <- subset(results(dds.DESeq.tank, contrast=c("tank", "12", "24"), alpha=0.01), padj < 0.01))
(diffex.6_low.6 <- subset(results(dds.DESeq.tank, contrast=c("tank", "23", "24"), alpha=0.01), padj < 0.01))
diffex.tank.6_low <- unique(c(rownames(diffex.6_low.1), rownames(diffex.6_low.2), rownames(diffex.6_low.3), 
                     rownames(diffex.6_low.4), rownames(diffex.6_low.5), rownames(diffex.6_low.6)))

# 10C ambient pH
(diffex.10_amb.1 <- subset(results(dds.DESeq.tank, contrast=c("tank", "1", "2"), alpha=0.01), padj < 0.01))
(diffex.10_amb.2 <- subset(results(dds.DESeq.tank, contrast=c("tank", "1", "9"), alpha=0.01), padj < 0.01))
(diffex.10_amb.3 <- subset(results(dds.DESeq.tank, contrast=c("tank", "1", "10"), alpha=0.01), padj < 0.01))
(diffex.10_amb.4 <- subset(results(dds.DESeq.tank, contrast=c("tank", "2", "9"), alpha=0.01), padj < 0.01))
(diffex.10_amb.5 <- subset(results(dds.DESeq.tank, contrast=c("tank", "2", "10"), alpha=0.01), padj < 0.01))
(diffex.10_amb.6 <- subset(results(dds.DESeq.tank, contrast=c("tank", "9", "10"), alpha=0.01), padj < 0.01))
diffex.tank.10_amb <- unique(c(rownames(diffex.10_amb.1), rownames(diffex.10_amb.2), rownames(diffex.10_amb.3), 
                     rownames(diffex.10_amb.4), rownames(diffex.10_amb.5), rownames(diffex.10_amb.6)))

# 10C low pH
(diffex.10_low.1 <- subset(results(dds.DESeq.tank, contrast=c("tank", "11", "12"), alpha=0.01), padj < 0.01))
(diffex.10_low.2 <- subset(results(dds.DESeq.tank, contrast=c("tank", "11", "23"), alpha=0.01), padj < 0.01))
(diffex.10_low.3 <- subset(results(dds.DESeq.tank, contrast=c("tank", "11", "24"), alpha=0.01), padj < 0.01))
(diffex.10_low.4 <- subset(results(dds.DESeq.tank, contrast=c("tank", "12", "23"), alpha=0.01), padj < 0.01))
(diffex.10_low.5 <- subset(results(dds.DESeq.tank, contrast=c("tank", "12", "24"), alpha=0.01), padj < 0.01))
(diffex.10_low.6 <- subset(results(dds.DESeq.tank, contrast=c("tank", "23", "24"), alpha=0.01), padj < 0.01))
diffex.tank.10_low <- unique(c(rownames(diffex.10_low.1), rownames(diffex.10_low.2), rownames(diffex.10_low.3), 
                     rownames(diffex.10_low.4), rownames(diffex.10_low.5), rownames(diffex.10_low.6)))
```

How many genes removed from DEG sets? 

```{r}
# How many genes are going to be removed due to outlier effects?
c(outs.diffex.3.ph$gene, outs.diffex.6.ph$gene, outs.diffex.10.ph$gene, 
  outs.diffex.3.6.amb$gene, outs.diffex.6.10.amb$gene, 
  outs.diffex.3.6.low$gene, outs.diffex.6.10.low$gene,
  outs.diffex.both.3low.6amb, outs.diffex.both.10low.6amb) %>% unique() %>% length()

# How many genes are going to be removed due to "tank effect?"
c(diffex.tank.3_low, diffex.tank.6_amb, diffex.tank.6_low, diffex.tank.10_amb, diffex.tank.10_low) %>% unique() %>% length()
mean(c(length(diffex.tank.3_low), length(diffex.tank.6_amb), length(diffex.tank.6_low), length(diffex.tank.10_amb), length(diffex.tank.10_low)))
sd(c(length(diffex.tank.3_low), length(diffex.tank.6_amb), length(diffex.tank.6_low), length(diffex.tank.10_amb), length(diffex.tank.10_low)))
```
```{r}
# Remove genes from DEG lists that: 
#  1)  contain iLOO-identified outliers from DEG objects, and 
#  2)  are DEGs among tanks within treatments. 
# then save to file. This is the final list of DEGs I will use for downstream analyses.

(diffex.pH.3.filt <- subset(diffex.pH.3, rownames(diffex.pH.3) %!in% c(outs.diffex.3.ph$gene, diffex.tank.3_low, genes.length)))
(diffex.pH.6.filt <- subset(diffex.pH.6, rownames(diffex.pH.6) %!in% c(outs.diffex.6.ph$gene, diffex.tank.6_amb, diffex.tank.6_low, genes.length)))
(diffex.pH.10.filt <- subset(diffex.pH.10, rownames(diffex.pH.10) %!in% c(outs.diffex.10.ph$gene, diffex.tank.10_amb, diffex.tank.10_low, genes.length)))

(diffex.temp.3.6.amb.filt <- subset(diffex.temp.3.6.amb, rownames(diffex.temp.3.6.amb) %!in% c(outs.diffex.3.6.amb$gene, diffex.tank.6_amb, genes.length)))
#(diffex.temp.3.10.amb.filt <- subset(diffex.temp.3.10.amb, rownames(diffex.temp.3.10.amb) %!in% c(outs.diffex.3.10.amb$gene, diffex.tank.10_amb, genes.length)))
(diffex.temp.6.10.amb.filt <- subset(diffex.temp.6.10.amb, rownames(diffex.temp.6.10.amb) %!in% c(outs.diffex.6.10.amb$gene, diffex.tank.6_amb, diffex.tank.10_amb, genes.length)))

(diffex.temp.3.6.low.filt <- subset(diffex.temp.3.6.low, rownames(diffex.temp.3.6.low) %!in% c(outs.diffex.3.6.low$gene, diffex.tank.3_low, diffex.tank.6_low, genes.length)))
#(diffex.temp.3.10.low.filt <- subset(diffex.temp.3.10.low, rownames(diffex.temp.3.10.low) %!in% c(outs.diffex.3.10.low$gene, diffex.tank.3_low, diffex.tank.10_low, genes.length)))
(diffex.temp.6.10.low.filt <- subset(diffex.temp.6.10.low, rownames(diffex.temp.6.10.low) %!in% c(outs.diffex.6.10.low$gene, diffex.tank.6_low, diffex.tank.10_low, genes.length)))


(diffex.both.3low.6amb.filt <- subset(diffex.both.3low.6amb, rownames(diffex.both.3low.6amb) %!in% c(outs.diffex.both.3low.6amb$gene, diffex.tank.3_low, diffex.tank.6_amb, genes.length)))
(diffex.both.10low.6amb.filt <- subset(diffex.both.10low.6amb, rownames(diffex.both.10low.6amb) %!in% c(outs.diffex.both.10low.6amb$gene, diffex.tank.10_low, diffex.tank.6_amb, genes.length)))

save(diffex.pH.3.filt, file = "../results/deseq2/diffex.pH.3.filt")
save(diffex.pH.6.filt, file = "../results/deseq2/diffex.pH.6.filt")
save(diffex.pH.10.filt, file = "../results/deseq2/diffex.pH.10.filt")
save(diffex.temp.3.6.amb.filt, file = "../results/deseq2/diffex.temp.3.6.amb.filt")
#save(diffex.temp.3.10.amb.filt, file = "../results/deseq2/diffex.temp.3.10.amb.filt")
save(diffex.temp.6.10.amb.filt, file = "../results/deseq2/diffex.temp.6.10.amb.filt")
save(diffex.temp.3.6.low.filt, file = "../results/deseq2/diffex.temp.3.6.low.filt")
#save(diffex.temp.3.10.low.filt, file = "../results/deseq2/diffex.temp.3.10.low.filt")
save(diffex.temp.6.10.low.filt, file = "../results/deseq2/diffex.temp.6.10.low.filt")
save(diffex.both.3low.6amb.filt, file = "../results/deseq2/diffex.both.3low.6amb.filt")
save(diffex.both.10low.6amb.filt, file = "../results/deseq2/diffex.both.10low.6amb.filt")
```
### Summary stats for the number of Differentially Expressed Genes (DEGs)

#### Bar plot of the number of DEGs identified in each contrast 

```{r}
# write simple function to return the name of an object as a string

# Here is a list containing all DEG sets - this will be used in subsequent code chunks, too! 
degs <- list(
diffex.pH.3.filt, diffex.pH.6.filt, diffex.pH.10.filt,
diffex.temp.3.6.amb.filt, diffex.temp.6.10.amb.filt,
diffex.temp.3.6.low.filt, diffex.temp.6.10.low.filt,
#diffex.temp.3.10.amb.filt, diffex.temp.3.10.low.filt, 
diffex.both.3low.6amb.filt, diffex.both.10low.6amb.filt)

names(degs) <- c("Amb vs. Low pH @3C", "Amb vs. Low pH @6C", "Amb vs. Low pH @ 10C", 
                 "3 vs. 6C @Amb pH", "6 vs. 10C @Amb pH", 
                 "3 vs. 6C @Low pH", "6 vs. 10C @Low pH",
#                 "3 vs. 10C @Amb pH",  "3 vs. 10C @Low pH",
                 "6 amb vs. 3C low", "6 amb vs. 10 low")
grp <- c(rep("pH contrast", times=3), 
         rep("temperature contrast, @ amb pH", times=3), 
         rep("temperature contrast, @ low pH", times=3),
         rep("combined stressors", times=2))


n.degs <- data.frame(matrix(NA, nrow = length(degs), ncol = 4))
names(n.degs) <- c("contrast", "n.degs", "perc", "grp")
for (i in (1:length(degs))) {
  n.degs[i,1] <- names(degs[i])
  n.degs[i,2] <- nrow(degs[[i]])
  n.degs[i,3] <- 100*round(nrow(degs[[i]])/ncol(counts.ts), digits = 3)
  n.degs[i,4] <- grp[i]
}

n.degs$contrast <- factor(n.degs$contrast, levels=n.degs$contrast, ordered = T)

save(degs, file = "../results/deseq2/degs")
n.degs.larvae <- n.degs
save(n.degs.larvae, file = "../results/deseq2/n.degs.larvae")
```
```{r}
round(3411/21076, digits=3)
round(2890/21076, digits=3)
round(1990/21076, digits=3)
round(1803/21076, digits=3)
round(1259/21076, digits=3)
```

Explore biological variation within treatments (and tanks) for just DEGs 

```{r}
degs.vector <- unique(c(rownames(diffex.pH.6.filt), #OA
rownames(diffex.temp.3.6.amb.filt), #cold 
rownames(diffex.temp.6.10.amb.filt), #warm
rownames(diffex.both.3low.6amb.filt), #cold-OA
rownames(diffex.both.10low.6amb.filt))) #warm-OA

length(degs.vector)
var_df_degs_treat <- gene_var_by_treatment %>% filter(gene %in% degs.vector) %>% 
  mutate(log_var=log(variance))
var_df_degs_treat %>% group_by(treatment) %>% summarize(mean=round(mean(variance), digits = 3), sd=round(sd(variance), digits=3)) 

var_df_degs_tank <- gene_var_by_tank %>% filter(gene %in% degs.vector) %>% 
  mutate(log_var=log(variance)) 
var_df_degs_tank %>% group_by(treatment, tank) %>% summarize(mean=round(mean(variance), digits = 3), sd=round(sd(variance), digits=3)) %>% ungroup() %>% 
  summarize(mean_var_tank=mean(mean, na.rm=T),sd_var_tank=sd(mean, na.rm=T))


# normal distribution after log transformation? 
hist(var_df_degs$log_var)

# Boxplot of variance per treatment
var_df_degs %>% filter(log_var!="-Inf") %>% 
ggplot(aes(x = treatment, y = log_var)) +
  geom_boxplot() + 
  theme_minimal() +
  labs(title = "Log-variance of Gene Expression by Treatment", y = "Gene Expression Variance") + 
  geom_point(data=var_df_degs %>% group_by(treatment) %>% summarize(mean=round(mean(variance), digits = 3), sd=round(sd(variance), digits=3)),
             aes(x=treatment, y=log(mean)), size=4)

summary(aov(log_var ~ treatment, data=var_df_degs%>% filter(log_var!="-Inf")))
TukeyHSD(aov(log_var ~ treatment, data=var_df_degs%>% filter(log_var!="-Inf")))
```
PCA using just DEGs 

### PCAs using prcomp

This enables screeplot to ID significant PCs, calculates variance, etc. 

```{r, warning=F, message=F}
#pca.princomp <- prcomp(cov(assay(vsd.treatment)), scale=F) #scale=F for variance-covariance matrix
pca.tank.degs <- prcomp(t(assay(vsd.treatment.tank)[degs.vector,])) # using the same code that's under the hood of PlotPCA but using all genes 
pca.eigenval(pca.tank.degs) #The Proporation of Variance = %variance 
perc.tank.degs <- pca.eigenval(pca.tank.degs)[2,1:7]*100
screeplot(pca.tank.degs, bstick=TRUE) #looks like PC 1-34 are significant 
perc.tank.degs[1:2] %>% sum()

#### Generate dataframe with prcomp results 
tab.tank.degs <- data.frame(sample.id = colnames(assay(vsd.treatment.tank)),
    EV1.all = pca.tank.degs$x[,1],    # the first eigenvector
    EV2.all = pca.tank.degs$x[,2],    # the second eigenvector
    EV3.all = pca.tank.degs$x[,3],    # the third eigenvector
    EV4.all = pca.tank.degs$x[,4],    # the fourth eigenvector
    stringsAsFactors = FALSE)
tab.tank.annot.degs <- left_join(tab.tank.degs, sample.info, by=c("sample.id"="vial_label")) %>% droplevels()

#pdf(file="../manuscript/CJFAS-submission/Fig1.pdf", height = 3.5, width = 5)
tab.tank.annot.degs %>% mutate(tank_rep=case_when(
  tank %in% c(1,11,19,13,3,7)~"1",
  tank %in% c(2,12,14,4,8)~"2",
  tank %in% c(9,23,21,17,15)~"3",
  tank %in% c(10,24,22,18,16)~"4")) %>% 
  ggplot(aes(x=treatment, y=EV1.all, color=treatment, shape=tank_rep)) + 
  geom_boxplot(outlier.shape = NULL) +
  geom_jitter(width=0.25)


ggscatter(tab.tank.annot.degs %>% 
            mutate(tank_rep=case_when(
  tank %in% c(1,11,19,13,3,7)~"1",
  tank %in% c(2,12,14,4,8)~"2",
  tank %in% c(9,23,21,17,15)~"3",
  tank %in% c(10,24,22,18,16)~"4")) %>%
            mutate(treatment=factor(treatment, 
                             levels=c("6_amb", "6_low", "10_amb", 
                                      "10_low", "3_amb", "3_low"), ordered = T)), 
          group=c("temperature", "tank"),
          x="EV1.all", y="EV2.all", col="treatment", shape="tank_rep", size=3.5, alpha=0.6, 
          ellipse = FALSE, star.plot = T, star.size=0.1) +
  theme_minimal() + ggtitle("Differential gene expression PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
  theme(axis.text = element_text(size=8), title = element_text(size=8),
        legend.position = "right", legend.text=element_text(size=7), legend.title=element_text(size=7)) + 
  scale_color_manual(name=NULL,
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25),
                              "10_amb"=lighten("firebrick4", 0.45),
                              "10_low"="firebrick4",
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="Acidified",
                              "10_amb"="Warm",
                              "10_low"="Warm + Acidified",
                              "3_amb"="Cold",
                              "3_low"="Cold + Acidified")) +
  scale_shape_manual(values=c(15,16,17,18)) + #guide="none", 
  guides(colour = guide_legend(override.aes =
         list(size=2.5, linetype="blank", shape=c(19, 17, 19, 17, 19, 17))))

centroids.degs <- tab.tank.annot.degs %>%
  group_by(treatment) %>%
  summarise(across(c(EV1.all, EV2.all), mean), .groups = "drop")

# Compute Euclidean distance to centroid
tab.tank.annot.degs <- tab.tank.annot.degs %>% mutate(tank_rep=case_when(
  tank %in% c(1,11,19,13,3,7)~"1",
  tank %in% c(2,12,14,4,8)~"2",
  tank %in% c(9,23,21,17,15)~"3",
  tank %in% c(10,24,22,18,16)~"4")) %>%
  left_join(centroids.degs, by = "treatment", suffix = c("", "_centroid")) %>%
  mutate(euclidean_dist = sqrt((EV1.all - EV1.all_centroid)^2 + (EV2.all - EV2.all_centroid)^2)) %>% 
  mutate(treatment=factor(treatment, 
                 levels=c("6_amb", "6_low", "10_amb", 
                          "10_low", "3_amb", "3_low"), ordered = T))

hist(log(tab.tank.annot.degs$euclidean_dist))
summary(aov(log(euclidean_dist)~treatment, tab.tank.annot.degs))
TukeyHSD(aov(log(euclidean_dist)~treatment, tab.tank.annot.degs))

tab.tank.annot.degs %>%
  ggplot(aes(x=treatment, color=treatment, y=euclidean_dist)) + 
  geom_boxplot(outlier.shape = NA) + geom_jitter(aes(shape=tank_rep), 
                                                 alpha=0.75, size=2, width = 0.25) + 
  theme_minimal() +
    scale_color_manual(name=NULL,
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25),
                              "10_amb"=lighten("firebrick4", 0.45),
                              "10_low"="firebrick4",
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="Acidified",
                              "10_amb"="Warm",
                              "10_low"="Warm + Acidified",
                              "3_amb"="Cold",
                              "3_low"="Cold + Acidified")) +
  ggtitle("Distance to treatment centroid, PC1xPC2") + 
  ylab("Euclidean distance") + xlab("Treatment") + 
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())

#dev.off()
```


Barplot of the number of DEGs 

```{r}
# single stressors vs. multiple stressors 
#pdf(file="../results/deseq2/DEGs_barplot_single-vs-double-stressor.pdf", height = 2, width = 7)
ggplot(n.degs %>% 
         filter(contrast %in% c("Amb vs. Low pH @6C",
                                "3 vs. 6C @Amb pH", "6 amb vs. 3C low",
                                "6 vs. 10C @Amb pH", "6 amb vs. 10 low")) %>%
         mutate(effect_of=case_when(
           grepl("Amb vs. Low pH @6C", contrast) ~ "High pCO2",
           grepl("3 vs. 6C", contrast) ~ "Cold temperature",
           grepl("6 amb vs. 3C low", contrast) ~ "Cold temperature + High pCO2",
           grepl("6 amb vs. 10 low", contrast) ~ "Warm temperature + High pCO2",
           grepl("6 vs. 10C", contrast) ~ "Warm temperature")) %>% 
        mutate(effect_of=fct_relevel(effect_of,
           c("Cold temperature + High pCO2", "Cold temperature",
             "Warm temperature + High pCO2", "Warm temperature",
             "High pCO2"))), 
       aes(x=effect_of, y=n.degs, fill=effect_of)) + 
  geom_bar(aes(alpha=effect_of), stat = "identity", color="gray20", size=0.1) +
  geom_text(hjust=-0.1, size=3.5, aes(label=paste(comma(n.degs), " (", perc, "%)", sep=""))) + 
  xlab("Contrast") + ylab("Number Differentially Expressed Genes (% of all genes)") + 
  xlab(NULL) + ylab(NULL) + coord_flip() + ggtitle("Number Differentially Expressed Genes (% of all genes)") +
  theme_minimal() + ylim(c(0,5000)) + 
  theme(legend.position = "none", plot.title = element_text(size = 10)) +
  scale_fill_manual(name=NULL, values=c( "navyblue", "navyblue", "firebrick4", "firebrick4","darkorange2")) +
  scale_alpha_manual(name=NULL, values=c(.7, .4, .7, .4, .4)) +
  guides(fill=guide_legend(nrow=3,byrow=TRUE,reverse=TRUE)) 
#dev.off()
```
MAKE BARPLOT FOR PRESENTATION THAT ONLY INCLUDES WARMING AND WARMING+ACIDIFICATION

```{r}
ggplot(n.degs %>% 
         filter(contrast %in% c("Amb vs. Low pH @6C",
                                "6 vs. 10C @Amb pH", "6 amb vs. 10 low")) %>%
         mutate(effect_of=case_when(
           grepl("Amb vs. Low pH @6C", contrast) ~ "High pCO2",
           grepl("6 amb vs. 10 low", contrast) ~ "Warm temperature + High pCO2",
           grepl("6 vs. 10C", contrast) ~ "Warm temperature")) %>% 
        mutate(effect_of=fct_relevel(effect_of,
           c("High pCO2", "Warm temperature + High pCO2", "Warm temperature"))), 
       aes(x=effect_of, y=n.degs, fill=effect_of)) + 
  geom_bar(aes(alpha=effect_of), stat = "identity", color="gray20", size=0.1) +
  geom_text(hjust=-0.1, size=3.5, aes(label=paste(comma(n.degs), " (", perc, "%)", sep=""))) + 
  xlab("Contrast") + ylab("Number Differentially Expressed Genes (% of all genes)") + 
  xlab(NULL) + ylab(NULL) + coord_flip() + ggtitle("Number Differentially Expressed Genes (% of all genes)") +
  theme_minimal() + ylim(c(0,5000)) + 
  theme(legend.position = "none", plot.title = element_text(size = 10)) +
  scale_fill_manual(name=NULL, values=c("darkorange2", "firebrick4", "firebrick4")) +
  scale_alpha_manual(name=NULL, values=c(.4, .7, .4)) +
  guides(fill=guide_legend(nrow=3,byrow=TRUE,reverse=TRUE)) 
```

Regenerate barplot for paper (Figure 3) with the number of upregulated / downregualted DEGs by treatment 

```{r}
degs.updown <- list(
diffex.pH.6.filt %>% data.frame() %>% filter(log2FoldChange > 0),  #downregulated in OA
diffex.pH.6.filt %>% data.frame() %>% filter(log2FoldChange < 0), #upregulated in OA
diffex.temp.3.6.amb.filt %>% data.frame() %>% filter(log2FoldChange > 0),  #upregulated in Cold
diffex.temp.3.6.amb.filt %>% data.frame() %>% filter(log2FoldChange < 0), #downregulated in Cold
diffex.temp.6.10.amb.filt %>% data.frame() %>% filter(log2FoldChange > 0),  #downregulated in Warm
diffex.temp.6.10.amb.filt %>% data.frame() %>% filter(log2FoldChange < 0), #upregulated in Warm
diffex.both.3low.6amb.filt %>% data.frame() %>% filter(log2FoldChange > 0),  #upregulated in Cold+OA
diffex.both.3low.6amb.filt %>% data.frame() %>% filter(log2FoldChange < 0), #downregulated in Cold+OA
diffex.both.10low.6amb.filt %>% data.frame() %>% filter(log2FoldChange > 0),  #downregulated in Warm+OA
diffex.both.10low.6amb.filt %>% data.frame() %>% filter(log2FoldChange < 0)) #upregulated in Warm+OA

names(degs.updown) <- c("High pCO2", "High pCO2", 
                        "Cold temperature", "Cold temperature", 
                        "Warm temperature", "Warm temperature", 
                        "Cold temperature + High pCO2", "Cold temperature + High pCO2", 
                        "Warm temperature + High pCO2", "Warm temperature + High pCO2")

response <- c("downregulated", "upregulated", "upregulated", "downregulated", "downregulated", "upregulated", "upregulated", "downregulated", "downregulated", "upregulated")


n.degs.updown <- data.frame(matrix(NA, nrow = length(degs.updown), ncol = 4))
names(n.degs.updown) <- c("stressor", "n.degs", "perc", "response")
for (i in (1:length(degs.updown))) {
  n.degs.updown[i,1] <- names(degs.updown[i])
  n.degs.updown[i,2] <- nrow(degs.updown[[i]])
  n.degs.updown[i,3] <- 100*round(nrow(degs.updown[[i]])/ncol(counts.ts), digits = 2)
  n.degs.updown[i,4] <- response[i]
}

n.degs.updown <- n.degs.updown %>%
  mutate(stressor=factor(stressor, levels=
                           c("Cold temperature + High pCO2","Cold temperature",
                             "Warm temperature + High pCO2","Warm temperature",
                             "High pCO2")),
         response=factor(response, ordered = T))

n.degs.updown

# Barplot with no. of DEGs by single stressors vs. multiple stressors, including up/down regulation 
#pdf(file="../results/deseq2/DEGs_barplot_single-vs-double-stressor-up-down.pdf", height = 2, width = 7)
ggplot(n.degs.updown, 
       aes(x=stressor, y=n.degs, fill=stressor)) + 
  geom_bar_pattern(aes(alpha=stressor, pattern=response), stat = "identity", color="gray20", size=0.05,
                   pattern_fill="black", pattern_alpha=0.2, pattern_density=0.01, pattern_spacing=0.025) +
  xlab("Contrast") + ylab("Number Differentially Expressed Genes (% of all genes)") + 
  xlab(NULL) + ylab(NULL) + coord_flip() + ggtitle("Number Differentially Expressed Genes (% of all genes)") +
  theme_pubr() + scale_y_continuous(label=comma, limits = c(0,4000)) +
  theme(legend.position = "none", plot.title = element_text(size = 10)) +
  scale_fill_manual(name=NULL, values=c( "navyblue", "navyblue", "firebrick4", "firebrick4","darkorange2")) +
  scale_alpha_manual(name=NULL, values=c(.7, .4, .7, .4, .4)) +
  scale_pattern_manual(values = c("stripe","none")) +
  guides(fill=guide_legend(nrow=3,byrow=TRUE,reverse=TRUE)) 
#dev.off()
```
Barplot showing abs( log fold change ) with error bars 

```{r}
bind_rows(diffex.temp.6.10.amb.filt %>% as.data.frame() %>% mutate(stressor="Warm temperature"), 
          diffex.both.10low.6amb.filt %>% as.data.frame() %>% mutate(stressor="Warm temperature + High pCO2"),   
          diffex.pH.6.filt %>% as.data.frame() %>% mutate(stressor="High pCO2"), 
          diffex.temp.3.6.amb.filt %>% as.data.frame() %>% mutate(stressor="Cold temperature"), 
          diffex.both.3low.6amb.filt %>% as.data.frame() %>% mutate(stressor="Cold temperature + High pCO2")) %>% 
  mutate(response=case_when(log2FoldChange>0~"upregulated", log2FoldChange<0~"downregulated"),
         SE = (2^(log2FoldChange + lfcSE) - 2^(log2FoldChange - lfcSE)) / 2) %>% 
  group_by(stressor, response) %>% 
  summarize(mean.lfc=mean(log2FoldChange), mean.lfcse=mean(lfcSE), mean.se=mean(SE)) %>% 
  mutate(mean.effect=2^abs(mean.lfc)) %>% 
  arrange(response) %>% ungroup() %>%
  mutate(mean.effect.plot=case_when(response=="downregulated"~-1*mean.effect, TRUE~mean.effect)) %>% 
  left_join(n.degs.updown, by=c("stressor", "response")) %>% 
  mutate(mean.effect.plot=case_when(response=="downregulated"~-1*(mean.effect-1), response=="upregulated"~mean.effect-1)) %>%
  mutate(response=case_when(response=="downregulated"~"Lower gene expression", TRUE~"Higher gene expression")) %>% 
  mutate(stressor=factor(stressor, ordered = T, levels=c("High pCO2", "Warm temperature", "Warm temperature + High pCO2", "Cold temperature", "Cold temperature + High pCO2"))) %>%
   mutate(treat=factor(case_when(
   stressor=="High pCO2"~"OA",
   stressor=="Cold temperature"~"Cold",
   stressor=="Warm temperature"~"Warm",
   stressor=="Cold temperature + High pCO2"~"Cold+OA",
   stressor=="Warm temperature + High pCO2"~"Warm+OA"),
   ordered=T, levels=c("Cold+OA","Cold","OA","Warm+OA","Warm"))) %>% 
  
  ggplot(aes(x=n.degs, y=mean.effect, color=stressor)) + geom_point(size=4.5, alpha=0.75) +
  geom_errorbar(aes(ymin=mean.effect-mean.se, ymax=mean.effect+mean.se), position=position_dodge(0.05)) + 
  xlab("Number of differentially expressed genes") + 
  ylab("Effect size (fold change, mean \u00B1 SE)") + #scale_y_continuous(labels = percent_format()) +
  
      scale_color_manual(name=NULL,
                     values=c("High pCO2"=lighten("darkorange2", 0.25),
                              "Warm temperature"=lighten("firebrick4", 0.45),
                              "Warm temperature + High pCO2"="firebrick4",
                              "Cold temperature"=lighten("navyblue", 0.45),
                              "Cold temperature + High pCO2"="navyblue"),
                     labels=c("High pCO2"="Acidified",
                              "Warm temperature"="Warm",
                              "Warm temperature + High pCO2"="Warm + Acidified",
                              "Cold temperature"="Cold",
                              "Cold temperature + High pCO2"="Cold + Acidified")) +
  scale_shape_manual(name=NULL, values=c(17, 19)) + 
  scale_x_continuous(breaks=c(500,1000,1500,2000), labels=c(500,1000,1500,2000), limits = c(500, 2000)) + 
  facet_wrap(~response, scales = "free_y", nrow=2) + theme_minimal() + 
  theme(axis.title.x = element_text(margin = margin(t = 10)))
```
```{r}
Plot2a <- n.degs.updown %>% 
         mutate(treat=factor(case_when(
           stressor=="High pCO2"~"OA",
           stressor=="Cold temperature"~"Cold",
           stressor=="Warm temperature"~"Warm",
           stressor=="Cold temperature + High pCO2"~"Cold+OA",
           stressor=="Warm temperature + High pCO2"~"Warm+OA"),
           ordered=T, levels=c("Cold+OA","Cold","OA","Warm+OA","Warm"))) %>% 
  ggplot(aes(x=treat, y=n.degs, fill=treat)) + 
  geom_bar_pattern(aes(alpha=treat, pattern=rev(response)), stat = "identity", color="gray20", size=0.05,
                   pattern_fill="black", pattern_alpha=0.2, pattern_density=0.01, pattern_spacing=0.025) +
  xlab(NULL) + ylab("Number DEGs (% of all genes)") + ggtitle("Number Differentially Expressed Genes") + 
  theme_pubr() + scale_y_continuous(label=comma, limits = c(0,4000)) +
  theme(legend.position = "none", plot.title = element_text(hjust = 0.5)) +
  scale_fill_manual(name=NULL, #labels=c("Cold", "Cold+OA","Warm+OA","Warm","OA"), 
                    values=c("Cold+OA"="navyblue", "Cold"="navyblue","OA"="darkorange2","Warm+OA"="firebrick4","Warm"="firebrick4")) +
  scale_alpha_manual(name=NULL, values=c(.7, .4, .7, .7, .4)) +
  scale_pattern_manual(values = c("none","stripe")) +
  guides(fill=guide_legend(nrow=3,byrow=TRUE,reverse=TRUE)) 
```

```{r}
Plot2b <- 
  bind_rows(diffex.temp.6.10.amb.filt %>% as.data.frame() %>% mutate(stressor="Warm temperature"), 
          diffex.both.10low.6amb.filt %>% as.data.frame() %>% mutate(stressor="Warm temperature + High pCO2"),   
          diffex.pH.6.filt %>% as.data.frame() %>% mutate(stressor="High pCO2"), 
          diffex.temp.3.6.amb.filt %>% as.data.frame() %>% mutate(stressor="Cold temperature"), 
          diffex.both.3low.6amb.filt %>% as.data.frame() %>% mutate(stressor="Cold temperature + High pCO2")) %>% 
  mutate(response=case_when(log2FoldChange>0~"upregulated", log2FoldChange<0~"downregulated"),
         SE = (2^(log2FoldChange + lfcSE) - 2^(log2FoldChange - lfcSE)) / 2) %>% 
  group_by(stressor, response) %>% 
  summarize(mean.lfc=mean(log2FoldChange), mean.lfcse=mean(lfcSE), mean.se=mean(SE)) %>% 
  mutate(mean.effect=2^abs(mean.lfc)) %>% 
  arrange(response) %>% ungroup() %>%
  mutate(mean.effect.plot=case_when(response=="downregulated"~-1*mean.effect, TRUE~mean.effect)) %>% 
  left_join(n.degs.updown, by=c("stressor", "response")) %>% 
  mutate(mean.effect.plot=case_when(response=="downregulated"~-1*(mean.effect-1), response=="upregulated"~mean.effect-1)) %>%
  mutate(response=case_when(response=="downregulated"~"Lower gene expression", TRUE~"Higher gene expression")) %>% 
  mutate(stressor=factor(stressor, ordered = T, levels=c("High pCO2", "Warm temperature", "Warm temperature + High pCO2", "Cold temperature", "Cold temperature + High pCO2"))) %>%
   mutate(treat=factor(case_when(
   stressor=="High pCO2"~"OA",
   stressor=="Cold temperature"~"Cold",
   stressor=="Warm temperature"~"Warm",
   stressor=="Cold temperature + High pCO2"~"Cold+OA",
   stressor=="Warm temperature + High pCO2"~"Warm+OA"),
   ordered=T, levels=c("Cold+OA","Cold","OA","Warm+OA","Warm"))) %>% 
  
  ggplot(aes(x=treat, y=mean.effect, color=treat, fill=treat)) + 
  geom_errorbar(aes(ymin=mean.effect-mean.se, ymax=mean.effect+mean.se), position=position_dodge(0.05), width=0.2) + 
  geom_tile_pattern(aes(pattern=response, pattern_fill=response, color=treat, fill=treat), width=0.3, height=0.02, fill="white", size=0.3,
                   pattern_alpha=0.2, pattern_density=0.01, pattern_spacing=0.025) +
  xlab(NULL) + ggtitle("Effect size") + 
  ylab("Fold change, mean \u00B1 SE") + #scale_y_continuous(labels = percent_format()) +
  scale_color_manual(name=NULL,
       values=c("OA"=lighten("darkorange2", 0.25),
                "Warm"=lighten("firebrick4", 0.45),
                "Warm+OA"="firebrick4",
                "Cold"=lighten("navyblue", 0.45),
                "Cold+OA"="navyblue")) +
  scale_pattern_manual(values = c("none", "stripe")) +
  scale_pattern_fill_manual(values = c("OA"=lighten("darkorange2", 0.25),
                "Warm"=lighten("firebrick4", 0.45),
                "Warm+OA"="firebrick4",
                "Cold"=lighten("navyblue", 0.45),
                "Cold+OA"="navyblue")) +
  facet_wrap(~response, scales = "free_y", nrow=2) + theme_pubr() + 
  theme(legend.position = "none", axis.title.x = element_text(margin = margin(t = 10)),
        plot.title = element_text(hjust = 0.5))
```

```{r}
pdf(file = "../manuscript/CJFAS-revision2/Figure-2-replaced4.pdf", width=8, height=5.5)
ggarrange(Plot2a, Plot2b, nrow = 1)
dev.off
```


### Heat maps of DEGs 

#### Cold treatments 

```{r}
diffex.cold.heatmap <- subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% unique(c(rownames(diffex.temp.3.6.amb.filt),rownames(diffex.both.3low.6amb.filt))))
nrow(diffex.cold.heatmap)
# Create a dataframe to annotate heatmap with treatments 
dds.cold.df <- sample.info[match(colnames(diffex.cold.heatmap), sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")] %>%
  remove_rownames() %>% column_to_rownames(var = "vial_label")

# Make sure treatment order is correct
all(colnames(diffex.cold.heatmap) == rownames(dds.cold.df)) #double check that samples are still in same order 

# All DEGs among any pH treatment (within temperatures)
pheatmap(diffex.cold.heatmap[,(sample.info%>%filter(treatment%in%c("6_amb","3_amb","3_low")))$vial_label], 
         cluster_rows=TRUE, cluster_cols=F, treeheight_row = 0, #show_colnames = F,
         show_rownames=FALSE, na.rm=TRUE, 
         annotation_colors = list(treatment= 
                              c("6_amb"="gray65","6_low"=lighten("darkorange2", 0.25),
                              "10_amb"=lighten("firebrick4", 0.45),
                              "10_low"="firebrick4",
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue")),
         scale="row", main = "Differentially expressed genes (DEGs), cold treatments", 
         annotation_col=dds.pH.df[,"treatment", drop=FALSE], fontsize = 8,
         cutree_rows = 2,  border_color=F)

```

Warm Treatments 
```{r}
diffex.warm.heatmap <- subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% unique(c(rownames(diffex.temp.6.10.amb.filt),rownames(diffex.both.10low.6amb.filt))))
nrow(diffex.warm.heatmap)

# Create a dataframe to annotate heatmap with treatments 
dds.warm.df <- sample.info[match(colnames(diffex.warm.heatmap), sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")] %>%
  remove_rownames() %>% column_to_rownames(var = "vial_label")

# Make sure treatment order is correct
all(colnames(diffex.warm.heatmap) == rownames(dds.warm.df)) #double check that samples are still in same order 

# All DEGs among any pH treatment (within temperatures)
pheatmap(diffex.warm.heatmap[,(sample.info%>% ungroup() %>% filter(treatment%in%c("6_amb","10_amb","10_low")) %>% select(treatment, vial_label) %>% 
    arrange(desc(treatment)))$vial_label ], 
         cluster_rows=TRUE, cluster_cols=F, treeheight_row = 0, #show_colnames = F,
         show_rownames=FALSE, na.rm=TRUE, 
         annotation_colors = list(treatment= 
                              c("6_amb"="gray65","6_low"=lighten("darkorange2", 0.25),
                              "10_amb"=lighten("firebrick4", 0.45),
                              "10_low"="firebrick4",
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue")),
         scale="row", main = "Differentially expressed genes (DEGs), warm treatments vs. control", 
         annotation_col=dds.pH.df[,"treatment", drop=FALSE], fontsize = 8,
         cutree_rows = 2,  border_color=F)
```

Acidified Treatment 
```{r}
diffex.acid.heatmap <- subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% unique(c(rownames(diffex.pH.6.filt))))
nrow(diffex.acid.heatmap)

# Create a dataframe to annotate heatmap with treatments 
dds.acid.df <- sample.info[match(colnames(diffex.acid.heatmap), sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")] %>%
  remove_rownames() %>% column_to_rownames(var = "vial_label")

# Make sure treatment order is correct
all(colnames(diffex.acid.heatmap) == rownames(dds.acid.df)) #double check that samples are still in same order 

# All DEGs among any pH treatment (within temperatures)
pheatmap(diffex.warm.heatmap[,(sample.info%>% ungroup() %>% filter(treatment%in%c("6_amb","6_low")) %>% select(treatment, vial_label) %>% 
    arrange(treatment))$vial_label ], 
         cluster_rows=TRUE, cluster_cols=F, treeheight_row = 0, #show_colnames = F,
         show_rownames=FALSE, na.rm=TRUE, 
         annotation_colors = list(treatment= 
                              c("6_amb"="gray65","6_low"=lighten("darkorange2", 0.25),
                              "10_amb"=lighten("firebrick4", 0.45),
                              "10_low"="firebrick4",
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue")),
         scale="row", main = "Differentially expressed genes (DEGs), acidified treatments vs. control", 
         annotation_col=dds.pH.df[,"treatment", drop=FALSE], fontsize = 8,
         cutree_rows = 2,  border_color=F)
```





#### Single heatmap with all DEGs among any pH treatment 
NOTE: There are 4,525 DEGs among pH treatment in the 6C group, but only 2 and 58 among the 3C and 10C groups, respectively. This is visible in the heat map- there are clear differences in expression in the 6_amb (green) and 6_low (yellow) treatments. 

Kinda interesting - below are only DEGs among pH treatment (mostly 6amb and 6low), but the expressoin patterns look similar among 6low, 10amb and 10low, compared to 6amb, 3amb and 3low. 

```{r}
# NOTE THIS DOESN'T EXCLUDE LENGTH-ASSOCIATED GENES
# Generate counts matrix With DEGs among any pH treatment 

diffex.pH.counts <- subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% unique(c(rownames(diffex.pH.3.filt),rownames(diffex.pH.6.filt),rownames(diffex.pH.10.filt))))

# Create a dataframe to annotate heatmap with treatments 
dds.pH.df <- sample.info[match(colnames(diffex.pH.counts), sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")] %>%
  remove_rownames() %>% column_to_rownames(var = "vial_label")

# Make sure treatment order is correct
all(colnames(diffex.pH.counts) == rownames(dds.pH.df)) #double check that samples are still in same order 

# All DEGs among any pH treatment (within temperatures)
pheatmap(diffex.pH.counts, cluster_rows=TRUE, cluster_cols=FALSE,
         show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=
             c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen",
               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")),
         scale="row", main = "Differentially expressed genes (DEGs) among any pH treatment", 
         annotation_col=dds.pH.df[,"treatment", drop=FALSE], fontsize = 8,
         cutree_rows = 2,  border_color=F)
```

#### Single heatmap with all DEGs among any temperature treatment, ambient pH only

```{r}
# NOTE THIS DOESN'T EXCLUDE LENGTH-ASSOCIATED GENES
ph.amb <- (sample.info %>% filter(ph=="amb") %>% arrange(temperature))$vial_label

# Generate counts matrix With DEGs among any tempreature treatment, ambient pH only
diffex.temp.counts.amb <-  
  (subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% 
            unique(c(rownames(diffex.temp.3.6.amb.filt),
                     #rownames(diffex.temp.3.10.amb.filt),
                     rownames(diffex.temp.6.10.amb.filt)))) %>% 
     as.data.frame())[ph.amb] %>% as.matrix()
  
# Create a dataframe to annotate heatmap with treatments 
dds.temp.df <- sample.info[match(colnames(diffex.temp.counts.amb), sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")] %>%
  remove_rownames() %>% column_to_rownames(var = "vial_label")

# Make sure treatment order is correct
all(colnames(diffex.temp.counts.amb) == rownames(dds.temp.df)) #double check that samples are still in same order 

# All DEGs among any temp treatment (within temperatures)
pheatmap(diffex.temp.counts.amb, cluster_rows=TRUE, cluster_cols=FALSE,
         show_rownames=FALSE, na.rm=TRUE, 
         annotation_colors = list(temperature = c("3"= "navyblue", "6"="darkgreen","10"="firebrick4")),
         scale="row", 
         main = "Differentially expressed genes (DEGs) among any temperature treatment\nAmbient pH only", 
         annotation_col=dds.temp.df[,"temperature", drop=FALSE], fontsize = 8,
         cutree_rows = 2,  border_color=F)
```

#### Single heatmap with all DEGs among any temperature treatment, low pH only

```{r}
# NOTE THIS DOESN'T EXCLUDE LENGTH-ASSOCIATED GENES
ph.low <- sample.info[sample.info$ph=="low",]$vial_label
ph.low <- (sample.info %>% filter(ph=="low") %>% arrange(temperature))$vial_label

# Generate counts matrix With DEGs among any tempreature treatment, low pH only
diffex.temp.counts.low <-  
  (subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% 
            unique(c(rownames(diffex.temp.3.6.low.filt),
                     #rownames(diffex.temp.3.10.low.filt),
                     rownames(diffex.temp.6.10.low.filt)))) %>% 
     as.data.frame())[ph.low] %>% as.matrix()

# Create a dataframe to annotate heatmap with treatments 
dds.temp.df <- sample.info[match(colnames(diffex.temp.counts.low), sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")] %>%
  remove_rownames() %>% column_to_rownames(var = "vial_label")

# Make sure treatment order is correct
all(colnames(diffex.temp.counts.low) == rownames(dds.temp.df)) #double check that samples are still in same order 

# All DEGs among any temp treatment (within temperatures)
pheatmap(diffex.temp.counts.low, cluster_rows=TRUE, cluster_cols=FALSE,
         show_rownames=FALSE, na.rm=TRUE, 
         annotation_colors = list(temperature = c("3"="royalblue1","6"="yellow3","10"="sienna2")),
         scale="row", 
         main = "Differentially expressed genes (DEGs) among any temperature treatment\nLow pH only", 
         annotation_col=dds.temp.df[,"temperature", drop=FALSE], fontsize = 8,
         cutree_rows = 2,  border_color=F)
```



### Venn Diagrams of DEGs by pairwise contrast


## MOST important venn diagram, potentially for paper - Show effect of warming, cooling, and high pCO2 all in contrast to the optimal conditions (6C-ambient)

```{r, include=F, echo=F, message=F}

#install.packages("eulerr")
library(eulerr)

`Cold Temperature` <- row.names(degs$`3 vs. 6C @Amb pH`)
`Warm Temperature`  <- row.names(degs$`6 vs. 10C @Amb pH`)
`High pCO2` <- row.names(degs$`Amb vs. Low pH @6C`)

venn.stressors <- data.frame(dat = 
             unique(c(`Cold Temperature`, `Warm Temperature`, `High pCO2`))) %>%
  mutate(`Cold Temperature` = dat %in% `Cold Temperature`,
         `Warm Temperature` = dat %in% `Warm Temperature`,
         `High pCO2` = dat %in% `High pCO2`) %>%
  select(`Cold Temperature`, `Warm Temperature`, `High pCO2`) %>%
  euler()

pdf(file="../results/deseq2/venn_single-stressors_no-values.pdf", height = 5.2, width = 6)
eulerr:::plot.euler(venn.stressors, counts = T, quantities = T, 
                      legend = T, edges=F, fontzie=8,
                      font=1, cex=1, alpha=0.5,
                      #fills=list(fill=c("navyblue","firebrick4", "darkorange2")))
                      fills=list(fill=c("#4B4B8C","#BB4D4D", "#FF955E")))
dev.off()

# lighten("navyblue", amount = .25)
# lighten("firebrick4", amount = .25)
# lighten("darkorange2", amount = .25)
```

Explore whether adding on pCO2 stressor exacerbated effects of temperature stressors 

```{r}
# Cold vs. Cold+High pCO2
`Cold + High pCO2` <- row.names(diffex.both.3low.6amb.filt)
`Cold only`  <- row.names(diffex.temp.3.6.amb.filt)

venn.stressors.cold <- data.frame(dat = 
             unique(c(`Cold + High pCO2`, `Cold only`))) %>%
  mutate(`Cold + High pCO2` = dat %in% `Cold + High pCO2`,
         `Cold only` = dat %in% `Cold only`) %>%
  select(`Cold + High pCO2`, `Cold only`) %>%
  euler()

eulerr:::plot.euler(venn.stressors.cold, counts = T, quantities = T, 
                      legend = T, edges=F, fontzie=8,
                      font=1, cex=1, alpha=0.6,
                      fills=list(fill=c("orange","navyblue")))


# Warm vs. Warm+High pCO2
`Warm + High pCO2` <- row.names(diffex.both.10low.6amb.filt)
`Warm only`  <- row.names(diffex.temp.6.10.amb.filt)

venn.stressors.warm <- data.frame(dat = 
             unique(c(`Warm + High pCO2`, `Warm only`))) %>%
  mutate(`Warm + High pCO2` = dat %in% `Warm + High pCO2`,
         `Warm only` = dat %in% `Warm only`) %>%
  select(`Warm + High pCO2`, `Warm only`) %>%
  euler()

eulerr:::plot.euler(venn.stressors.warm, counts = T, quantities = T, 
                      legend = T, edges=F, fontzie=8,
                      font=1, cex=1, alpha=0.6,
                      fills=list(fill=c("orange","firebrick4")))

# all single and double stressors 
venn.stressors.warm.cold <- data.frame(dat = 
             unique(c(`Cold + High pCO2`, `Cold only`,
                      `Warm + High pCO2`, `Warm only`))) %>%
  mutate(`Cold + High pCO2` = dat %in% `Cold + High pCO2`,
         `Cold only` = dat %in% `Cold only`,
         `Warm + High pCO2` = dat %in% `Warm + High pCO2`,
         `Warm only` = dat %in% `Warm only`) %>%
  select(`Cold + High pCO2`, `Cold only`,
         `Warm + High pCO2`, `Warm only`) %>%
  euler()

eulerr:::plot.euler(venn.stressors.warm.cold, counts = T, quantities = T, 
                      legend = T, edges=F, fontsize=8,
                      font=1, cex=1, alpha=0.6,
                      fills=list(fill=c("deepskyblue1","navyblue",
                                        "orange","firebrick4")))

```
MAKE VENN FOR PRESENTATION SHOWING ONLY EFFECTS OF WARMING AND ACIDIFICATION

```{r}
venn.stressors.presentation <- data.frame(dat = 
             unique(c(`Warm Temperature`, `High pCO2`, `Warm + High pCO2`))) %>%
  mutate(`Warm Temperature` = dat %in% `Warm Temperature`,
         `High pCO2` = dat %in% `High pCO2`,
         `Warm + High pCO2` = dat %in% `Warm + High pCO2`) %>%
  select(`Warm Temperature`, `High pCO2`, `Warm + High pCO2`) %>%
  euler()

pdf(file="../results/deseq2/venn_presentation.pdf", height = 5.2, width = 6)
eulerr:::plot.euler(venn.stressors.presentation, counts = T, quantities = T, 
                      legend = T, edges=T, fontzie=8,
                      font=1, cex=1, alpha=0.65,
                      fills=list(fill=c(lighten("firebrick4", amount=0.55), "#FF955E", "firebrick4")))
dev.off()

```


###  DAVID enrichment analysis

It is a little clunky, but the DAVID functional analysis tool is one of the best ways to perform enrichment analysis (that I have found to date). To use, you copy a list of Uniprot IDs (or other gene identifier) to clipboard, then paste it into their tool. You then do the same for all genes in the analysis, providing an accurate background list of genes.  Here is code to copy all genes, then upregulated and downregulated DEGs identified by each contrast.

#### GENERATE LIST OF DEGS WITH ABS(L2FC)>0.5

```{r, include=F, echo=F, message=F}
names(degs)

# BACKGROUND - all genes submitted to DESeq2 analysis 
enrich.background <- res.pH.6 %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>% View()
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names, protein_names), by = "gene_gadmor") %>%
  dplyr::select(spid) %>% na.omit() %>% unlist() %>% as.vector()

# ------- "Effect of low pH"
print(paste("DEGs, ", names(degs[2])))
enrich.Amb.Low.6 <- degs[2] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  filter(abs(select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor")
#enrich.Amb.Low.6  %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# ------- "Effect of Cold Treatment"
print(paste("DEGs, ", names(degs[4])))
enrich.3.6.Amb <- degs[4] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  filter(abs(select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor")
#enrich.3.6.Amb %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# ------- "Effect of Warm Treatment"
print(paste("DEGs, ", names(degs[5])))
enrich.6.10.Amb <- degs[5] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  filter(abs(select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor")
#enrich.6.10.Amb %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# ------- "Effect of combined Cold + Low pH"
print(paste("DEGs, ", names(degs[8])))
enrich.both.3low.6amb <- degs[8] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  filter(abs(select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor")
#enrich.both.3low.6amb %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# ------- "Effect of combined Warm + Low pH"
print(paste("DEGs, ", names(degs[9])))
enrich.both.10low.6amb <- degs[9] %>% as.data.frame() %>% rownames_to_column("gene_gadmor") %>%
  filter(abs(select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names), by = "gene_gadmor")
#enrich.both.10low.6amb %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
```

#### FIGURE OUT UP/DOWN-REGULATED GENES 

Here I determine which LFC direction (>0 or <0) refers to higher/lower expression for treatments in each contrast

```{r}
test <- ((
  enrich.both.10low.6amb %>% 
#    filter(select(., contains("log2FoldChange"))>.5))$gene_gadmor)[1:200]
    filter(select(., contains("log2FoldChange"))<(-0.5)))$gene_gadmor)[1:100]


a <- c("10_low", "6_amb")
c <- data.frame(gene=character(), treatment=character(), mean=numeric())
d <- list()

for (i in 1:length(test)) {
b <-  plotCounts(dds.DESeq.treatment, gene=test[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  filter(treatment %in% a)

c[1:2,1] <- test[i]
c[1:2,2:3] <- summarySE(measurevar = "count", groupvars = "treatment", data = b)[,c("treatment", "count")] %>% 
  mutate(treatment=as.character(treatment))

d[[i]] <- c
}

# 
bind_rows(d, .id = "i") %>% mutate(treat.num=factor(treatment, levels = a, ordered = T) %>% as.numeric()) %>%
  ggplot(aes(x=treat.num, y=log(mean, base = 2), group=gene)) + geom_line(size=.01) +
  theme_minimal() + 
  scale_x_continuous(breaks=c(1,2), labels=a) +
  theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2))

# RESULTS
# enrich.Amb.Low.6 - LFC > 0.5 - LOWER IN LOW PH
# enrich.Amb.Low.6 - LFC < (-0.5) - HIGHER IN LOW PH
# enrich.3.6.Amb - LFC > 0.5 - HIGHER IN 3C
# enrich.3.6.Amb - LFC < (-0.5) - LOWER IN 3C 
# enrich.6.10.Amb - LFC > 0.5 - LOWER IN 10C
# enrich.6.10.Amb - LFC < (-0.5) - HIGHER IN 10C
# enrich.both.3low.6amb - LFC > 0.5 - HIGHER IN 3C-LOW PH
# enrich.both.3low.6amb - LFC < (-0.5) - LOWER IN 3C-LOW PH
# enrich.both.10low.6amb - LFC > 0.5 - LOWER IN 10C-LOW PH
# enrich.both.10low.6amb - LFC < (-0.5) - HIGHER IN 10C-LOW PH

# Others not used in analysis
# enrich.3.10.Amb - LFC > 0.5 - LOWER IN 10C
# enrich.3.10.Amb - LFC < (-0.5) - HIGHER IN 10C
# enrich.3.6.Low - LFC > 0.5 - HIGHER IN 3C
# enrich.3.6.Low - LFC < (-0.5) - LOWER IN 3C
# enrich.3.10.Low - LFC > 0.5 - LOWER IN 10C
# enrich.3.10.Low - LFC < (-0.5) - HIGHER IN 10C
# enrich.6.10.Low - LFC > 0.5 - LOWER IN 10C
# enrich.6.10.Low - LFC < (-0.5) - HIGHER IN 10C
```

Generate .tab file with lists of DEGs that are upregulated/downregulated in response to each treatment for DAVID enrichment analysis

```{r}
DEGs.4David <- list(
  
  "OA_Downregulated" = 
    # enrich.Amb.Low.6 - LFC>0.5 - LOWER IN LOW PH - Downregulated in OA
    enrich.Amb.Low.6 %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "OA_Upregulated" =  
    # enrich.Amb.Low.6 - LFC<0.5 - HIGHER IN LOW PH - Upregulated in OA
    enrich.Amb.Low.6 %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(), 

  "Cold_Upregulated" = 
    # enrich.3.6.Amb - LFC>0.5 - HIGHER IN 3C - Upregulated in Cold Temperature
    enrich.3.6.Amb %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "Cold_Downregulated"  =
    # enrich.3.6.Amb - LFC<0.5 - LOWER IN 3C - Downregulated in Cold Temperature
    enrich.3.6.Amb %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "Warm_Downregulated" = 
    # enrich.6.10.Amb - LFC>0.5 - LOWER IN 10C - Downregulated in Warm Temperature
    enrich.6.10.Amb %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "Warm_Upregulated" = 
    # enrich.6.10.Amb - LFC<0.5 - HIGHER IN 10C - Upregulated in Warm Temperature
    enrich.6.10.Amb %>%
     filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "Cold-OA_Upregulated" = 
    # enrich.both.3low.6amb - LFC>0.5 - HIGHER IN 3C-LOW PH - Upregulated in Cold+OA
    enrich.both.3low.6amb %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "Cold-OA_Downregulated" = 
    # enrich.both.3low.6amb - LFC<0.5 - LOWER IN 3C-LOW PH - Downregulated in Cold+OA
    enrich.both.3low.6amb %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "Warm-OA_Upregualted" = 
    # enrich.both.10low.6amb - lfc<0.5 - HIGHER IN 10C-LOW PH - Upregulated in Warm+OA
    enrich.both.10low.6amb %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "Warm-OA_Downregulated" = 
    # enrich.both.10low.6amb - lfc>0.5 - LOWER IN 10C-LOW PH - Downregulated in Warm+OA
    enrich.both.10low.6amb %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector()
  ) 

# Write tab-delimited file with each column containing Uniprot IDs for each set of DEGs 
write_delim(x = ldply(DEGs.4David, rbind) %>% t() %>% as.data.frame(), delim = "\t", col_names = F, na="", 
            file = "../results/GO-enrichment/DEGs-4David-final.tab")
            
# BACKGROUND - all genes submitted to DESeq2 analysis 
enrich.background %>% write_clip(allow_non_interactive = TRUE)
```


REDO with Entrez Gene IDs (i.e. NCBI gene IDs)

```{r}
# BACKGROUND - all genes submitted to DESeq2 analysis
enrich.background.ncbi <- res.pH.6 %>% as.data.frame() %>% rownames_to_column("gene") %>% left_join(gadmor.gff %>% filter(type=="gene")) %>%
  dplyr::select(ncbi_id) %>% na.omit() %>% unlist() %>% as.vector()

enrich.Amb.Low.6.ncbi <- degs[2] %>% as.data.frame() %>% rownames_to_column("gene") %>%
  filter(abs(dplyr::select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(gadmor.gff %>% filter(type=="gene"))

enrich.3.6.Amb.ncbi <- degs[4] %>% as.data.frame() %>% rownames_to_column("gene") %>%
  filter(abs(dplyr::select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(gadmor.gff %>% filter(type=="gene"))

enrich.6.10.Amb.ncbi <- degs[5] %>% as.data.frame() %>% rownames_to_column("gene") %>%
  filter(abs(dplyr::select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(gadmor.gff %>% filter(type=="gene"))

enrich.both.3low.6amb.ncbi <- degs[8] %>% as.data.frame() %>% rownames_to_column("gene") %>%
  filter(abs(dplyr::select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(gadmor.gff %>% filter(type=="gene"))

enrich.both.10low.6amb.ncbi <- degs[9] %>% as.data.frame() %>% rownames_to_column("gene") %>%
  filter(abs(dplyr::select(., contains("log2FoldChange")))>0.5) %>% #filter for genes with abs(L2FC)>0.5
  left_join(gadmor.gff %>% filter(type=="gene"))

DEGs.4David.ncbi <- list(

  "OA_Downregulated" =
    # enrich.Amb.Low.6 - LFC>0.5 - LOWER IN LOW PH - Downregulated in OA
    enrich.Amb.Low.6.ncbi %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(ncbi_id) %>%  na.omit() %>% unlist() %>% as.vector(),

  "OA_Upregulated" =
    # enrich.Amb.Low.6 - LFC<0.5 - HIGHER IN LOW PH - Upregulated in OA
    enrich.Amb.Low.6.ncbi %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>%
    dplyr::select(ncbi_id) %>%  na.omit() %>% unlist() %>% as.vector(),

  "Cold_Upregulated" =
    # enrich.3.6.Amb - LFC>0.5 - HIGHER IN 3C - Upregulated in Cold Temperature
    enrich.3.6.Amb.ncbi %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(ncbi_id) %>%  na.omit() %>% unlist() %>% as.vector(),

  "Cold_Downregulated"  =
    # enrich.3.6.Amb - LFC<0.5 - LOWER IN 3C - Downregulated in Cold Temperature
    enrich.3.6.Amb.ncbi %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>%
    dplyr::select(ncbi_id) %>%  na.omit() %>% unlist() %>% as.vector(),

  "Warm_Downregulated" =
    # enrich.6.10.Amb - LFC>0.5 - LOWER IN 10C - Downregulated in Warm Temperature
    enrich.6.10.Amb.ncbi %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(ncbi_id) %>%  na.omit() %>% unlist() %>% as.vector(),

  "Warm_Upregulated" =
    # enrich.6.10.Amb - LFC<0.5 - HIGHER IN 10C - Upregulated in Warm Temperature
    enrich.6.10.Amb.ncbi %>%
     filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>%
    dplyr::select(ncbi_id) %>%  na.omit() %>% unlist() %>% as.vector(),

  "Cold-OA_Upregulated" =
    # enrich.both.3low.6amb - LFC>0.5 - HIGHER IN 3C-LOW PH - Upregulated in Cold+OA
    enrich.both.3low.6amb.ncbi %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(ncbi_id) %>%  na.omit() %>% unlist() %>% as.vector(),

  "Cold-OA_Downregulated" =
    # enrich.both.3low.6amb - LFC<0.5 - LOWER IN 3C-LOW PH - Downregulated in Cold+OA
    enrich.both.3low.6amb.ncbi %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>%
    dplyr::select(ncbi_id) %>%  na.omit() %>% unlist() %>% as.vector(),

  "Warm-OA_Upregualted" =
    # enrich.both.10low.6amb - lfc<0.5 - HIGHER IN 10C-LOW PH - Upregulated in Warm+OA
    enrich.both.10low.6amb.ncbi %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>%
    dplyr::select(ncbi_id) %>%  na.omit() %>% unlist() %>% as.vector(),

  "Warm-OA_Downregulated" =
    # enrich.both.10low.6amb - lfc>0.5 - LOWER IN 10C-LOW PH - Downregulated in Warm+OA
    enrich.both.10low.6amb.ncbi %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    dplyr::select(ncbi_id) %>%  na.omit() %>% unlist() %>% as.vector()
  )

# Write tab-delimited file with each column containing Uniprot IDs for each set of DEGs
write_delim(x = ldply(DEGs.4David.ncbi, rbind) %>% t() %>% as.data.frame(), delim = "\t", col_names = F, na="",
            file = "../results/GO-enrichment/DEGs-4David-ncbi.txt")

# BACKGROUND - all genes submitted to DESeq2 analysis
enrich.background.ncbi %>% write_clip(allow_non_interactive = TRUE)
```



NOTE: outside of R I created a .tab file with the first column containing file names for enrichment results, and second column containing the contrast + which treatment was upregulated (for grouping in R). 
Example: the following row refers to the file contains enrichment results for genes upregulated in 10C in the contrast 6C vs. 10C at Low pH

`GO-enrich_6Cvs10C-Low-pH_up-in-10.txt	6Cvs10C-Low-pH_10`

```{r}
#### Read in enriched biological functions for all contrasts

filenames <- read_delim(file="../results/GO-enrichment/Final-DAVID/GO_filenames.txt", delim = "\t", col_names = c("filename", "gene_set"))
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=paste("../results/GO-enrichment/Final-DAVID/", filenames[i,"filename"], sep=""), delim = "\t"))}


david <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term", "process"), sep = "~") %>%
    separate(gene_set, sep = "_", into = c("stress", "response"), remove = F) %>%
    mutate(stress=factor(stress, ordered = TRUE, levels=c("OA","Warm", "Warm-OA", "Cold","Cold-OA")),
           response=as.factor(response)) %>% 
  tidyr::complete(stress, response, category) # add rows for gene_sets that are missing from dataframe

save(david, file="../results/GO-enrichment/Final-DAVID/david")
```

Copy DAVID results (Uniprot Kewords and Gene Ontology terms) to paste into Google Spreadsheet 

```{r}
# Uniprot KW
david %>% 
  filter(p_value<0.05) %>%
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
  arrange(response, stress, p_value) %>%
  mutate_at(c("p_value", "fdr"), funs(signif(.,2))) %>%
  dplyr::select(response, stress, term, process, count, p_value, fdr, genes) %>%
  write_clip()

# GO terms
david %>% 
  filter(p_value<0.05) %>%
  filter(category=="GOTERM_BP_DIRECT") %>%
  arrange(response, stress, p_value) %>%
  mutate_at(c("p_value", "fdr"), funs(signif(.,2))) %>%
  dplyr::select(response, stress, term, process, count, p_value, fdr, genes) %>%
  write_clip()
```

Up-regulated Uniprot Keywords

```{r}
# Bubble plot, Upregulated processes

pdf(file="../results/GO-enrichment/Final-DAVID/Upregulated-KW-bubble.pdf", height = 3.05, width = 4)

david %>% 
  #filter(count>=3) %>% 
  filter(p_value<0.05) %>%
#  filter(category=="GOTERM_BP_DIRECT") %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(stress, response, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david %>% 
              dplyr::select(stress, response, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("stress", "response", "genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%
  filter(response=="Upregulated") %>% 
  filter(process%!in%c("DNA integration", "DNA recombination")) %>%

  ungroup() %>% arrange(stress, desc(fold_enrichment)) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=fold_enrichment)) +  #aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4", 
                              "Cold"=lighten("navyblue", 0.45),
                              "Cold-OA"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Upregulated Biological Processes")
dev.off()
```

Up-regulated Gene Ontology terms

```{r}
# Bubble plot, Upregulated processes
pdf(file="../results/GO-enrichment/Final-DAVID/Upregulated-GO-bubble.pdf", height = 9, width = 7.5)
david %>% 
#  filter(fdr<0.2) %>% 
  filter(p_value<0.01) %>% filter(count>=3) %>%
  filter(category=="GOTERM_BP_DIRECT") %>% 
#  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
  filter(process %!in% c("DNA integration", "DNA recombination")) %>%

  group_by(stress, response, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david %>% 
              dplyr::select(stress, response, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("stress", "response", "genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%
  filter(response=="Upregulated") %>% 

  ungroup() %>% arrange(stress, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4", 
                              "Cold"=lighten("navyblue", 0.45),
                              "Cold-OA"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), breaks = c(2,4,6,8),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Lower gene expression")
```

Down-regulated Uniprot Keywords

```{r}
# Bubble plot, downregulated processes
pdf(file="../results/GO-enrichment/Final-DAVID/Downregulated-KW-bubble.pdf", height = 1.75, width = 2.8)
david %>% 
  #filter(count>=3) %>% 
  filter(p_value<0.05) %>%
#  filter(category=="GOTERM_BP_DIRECT") %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(stress, response, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david %>% 
              dplyr::select(stress, response, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("stress", "response", "genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%
  filter(response=="Downregulated") %>% 

  ungroup() %>% arrange(stress, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>% View()

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4", 
                              "Cold"=lighten("navyblue", 0.45),
                              "Cold-OA"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "none", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Downregulated Biological Processes")
dev.off()
```
Down-regulated Gene Ontology terms

```{r}
# Bubble plot, downregulated processes
pdf(file="../results/GO-enrichment/Final-DAVID/Downregulated-GO-bubble.pdf", height = 4, width = 5.75)
david %>% 
  filter(count>=3) %>% 
  filter(p_value<0.05) %>%
  filter(category=="GOTERM_BP_DIRECT") %>%
#  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(stress, response, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david %>% 
              dplyr::select(stress, response, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("stress", "response", "genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%
  filter(response=="Downregulated") %>% 

  ungroup() %>% arrange(stress, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4", 
                              "Cold"=lighten("navyblue", 0.45),
                              "Cold-OA"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Downregulated Biological Processes\n(GO terms)")
dev.off()
```
MAKE UNIPROT BUBBLE PLOTS FOR PRESENTATION - ONLY WARMING AND ACIDIFICATION TREATMENTS

```{r}
david %>% 
  filter(p_value<0.05) %>%
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(stress, response, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david %>% 
              dplyr::select(stress, response, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("stress", "response", "genes", "p_value", "count")) %>%  #re-add data
  filter(response=="Upregulated") %>% 

  ungroup() %>% arrange(stress, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  filter(stress %!in% c("Cold", "Cold-OA")) %>%
  mutate(stress = factor(stress, c("Warm", "Warm-OA", "OA"), ordered = TRUE)) %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Upregulated Biological Processes")



david %>% 
  filter(p_value<0.05) %>%
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
  group_by(stress, response, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david %>% 
              dplyr::select(stress, response, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("stress", "response", "genes", "p_value", "count")) %>%  #re-add data
  filter(response=="Downregulated") %>% 
  ungroup() %>% arrange(stress, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  filter(stress %!in% c("Cold", "Cold-OA")) %>%
  mutate(stress = factor(stress, c("Warm", "Warm-OA", "OA"), ordered = TRUE)) %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4"),
             guide = guide_legend(order = 1, override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        #legend.position = "none", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Downregulated Biological Processes")
```









Enriched GO biological processes - Network Diagrams using metacoder

Here I prep R objects from GO enrichment results to visualize in network plots - the visualization is done in RStudio Cloud, as it requires a specific older version of R and some R packages. 

```{r}
filenames <- read_delim(file="../results/GO-enrichment/GO_filenames_upregulated.txt", delim = "\t", col_names = c("filename", "gene_set"))
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=paste("../results/GO-enrichment/", filenames[i,"filename"], sep=""), delim = "\t"))}

enriched.bp.degs.network <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("go", "process"), sep = "~") %>%
    separate(gene_set, sep = "_", into = c("contrast", "upregulated"), remove = F) %>%
    mutate(contrast=factor(contrast, ordered = TRUE,
                           levels=c("AmbvsLow.3C", "AmbvsLow.6C", "AmbvsLow.10C", 
                                    "3vs6.Amb-pH", "3vs6.Low-pH",
                                    "3vs10.Amb-pH", "3vs10.Low-pH",
                                    "6vs10.Amb-pH", "6vs10.Low-pH"))) %>% #Convert gene_set column to factor, define all gene_sets levels
  #tidyr::complete(contrast, upregulated, category) %>% # add rows for gene_sets that are missing from dataframe
  mutate(upregulated=factor(upregulated, levels = c("3C", "6C", "10C", "Ambient-pH", "Low-pH"), ordered = T)) %>%
# 
#   # remove upregulated treatment that doesn't exist in given contrast
#   filter(!( (contrast=="AmbvsLow.3C" | contrast=="AmbvsLow.6C"| contrast=="AmbvsLow.10C") & (upregulated=="3C" | upregulated=="6C" | upregulated=="10C"))) %>%
#   filter(!( (contrast=="3vs6.Amb-pH" | contrast=="3vs10.Amb-pH" | contrast=="6vs10.Amb-pH" | contrast=="3vs6.Low-pH" | contrast=="3vs10.Low-pH" | contrast=="6vs10.Low-pH") & (upregulated=="Ambient-pH" | upregulated=="Low-pH"))) %>%
#   filter(!( (contrast=="3vs6.Amb-pH" | contrast=="3vs6.Low-pH") & upregulated=="10C")) %>%
#   filter(!( (contrast=="3vs10.Amb-pH" | contrast=="3vs10.Low-pH") & upregulated=="6C")) %>%
#   filter(!( (contrast=="6vs10.Amb-pH" | contrast=="6vs10.Low-pH") & upregulated=="3C")) %>%
  
  separate(contrast, sep="\\.", into = c("comparison", "background"), remove = FALSE) %>% select(-comparison) %>%
  mutate(background=factor(background, ordered = T, levels=c("Low-pH", "Amb-pH", "3C", "6C", "10C")))

# Save R object to then import to RStudio Cloud for use in metacoder 
save(enriched.bp.degs.network, file="../results/GO-enrichment/enriched.bp.degs.network")

enriched.bp.degs.network
enriched.bp.linear.nonlinear_all
```

Which biological processes are enriched in genes associated with fish length? 

```{r}
# Copy Uniprot IDs for length-associated genes
gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, spid, protein_names) %>%
  filter(gene_gadmor %in% genes.length) %>% View()
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# BACKGROUND - all genes submitted to DESeq2 analysis 
enrich.background %>% write_clip(allow_non_interactive = TRUE)

david.length <- data.frame(read_delim(file="../results/GO-enrichment/Final-DAVID/Length-associated-genes.txt", delim = "\t")) %>% 
  clean_names() %>% 
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term", "process"), sep = "~")

# Look at enriched Uniprot Keywords or Gene Ontology terms (biological processes)
david.length %>% 
  #filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
  filter(category=="GOTERM_BP_DIRECT", count>=3) %>%  
  arrange(p_value) %>%
  mutate_at(c("p_value", "fdr"), funs(signif(.,2))) %>%
  dplyr::select(term, process, count, p_value, fdr, genes) %>%
  write_clip()
```

```{r}
david %>% 
  filter(gene_set%in%c("Cold_Downregulated", "Cold_Upregulated")) %>%
  View()
```




# Lots of old bubble plot code 

### Bubble plot of enriched GO term Biological Processes

single stressor vs. combined stressor, DOWNREGULATED 

```{r}
pdf(file="../results/GO-enrichment/GO-enriched-BP-DOWNREGULATED-bubble.pdf", height = 7, width = 7)

enriched.degs.upregulated %>% filter(count>=3) %>% 
#  filter(fdr<0.1) %>%
  filter(p_value<0.05) %>%
  filter(category=="GOTERM_BP_DIRECT") %>% 
  #filter(category=="GOTERM_MF_DIRECT") %>%

  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.degs.upregulated %>% select(contrast, upregulated, category, background, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>% #dplyr::slice(1:15) %>%

  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.6C" ~ "High pCO2",
      x == "3vs6.Amb-pH" ~ "Cold Temperature",
      x == "6vs10.Amb-pH" ~ "Warm Temperature",
      x == "6ambvs3low" ~ "Cold Temperature + High pCO2", 
      x == "6ambvs10low" ~ "Warm Temperature + High pCO2")}) %>%
  mutate(group=factor(group, ordered=TRUE, 
                      levels=c("High pCO2", "Warm Temperature", "Warm Temperature + High pCO2",
                               "Cold Temperature", "Cold Temperature + High pCO2"))) %>% 

  mutate(upregulated=factor(upregulated, ordered = TRUE)) %>% 
  
  # Add column indicating that all responses upregulated in 6amb are downregulated in the stressors 
  mutate(Response=case_when(upregulated == "6amb" ~ "Downregulated", TRUE ~ "Upregulated")) %>%

  filter(upregulated=="6amb") %>% # Filter for only genes that were upregulated in optimal conditions (6amb) aka downregulated in stressor

  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
#  ungroup() %>% arrange(term) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=group, col=group)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("High pCO2"=lighten("darkorange2", 0.25), 
                              "Warm Temperature"=lighten("firebrick4", 0.45), 
                              "Warm Temperature + High pCO2"="firebrick4", 
                              "Cold Temperature"=lighten("navyblue", 0.45),
                              "Cold Temperature + High pCO2"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Downregulated Biological Processes\n(GO terms)") +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) #+
  coord_cartesian(clip = "off", ylim = c(-3, 53)) +
  annotate("text", x = 1, y = -1.3, label = "OA", size=2.5, col="gray15") +
  annotate("text", x = 2, y = -1.3, label = "Warm", size=2.5, col="gray15") +
  annotate("text", x = 3, y = -1.3, label = "Warm +\nOA", size=2.5, col="gray15") +
  annotate("text", x = 4, y = -1.3, label = "Cold", size=2.5, col="gray15") +
  annotate("text", x = 5, y = -1.3, label = "Cold +\nOA", size=2.5, col="gray15") +

  annotate("text", x = 3, y = 52, label = "Downregulated Biological Processes\n(GO terms)", size=3, col="gray15")
dev.off()
```

single stressor vs. combined stressor, UPREGULATED 

```{r}
pdf(file="../results/GO-enrichment/GO-enriched-BP-UPREGULATED-bubble.pdf", height = 7, width = 7)

enriched.degs.upregulated %>% filter(count>=3) %>% 
#  filter(fdr<0.1) %>%
  filter(p_value<0.01) %>%
  filter(category=="GOTERM_BP_DIRECT") %>% 
  #filter(category=="GOTERM_MF_DIRECT") %>%

  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.degs.upregulated %>% select(contrast, upregulated, category, background, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>% #dplyr::slice(1:15) %>%

  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.6C" ~ "High pCO2",
      x == "3vs6.Amb-pH" ~ "Cold Temperature",
      x == "6vs10.Amb-pH" ~ "Warm Temperature",
      x == "6ambvs3low" ~ "Cold Temperature + High pCO2", 
      x == "6ambvs10low" ~ "Warm Temperature + High pCO2")}) %>%
  mutate(group=factor(group, ordered=TRUE, 
                      levels=c("High pCO2", "Warm Temperature", "Warm Temperature + High pCO2",
                               "Cold Temperature", "Cold Temperature + High pCO2"))) %>% 

  mutate(upregulated=factor(upregulated, ordered = TRUE)) %>% #, levels = c("Low-pH", "Ambient-pH")
  
  # Add column indicating that all responses upregulated in 6amb are downregulated in the stressors 
  mutate(Response=case_when(upregulated == "6amb" ~ "Downregulated", TRUE ~ "Upregulated")) %>%

  filter(upregulated !="6amb") %>% # Filter for only genes that were NOT upregulated in optimal conditions (6amb) aka upregulated in stressor

  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
#  ungroup() %>% arrange(term) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=group, col=group)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("High pCO2"=lighten("darkorange2", 0.25), 
                              "Warm Temperature"=lighten("firebrick4", 0.45), 
                              "Warm Temperature + High pCO2"="firebrick4", 
                              "Cold Temperature"=lighten("navyblue", 0.45),
                              "Cold Temperature + High pCO2"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Upregulated Biological Processes\n(GO terms)") +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) #+
  coord_cartesian(clip = "off", ylim = c(-3, 53)) +
  annotate("text", x = 1, y = -1.3, label = "OA", size=2.5, col="gray15") +
  annotate("text", x = 2, y = -1.3, label = "Warm", size=2.5, col="gray15") +
  annotate("text", x = 3, y = -1.3, label = "Warm +\nOA", size=2.5, col="gray15") +
  annotate("text", x = 4, y = -1.3, label = "Cold", size=2.5, col="gray15") +
  annotate("text", x = 5, y = -1.3, label = "Cold +\nOA", size=2.5, col="gray15") +

  annotate("text", x = 3, y = 52, label = "Upregulated Biological Processes\n(GO terms)", size=3, col="gray15")
dev.off()
```

```{r}
enriched.degs.upregulated %>% filter(count>=3) %>% 
#  filter(fdr<0.1) %>%
  filter(p_value<0.05) %>%
  filter(category=="GOTERM_BP_DIRECT") %>% View()
  filter(contrast=="AmbvsLow.6C", upregulated=="6low")
```


### Bubble plot of enriched Uniprot Keyword Biological Processes

single stressor vs. combined stressor, up & down regulated

```{r}
pdf(file="../results/GO-enrichment/Enriched-Uniprot-Keywords_single-vs-double-stressors_Bubble-plot.pdf", height = 8, width = 4.25)

enriched.degs.upregulated %>% filter(count>=3) %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
  #filter(category=="UP_KW_CELLULAR_COMPONENT") %>%

  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.degs.upregulated %>% select(contrast, upregulated, category, background, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>%
  #dplyr::slice(1:15) %>%

  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.6C" ~ "High pCO2",
      x == "3vs6.Amb-pH" ~ "Cold Temperature",
      x == "6vs10.Amb-pH" ~ "Warm Temperature",
      x == "6ambvs3low" ~ "Cold Temperature + High pCO2", 
      x == "6ambvs10low" ~ "Warm Temperature + High pCO2")}) %>%
  mutate(group=factor(group, ordered=TRUE, 
                      levels=c("Cold Temperature", "Cold Temperature + High pCO2",
                               "Warm Temperature", "Warm Temperature + High pCO2",
                               "High pCO2"))) %>% 

  mutate(upregulated=factor(upregulated, ordered = TRUE)) %>% #, levels = c("Low-pH", "Ambient-pH")

  # Add column indicating that all responses upregulated in 6amb are downregulated in the stressors 
  mutate(Response=case_when(upregulated == "6amb" ~ "Downregulated", TRUE ~ "Upregulated")) %>%

  #ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
  ungroup() %>% arrange(upregulated, term) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=contrast, col=Response)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2, 
            labeller = as_labeller(c(`GOTERM_BP_DIRECT`="Enriched Uniprot Keywords\nBiological Processes"))) +
 scale_color_manual(name="Response to Stressor",
                    values=c(`Upregulated`="dodgerblue4",
                             `Downregulated`="firebrick4"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) +
  coord_cartesian(clip = "off", ylim = c(-2, 53.5)) +
  ggtitle(NULL) +
  geom_vline(xintercept = c(1.5, 2.5, 3.5, 4.5), size=.1, color="gray75") +
  annotate("text", x = 1, y = -1, label = "OA", size=2.5, col="gray15") +
  annotate("text", x = 2, y = -1, label = "Cold", size=2.5, col="gray15") +
  annotate("text", x = 3, y = -1, label = "Cold\n+OA", size=2.5, col="gray15") +
  annotate("text", x = 4, y = -1, label = "Warm", size=2.5, col="gray15") +
  annotate("text", x = 5, y = -1, label = "Warm\n+OA", size=2.5, col="gray15") +
  ggtitle("Enriched Uniprot Keywords\nBiological Processes") #+
  # annotate("text", x = 3, y = 55, label = "Enriched Uniprot Keywords\nBiological Processes", size=3, col="gray15") #+
  # annotate("text", x = 2, y = 53.5, label = "Ambient vs. Low pH contrasts", size=3.25, col="gray15", fontface = 'bold') +
  # annotate("text", x = 2, y = 52, label = "by temperature", size=3, col="gray15", fontface = 'italic')
dev.off()
```

single stressor vs. combined stressor

DOWNREGULATED ONLY 

```{r}
pdf(file="../results/GO-enrichment/Enriched-Uniprot-Keywords_single-vs-double-stressors_Bubble-plot-DOWNREGULATED.pdf", height = 5.85, width = 4.8)

enriched.degs.upregulated %>% filter(count>=3) %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
  #filter(category=="UP_KW_CELLULAR_COMPONENT") %>%

  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.degs.upregulated %>% select(contrast, upregulated, category, background, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>%
  #dplyr::slice(1:15) %>%

  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.6C" ~ "High pCO2",
      x == "3vs6.Amb-pH" ~ "Cold Temperature",
      x == "6vs10.Amb-pH" ~ "Warm Temperature",
      x == "6ambvs3low" ~ "Cold Temperature + High pCO2", 
      x == "6ambvs10low" ~ "Warm Temperature + High pCO2")}) %>%
  mutate(group=factor(group, ordered=TRUE, 
                      levels=c("High pCO2", "Warm Temperature", "Warm Temperature + High pCO2",
                               "Cold Temperature", "Cold Temperature + High pCO2"))) %>% 

  mutate(upregulated=factor(upregulated, ordered = TRUE)) %>% #, levels = c("Low-pH", "Ambient-pH")

  # Add column indicating that all responses upregulated in 6amb are downregulated in the stressors 
  mutate(Response=case_when(upregulated == "6amb" ~ "Downregulated", TRUE ~ "Upregulated")) %>%

  filter(upregulated=="6amb") %>% # Filter for only genes that were upregulated in optimal conditions (6amb) aka downregulated in stressor

  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
  #ungroup() %>% arrange(upregulated, term) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  
ggplot(aes(y = process, x=group, col=group)) + 
 geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("High pCO2"=lighten("darkorange2", 0.25), 
                              "Warm Temperature"=lighten("firebrick4", 0.45), 
                              "Warm Temperature + High pCO2"="firebrick4", 
                              "Cold Temperature"=lighten("navyblue", 0.45),
                              "Cold Temperature + High pCO2"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) +
  guides(colour = guide_legend(override.aes = list(size=4, linetype="blank"), order=1)) +
  coord_cartesian(clip = "off", ylim = c(-2, 35.5)) +
  ggtitle(NULL) +
  geom_vline(xintercept = c(1.5, 2.5, 3.5, 4.5), size=.1, color="gray75") +
  annotate("text", x = 1, y = -1, label = "OA", size=2.5, col="gray15") +
  annotate("text", x = 2, y = -1, label = "Warm", size=2.5, col="gray15") +
  annotate("text", x = 3, y = -1, label = "Warm\n+OA", size=2.5, col="gray15") +
  annotate("text", x = 4, y = -1, label = "Cold", size=2.5, col="gray15") +
  annotate("text", x = 5, y = -1, label = "Cold\n+OA", size=2.5, col="gray15") +
  ggtitle("Downregulated Biological Processes\n(Enriched Uniprot Keywords)")
dev.off()
```

UPREGULATED ONLY 

```{r}
pdf(file="../results/GO-enrichment/Enriched-Uniprot-Keywords_single-vs-double-stressors_Bubble-plot-UPREGULATED.pdf", height = 4.4, width = 4.5)

enriched.degs.upregulated %>% filter(count>=3) %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>% View()
  #filter(category=="UP_KW_CELLULAR_COMPONENT") %>%

  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.degs.upregulated %>% select(contrast, upregulated, category, background, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>%
  #dplyr::slice(1:15) %>%

  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.6C" ~ "High pCO2",
      x == "3vs6.Amb-pH" ~ "Cold Temperature",
      x == "6vs10.Amb-pH" ~ "Warm Temperature",
      x == "6ambvs3low" ~ "Cold Temperature + High pCO2", 
      x == "6ambvs10low" ~ "Warm Temperature + High pCO2")}) %>%
  mutate(group=factor(group, ordered=TRUE, 
                      levels=c("High pCO2", "Warm Temperature", "Warm Temperature + High pCO2",
                               "Cold Temperature", "Cold Temperature + High pCO2"))) %>% 

  mutate(upregulated=factor(upregulated, ordered = TRUE)) %>% #, levels = c("Low-pH", "Ambient-pH")

  # Add column indicating that all responses upregulated in 6amb are downregulated in the stressors 
  mutate(Response=case_when(upregulated == "6amb" ~ "Downregulated", TRUE ~ "Upregulated")) %>%

  filter(upregulated!="6amb") %>% # Filter for only genes that were upregulated in optimal conditions (6amb) aka downregulated in stressor

  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
  #ungroup() %>% arrange(upregulated, term) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  
ggplot(aes(y = process, x=group, 
           #col=Response,
           col=group)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("High pCO2"=lighten("darkorange2", 0.25), 
                              "Warm Temperature"=lighten("firebrick4", 0.45), 
                              "Warm Temperature + High pCO2"="firebrick4", 
                              "Cold Temperature"=lighten("navyblue", 0.45),
                              "Cold Temperature + High pCO2"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) +
  coord_cartesian(clip = "off", ylim = c(-2, 24.5)) +
  ggtitle(NULL) +
  geom_vline(xintercept = c(1.5, 2.5, 3.5, 4.5), size=.1, color="gray75") +
  annotate("text", x = 1, y = -1, label = "OA", size=2.5, col="gray15") +
  annotate("text", x = 2, y = -1, label = "Warm", size=2.5, col="gray15") +
  annotate("text", x = 3, y = -1, label = "Warm\n+OA", size=2.5, col="gray15") +
  annotate("text", x = 4, y = -1, label = "Cold", size=2.5, col="gray15") +
  annotate("text", x = 5, y = -1, label = "Cold\n+OA", size=2.5, col="gray15") +
  ggtitle("Upregulated Biological Processes\n(Enriched Uniprot Keywords)")
dev.off()
```

Enriched GO biological processes - Network Diagrams using metacoder

Here I save an R objects from GO enrichment results of the DEGs in response to single and combined stressors to visualize in network plots - the visualization is done in RStudio Cloud, as it requires a specific older version of R and some R packages. 

```{r}
# Save R object to then import to RStudio Cloud for use in metacoder 
save(enriched.degs.upregulated, file="../results/GO-enrichment/enriched.degs.upregulated")
```

```{r}
enriched.degs.upregulated %>% filter(category=="GOTERM_BP_DIRECT") %>% filter(count>=3, p_value<0.05) %>%
  filter(contrast %in% c("AmbvsLow.6C", "6vs10.Amb-pH")) %>% 
#  group_by(contrast, upregulated) %>% tally() 
# filter(contrast=="6vs10.Amb-pH", upregulated=="10amb") %>% arrange(p_value) #upregulated in warming
# filter(contrast=="6vs10.Amb-pH", upregulated=="6amb") %>% arrange(p_value)  #downregulated in warming
#  filter(contrast=="AmbvsLow.6C", upregulated=="6low") %>% arrange(p_value) #upregulated in OA
  filter(contrast=="AmbvsLow.6C", upregulated=="6amb") %>% arrange(p_value) #downregulated in OA
```

#### Temperature contrasts for each pH 

```{r}
pdf(file="../results/GO-enrichment/BP-temperature-upregulated-enriched-bubble.pdf", height = 16, width = 8)

enriched.bp.degs.upregulated %>% 
  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.degs.upregulated %>% select(contrast, upregulated, category, background, go, process, count, percent, p_value, fold_enrichment, fdr, genes), by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>% dplyr::slice(1:15) %>%
  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.3C" ~ "pH",
      x == "AmbvsLow.6C" ~ "pH",
      x == "AmbvsLow.10C" ~ "pH",
      x == "3vs6.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "6vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs6.Low-pH" ~ "Temp, Low pH",
      x == "3vs10.Low-pH" ~ "Temp, Low pH",
      x == "6vs10.Low-pH" ~ "Temp, Low pH")}) %>%
  mutate(group=factor(group, ordered=TRUE, levels=c("pH", "Temp, Amb pH", "Temp, Low pH"))) %>% 
  filter(background %in% c("Low-pH","Amb-pH")) %>%
  #mutate(upregulated=factor(upregulated, ordered = TRUE, levels = c("Low-pH", "Ambient-pH"))) %>%
  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  mutate(upreg_ph=factor(paste(upregulated, background, sep = " "))) %>%
  
ggplot(aes(y = process, x=contrast, col=upreg_ph)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2, labeller = as_labeller(c(`GOTERM_BP_DIRECT`="Enriched Biological Processes\nAmbient vs. Low pH\nby temperature"))) +
  scale_color_manual(name="Upregulated",
                     values=c("3C Amb-pH"= "navyblue", "3C Low-pH"="royalblue1",
                              "6C Amb-pH"="darkgreen", "6C Low-pH"="yellow3",
                              "10C Amb-pH"="firebrick4", "10C Low-pH"="sienna2"),
                     guide = guide_legend(override.aes = list(size=4))) +
  #scale_color_manual(guide="none", values=c("#ffffb3","#8dd3c7","#bebada")) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) +
  coord_cartesian(clip = "off", ylim = c(-6, 108)) +
  ggtitle(NULL) +
  annotate("text", x = 1, y = -1.5, label = "Amb", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 2, y = -1.5, label = "Low", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 3, y = -1.5, label = "Amb", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 4, y = -1.5, label = "Low", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 5, y = -1.5, label = "Amb", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 6, y = -1.5, label = "Low", size=2.75, col="gray15", angle=90) +

  annotate("text", x = 1.5, y = -5, label = "3C vs. 6C", size=2.75, col="gray15", fontface = 'bold') +
  annotate("text", x = 3.5, y = -5, label = "3C vs. 10C", size=2.75, col="gray15", fontface = 'bold') +
  annotate("text", x = 5.5, y = -5, label = "6C vs. 10C", size=2.75, col="gray15", fontface = 'bold') +  
  
  annotate("text", x = 3.5, y = 108, label = "Enriched Biological Processes", size=3, col="gray15") +
  annotate("text", x = 3.5, y = 106.5, label = "Temperature contrasts", size=3.25, col="gray15", fontface = 'bold') +
  annotate("text", x = 3.5, y = 105, label = "by pH", size=3, col="gray15", fontface = 'italic') +
  geom_segment(aes(x = 2.5, y = -6, xend = 2.5, yend = 104), linetype="dashed", color = "gray65", size=0.75) +
  geom_segment(aes(x = 4.5, y = -6, xend = 4.5, yend = 104), linetype="dashed", color = "gray65", size=0.75)
dev.off()
```

Wow- that bubble plot showing enriched bio. functions among temperature contrasts is huge. Break it up into 3 diff. temperature plots? 

```{r}
# 3C vs. 6C contrasts 
pdf(file="../results/GO-enrichment/BP-3vs6-upregulated-enriched-bubble.pdf", height = 9, width = 6.5)

enriched.bp.degs.upregulated %>% 
  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.degs.upregulated %>% select(contrast, upregulated, category, background, go, process, count, percent, p_value, fold_enrichment, fdr, genes), by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>% dplyr::slice(1:15) %>%
  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.3C" ~ "pH",
      x == "AmbvsLow.6C" ~ "pH",
      x == "AmbvsLow.10C" ~ "pH",
      x == "3vs6.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "6vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs6.Low-pH" ~ "Temp, Low pH",
      x == "3vs10.Low-pH" ~ "Temp, Low pH",
      x == "6vs10.Low-pH" ~ "Temp, Low pH")}) %>%
  mutate(group=factor(group, ordered=TRUE, levels=c("pH", "Temp, Amb pH", "Temp, Low pH"))) %>% 
  filter(background %in% c("Low-pH","Amb-pH")) %>%
  #mutate(upregulated=factor(upregulated, ordered = TRUE, levels = c("Low-pH", "Ambient-pH"))) %>%
  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  mutate(upreg_ph=factor(paste(upregulated, background, sep = " "))) %>%

  filter(grepl("3", contrast)) %>% filter(grepl("6", contrast)) %>%  #filter for 3 vs 6 contrasts
  
ggplot(aes(y = process, x=contrast, col=upreg_ph)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Upregulated",
                     values=c("3C Amb-pH"= "navyblue", "3C Low-pH"="royalblue1",
                              "6C Amb-pH"="darkgreen", "6C Low-pH"="yellow3",
                              "10C Amb-pH"="firebrick4", "10C Low-pH"="sienna2"),
                     guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) +
  coord_cartesian(clip = "off", ylim = c(-2, 50)) +
  ggtitle(NULL) +
  annotate("text", x = 1, y = -.5, label = "Amb", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 2, y = -.5, label = "Low", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 1.5, y = -1.75, label = "3C vs. 6C", size=2.75, col="gray15", fontface = 'bold') +
  annotate("text", x = 1.5, y = 50, label = "Enriched Biological Processes", size=3, col="gray15") +
  annotate("text", x = 1.5, y = 48.5, label = "3C vs 6C by pH", size=3.25, col="gray15", fontface = 'bold')
dev.off()
```

```{r}
# 3C vs. 10C contrasts 
pdf(file="../results/GO-enrichment/BP-3vs10-upregulated-enriched-bubble.pdf", height = 9, width = 6.5)

enriched.bp.degs.upregulated %>% 
  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.degs.upregulated %>% select(contrast, upregulated, category, background, go, process, count, percent, p_value, fold_enrichment, fdr, genes), by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>% dplyr::slice(1:15) %>%
  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.3C" ~ "pH",
      x == "AmbvsLow.6C" ~ "pH",
      x == "AmbvsLow.10C" ~ "pH",
      x == "3vs6.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "6vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs6.Low-pH" ~ "Temp, Low pH",
      x == "3vs10.Low-pH" ~ "Temp, Low pH",
      x == "6vs10.Low-pH" ~ "Temp, Low pH")}) %>%
  mutate(group=factor(group, ordered=TRUE, levels=c("pH", "Temp, Amb pH", "Temp, Low pH"))) %>% 
  filter(background %in% c("Low-pH","Amb-pH")) %>%
  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  mutate(upreg_ph=factor(paste(upregulated, background, sep = " "))) %>%

  filter(grepl("3", contrast)) %>% filter(grepl("10", contrast)) %>%  #filter for 3 vs 10 contrasts
  
ggplot(aes(y = process, x=contrast, col=upreg_ph)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + 
 facet_wrap(~category,scales="free", nrow = 2) + 
  scale_color_manual(name="Upregulated",
                     values=c("3C Amb-pH"= "navyblue", "3C Low-pH"="royalblue1",
                              "6C Amb-pH"="darkgreen", "6C Low-pH"="yellow3",
                              "10C Amb-pH"="firebrick4", "10C Low-pH"="sienna2"),
                     guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), 
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) +
  coord_cartesian(clip = "off", ylim = c(-2, 45)) +
  ggtitle(NULL) +
  annotate("text", x = 1, y = -.5, label = "Amb", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 2, y = -.5, label = "Low", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 1.5, y = -1.75, label = "3C vs. 10C", size=2.75, col="gray15", fontface = 'bold') +
  annotate("text", x = 1.5, y = 45, label = "Enriched Biological Processes", size=3, col="gray15") +
  annotate("text", x = 1.5, y = 43.5, label = "3C vs 10C by pH", size=3.25, col="gray15", fontface = 'bold')
dev.off()
```


```{r}
# 6C vs. 10C contrasts 
pdf(file="../results/GO-enrichment/BP-6vs10-upregulated-enriched-bubble.pdf", height = 9, width = 6.5)

enriched.bp.degs.upregulated %>% 
  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.degs.upregulated %>% select(contrast, upregulated, category, background, go, process, count, percent, p_value, fold_enrichment, fdr, genes), by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>% dplyr::slice(1:15) %>%
  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.3C" ~ "pH",
      x == "AmbvsLow.6C" ~ "pH",
      x == "AmbvsLow.10C" ~ "pH",
      x == "3vs6.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "6vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs6.Low-pH" ~ "Temp, Low pH",
      x == "3vs10.Low-pH" ~ "Temp, Low pH",
      x == "6vs10.Low-pH" ~ "Temp, Low pH")}) %>%
  mutate(group=factor(group, ordered=TRUE, levels=c("pH", "Temp, Amb pH", "Temp, Low pH"))) %>% 
  filter(background %in% c("Low-pH","Amb-pH")) %>%
  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  mutate(upreg_ph=factor(paste(upregulated, background, sep = " "))) %>%

  filter(grepl("6", contrast)) %>% filter(grepl("10", contrast)) %>%  #filter for 6 vs 10 contrasts
  
ggplot(aes(y = process, x=contrast, col=upreg_ph)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) +
 facet_wrap(~category,scales="free", nrow = 2) + 
  scale_color_manual(name="Upregulated",
                     values=c("3C Amb-pH"= "navyblue", "3C Low-pH"="royalblue1",
                              "6C Amb-pH"="darkgreen", "6C Low-pH"="yellow3",
                              "10C Amb-pH"="firebrick4", "10C Low-pH"="sienna2"),
                     guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), 
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) +

  coord_cartesian(clip = "off", ylim = c(-2, 59.5)) + #set plot dimensions
  
  ggtitle(NULL) +
  annotate("text", x = 1, y = -.5, label = "Amb", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 2, y = -.5, label = "Low", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 1.5, y = -1.75, label = "6C vs. 10C", size=2.75, col="gray15", fontface = 'bold') +
  annotate("text", x = 1.5, y = 59.5, label = "Enriched Biological Processes", size=3, col="gray15") +
  annotate("text", x = 1.5, y = 58, label = "6C vs 10C by pH", size=3.25, col="gray15", fontface = 'bold')
dev.off()
```

#### Enriched Uniprot Keywords 

```{r}
enriched.results <- function(contrast1, upreg, david.category, GO.results, GO.type, alpha, ngenes) {
  
  print(GO.results %>%
      filter(category == david.category) %>% 
      filter(contrast %in% contrast1) %>% 
      filter(upregulated %in% upreg) %>%
      filter(count>=ngenes) %>%
      filter(fdr<alpha) %>% 
      select(contrast, upregulated, go, process, p_value) %>% 
      arrange(upregulated, p_value))
}

enriched.results(contrast1 = ,)

# Uniprot Keywords: "UP_KW_BIOLOGICAL_PROCESS"
# GO Biological Processes: "GOTERM_BP_DIRECT"

enriched.results(contrast1 = "AmbvsLow.6C", upreg = c("Ambient-pH", "Low-pH"), david.category = "UP_KW_BIOLOGICAL_PROCESS", GO.results = enriched.bp.degs.network, alpha = 0.1, ngenes=3) %>% write_clip()

enriched.results(contrast1 = "3vs6.Amb-pH", upreg = c("3C", "6C"), david.category = "UP_KW_BIOLOGICAL_PROCESS", GO.results = enriched.bp.degs.network, alpha = 0.1, ngenes=3) %>% write_clip()
enriched.results(contrast1 = "3vs6.Low-pH", upreg = c("3C", "6C"), david.category = "UP_KW_BIOLOGICAL_PROCESS", GO.results = enriched.bp.degs.network, alpha = 0.1, ngenes=3) %>% write_clip()

enriched.results(contrast1 = "3vs10.Amb-pH", upreg = c("3C", "10C"), david.category = "UP_KW_BIOLOGICAL_PROCESS", GO.results = enriched.bp.degs.network, alpha = 0.1, ngenes=3) %>% write_clip()
enriched.results(contrast1 = "3vs10.Low-pH", upreg = c("3C", "10C"), david.category = "UP_KW_BIOLOGICAL_PROCESS", GO.results = enriched.bp.degs.network, alpha = 0.1, ngenes=3) %>% write_clip()

enriched.results(contrast1 = "6vs10.Amb-pH", upreg = c("6C", "10C"), david.category = "UP_KW_BIOLOGICAL_PROCESS", GO.results = enriched.bp.degs.network, alpha = 0.1, ngenes=3) %>% write_clip()
enriched.results(contrast1 = "6vs10.Low-pH", upreg = c("6C", "10C"), david.category = "UP_KW_BIOLOGICAL_PROCESS", GO.results = enriched.bp.degs.network, alpha = 0.1, ngenes=3) %>% write_clip()


```

### Uniprot Keyword Bubble Plots

```{r}
#### Read in enriched Uniprot Keyword biological processes for all contrasts

filenames <- read_delim(file="../results/GO-enrichment/GO_filenames_upregulated.txt", delim = "\t", col_names = c("filename", "gene_set"))
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=paste("../results/GO-enrichment/", filenames[i,"filename"], sep=""), delim = "\t"))}

enriched.kw.degs.upregulated <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("kw_id", "kw"), sep = "~") %>%
    separate(gene_set, sep = "_", into = c("contrast", "upregulated"), remove = F) %>%
    mutate(contrast=factor(contrast, ordered = TRUE,
                           levels=c("AmbvsLow.3C", "AmbvsLow.6C", "AmbvsLow.10C", 
                                    "3vs6.Amb-pH", "3vs6.Low-pH",
                                    "3vs10.Amb-pH", "3vs10.Low-pH",
                                    "6vs10.Amb-pH", "6vs10.Low-pH"))) %>% #Convert gene_set column to factor, define all gene_sets levels
  tidyr::complete(contrast, upregulated, category) %>% # add rows for gene_sets that are missing from dataframe
  mutate(upregulated=factor(upregulated, levels = c("3C", "6C", "10C", "Ambient-pH", "Low-pH"), ordered = T)) %>%

    # remove upregulated treatment that doesn't exist in given contrast
  filter(!( (contrast=="AmbvsLow.3C" | contrast=="AmbvsLow.6C"| contrast=="AmbvsLow.10C") & (upregulated=="3C" | upregulated=="6C" | upregulated=="10C"))) %>%
  filter(!( (contrast=="3vs6.Amb-pH" | contrast=="3vs10.Amb-pH" | contrast=="6vs10.Amb-pH" | contrast=="3vs6.Low-pH" | contrast=="3vs10.Low-pH" | contrast=="6vs10.Low-pH") & (upregulated=="Ambient-pH" | upregulated=="Low-pH"))) %>%
  filter(!( (contrast=="3vs6.Amb-pH" | contrast=="3vs6.Low-pH") & upregulated=="10C")) %>%
  filter(!( (contrast=="3vs10.Amb-pH" | contrast=="3vs10.Low-pH") & upregulated=="6C")) %>%
  filter(!( (contrast=="6vs10.Amb-pH" | contrast=="6vs10.Low-pH") & upregulated=="3C")) %>%
  
  separate(contrast, sep="\\.", into = c("comparison", "background"), remove = FALSE) %>% select(-comparison) %>%
  mutate(background=factor(background, ordered = T, levels=c("Low-pH", "Amb-pH", "3C", "6C", "10C")))

```

#### Enriched Uniprot Keyword Bubble Plot - Ambient vs Low pH @ 6C 

```{r}
pdf(file="../results/GO-enrichment/Keywords-pH-upregulated-enriched-bubble.pdf", height = 5, width = 4)

enriched.kw.degs.upregulated %>% 
  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.kw.degs.upregulated %>% select(contrast, upregulated, category, background, kw_id, kw, count, percent, p_value, fold_enrichment, fdr, genes), by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>% #dplyr::slice(1:25) %>%
  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.3C" ~ "pH",
      x == "AmbvsLow.6C" ~ "pH",
      x == "AmbvsLow.10C" ~ "pH",
      x == "3vs6.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "6vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs6.Low-pH" ~ "Temp, Low pH",
      x == "3vs10.Low-pH" ~ "Temp, Low pH",
      x == "6vs10.Low-pH" ~ "Temp, Low pH")}) %>%
  mutate(group=factor(group, ordered=TRUE, levels=c("pH", "Temp, Amb pH", "Temp, Low pH"))) %>% 
  filter(background %in% c("3C","6C","10C")) %>%
  mutate(upregulated=factor(upregulated, ordered = TRUE, levels = c("Low-pH", "Ambient-pH"))) %>%
  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
  mutate(kw = factor(kw, rev(unique(kw)), ordered = TRUE)) %>%

ggplot(aes(y = kw, x=contrast, col=upregulated)) + 
  geom_point(alpha=0.5, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Upregulated", values=c(`Ambient-pH`="#2c7bb6", `Low-pH`="#d7191c"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Uniprot Keywords") + xlab(NULL) +
  
  # Set plot dimensions
  coord_cartesian(clip = "off", ylim = c(-.1, 28)) +
  
  ggtitle(NULL) +
  annotate("text", x = 1, y = -.05, label = "3C", size=2.5, col="gray15") +
  annotate("text", x = 2, y = -.05, label = "6C", size=2.5, col="gray15") +
  annotate("text", x = 3, y = -.05, label = "10C", size=2.5, col="gray15") +
  annotate("text", x = 2, y = 28, label = "Enriched Uniprot Keywords", size=3, col="gray15") + 
  annotate("text", x = 2, y = 27, label = "Ambient vs. Low pH contrasts", size=3.25, col="gray15", fontface = 'bold') + 
  annotate("text", x = 2, y = 26, label = "by temperature", size=3, col="gray15", fontface = 'italic') +
  geom_segment(aes(x = 1.5, y = -.5, xend = 1.5, yend = 25), linetype="solid", color = "gray65", size=0.25) +
  geom_segment(aes(x = 2.5, y = -.5, xend = 2.5, yend = 25), linetype="solid", color = "gray65", size=0.25)
dev.off()
```

#### Enriched Uniprot Keyword Bubble Plot - 3C vs. 6C

```{r}
# 3C vs. 6C contrasts 
#pdf(file="../results/GO-enrichment/Keywords-3vs6-upregulated-enriched-bubble.pdf", height = 6.25, width = 4)
pdf(file="../results/GO-enrichment/Keywords-3vs6-upregulated-enriched-bubble-simplecolors.pdf", height = 6.25, width = 4)

# prepare data for bubble plot
enriched.kw.degs.upregulated %>% 
  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.kw.degs.upregulated %>% select(contrast, upregulated, category, background, kw_id, kw, count, percent, p_value, fold_enrichment, fdr, genes), by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>% #dplyr::slice(1:10) %>%
  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.3C" ~ "pH",
      x == "AmbvsLow.6C" ~ "pH",
      x == "AmbvsLow.10C" ~ "pH",
      x == "3vs6.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "6vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs6.Low-pH" ~ "Temp, Low pH",
      x == "3vs10.Low-pH" ~ "Temp, Low pH",
      x == "6vs10.Low-pH" ~ "Temp, Low pH")}) %>%
  filter(background %in% c("Low-pH","Amb-pH")) %>%
  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
  mutate(kw = factor(kw, rev(unique(kw)), ordered = TRUE)) %>%
  mutate(upreg_ph=factor(paste(upregulated, background, sep = " "))) %>%
  filter(grepl("3", contrast)) %>% filter(grepl("6", contrast)) %>% droplevels() %>% #filter for 3 vs 6 contrasts
  mutate(upreg_ph=factor(upreg_ph, ordered=TRUE, levels=c("3C Amb-pH", "6C Amb-pH", "3C Low-pH", "6C Low-pH"))) %>% 

# Plot  
ggplot(aes(y = kw, x=contrast, 
#           col=upreg_ph)) + 
           col=upregulated)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) +
 facet_wrap(~category,scales="free", nrow = 2) + 
  scale_color_manual(
       # name="Upregulated",
       # values=c("3C Amb-pH"= "navyblue", "3C Low-pH"="royalblue1",
       # "6C Amb-pH"="darkgreen", "6C Low-pH"="yellow3",
       # "10C Amb-pH"="firebrick4", "10C Low-pH"="sienna2"),
       name=NULL,
       values=c("3C"= "navyblue", "6C"="darkgreen"), 
       labels=c("3C"="Upregulated in 3C", "6C"="Downregulated in 3C"),
       guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), 
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Uniprot Keyword") + xlab(NULL) +

  coord_cartesian(clip = "off", ylim = c(-1.75, 36)) + #set plot dimensions
  
  ggtitle(NULL) +
  annotate("text", x = 1, y = -.75, label = "Amb pH", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 2, y = -.75, label = "Low pH", size=2.75, col="gray15", angle=90) +
  #annotate("text", x = 1.5, y = -1.75, label = "3C vs. 6C", size=2.75, col="gray15", fontface = 'bold') +
  annotate("text", x = 1.5, y = 36, label = "Enriched Uniprot Keywords", size=3, col="gray15") +
  annotate("text", x = 1.5, y = 34.5, label = "3C vs 6C by pH", size=3.25, col="gray15", fontface = 'bold')
dev.off()
```

#### Enriched Uniprot Keyword Bubble Plot - 3C vs. 10C

```{r}
# # 3C vs. 10C contrasts 
# pdf(file="../results/GO-enrichment/Keywords-3vs10-upregulated-enriched-bubble.pdf", height = 8.5, width = 4)
# 
# # prepare data for bubble plot
# enriched.kw.degs.upregulated %>% 
#   group_by(contrast, upregulated, genes, count) %>%
#   dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
#   left_join(enriched.kw.degs.upregulated %>% select(contrast, upregulated, category, background, kw_id, kw, count, percent, p_value, fold_enrichment, fdr, genes), by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
#   ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>% #dplyr::slice(1:15) %>%
#   mutate(group=contrast) %>%
#   mutate_at("group", function(x) {
#     case_when(
#       x == "AmbvsLow.3C" ~ "pH",
#       x == "AmbvsLow.6C" ~ "pH",
#       x == "AmbvsLow.10C" ~ "pH",
#       x == "3vs6.Amb-pH" ~ "Temp, Amb pH",
#       x == "3vs10.Amb-pH" ~ "Temp, Amb pH",
#       x == "6vs10.Amb-pH" ~ "Temp, Amb pH",
#       x == "3vs6.Low-pH" ~ "Temp, Low pH",
#       x == "3vs10.Low-pH" ~ "Temp, Low pH",
#       x == "6vs10.Low-pH" ~ "Temp, Low pH")}) %>%
#   filter(background %in% c("Low-pH","Amb-pH")) %>%
#   ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
#   mutate(kw = factor(kw, rev(unique(kw)), ordered = TRUE)) %>%
#   mutate(upreg_ph=factor(paste(upregulated, background, sep = " "))) %>%
#   filter(grepl("3", contrast)) %>% filter(grepl("10", contrast)) %>% droplevels() %>% #filter for 3 vs 10 contrasts
#   mutate(upreg_ph=factor(upreg_ph, ordered=TRUE, levels=c("10C Amb-pH", "3C Amb-pH", "10C Low-pH", "3C Low-pH"))) %>% 
# 
# # Plot  
# ggplot(aes(y = kw, x=contrast, col=upreg_ph)) + 
#   geom_point(alpha=0.65, aes(size=-log10(p_value))) +
#  facet_wrap(~category,scales="free", nrow = 2) + 
#   scale_color_manual(name="Upregulated",
#                      values=c("3C Amb-pH"= "navyblue", "3C Low-pH"="royalblue1",
#                               "6C Amb-pH"="darkgreen", "6C Low-pH"="yellow3",
#                               "10C Amb-pH"="firebrick4", "10C Low-pH"="sienna2"),
#                      guide = guide_legend(override.aes = list(size=4))) +
#   scale_size("-Log10 P-value", range = c(2,8), 
#              guide = guide_legend(override.aes = list(col="gray50"))) +
#   scale_x_discrete(drop=T) + #Do drop empty factors
#   theme_cleveland() + 
#   theme(
#         legend.position = "right", 
#         axis.text.x=element_blank(),
#         axis.ticks.x = element_blank(),
#         axis.text.y=element_text(size=7.25), 
#         plot.title = element_text(size=10),
#         legend.text=element_text(size=6), 
#         legend.title = element_text(size=6),
#         strip.background = element_blank(),
#         strip.text.x = element_blank()) +
#   ylab("Enriched Uniprot Keyword") + xlab(NULL) +
# 
#   coord_cartesian(clip = "off", ylim = c(-2, 56)) + #set plot dimensions
#   
#   ggtitle(NULL) +
#   annotate("text", x = 1, y = -.75, label = "Amb pH", size=2.75, col="gray15", angle=90) +
#   annotate("text", x = 2, y = -.75, label = "Low pH", size=2.75, col="gray15", angle=90) +
#   #annotate("text", x = 1.5, y = -1.75, label = "3C vs. 10C", size=2.75, col="gray15", fontface = 'bold') +
#   annotate("text", x = 1.5, y = 56, label = "Enriched Uniprot Keywords", size=3, col="gray15") +
#   annotate("text", x = 1.5, y = 54.5, label = "3C vs 10C by pH", size=3.25, col="gray15", fontface = 'bold')
# dev.off()
```

```{r}
# 6C vs. 10C contrasts 
#pdf(file="../results/GO-enrichment/Keywords-6vs10-upregulated-enriched-bubble.pdf", height = 6.5, width = 4)
pdf(file="../results/GO-enrichment/Keywords-6vs10-upregulated-enriched-bubble-simplecolors.pdf", height = 6.5, width = 4)

# prepare data for bubble plot
enriched.kw.degs.upregulated %>% 
  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.kw.degs.upregulated %>% select(contrast, upregulated, category, background, kw_id, kw, count, percent, p_value, fold_enrichment, fdr, genes), by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>% #dplyr::slice(1:15) %>%
  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.3C" ~ "pH",
      x == "AmbvsLow.6C" ~ "pH",
      x == "AmbvsLow.10C" ~ "pH",
      x == "3vs6.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "6vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs6.Low-pH" ~ "Temp, Low pH",
      x == "3vs10.Low-pH" ~ "Temp, Low pH",
      x == "6vs10.Low-pH" ~ "Temp, Low pH")}) %>%
  filter(background %in% c("Low-pH","Amb-pH")) %>%
  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
  mutate(kw = factor(kw, rev(unique(kw)), ordered = TRUE)) %>%
  mutate(upreg_ph=factor(paste(upregulated, background, sep = " "))) %>%
  filter(grepl("6", contrast)) %>% filter(grepl("10", contrast)) %>% droplevels() %>% #filter for 6 vs 10 contrasts
  mutate(upreg_ph=factor(upreg_ph, ordered=TRUE, levels=c("10C Amb-pH", "6C Amb-pH", "10C Low-pH", "6C Low-pH"))) %>% 

# Plot  
ggplot(aes(y = kw, x=contrast, 
#           col=upreg_ph)) + 
           col=upregulated)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) +
 facet_wrap(~category,scales="free", nrow = 2) + 
  scale_color_manual(
   #  name="Upregulated",
   #  values=c("3C Amb-pH"= "navyblue", "3C Low-pH"="royalblue1",
   #           "6C Amb-pH"="darkgreen", "6C Low-pH"="yellow3",
   #            "10C Amb-pH"="firebrick4", "10C Low-pH"="sienna2"),
       name=NULL,
       values=c("10C"= "firebrick4", "6C"="darkgreen"), 
       labels=c("10C"="Upregulated in 10C", "6C"="Downregulated in 10C"),
       guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), 
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Uniprot Keyword") + xlab(NULL) +

  coord_cartesian(clip = "off", ylim = c(-2.5, 45)) + #set plot dimensions
  
  ggtitle(NULL) +
  annotate("text", x = 1, y = -1, label = "Amb pH", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 2, y = -1, label = "Low pH", size=2.75, col="gray15", angle=90) +
  #annotate("text", x = 1.5, y = -1.75, label = "6C vs. 6C", size=2.75, col="gray15", fontface = 'bold') +
  annotate("text", x = 1.5, y = 45, label = "Enriched Uniprot Keywords", size=3, col="gray15") +
  annotate("text", x = 1.5, y = 43.5, label = "6C vs 10C by pH", size=3.25, col="gray15", fontface = 'bold')
dev.off()
```

```{r}
# Single vs. Double Stressors 
#pdf(file="../results/GO-enrichment/Keywords-6vs10-upregulated-enriched-bubble.pdf", height = 6.5, width = 4)
#pdf(file="../results/GO-enrichment/Keywords-6vs10-upregulated-enriched-bubble-simplecolors.pdf", height = 6.5, width = 4)

# prepare data for bubble plot
enriched.kw.degs.upregulated %>% 
  group_by(contrast, upregulated, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.kw.degs.upregulated %>% select(contrast, upregulated, category, background, kw_id, kw, count, percent, p_value, fold_enrichment, fdr, genes), by = c("contrast", "upregulated", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(contrast, upregulated, p_value) %>% group_by(contrast, upregulated) %>% #dplyr::slice(1:15) %>%
  mutate(group=contrast) %>%
  mutate_at("group", function(x) {
    case_when(
      x == "AmbvsLow.3C" ~ "pH",
      x == "AmbvsLow.6C" ~ "pH",
      x == "AmbvsLow.10C" ~ "pH",
      x == "3vs6.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "6vs10.Amb-pH" ~ "Temp, Amb pH",
      x == "3vs6.Low-pH" ~ "Temp, Low pH",
      x == "3vs10.Low-pH" ~ "Temp, Low pH",
      x == "6vs10.Low-pH" ~ "Temp, Low pH")}) %>%
  #filter(background %in% c("Low-pH","Amb-pH")) %>%
  ungroup() %>% arrange(group, contrast, upregulated, p_value) %>%
  mutate(kw = factor(kw, rev(unique(kw)), ordered = TRUE)) %>%
  mutate(upreg_ph=factor(paste(upregulated, background, sep = " "))) %>%
  filter(grepl("AmbvsLow.6C|3vs6.Amb-pH|6vs10.Amb-pH", contrast)) %>%  #filter for contrasts of interest
    select(contrast) %>% unique()

  mutate(upreg_ph=factor(upreg_ph, ordered=TRUE, levels=c("10C Amb-pH", "6C Amb-pH", "10C Low-pH", "6C Low-pH"))) %>% 

# Plot  
ggplot(aes(y = kw, x=contrast, 
#           col=upreg_ph)) + 
           col=upregulated)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) +
 facet_wrap(~category,scales="free", nrow = 2) + 
  scale_color_manual(
   #  name="Upregulated",
   #  values=c("3C Amb-pH"= "navyblue", "3C Low-pH"="royalblue1",
   #           "6C Amb-pH"="darkgreen", "6C Low-pH"="yellow3",
   #            "10C Amb-pH"="firebrick4", "10C Low-pH"="sienna2"),
       name=NULL,
       values=c("10C"= "firebrick4", "6C"="darkgreen"), 
       labels=c("10C"="Upregulated in 10C", "6C"="Downregulated in 10C"),
       guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), 
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Uniprot Keyword") + xlab(NULL) +

  coord_cartesian(clip = "off", ylim = c(-2.5, 45)) + #set plot dimensions
  
  ggtitle(NULL) +
  annotate("text", x = 1, y = -1, label = "Amb pH", size=2.75, col="gray15", angle=90) +
  annotate("text", x = 2, y = -1, label = "Low pH", size=2.75, col="gray15", angle=90) +
  #annotate("text", x = 1.5, y = -1.75, label = "6C vs. 6C", size=2.75, col="gray15", fontface = 'bold') +
  annotate("text", x = 1.5, y = 45, label = "Enriched Uniprot Keywords", size=3, col="gray15") +
  annotate("text", x = 1.5, y = 43.5, label = "6C vs 10C by pH", size=3.25, col="gray15", fontface = 'bold')
dev.off()
```

## Deeper dive into effect of Warming, OA, Warming+OA using stricter Log2FC 

```{r, include=F, echo=F, message=F}
# BACKGROUND - all genes submitted to DESeq2 analysis 
enrich.background %>% write_clip(allow_non_interactive = TRUE)

### Effect of ACIDIFICATION as sole stressor 

# enrich.Amb.Low.6 - LFC > 1.0 - LOWER IN LOW PH
enrich.Amb.Low.6 %>%
  filter(select(., contains("log2FoldChange"))>1) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# enrich.Amb.Low.6 - LFC < -1.0 - HIGHER IN LOW PH
enrich.Amb.Low.6 %>%
  filter(select(., contains("log2FoldChange"))<(-1)) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

### Effect of WARMING as sole stressor 

# enrich.6.10.Amb - LFC > 1.0 - LOWER IN 10C
enrich.6.10.Amb %>%
  filter(select(., contains("log2FoldChange"))>1) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# enrich.6.10.Amb - LFC < -1.0 - HIGHER IN 10C
enrich.6.10.Amb %>%
  filter(select(., contains("log2FoldChange"))<(-1)) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

### Effect of WARMING + ACIDIFICATION as combined stressors 

# enrich.both.10low.6amb - lfc < -1.0 - HIGHER IN 10C-LOW PH
enrich.both.10low.6amb %>% 
  filter(select(., contains("log2FoldChange"))<(-1)) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# enrich.both.10low.6amb - lfc > 1.0 - LOWER IN 10C-LOW PH
enrich.both.10low.6amb %>%
  filter(select(., contains("log2FoldChange"))>1) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

### Effect of COOLING as sole stressor 

# enrich.3.6.Amb - LFC > 1.0 - HIGHER IN 3C
enrich.3.6.Amb %>%
  filter(select(., contains("log2FoldChange"))>1) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# enrich.3.6.Amb - LFC < -1.0 - LOWER IN 3C 
enrich.3.6.Amb %>%
  filter(select(., contains("log2FoldChange"))<(-1)) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

### Effect of COOLING + ACIDIFICATION as combined stressors 

# enrich.both.3low.6amb - LFC > 1.0 - HIGHER IN 3C-LOW PH
enrich.both.3low.6amb %>%
  filter(select(., contains("log2FoldChange"))>1) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# enrich.both.3low.6amb - LFC < 1.0 - LOWER IN 3C-LOW PH
enrich.both.3low.6amb %>%
  filter(select(., contains("log2FoldChange"))<(-1)) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

```

```{r}
#### Read in enriched biological functions for all contrasts

filenames <- read_delim(file="../results/GO-enrichment/DAVID_degs_2x/GO_filenames_2x.txt", delim = "\t", col_names = c("filename", "gene_set"))
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=paste("../results/GO-enrichment/DAVID_degs_2x/", filenames[i,"filename"], sep=""), delim = "\t"))}

enriched.degs.2x <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("ID", "term"), sep = "~") %>%
    separate(gene_set, sep = "_", into = c("stress", "response"), remove = F) %>%
    mutate(stress=factor(stress, ordered = TRUE, levels=c("OA", "Warm", "Warm.OA", "Cold", "Cold.OA")),
           mutate(response=factor(response))) #Convert stressor column to factor, define all gene_sets levels

# Save R object to then import to RStudio Cloud for use in metacoder 
save(enriched.degs.2x, file="../results/GO-enrichment/DAVID_degs_2x/enriched.degs.2x")

enriched.degs.2x %>% filter(category=="UP_KW_BIOLOGICAL_PROCESS", response=="Upregulated") %>% View()
```


# BONEYARD

<!-- #### Read in Enriched results for all pairwise DEG contrasts -->

<!-- ```{r} -->

<!-- # Enriched GO Biological Processes -->
<!-- filenames <- read_delim(file="../results/GO-enrichment/GO_filenames.txt", delim = "\t", col_names = c("filename", "gene_set")) -->
<!-- file_list <- vector(mode = "list", length = nrow(filenames)) -->
<!-- names(file_list) <- c(filenames$gene_set) -->

<!-- for (i in 1:nrow(filenames)) { -->
<!--     file_list[[i]] <- data.frame(read_delim(file=paste("../results/GO-enrichment/", filenames[i,"filename"], sep=""), delim = "\t"))} -->

<!-- enriched.degs <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>% # Bind dataframe for each gene_set together -->
<!--     #filter(category=="GOTERM_BP_DIRECT") %>% -->
<!--     filter(p_value<0.05) %>% -->
<!--     dplyr::rename(percent=x) %>% -->
<!--     separate(term, into = c("term", "process"), sep = "~") %>% -->
<!--     mutate(gene_set=factor(gene_set,  -->
<!--                            levels=c("Amb-vs-Low-3C", "Amb-vs-Low-6C", "Amb-vs-Low-10C",  -->
<!--                                     "3Cvs6C-Amb-pH", "3Cvs10C-Amb-pH", "6Cvs10C-Amb-pH", -->
<!--                                     "3Cvs6C-Low-pH", "3Cvs10C-Low-pH", "6Cvs10C-Low-pH", -->
<!--                                     "6AMBvs3LOW", "6AMBvs10LOW"))) %>% #Convert gene_set column to factor, define all gene_sets levels -->
<!--     tidyr::complete(gene_set, category) # add rows for gene_sets that are missing from dataframe -->

<!-- #enriched.degs %>% write_clip() #copy to clipboard to paste into google spreadsheet  -->
<!-- ``` -->

<!-- ### Bubble plot of Enriched Biological Process GO terms  -->

<!-- ```{r} -->
<!-- #pdf(file="../results/GO-enrichment/BP-pH-enriched-bubble.pdf", height = 7, width = 7) -->

<!-- enriched.degs %>% filter(count>=3) %>%  -->
<!--   filter(category=="GOTERM_BP_DIRECT") %>% -->
<!--   #filter(category=="GOTERM_MF_DIRECT") %>% -->
<!--   filter(gene_set %in% c("Amb-vs-Low-6C",  -->
<!--                          "3Cvs6C-Amb-pH", "6AMBvs3LOW", -->
<!--                          "6Cvs10C-Amb-pH", "6AMBvs10LOW")) %>% # keep single and double stressor contrasts -->
<!--   group_by(gene_set, genes, count) %>% -->
<!--   dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate genes, select one with lowest p-value -->
<!--   left_join(enriched.degs %>% select(gene_set, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes),  -->
<!--             by = c("gene_set", "genes", "p_value", "count")) %>%  #re-add data -->
<!--   ungroup() %>% arrange(gene_set, p_value) %>% group_by(gene_set) %>% dplyr::slice(1:20) %>% -->
<!--   mutate(gene_set=factor(gene_set, ordered = TRUE, -->
<!--                            levels=c("3Cvs6C-Amb-pH", "6AMBvs3LOW", -->
<!--                                     "6Cvs10C-Amb-pH", "6AMBvs10LOW", -->
<!--                                     "Amb-vs-Low-6C"))) %>% -->
<!--   mutate(group=gene_set) %>% -->
<!--   mutate_at("group", function(x) { -->
<!--     case_when( -->
<!--       x == "Amb-vs-Low-6C" ~ "High pCO2", -->
<!--       x == "3Cvs6C-Amb-pH" ~ "Cold Temperature", -->
<!--       x == "6Cvs10C-Amb-pH" ~ "Warm Temperature", -->
<!--       x == "6AMBvs3LOW" ~ "Cold Temperature + High pCO2",  -->
<!--       x == "6AMBvs10LOW" ~ "Warm Temperature + High pCO2")}) %>% -->
<!--   mutate(group=factor(group, ordered=TRUE,  -->
<!--                       levels=c("Cold Temperature", "Cold Temperature + High pCO2", -->
<!--                                "Warm Temperature", "Warm Temperature + High pCO2", -->
<!--                                "High pCO2"))) %>%  -->
<!-- #  filter(!grepl('Amb-pH|Low-pH', gene_set)) %>% # Filter for only the pH contrasts -->
<!-- #  ungroup() %>% arrange(group, gene_set, p_value) %>% -->
<!--   ungroup() %>% arrange(term) %>% -->
<!--   mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>% -->

<!-- ggplot(aes(y = process, x=group, col=group)) +  -->
<!--   geom_point(aes(size=-log10(p_value))) + #size=count,  alpha=p_value,  -->
<!--  facet_wrap(~category,scales="free", nrow = 2,  -->
<!--             labeller = as_labeller(c(`GOTERM_BP_DIRECT`="Enriched Biological Processes\nAmbient vs. Low pH\nby temperature"))) + -->
<!--   # scale_color_manual(name="Upregulated", values=c(Ambient="#2c7bb6", Moderate="#fdae61", Severe="#d7191c"), -->
<!--   #            guide = guide_legend(override.aes = list(size=4))) + -->
<!--   scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10), -->
<!--              guide = guide_legend(override.aes = list(col="gray50"))) + -->
<!--   scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts -->
<!--   theme_cleveland() +  -->
<!--   theme( -->
<!--         legend.position = "right",  -->
<!--         axis.text.x=element_blank(), -->
<!--         axis.ticks.x = element_blank(), -->
<!--         axis.text.y=element_text(size=7.25),  -->
<!--         plot.title = element_text(size=10), -->
<!--         legend.text=element_text(size=6),  -->
<!--         legend.title = element_text(size=6), -->
<!--         strip.background = element_blank(), -->
<!--         strip.text.x = element_blank()) + -->
<!--   ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) #+ -->
<!--   # coord_cartesian(clip = "off", ylim = c(-.7, 35)) + -->
<!--   # ggtitle(NULL) + -->
<!--   # annotate("text", x = 1, y = -.15, label = "3C", size=3.5, col="gray15") + -->
<!--   # annotate("text", x = 2, y = -.15, label = "6C", size=3.5, col="gray15") + -->
<!--   # annotate("text", x = 3, y = -.15, label = "10C", size=3.5, col="gray15") + -->
<!--   # annotate("text", x = 2, y = 35, label = "Enriched Biological Processes", size=3, col="gray15") +  -->
<!--   # annotate("text", x = 2, y = 33.5, label = "Ambient vs. Low pH contrasts", size=3.25, col="gray15", fontface = 'bold') +  -->
<!--   # annotate("text", x = 2, y = 32, label = "by temperature", size=3, col="gray15", fontface = 'italic') -->
<!-- #dev.off() -->
<!-- ``` -->

<!-- ## Bubble plot of Enriched Uniprot Keywords - Biological Processes -->

<!-- ```{r} -->
<!-- enriched.degs %>% -->
<!--   filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(count>2) %>% -->
<!--   #filter(category=="UP_KW_CELLULAR_COMPONENT") %>% -->
<!--   filter(gene_set %in% c("Amb-vs-Low-6C",  -->
<!--                          "3Cvs6C-Amb-pH", "6AMBvs3LOW", -->
<!--                          "6Cvs10C-Amb-pH", "6AMBvs10LOW")) %>% # keep single and double stressor contrasts -->
<!--   group_by(gene_set, genes, count) %>% -->
<!--   dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate genes, select one with lowest p-value -->
<!--   left_join(enriched.degs %>% select(gene_set, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes),  -->
<!--             by = c("gene_set", "genes", "p_value", "count")) %>%  #re-add data -->
<!--   ungroup() %>% arrange(gene_set, p_value) %>% group_by(gene_set) %>% #dplyr::slice(1:20) %>% -->
<!--   mutate(gene_set=factor(gene_set, ordered = TRUE, -->
<!--                            levels=c("3Cvs6C-Amb-pH", "6AMBvs3LOW", -->
<!--                                     "6Cvs10C-Amb-pH", "6AMBvs10LOW", -->
<!--                                     "Amb-vs-Low-6C"))) %>% -->
<!--   mutate(group=gene_set) %>% -->
<!--   mutate_at("group", function(x) { -->
<!--     case_when( -->
<!--       x == "Amb-vs-Low-6C" ~ "High pCO2", -->
<!--       x == "3Cvs6C-Amb-pH" ~ "Cold Temperature", -->
<!--       x == "6Cvs10C-Amb-pH" ~ "Warm Temperature", -->
<!--       x == "6AMBvs3LOW" ~ "Cold Temperature + High pCO2",  -->
<!--       x == "6AMBvs10LOW" ~ "Warm Temperature + High pCO2")}) %>% -->
<!--   mutate(group=factor(group, ordered=TRUE,  -->
<!--                       levels=c("Cold Temperature", "Cold Temperature + High pCO2", -->
<!--                                "Warm Temperature", "Warm Temperature + High pCO2", -->
<!--                                "High pCO2"))) %>%  -->
<!-- #  filter(!grepl('Amb-pH|Low-pH', gene_set)) %>% # Filter for only the pH contrasts -->
<!--   #ungroup() %>% arrange(group, gene_set, p_value) %>%  #sort rows by treatment and p-value  -->
<!--    ungroup() %>% arrange(term) %>%  #sort rows by treatment and p-value  -->
<!--    mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%  -->

<!-- ggplot(aes(y = process, x=group, col=group)) +  -->
<!--   geom_point(aes(size=-log10(p_value))) + #size=count,  alpha=p_value,  -->
<!--  facet_wrap(~category,scales="free", nrow = 2,  -->
<!--             labeller = as_labeller(c(`GOTERM_BP_DIRECT`="Enriched Biological Processes\nAmbient vs. Low pH\nby temperature"))) + -->
<!--   # scale_color_manual(name="Upregulated", values=c(Ambient="#2c7bb6", Moderate="#fdae61", Severe="#d7191c"), -->
<!--   #            guide = guide_legend(override.aes = list(size=4))) + -->
<!--   scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10), -->
<!--              guide = guide_legend(override.aes = list(col="gray50"))) + -->
<!--   scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts -->
<!--   theme_cleveland() +  -->
<!--   theme( -->
<!--         legend.position = "right",  -->
<!--         axis.text.x=element_blank(), -->
<!--         axis.ticks.x = element_blank(), -->
<!--         axis.text.y=element_text(size=7.25),  -->
<!--         plot.title = element_text(size=10), -->
<!--         legend.text=element_text(size=6),  -->
<!--         legend.title = element_text(size=6), -->
<!--         strip.background = element_blank(), -->
<!--         strip.text.x = element_blank()) + -->
<!--   ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) #+ -->
<!--   # coord_cartesian(clip = "off", ylim = c(-.7, 35)) + -->
<!--   # ggtitle(NULL) + -->
<!--   # annotate("text", x = 1, y = -.15, label = "3C", size=3.5, col="gray15") + -->
<!--   # annotate("text", x = 2, y = -.15, label = "6C", size=3.5, col="gray15") + -->
<!--   # annotate("text", x = 3, y = -.15, label = "10C", size=3.5, col="gray15") + -->
<!--   # annotate("text", x = 2, y = 35, label = "Enriched Biological Processes", size=3, col="gray15") +  -->
<!--   # annotate("text", x = 2, y = 33.5, label = "Ambient vs. Low pH contrasts", size=3.25, col="gray15", fontface = 'bold') +  -->
<!--   # annotate("text", x = 2, y = 32, label = "by temperature", size=3, col="gray15", fontface = 'italic') -->
<!-- ``` -->


<!-- #### Temperature contrasts for each pH  -->

<!-- ```{r} -->
<!-- pdf(file="../results/GO-enrichment/BP-temperature-enriched-bubble.pdf", height = 11, width = 8) -->

<!-- enriched.bp.degs %>%  -->
<!--   group_by(gene_set, genes, count) %>% -->
<!--   dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value -->
<!--   left_join(enriched.bp.degs %>% select(gene_set, category, go, process, count, percent, p_value, fold_enrichment, fdr, genes), by = c("gene_set", "genes", "p_value", "count")) %>%  #re-add data -->
<!--   ungroup() %>% arrange(gene_set, p_value) %>% group_by(gene_set) %>% dplyr::slice(1:20) %>% -->
<!--   mutate(gene_set=factor(gene_set, ordered = TRUE, -->
<!--                            levels=c("Amb-vs-Low-3C", "Amb-vs-Low-6C", "Amb-vs-Low-10C",  -->
<!--                                     "3Cvs6C-Amb-pH", "3Cvs6C-Low-pH", -->
<!--                                     "3Cvs10C-Amb-pH", "3Cvs10C-Low-pH",  -->
<!--                                     "6Cvs10C-Amb-pH", "6Cvs10C-Low-pH"))) %>% -->
<!--   mutate(group=gene_set) %>% -->
<!--   mutate_at("group", function(x) { -->
<!--     case_when( -->
<!--       x == "Amb-vs-Low-3C" ~ "pH", -->
<!--       x == "Amb-vs-Low-6C" ~ "pH", -->
<!--       x == "Amb-vs-Low-10C" ~ "pH", -->
<!--       x == "3Cvs6C-Amb-pH" ~ "Temp, Amb pH", -->
<!--       x == "3Cvs10C-Amb-pH" ~ "Temp, Amb pH", -->
<!--       x == "6Cvs10C-Amb-pH" ~ "Temp, Amb pH", -->
<!--       x == "3Cvs6C-Low-pH" ~ "Temp, Low pH", -->
<!--       x == "3Cvs10C-Low-pH" ~ "Temp, Low pH", -->
<!--       x == "6Cvs10C-Low-pH" ~ "Temp, Low pH")}) %>% -->
<!--   mutate(group=factor(group, ordered=TRUE, levels=c("pH", "Temp, Amb pH", "Temp, Low pH"))) %>%  -->

<!--   filter(grepl('Amb-pH|Low-pH', gene_set)) %>% # Filter for only the pH contrasts -->
<!--   ungroup() %>% arrange(group, gene_set, p_value) %>% -->
<!--   mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%  -->

<!-- ggplot(aes(y = process, x=gene_set, col=gene_set)) +  -->
<!--   geom_point(aes(size=-log10(p_value))) + #size=count,  alpha=p_value,  -->
<!--  facet_wrap(~category,scales="free", nrow = 2, labeller = as_labeller(c(`GOTERM_BP_DIRECT`="Enriched Biological Processes\nAmbient vs. Low pH\nby temperature"))) + -->
<!--   # scale_color_manual(name="Upregulated", values=c(Ambient="#2c7bb6", Moderate="#fdae61", Severe="#d7191c"), -->
<!--   #            guide = guide_legend(override.aes = list(size=4))) + -->
<!--   #scale_color_manual(guide="none", values=c("#ffffb3","#8dd3c7","#bebada")) + -->
<!--   scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10), -->
<!--              guide = guide_legend(override.aes = list(col="gray50"))) + -->
<!--   scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts -->
<!--   theme_cleveland() +  -->
<!--   theme( -->
<!--         legend.position = "right",  -->
<!--         axis.text.x=element_blank(), -->
<!--         axis.ticks.x = element_blank(), -->
<!--         axis.text.y=element_text(size=7.25),  -->
<!--         plot.title = element_text(size=10), -->
<!--         legend.text=element_text(size=6),  -->
<!--         legend.title = element_text(size=6), -->
<!--         strip.background = element_blank(), -->
<!--         strip.text.x = element_blank()) + -->
<!--   ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) + -->
<!--   coord_cartesian(clip = "off", ylim = c(-6, 80)) + -->
<!--   ggtitle(NULL) + -->
<!--   annotate("text", x = 1, y = -1.5, label = "Amb", size=2.75, col="gray15", angle=90) + -->
<!--   annotate("text", x = 2, y = -1.5, label = "Low", size=2.75, col="gray15", angle=90) + -->
<!--   annotate("text", x = 3, y = -1.5, label = "Amb", size=2.75, col="gray15", angle=90) + -->
<!--   annotate("text", x = 4, y = -1.5, label = "Low", size=2.75, col="gray15", angle=90) + -->
<!--   annotate("text", x = 5, y = -1.5, label = "Amb", size=2.75, col="gray15", angle=90) + -->
<!--   annotate("text", x = 6, y = -1.5, label = "Low", size=2.75, col="gray15", angle=90) + -->

<!--   annotate("text", x = 1.5, y = -5, label = "3C vs. 6C", size=2.75, col="gray15", fontface = 'bold') + -->
<!--   annotate("text", x = 3.5, y = -5, label = "3C vs. 10C", size=2.75, col="gray15", fontface = 'bold') + -->
<!--   annotate("text", x = 5.5, y = -5, label = "6C vs. 10C", size=2.75, col="gray15", fontface = 'bold') +   -->

<!--   annotate("text", x = 3.5, y = 79.5, label = "Enriched Biological Processes", size=3, col="gray15") + -->
<!--   annotate("text", x = 3.5, y = 78, label = "Temperature contrasts", size=3.25, col="gray15", fontface = 'bold') + -->
<!--   annotate("text", x = 3.5, y = 76.5, label = "by pH", size=3, col="gray15", fontface = 'italic') + -->
<!--   geom_segment(aes(x = 2.5, y = -6, xend = 2.5, yend = 73), linetype="dashed", color = "gray65", size=0.75) + -->
<!--   geom_segment(aes(x = 4.5, y = -6, xend = 4.5, yend = 73), linetype="dashed", color = "gray65", size=0.75) -->
<!-- dev.off() -->
<!-- ``` -->


# ```{r, include=F, echo=F, message=F}
# # BACKGROUND - all genes submitted to DESeq2 analysis 
# enrich.background %>% write_clip(allow_non_interactive = TRUE)
# 
# # NO ANNOTATED DEGS AT ABS(L2FC)>0.5 FOR AMB VS LOW @ 3C OR 10C
# 
# # enrich.Amb.Low.6 - LFC>0.5 - LOWER IN LOW PH
# enrich.Amb.Low.6 %>%
#   filter(select(., contains("log2FoldChange"))>0.5) %>%
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.Amb.Low.6 - LFC<0.5 - HIGHER IN LOW PH
# enrich.Amb.Low.6 %>%
#   filter(select(., contains("log2FoldChange"))<(-0.5)) %>% 
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.3.6.Amb - LFC>0.5 - HIGHER IN 3C
# enrich.3.6.Amb %>%
#   filter(select(., contains("log2FoldChange"))>0.5) %>%
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.3.6.Amb - LFC<0.5 - LOWER IN 3C 
# enrich.3.6.Amb %>%
#   filter(select(., contains("log2FoldChange"))<(-0.5)) %>% 
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # # enrich.3.10.Amb - LFC>0.5 - LOWER IN 10C
# # enrich.3.10.Amb %>%
# #   filter(select(., contains("log2FoldChange"))>0.5) %>%
# #   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# # 
# # # enrich.3.10.Amb - LFC<0.5 - HIGHER IN 10C
# # enrich.3.10.Amb %>%
# #   filter(select(., contains("log2FoldChange"))<(-0.5)) %>% 
# #   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.6.10.Amb - LFC>0.5 - LOWER IN 10C
# enrich.6.10.Amb %>%
#   filter(select(., contains("log2FoldChange"))>0.5) %>%
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.6.10.Amb - LFC<0.5 - HIGHER IN 10C
# enrich.6.10.Amb %>%
#   filter(select(., contains("log2FoldChange"))<(-0.5)) %>% 
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.3.6.Low - LFC>0.5 - HIGHER IN 3C
# enrich.3.6.Low %>%
#   filter(select(., contains("log2FoldChange"))>0.5) %>%
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.3.6.Low - LFC<0.5 - LOWER IN 3C
# enrich.3.6.Low %>%
#   filter(select(., contains("log2FoldChange"))<(-0.5)) %>% 
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # # enrich.3.10.Low - LFC>0.5 - LOWER IN 10C
# # enrich.3.10.Low %>%
# #   filter(select(., contains("log2FoldChange"))>0.5) %>%
# #   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# # 
# # # enrich.3.10.Low - LFC<0.5 - HIGHER IN 10C
# # enrich.3.10.Low %>%
# #   filter(select(., contains("log2FoldChange"))<(-0.5)) %>% 
# #   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.6.10.Low - LFC>0.5 - LOWER IN 10C
# enrich.6.10.Low %>%
#   filter(select(., contains("log2FoldChange"))>0.5) %>%
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.6.10.Low - LFC<0.5 - HIGHER IN 10C
# enrich.6.10.Low %>%
#   filter(select(., contains("log2FoldChange"))<(-0.5)) %>% 
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.both.3low.6amb - LFC>0.5 - HIGHER IN 3C-LOW PH
# enrich.both.3low.6amb %>%
#   filter(select(., contains("log2FoldChange"))>0.5) %>%
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.both.3low.6amb - LFC<0.5 - LOWER IN 3C-LOW PH
# enrich.both.3low.6amb %>%
#   filter(select(., contains("log2FoldChange"))<(-0.5)) %>% 
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.both.10low.6amb - lfc<0.5 - HIGHER IN 10C-LOW PH
# enrich.both.10low.6amb %>%
#   filter(select(., contains("log2FoldChange"))<(-0.5)) %>% 
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# 
# # enrich.both.10low.6amb - lfc>0.5 - LOWER IN 10C-LOW PH
# enrich.both.10low.6amb %>%
#   filter(select(., contains("log2FoldChange"))>0.5) %>%
#   dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
# ```
