---
title: "03-edgeR-and-nonlinear-analyses"
author: "Laura Spencer"
date: '2022-11-08'
output: html_document
---


```{r include = FALSE}
knitr::opts_chunk$set(echo=FALSE, warning = F, message = F)
```

```{r, message=FALSE, warning=FALSE, results=FALSE}
#### Load libraries and source scripts 

source("../references/biostats.R")

# Add all required libraries that are installed with install.packages() here
list.of.packages <- c("RCurl", "tidyverse", "vegan", "pheatmap", "pastecs", "factoextra", "FactoMineR", "RColorBrewer", "tibble", "reshape2", "plotly", "cowplot", "clipr", "janitor", "ggpubr", "forcats", "apeglm", "car", "vsn", "devtools", "grid", "gridGraphics", "Rfast", "dendextend", "RColorBrewer", "scales", "VennDiagram", "ape", "lmtest", "limma")

# Add all libraries that are installed using BiocManager here
bioconductor.packages <- c("DESeq2", "WGCNA", "edgeR")

# Install BiocManager if needed
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")

# Get names of all required packages that aren't installed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
new.bioc.packages <- bioconductor.packages[!(bioconductor.packages %in% installed.packages()[, "Package"])]
# Install all new packages
if(length(new.packages)) install.packages(new.packages)
if(length(new.bioc.packages)) BiocManager::install(new.bioc.packages)

# Load all required libraries
all.packages <- c(list.of.packages, bioconductor.packages)
lapply(all.packages, FUN = function(X) {
  do.call("require", list(X))
})

# Github packages 
#install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
library(pairwiseAdonis)

```

### Import counts summary from featurecounts

NOTE: looks like a good portion of reads were not included due to multimapping. Need to look into those! 

```{r}
# Extract sample names for each column from file path 
samples <- (read_delim(file = "../results/star/featurecounts/featurecounts_gene.summary", col_names = F)[1,-1] %>% as.data.frame() %>% t() %>% as.data.frame() %>%
                            mutate(V1=gsub("/home/lspencer/pcod-2022/aligned/star-gadmor/", "", V1)) %>%
                            mutate(V1=gsub(".Aligned.sortedByCoord.out.bam", "", V1)))$V1 %>% as.vector()

# Import count summary from featurecounts
counts.summary <- read_delim(file = "../results/star/featurecounts/featurecounts_gene.summary", col_names = F, skip=1)  %>% as.data.frame()

# Add sample names to count summary table 
colnames(counts.summary) <- c("results", samples)
#ggplotly(
counts.summary %>% 
  filter(rowSums(across(where(is.numeric)))!=0) %>%
pivot_longer(-results, names_to = "sample", values_to = "count") %>%
  mutate(results=fct_rev(factor(results, ordered = T))) %>%
  ggplot(aes(x=sample, y=count, fill=results)) + 
  geom_bar(position="fill", stat="identity") + scale_fill_brewer(palette="Dark2") +
  theme(axis.text.x = element_text(size=6, angle = 90, vjust = 0.5, hjust=1)) +
  theme(legend.position = "bottom") +
  guides(fill=guide_legend(nrow=2,byrow=TRUE, title = ""))#)

```

## Load counts

```{r}
load(file = "../results/star/featurecounts/counts.fc") #raw counts generated from featurecounts
load(file = "../results/counts")
load(file = "../results/counts.t")
load(file = "../results/counts.ts") #counts filtered for low frequency, transposed
counts.ts %>% head()
```

```{r}
# Load counts.treat object created in notebook 02
load(file = "../data/counts.treat")

counts.edgeR <- counts.treat %>% column_to_rownames("sample") %>%
  select(-genes.length) %>% #remove genes associated with fish length
  select(-tank, -temperature, -ph, -treatment) %>% #get rid of extraneous info
  t() %>% as.data.frame()

# double check samples (columns) are in the same order as the samples in the annotated count dataframe. Should = TRUE
all(colnames(counts.edgeR) == counts.treat$sample)

# Extract vectors of treatments
temperature <- counts.treat$temperature
temperature_2 <- temperature^2 #square temperature variable, for potential non-linear effects of temp.
ph <- counts.treat$ph
tank <- counts.treat$tank
treatment <- counts.treat$treatment
```

### Analysis based on MarineOmics tutorial analysis 
#### Non-linear effects, interactive effects 

```{r}
# Create a DGEList object 
y <- DGEList(counts.edgeR, group=treatment)

#NOTE: counts have already been filtered for low-frequency genes, see "01-Importing-data" notebook

# TMM normalization
y <- calcNormFactors(y)
y$samples #check out the normalization factor column 

# check out distribution of normalized, filtered read counts

# Calculate logCPM
df_log <- cpm(y, log = TRUE, prior.count = 2)

# Plot distribution of filtered logCPM values
ggplot(data = data.frame(rowMeans(df_log)), 
       aes(x = rowMeans.df_log.) ) +
  geom_histogram(fill = "grey", colour="black", bins = 50) +
  theme_classic() +
  labs(y = "Density", x = "Filtered read counts (logCPM)",
       title = "Distribution of normalized, filtered read counts")
```

The performance of the TMM normalization procedure can be examined using mean-difference (MD) plots. This visualizes the library size-adjusted log-fold change between two libraries (the difference) against the average log-expression across those libraries (the mean). The following MD plot is generated by comparing sample 1 against an artificial library constructed from the average of all other samples. Ideally, the bulk of genes should be centred at a log-fold change of zero. This indicates that any composition bias between libraries has been successfully removed.

```{r}
for (i in 1:nrow(y$samples)) {
  plotMD(cpm(y, log=T), column = i)
  abline(h=0, col="red", lty=2, lwd=2)
}
```

MDS plot

```{r}
points <- c(0,15,1,16,2,17)
colors <- c("goldenrod1", "goldenrod1", "steelblue1", "steelblue1", "tomato1", "tomato1")

# using top 500 genes (default)
plotMDS(y, col=colors[treatment], pch=points[treatment], cex=1.4, main="MDS plot, PC1 x PC2\ntop 500 genes only")
legend("topleft", legend=levels(treatment), pch=points, col=colors, ncol=2)

# using all genes
plotMDS(y, col=colors[treatment], pch=points[treatment], cex=1.4, top=nrow(y$counts), main="MDS plot, PC1 x PC2\nall genes")
legend("bottomleft", legend=levels(treatment), pch=points, col=colors, ncol=2)
```

## Another MDS plot option, from [MarineOmics tutorial](https://marineomics.github.io/DGE_comparison_v2.html#Intro_to_multifactorial_RNA-seq_models) 

```{r}
# Export pcoa loadings
dds.pcoa = pcoa(vegdist(t(df_log <- cpm(y, log = TRUE, prior.count = 2)),
                          method = "euclidean") / 1000)

# Create df of MDS vector loading
scores <- dds.pcoa$vectors

## Plot pcoa loadings of each sample, groouped by time point and pCO2 treatment

# Calculate % variation explained by each eigenvector
percent <- dds.pcoa$values$Eigenvalues
cumulative_percent_variance <- (percent / sum( percent)) * 100

# Prepare information for pcoa plot, then plot
#colors <- c("goldenrod1", "goldenrod1", "steelblue1", "steelblue1", "tomato1", "tomato1")

par(mfrow = c(1, 1))
plot(
  scores[, 1],
  scores[, 2],
  cex = .5,
  cex.axis = 1,
  cex.lab = 1.25,
  xlab = paste("PC1, ", round(cumulative_percent_variance[1], 2), "%"),
  ylab = paste("PC2, ", round(cumulative_percent_variance[2], 2), "%")
  )

# Add visual groupings to pcoa plot
ordihull(
  scores,
  treatment,
  border = NULL,
  lty = 2,
  lwd = .5,
  label = F,
  col = colors,
  draw = "polygon",
  alpha = 100,
  cex = .5
  )

ordispider(scores, treatment, label = F) # Vectors connecting samples in same ph x treatment group
#ordilabel(scores, cex = 0.4) # Label sample IDs
```

```{r}
# Fit multifactorial design matrix
design <-
  model.matrix(
    ~ 1 + temperature_2 + temperature + ph + temperature_2:ph + temperature:ph) # Generate multivariate edgeR glm

#design <- model.matrix(~ 0 + treatment) 
#colnames(design) <- levels(treatment)

# Check out coefficients 
head(design)
```

### Examine non-linear effects of treatments on gene expression

```{r}
# Estimate dispersion coefficients
y <- estimateDisp(y, design, robust=TRUE)
y$common.dispersion
plotBCV(y)
```

### Fit quasi-likelihood, neg binom linear regression with multifactorial design matrix
```{r}
fit <- glmQLFit(y, design, robust=TRUE)
head(fit$coefficients)
#plotQLDisp(fit)
```

### Test for effects of different treatments 

```{r}
# effect of temperature^2

# estimate sign. degs 
nl_temperature_2 <-
    glmQLFTest(fit,
    coef = 2,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs

# Make contrasts
DEGs_nl_temperature_2 <-
  decideTestsDGE(nl_temperature_2, adjust.method = "fdr", p.value = 0.05)

summary(DEGs_nl_temperature_2)
plotMD(nl_temperature_2, cex=0.6)
```

```{r}
# effect of temperature

# estimate sign. degs 
nl_temperature <-
    glmQLFTest(fit,
    coef = 3,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs

# Make contrasts
DEGs_nl_temperature <-
  decideTestsDGE(nl_temperature, adjust.method = "fdr", p.value = 0.05)

summary(DEGs_nl_temperature)
plotMD(nl_temperature, cex=0.6)
```

```{r}
# effect of low pH relative to ambient pH
# estimate sign. degs 
nl_ph <-
    glmQLFTest(fit,
    coef = 4,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs

DEGs_nl_ph <-
  decideTestsDGE(nl_ph, adjust.method = "fdr", p.value = 0.05)

summary(DEGs_nl_ph)
plotMD(nl_ph, cex=0.6)
```

```{r}
# interaction between temperature_2 & pH

# estimate sign. degs 
nl_interaction_2 <-
    glmQLFTest(fit,
    coef = 5,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs

DEGs_nl_interaction_2 <-
  decideTestsDGE(nl_interaction_2, adjust.method = "fdr", p.value = 0.05)

summary(DEGs_nl_interaction_2)
plotMD(nl_interaction_2, cex=0.6)
```

```{r}
# interaction between temperature & pH

# estimate sign. degs 
nl_interaction <-
    glmQLFTest(fit,
    coef = 6,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs

DEGs_nl_interaction <-
  decideTestsDGE(nl_interaction, adjust.method = "fdr", p.value = 0.05)

summary(DEGs_nl_interaction)
plotMD(nl_interaction, cex=0.6)
```

Checking to see distribution of residual - does our data fit a negative binomial distribution?

Residuals appear normally distributed!  That's great - our data fit the negative binomial distribution assumed by the GLM. Phew! 

```{r}
# Output observed
y_nl <- fit$counts

# Output fitted
mu_nl <- fit$fitted.values

# Output dispersion or coefficient of variation
phi_nl <- fit$dispersion

# Calculate denominator
v_nl <- mu_nl*(1+phi_nl*mu_nl)

# Calculate Pearson residual
resid.pearson <- (y_nl-mu_nl) / sqrt(v_nl)

# Plot distribution of Pearson residuals
ggplot(data = melt(as.data.frame(resid.pearson)), aes(x = value)) +
  geom_histogram(fill = "grey", color="black", bins=50) +
  xlim(-2.5, 5.0) +
  theme_classic() +
  labs(title = "Distribution of negative binomial GLM residuals",
       x = "Pearson residuals",
       y = "Density")
```


```{r}
## Bin genes based on (i) whether they have a significant positive or negative vertex and then (ii) whether they showed significant interactions between beta1 (vertex value) and ph

# Export diff expression data for transcripts with significant DE associated with temperature^2 parameter
nl_temperature_2_sig <-
  topTags(
  nl_temperature_2,
  n = sum(abs(DEGs_nl_temperature_2)),
  adjust.method = "BH",
  p.value = 0.05
  )
nl_temperature_2_sig_geneids <- row.names(nl_temperature_2_sig) #Output a list of geneids associated with sig temperature effect

# Export diff expression data for transcripts with significant DE associated with temperature parameter
nl_temperature_sig <-
  topTags(
  nl_temperature,
  n = sum(abs(DEGs_nl_temperature)),
  adjust.method = "BH",
  p.value = 0.05
  )
nl_temperature_sig_geneids <- row.names(nl_temperature_sig) #Output a list of geneids associated with sig temperature effect

# Export diff expression data for transcripts with significant DE associated with ph parameter
nl_ph_sig <-
  topTags(
  nl_ph,
  n = sum(abs(DEGs_nl_ph)),
  adjust.method = "BH",
  p.value = 0.05
  )
nl_ph_sig_geneids <- row.names(nl_ph_sig) #Output a list of geneids associated with sig ph effect

# Export diff expression data for transcripts with significant DE associated with ph parameter
nl_interaction_sig <-
  topTags(
  nl_interaction,
  n = sum(abs(DEGs_nl_interaction)),
  adjust.method = "BH",
  p.value = 0.05
  )
nl_interaction_sig_geneids <- row.names(nl_interaction_sig) #Output a list of geneids associated with sig temperature:ph interactiom  effect

# Create tabulated dataframe of mean expression across each temperature level with metadata for transcript ID and ph
tab_exp_df <- as.data.frame(df_log) %>% rownames_to_column("geneid") %>%
  pivot_longer(cols = -geneid) %>% 
  left_join(counts.treat[c("sample", "temperature", "ph")], 
            by=c("name"="sample")) %>%
  mutate(temperature_2=temperature^2)

# Create binary variable in df_all_log for significant expression
tab_exp_df$temperature_sig <-
  ifelse(tab_exp_df$geneid %in% nl_temperature_sig_geneids, "Yes", "No")
tab_exp_df$temperature_2_sig <-
  ifelse(tab_exp_df$geneid %in% nl_temperature_2_sig_geneids, "Yes", "No")
tab_exp_df$ph_sig <-
  ifelse(tab_exp_df$geneid %in% nl_ph_sig_geneids, "Yes", "No")
tab_exp_df$interact_sig <-
  ifelse(tab_exp_df$geneid %in% nl_interaction_sig_geneids, "Yes", "No")

# Create a binary variable related to up or down-regulation
up_genes <- filter(nl_temperature_sig$table, logFC > 0)
tab_exp_df$logFC_dir <-
  ifelse(tab_exp_df$geneid %in% row.names(up_genes), "Up", "Down")

# Add geneid to edgeR interaction results
nl_interaction$coefficients$geneid <- row.names(nl_interaction$coefficients)
nl_interaction_2$coefficients$geneid <- row.names(nl_interaction_2$coefficients)

# Estimate average logCPM per gene per pH treatment
library(Rmisc)
tab_exp_avg <- summarySE(
  measurevar = "value",
  groupvars = c("temperature","ph","geneid","temperature_sig", "temperature_2_sig", "logFC_dir"),
  data = tab_exp_df)

# First exploratory plot of non-linear expression across temperatures grouped by pH (aka pco2) and direction of differential expression
ggplot(data = filter(tab_exp_avg, temperature_2_sig == "Yes"), 
       aes(x = temperature, y = value)) +
  geom_path(
    alpha = 0.05,
    size = 0.25,
    stat = "identity",
    aes(group = as.factor(geneid))) +
  scale_x_continuous(breaks=c(3,6,10), labels=c(3,6,10), limits=c(2,11)) +
  facet_grid(logFC_dir ~ ph) +
  theme_classic() +
  theme(strip.background = element_blank()) +
  labs(y = "Avg logCPM", title = "Non-linear changes in GE output by edgeR")
```

### How many genes are identified as having non-linear expression patterns according to edgeR? 

```{r}
filter(tab_exp_df, temperature_2_sig == "Yes") %>% 
  select(geneid) %>% distinct(geneid) %>% nrow() #458
```
Determine whether it is more probable that gene's expression is linear or non-linear relative to a continuous predictor. We can calculate this relative probability using a likelihood ratio test (LRT). In the code chunk below, we will fit linear and 2nd order non-linear models to the expression of each gene before applying LRTs to each transcript. 

```{r}
## Using dlply, fit linear and non-linear models to each gene

# Create temperature^2 variable in df_all_log
tab_exp_df$temperature_2 <- tab_exp_df$temperature^2

# Fit linear models - should take about 4 minutes
lms <- dlply(tab_exp_df, c("geneid"), function(df) 
lm(value ~ temperature + ph + temperature:ph, data = df))

# Fit non-linear models - should take about 2 minutes
nlms <- dlply(tab_exp_df, c("geneid"), function(df) 
lm(value ~ temperature + temperature_2 + ph + temperature:ph + temperature_2:ph, data = df))

# Output nlm coefficients into dataframe
nlms_coeff <- ldply(nlms, coef)
head(nlms_coeff)

# Output lm coefficients into dataframe
lms_coeff <- ldply(lms, coef)
head(lms_coeff)

## Apply LRTs to lm's and nlm's for each transcript - should take about 2 minutes
lrts <- list() # Create list to add LRT results to

for (i in 1:length(lms)) {
  lrts[[i]] <- lrtest(lms[[i]], nlms[[i]]) # Apply LRTs with for loop
}

## Filter lrt results for transcripts with significantly higher likelihoods of nl model
lrt_dfs <- list()

# Turn list of LRT outputs into list of dataframes containing output info
for (i in 1:length(lrts)) {
  lrt_dfs[[i]] <- data.frame(lrts[i])
}

# Create singular dataframe with geneids and model outputs for chi-squared and LRT p-value
lrt_coeff_df  <- na.omit(bind_rows(lrt_dfs, .id = "column_label")) # na.omit removes first row of each df, which lacks these data

# Add geneid based on element number from original list of LRT outputs
lrt_coeff_df <- merge(lrt_coeff_df,
                      data.frame(geneid = names(nlms),
                      column_label = as.character(seq(length(
                      nlms
                      )))),
                      by = "column_label")
                      
# Apply FDR adjustment to LRT p-values before filtering for sig non-linear effects
lrt_coeff_df$FDR <- p.adjust(lrt_coeff_df$Pr..Chisq., method = "fdr")

# Filter LRT results for sig FDR coeff
lrt_filt <- filter(lrt_coeff_df, FDR < 0.01)

## Plot sig nl genes according to LRT, grouped by ph and direction of beta 1 coefficient
# Add beta coefficients to logCPM df
temperature_pos <- filter(nlms_coeff, temperature > 0)
temperature_2_pos <- filter(nlms_coeff, temperature_2 > 0)
temperature_pos_linear <- filter(lms_coeff, temperature > 0)

# Bin genes based on positive or negative temperature and temperature^2 betas
tab_exp_avg$temperature_binom <- ifelse(tab_exp_avg$geneid %in% temperature_pos$geneid, "Positive", "Negative")
tab_exp_avg$temperature_2_binom <- ifelse(tab_exp_avg$geneid %in% temperature_2_pos$geneid, "Concave", "Convex")
tab_exp_avg$temperature_binom_linear <- ifelse(tab_exp_avg$geneid %in% temperature_pos_linear$geneid, "Positive", "Negative")

# Filter for how many gene id's with significant likelihood of nl effect in LRT
LRT_filt_df <- filter(tab_exp_avg, geneid %in% lrt_filt$geneid)

# Plot
ggplot(data = LRT_filt_df,
       aes(x = temperature, y = value)) +
  geom_path(
    alpha = .025,
    size = 0.25,
    stat = "identity",
    aes(group = as.factor(geneid))
    ) +
  facet_grid(temperature_2_binom ~ ph) +
  #geom_smooth(method = "loess", se = TRUE, span = 1) +
  theme_classic() +
  theme(strip.background = element_blank()) +
  labs(y = "Avg logCPM", title = "Non-linear changes in GE output by LRTs")
```
### How many genes are identified as having non-linear expression patterns according to the LRT tests? 1389

```{r}
nrow(as.data.frame(unique(LRT_filt_df$geneid)))
```
```{r, warning=F, message=FALSE}
# Filter down df for gene id's exhibit temperature significant effect in edgeR and significant likelihood of nl effect in LRT 
edgeR_LRT_df <- filter(tab_exp_avg, geneid %in% lrt_filt$geneid & temperature_sig == "Yes" |
                         geneid %in% lrt_filt$geneid & temperature_2_sig == "Yes")

# Plot non-linear effects of temperature, by ph
ggplot(data = edgeR_LRT_df, 
       aes(x = temperature, y = value)) +
  geom_path(
    alpha = .25,
    size = 0.25,
    stat = "identity",
    aes(group = as.factor(geneid))
    ) +
  facet_grid(temperature_2_binom~ph) +
  geom_smooth(method = "loess", se = TRUE, span = 1) +
  theme_classic() +
  theme(strip.background = element_blank()) +
  labs(y = "Avg logCPM", title = "Non-linear changes in GE output by edgeR & LRTs")
```

### How many genes are identified as having non-linear expression according to both edgeR and LRT and also a DEG? 287 

```{r}
edgeR_LRT_df %>% select(geneid) %>%
  filter(geneid %in% unique(unlist(c(lapply(degs, rownames))))) %>% #filter for only DEGs from DESeq2
  distinct() %>% nrow()
```
```{r}
# Filter down df containing sign. non-linear genes (from edger AND LRT) for only those with sign. interaction betwen temp^2 and pH AND are DEGs from DESeq2
nonlinear_df <- edgeR_LRT_df %>%
  filter(geneid %in% unique(unlist(c(lapply(degs, rownames))))) #filter for only DEGs from DESeq2

nonlinear_df$gene_id_ph <- paste(nonlinear_df$geneid,
                                           nonlinear_df$ph,
                                           sep = "_")

# Add annotation annotation information 
nonlinear_df <- nonlinear_df %>%
               left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names", "spid")],
                         by=c("geneid"="gene_gadmor"))

# How many genes have nonlinear expression
nonlinear_df %>% distinct(geneid) %>% nrow()
```
```{r}
nonlinear_df %>% filter(temperature_2_binom=="Concave") %>% distinct(geneid) #how many concave pattern?
nonlinear_df %>% filter(temperature_2_binom=="Convex") %>% distinct(geneid) #how many convex pattern?
```


```{r}
# Average logCPM across different groups according to temperature^2 estimate and ph
nonlinear_degs_avg <- 
  summarySE(measurevar = "value",
            groupvars = c("ph", "temperature", "temperature_2_binom"),
            data = nonlinear_df)  %>%
  mutate(temperature_2_binom = factor(ifelse(temperature_2_binom == "Concave", "V-shaped in ambient", "A-shaped in ambient")))

# Plot 
pdf(file="../results/edgeR/Nonlinear-response-genes-plot.pdf", height = 2.75, width = 3)
ggplot(data = nonlinear_degs_avg %>%
         filter(ph==370), 
       aes(
         x = temperature,
         y = value,
         #color = as.factor(ph),
         color = as.factor(temperature_2_binom),
         group = as.factor(temperature_2_binom)#,
         #shape=as.factor(ph)
         )) +
         geom_path(stat = "identity",
                   aes(group=as.factor(temperature_2_binom))
                   #aes(linetype=interaction(as.factor(ph):as.factor(temperature_2_binom)))
                   ) +
  geom_errorbar(aes(ymin = value - se, ymax = value + se), width = 0) +
  geom_point(size=2.25) +
  #facet_wrap(~temperature_2_binom) +
  theme_classic() +
  #xlim(3,10) + 
  #ylim(3.25,5) +
  scale_x_discrete(limits=c(3,6,10)) +
  theme(strip.background = element_blank()) +
  ggtitle(NULL) + #, "Average expression of genes with non-linear pattern across temperature\nthat is dependent on pCO2"
  labs(y = NULL, x=NULL)  +
  scale_color_manual(values=c(Increase="darkgreen", Decrease="purple4", 
                              `V-shaped in ambient`=lighten("darkgreen", 0.45), 
                              `A-shaped in ambient`=lighten("purple4", 0.45)), guide="none") #+
  # scale_shape_manual(values=c(19, 17), guide="none") +
  # scale_linetype_manual(values=c("solid", "solid", "dashed", "dashed"), 
  #                       #limits=c("370:Decrease","950:Decrease"),
  #                       # labels=c("370:Decrease"="Ambient pCO2", 
  #                       #          "370:Increase"=NA, "950:Increase"=NA,
  #                       #          "950:Decrease"="High pCO2")
  #                       guide="none") #+
#  guides(colour = guide_legend(order=1))
dev.off()

# NOTE - this shows the average expression across all genes by group, which was calculated from averages of each gene by group. could consider another way of showing this data that doesn't simplify as much.   
```
```{r}
nonlinear.4David <- list(
  
  "Convex" = 
  nonlinear_df %>% filter(temperature_2_binom=="Convex") %>% distinct(geneid) %>%
  left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names", "spid")], #annotate with blast results
                         by=c("geneid"="gene_gadmor")) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  "Concave" = 
  nonlinear_df %>% filter(temperature_2_binom=="Concave") %>% distinct(geneid) %>%
  left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names", "spid")], #annotate with blast results
                         by=c("geneid"="gene_gadmor")) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector()

  ) 

# Write tab-delimited file with each column containing Uniprot IDs for each set of DEGs 
write_delim(x = ldply(nonlinear.4David, rbind) %>% t() %>% as.data.frame(), delim = "\t", col_names = F, na="", 
            file = "../results/GO-enrichment/nonlinear-4David.tab")
            
# BACKGROUND - all genes submitted to nonlinear anlaysis
gadMor.blast.GO %>% filter(gene_gadmor %in% rownames(counts.edgeR[-1,])) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>%
  write_clip(allow_non_interactive = TRUE)
```

#### Read in enriched biological functions for all patterns

```{r}
filenames <- read_delim(file="../results/GO-enrichment/Final-DAVID/GO_filenames_nonlinear.txt", delim = "\t", col_names = c("filename", "gene_set"))
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=paste("../results/GO-enrichment/Final-DAVID/", filenames[i,"filename"], sep=""), delim = "\t"))}

david.nonlinear <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    separate(gene_set, sep = "_", into = c("shape", "pattern"), remove = F) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term_id", "term_process"), sep = "~") %>%
  mutate_at(vars(shape, pattern), as.factor)
```
#### Enriched Biological Process GO Terms / Uniprot Keywords - genes with non-linear expression patterns with temperature

```{r}
#pdf(file="../results/GO-enrichment/Final-DAVID/Nonlinear-GO-bubble.pdf", height = 3.75, width = 3.75)
pdf(file="../results/GO-enrichment/Final-DAVID/Nonlinear-KW-bubble.pdf", height = 2.30, width = 3.25)
david.nonlinear %>%
#  filter(category=="GOTERM_BP_DIRECT") %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
#  filter(category=="KEGG_PATHWAY") %>%
#  separate(term_id, into = c("term_id", "term_process"), sep = ":") %>%
  filter(p_value<0.05) %>%
#  filter(pattern=="Convex") %>% 
  group_by(pattern, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david.nonlinear %>% 
#              separate(term_id, into = c("term_id", "term_process"), sep = ":") %>% 
              select(pattern, category, term_id, term_process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("pattern", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(pattern, p_value) %>% group_by(pattern) %>% #dplyr::slice(1:20) %>% # select the top n terms per group
  mutate(pattern=factor(pattern, levels=c("Convex", "Concave"), ordered = TRUE)) %>%
  arrange(pattern, p_value) %>%
#  arrange(pattern, term_id) %>%
#  ungroup() %>% arrange(term_id) %>%
  mutate(term_process = str_wrap(term_process, width = 40)) %>%
  mutate(term_process = factor(term_process, rev(unique(term_process)), ordered = TRUE)) %>%

  # plot
ggplot(aes(y = term_process, x=pattern, col=pattern)) + 
  geom_point(alpha=.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Pattern across\ntemperature",
                     values=c(Increase="darkgreen", Decrease="purple4", 
                              Concave=lighten("darkgreen", 0.45), Convex=lighten("purple4", 0.45)),
             guide = guide_legend(order = 1, override.aes = list(size=4))) +
  # scale_shape_manual(name="Significant pH interaction effect", 
  #                    labels=c(`interaction`="Yes", `nointeraction`='No'),
  #                    values=c(`interaction`=19, `nointeraction`=17),
  #            guide = guide_legend(order = 2, override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab(NULL) + xlab(NULL) +
  # coord_cartesian(clip = "off", ylim = c(.5, 38)) +
   ggtitle("Biological Processes") #+
  # annotate("text", x = 1, y = -1.5, label = "Decrease with\ntemperature", size=2, col="gray15") +
  # annotate("text", x = 2, y = -1.5, label = "Increase with\ntemperature", size=2, col="gray15") +
  # annotate("text", x = 2.5, y = 38, label = "Enriched Biological Processes", size=2.5, col="gray15") +
  # annotate("text", x = 2.5, y = 37.25, label = "Non-linear response to temperature", size=2.5, col="gray15", fontface = 'bold') +
  # annotate("text", x = 2.5, y = 36.5, label = "by pH interaction effect ", size=2.5, col="gray15", fontface = 'italic')
dev.off()
```

Copy DAVID results (Uniprot Kewords and Gene Ontology terms) to paste into Google Spreadsheet 

```{r}
# Uniprot KW
david.nonlinear %>% 
  filter(p_value<0.05) %>%
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
  arrange(pattern, p_value) %>%
  mutate_at(c("p_value", "fdr"), funs(signif(.,2))) %>%
  dplyr::select(pattern, term_id, term_process, count, p_value, fdr, genes) %>%
  write_clip()

# GO terms
david.nonlinear %>% 
  filter(p_value<0.05) %>%
  filter(category=="GOTERM_BP_DIRECT") %>%
  arrange(pattern, p_value) %>%
  mutate_at(c("p_value", "fdr"), funs(signif(.,2))) %>%
  dplyr::select(pattern, term_id, term_process, count, p_value, fdr, genes) %>%
  write_clip()
```

Enriched GO biological processes - Network Diagrams using metacoder

Here I save an R objects from GO enrichment results of the linear/nonlinear response genes to visualize in network plots - the visualization is done in RStudio Cloud, as it requires a specific older version of R and some R packages. 

```{r}
# Save R object to then import to RStudio Cloud for use in metacoder 
save(david.nonlinear, file="../results/GO-enrichment/Final-DAVID/david.nonlinear")
```









### More analysis, looking at interaction effect and linear response genes - not used in paper!

But for which genes is there a significant interaction between pH and temperature^2? 

```{r}
# Export diff expression data for genes with significant DE associated with interaction between temperature^2 and ph
nl_interaction_2_sig <- topTags(nl_interaction_2, n = sum(abs(DEGs_nl_interaction_2)), adjust.method = "BH",p.value = 0.05)
nl_interaction_2_sig_geneids <- row.names(nl_interaction_2_sig) #Output a list of geneids associated with sig temperature^2 x ph interaction

# genes which are non-linear with temperature AND for which expression pattern depends on pH
edgeR_LRT_df %>%  # dataframe with genes which are deemed non-linear as per edger AND LRT
    filter(geneid %in% nl_interaction_2_sig_geneids) %>% # genes with significant interaction between temp^2 and pH as per edgeR analysis
    filter(geneid %in% unique(unlist(c(lapply(degs, rownames))))) %>% #filter for only DEGs from DESeq2
    select(geneid) %>% unique() %>% nrow() # count those 

# Filter down df containing sign. non-linear genes (from edger AND LRT) for only those with sign. interaction betwen temp^2 and pH AND are DEGs from DESeq2
nonlinear_interaction_df <- filter(edgeR_LRT_df, geneid %in% nl_interaction_2_sig_geneids ) %>%
                            filter(geneid %in% unique(unlist(c(lapply(degs, rownames))))) #filter for only DEGs from DESeq2

nonlinear_interaction_df$gene_id_ph <- paste(nonlinear_interaction_df$geneid,
                                           nonlinear_interaction_df$ph,
                                           sep = "_")

# Add annotation annotation information 
nonlinear_interaction_df <- nonlinear_interaction_df %>%
               left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names", "spid")],
                         by=c("geneid"="gene_gadmor"))
```

```{r}
### Generate a lineplot showing average expression by temperature, ph, and expression pattern

# Average logCPM across different groups according to temperature^2 estimate and ph
edgeR_LRT_df_avg <- summarySE(measurevar = "value",
                                   groupvars = c("ph", "temperature", "temperature_2_binom"),
                                                 data = edgeR_LRT_df)

# Lineplot
ggplot(data = edgeR_LRT_df_avg, 
       aes(
         x = temperature,
         y = value,
         color = as.factor(ph),
         group = as.factor(temperature_2_binom)
         )) +
         geom_path(stat = "identity") +
  geom_errorbar(aes(ymin = value - se, ymax = value + se), width = 0) +
  geom_point() +
  facet_wrap(~ph) +
  theme_classic() +
  xlim(3,10) +
  scale_x_discrete(limits=c(3,6,10)) +
  theme(strip.background = element_blank()) +
  labs(y = "logCPM", color = "pCO2 treatment", 
       title = "Non-linear changes in GE across temperature\nby ph and expression pattern(edgeR & LRT consensus sets)") +
  scale_color_manual(values=c("blue", "red"))
```

Exploring through gene-level plots - genes that respond non-linearly to temperature and for which pH affects that expression:

```{r}
# Generate list of genes with interaction for a) those with A-shape in ambient pH, and b) those with V-shape in ambient pH 
  
# a) A-shape expression across temperature in ambient pH 
nonlinear_interaction_Ashaped <- 
  nl_interaction_2_sig %>% as.data.frame() %>% rownames_to_column("geneid") %>% #convert edgeR model results to dataframe to access stats
  filter(geneid %in% unique(nonlinear_interaction_df$geneid)) %>% # select only genes that are nonlinear+interaction 
  left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names", "spid")], #annotate with blast results
                         by=c("geneid"="gene_gadmor")) %>%
  filter(logFC > 0) #select genes that have A-shape expression pattern in ambient pH (convex shape)

# b) V-shape expression across temperature in ambient pH 
nonlinear_interaction_Vshaped <- 
  nl_interaction_2_sig %>% as.data.frame() %>% rownames_to_column("geneid") %>% #convert edgeR model results to dataframe to access stats
  filter(geneid %in% unique(nonlinear_interaction_df$geneid)) %>% # select only genes that are nonlinear+interaction 
  left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names", "spid")], #annotate with blast results
                         by=c("geneid"="gene_gadmor")) %>%
  filter(logFC < 0) #select genes that have A-shape expression pattern in ambient pH (convex shape)

# select top 25 by p-value that are annotated to plot 
genes <- nonlinear_interaction_Ashaped %>% 
  filter(!is.na(spid)) %>% #remove genes that are not annotated
  arrange(PValue) %>% dplyr::slice(1:25) %>% #arrange by pvalue, select top 25 with lowest pvalue 
  select(geneid) %>% unlist() %>% as.vector() #retain only geneid in vector object

genes <- nonlinear_interaction_Vshaped %>%
    filter(!is.na(spid)) %>% #remove genes that are not annotated
  arrange(PValue) %>% dplyr::slice(1:25) %>% #arrange by pvalue, select top 25 with lowest pvalue 
  select(geneid) %>% unlist() %>% as.vector() #retain only geneid in vector object

for (i in 1:length(genes)) {
  # Plot individual genes 
print(ggplot(data = nonlinear_interaction_df %>% filter(geneid == genes[i]), 
       aes(
         x = temperature,
         y = value,
         color = as.factor(ph),
         group = as.factor(ph)
         )) +
         geom_path(stat = "identity") +
  geom_errorbar(aes(ymin = value - se, ymax = value + se), width = 0) +
  geom_point() +
  #facet_wrap(~temperature_2_binom) +
  theme_classic() +
  xlim(3,10) +
  #scale_x_continuous(limits=c(3,6,10)) +
  theme(strip.background = element_blank()) +
  labs(y = "logCPM", color = "pCO2 treatment", 
       title = paste(interact_top_25[i], "\n",
       str_wrap(nonlinear_interaction_df[nonlinear_interaction_df$geneid==interact_top_25[i], 
       "protein_names"] %>% unique()))) +
  scale_color_manual(values=c("blue", "red")))
}
```

### Look for possible patterns in genes that have nonlinear expression with interactive effects by counting the number of genes with DEGs among different treatments

```{r}
names(degs)

# For Scenarios, see slides - https://docs.google.com/presentation/d/1fdBQIfGAU1BG5Kba_EEjATBfKv_aw3PMj5P9NDMEE7M/edit?usp=sharing 
  
# Scenarios A. B. E. & I. - no genes
intersect(degs[["Amb vs. Low pH @3C"]] %>% rownames(),  degs[["Amb vs. Low pH @6C"]] %>% rownames()) 

# Scenarios L. N. O. P. - no genes
intersect(degs[["Amb vs. Low pH @3C"]] %>% rownames(),  degs[["Amb vs. Low pH @ 10C"]] %>% rownames())

# Scenario C. - 27 gene
Reduce(intersect, list(degs[["Amb vs. Low pH @6C"]] %>% rownames(),
          degs[["3 vs. 6C @Amb pH"]] %>% rownames(), degs[["6 vs. 10C @Amb pH"]] %>% rownames())) %>% length() 

# Scenario D. - 1 gene
Reduce(intersect, list(degs[["Amb vs. Low pH @6C"]] %>% rownames(),  degs[["Amb vs. Low pH @ 10C"]] %>% rownames(),
          degs[["3 vs. 6C @Amb pH"]] %>% rownames(), degs[["6 vs. 10C @Amb pH"]] %>% rownames(), 
          degs[["6 vs. 10C @Low pH"]] %>% rownames())) %>% length() 

# Scenario F. - 2 genes
Reduce(intersect, list(degs[["Amb vs. Low pH @6C"]] %>% rownames(),  degs[["Amb vs. Low pH @ 10C"]] %>% rownames(),
          degs[["3 vs. 6C @Amb pH"]] %>% rownames(), degs[["3 vs. 10C @Amb pH"]] %>% rownames())) %>% length() 

# Scenario G. - 325 genes
Reduce(intersect, list(degs[["Amb vs. Low pH @6C"]] %>% rownames(), 
          degs[["3 vs. 10C @Amb pH"]] %>% rownames(), degs[["6 vs. 10C @Amb pH"]] %>% rownames(),
          degs[["3 vs. 6C @Low pH"]] %>% rownames(), degs[["3 vs. 10C @Low pH"]] %>% rownames())) %>% length() 

# Scenario G. - 10 genes
Reduce(intersect, list(degs[["Amb vs. Low pH @6C"]] %>% rownames(), 
          degs[["3 vs. 6C @Amb pH"]] %>% rownames(), degs[["3 vs. 10C @Amb pH"]] %>% rownames(),
          degs[["3 vs. 10C @Low pH"]] %>% rownames(), degs[["6 vs. 10C @Low pH"]] %>% rownames())) %>% length() 

# Scenario J. -  3 genes
Reduce(intersect, list(degs[["Amb vs. Low pH @6C"]] %>% rownames(),  degs[["Amb vs. Low pH @ 10C"]] %>% rownames(),
          degs[["3 vs. 10C @Amb pH"]] %>% rownames(), degs[["6 vs. 10C @Amb pH"]] %>% rownames(),
          degs[["3 vs. 6C @Low pH"]] %>% rownames(), degs[["6 vs. 10C @Low pH"]] %>% rownames())) %>% length() 

# Scenario K, R - 89 genes
Reduce(intersect, list(degs[["Amb vs. Low pH @6C"]] %>% rownames(),
          degs[["3 vs. 6C @Low pH"]] %>% rownames(), degs[["6 vs. 10C @Low pH"]] %>% rownames())) %>% length() 

# Scenario M. - 0 genes
Reduce(intersect, list(degs[["Amb vs. Low pH @3C"]] %>% rownames(),
          degs[["3 vs. 6C @Amb pH"]] %>% rownames(), degs[["6 vs. 10C @Amb pH"]] %>% rownames(),
          degs[["3 vs. 10C @Low pH"]] %>% rownames(), degs[["6 vs. 10C @Low pH"]] %>% rownames())) %>% length() 

# Scenario S. - 2 genes
Reduce(intersect, list(degs[["Amb vs. Low pH @6C"]] %>% rownames(),  degs[["Amb vs. Low pH @ 10C"]] %>% rownames(),
          degs[["3 vs. 6C @Low pH"]] %>% rownames(), degs[["3 vs. 10C @Low pH"]] %>% rownames())) %>% length() 

# Scenario V. - 1 genes
Reduce(intersect, list(degs[["Amb vs. Low pH @6C"]] %>% rownames(),  degs[["Amb vs. Low pH @ 10C"]] %>% rownames(),
          degs[["3 vs. 6C @Amb pH"]] %>% rownames(), degs[["6 vs. 10C @Amb pH"]] %>% rownames(),
          degs[["3 vs. 10C @Low pH"]] %>% rownames(), degs[["6 vs. 10C @Low pH"]] %>% rownames())) %>% length() 
```

Subset nonlinear interactive gene set for only those that are also DEGs from DESeq2

```{r}
# a) A-shape expression across temperature in ambient pH 
nonlinear_interaction_degs_Ashaped <- nonlinear_interaction_Ashaped %>% 
  filter(geneid %in% c(lapply(degs, rownames) %>% unlist() %>% unique()))  #select genes that are DEGs 

# b) V-shape expression across temperature in ambient pH 
nonlinear_interaction_degs_Vshaped <- nonlinear_interaction_Vshaped %>% 
  filter(geneid %in% c(lapply(degs, rownames) %>% unlist() %>% unique()))   #select genes that are DEGs 

nrow(nonlinear_interaction_degs_Ashaped) #how many A-shaped in ambient pH? 334
nrow(nonlinear_interaction_degs_Vshaped) #how many V-shaped in ambient pH? 158
```

Here also is a list of genes that respond nonlinearly to temperature AND are a DEG, but for which pH does not impact the shape. 

```{r}
nonlinear_no.interaction_degs <- edgeR_LRT_df %>% filter(geneid %!in% 
                          c(nonlinear_interaction_Ashaped$geneid, nonlinear_interaction_Vshaped$geneid)) %>%
  filter(geneid %in%  c(lapply(degs, rownames) %>% unlist() %>% unique())) 

# How many are there? 33
nonlinear_no.interaction_degs$geneid %>% unique() %>% length()
 
# Plot to make sure convex / concave shapes correctly describe the data
genes <- (nonlinear_no.interaction_degs %>% filter(temperature_2_binom=="Convex"))$geneid %>% unique() 
length(genes)

genes <- (nonlinear_no.interaction_degs %>% filter(temperature_2_binom=="Concave"))$geneid %>% unique()  
length(genes)

for (i in 1:length(genes)) {
  # Plot individual genes 
print(ggplot(data = nonlinear_no.interaction_degs %>% 
               filter(geneid == genes[i]), 
       aes(
         x = temperature,
         y = value ,
         color = as.factor(ph),
         group = as.factor(ph)
         )) +
         geom_path(stat = "identity") +
  geom_errorbar(aes(ymin = value - se, ymax = value + se), width = 0) +
  geom_point() +
  #facet_wrap(~temperature_2_binom) +
  theme_classic() +
  xlim(3,10) +
  #scale_x_continuous(limits=c(3,6,10)) +
  theme(strip.background = element_blank()) +
  labs(y = "logCPM", color = "pCO2 treatment", 
       title = genes[i]) +
  scale_color_manual(values=c("blue", "red")))
}
```


```{r}
# Here I plot genes that have 
# 1) nonlinear expression pattern across temperature as deemed by edgeR & LRT, which is 
# 2) dependent on pH (significant interaction) as deemed by edger, and 
# 3) is differentially expressed among pH or temperature as deemed by DESeq2 
# 4) are in the top 25 genes - i.e. those with the lowest p-values (since there's too many to plot)

genes <- nonlinear_interaction_degs_Ashaped %>% 
  filter(!is.na(spid)) %>% #remove genes that are not annotated
  arrange(PValue) %>% dplyr::slice(1:50) #arrange by pvalue, select top 25 with lowest pvalue 

genes <- nonlinear_interaction_degs_Vshaped %>%
  filter(!is.na(spid)) %>% #remove genes that are not annotated
  arrange(PValue) %>% dplyr::slice(1:25) #arrange by pvalue, select top 25 with lowest pvalue 

for (i in 1:nrow(genes)) {
  # Plot individual genes 
print(ggplot(data = nonlinear_interaction_df %>% 
               filter(geneid == genes$geneid[i]), 
       aes(
         x = temperature,
         y = value,
         color = as.factor(ph),
         group = as.factor(ph)
         )) +
         geom_path(stat = "identity") +
  geom_errorbar(aes(ymin = value - se, ymax = value + se), width = 0) +
  geom_point() +
  #facet_wrap(~temperature_2_binom) +
  theme_classic() +
  xlim(3,10) +
  #scale_x_continuous(limits=c(3,6,10)) +
  theme(strip.background = element_blank()) +
  labs(y = "logCPM", color = "pCO2 treatment", 
       title = paste(genes$geneid[i], "\n",
       str_wrap(genes[genes$geneid==genes$geneid[i], 
       "protein_names"] %>% unique()))) +
  scale_color_manual(values=c("blue", "red")))
}
```

Generate heat map with those genes 

```{r}
ph.low <- sample.info[sample.info$ph=="low",]$vial_label
ph.ambient <- sample.info[sample.info$ph=="amb",]$vial_label

# Generate counts matrix with non-linear response genes for LOW pH treatment only 
counts.nl.interact.degs.low <-  
  (subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% 
            c(nonlinear_interaction_degs_Ashaped$geneid, 
             nonlinear_interaction_degs_Vshaped$geneid)) %>% 
     as.data.frame())[ph.low] %>% as.matrix()

# Create a dataframe to annotate heatmap with treatments 
counts.nl.interact.degs.low.df <- sample.info[match(colnames(counts.nl.interact.degs.low), sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")] %>%
  remove_rownames() %>% column_to_rownames(var = "vial_label")

# Make sure treatment order is correct
all(colnames(counts.nl.interact.degs.low) == rownames(counts.nl.interact.degs.low.df)) #double check that samples are still in same order 

# All DEGs among any pH treatment (within temperatures)
pheatmap(counts.nl.interact.degs.low, cluster_rows=TRUE, cluster_cols=FALSE,
         show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=
             c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen",
               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")),
         scale="row", main = "Differentially expressed genes (DEGs) among any pH treatment", 
         annotation_col=counts.nl.interact.degs.low.df[,"treatment", drop=FALSE], fontsize = 8,
         cutree_rows = 2,  border_color=F)


# Generate counts matrix with non-linear response genes for AMBIENT pH treatment only 
counts.nl.interact.degs.amb <-  
  (subset(assay(vsd.treatment), rownames(dds.DESeq.treatment) %in% 
            c(nonlinear_interaction_degs_Ashaped$geneid, 
             nonlinear_interaction_degs_Vshaped$geneid)) %>% 
     as.data.frame())[ph.amb] %>% as.matrix()

# Create a dataframe to annotate heatmap with treatments 
counts.nl.interact.degs.amb.df <- sample.info[match(colnames(counts.nl.interact.degs.amb), sample.info$vial_label),c("vial_label", "treatment", "tank", "temperature", "ph")] %>%
  remove_rownames() %>% column_to_rownames(var = "vial_label")

# Make sure treatment order is correct
all(colnames(counts.nl.interact.degs.amb) == rownames(counts.nl.interact.degs.amb.df)) #double check that samples are still in same order 

# All DEGs among any pH treatment (within temperatures)
pheatmap(counts.nl.interact.degs.amb, cluster_rows=TRUE, cluster_cols=FALSE,
         show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(treatment=
             c("3_amb"= "navyblue", "3_low"="royalblue1","6_amb"="darkgreen",
               "6_low"="yellow3","10_amb"="firebrick4", "10_low"="sienna2")),
         scale="row", main = "Differentially expressed genes (DEGs) among any pH treatment", 
         annotation_col=counts.nl.interact.degs.amb.df[,"treatment", drop=FALSE], fontsize = 8,
         cutree_rows = 2,  border_color=F)
```

Generate single plot with all genes for which expression patterns respond non-linearly to temperature with significant interaction AND are DEGs

```{r}
# Average logCPM across different groups according to temperature^2 estimate and ph
nonlinear_interaction_degs_avg <- 
  summarySE(measurevar = "value",
            groupvars = c("ph", "temperature", "temperature_2_binom"),
            data = nonlinear_interaction_df %>% 
              filter(geneid %in% c(nonlinear_interaction_degs_Ashaped$geneid, 
                                   nonlinear_interaction_degs_Vshaped$geneid)))  %>%
  mutate(temperature_2_binom = factor(ifelse(temperature_2_binom == "Concave", "V-shaped in ambient", "A-shaped in ambient")))

# Plot 
#pdf(file="../results/edgeR/Nonlinear-response-genes-plot.pdf", height = 2.75, width = 3)
ggplot(data = nonlinear_interaction_degs_avg %>%
         filter(ph==370), 
       aes(
         x = temperature,
         y = value,
         #color = as.factor(ph),
         color = as.factor(temperature_2_binom),
         group = as.factor(temperature_2_binom)#,
         #shape=as.factor(ph)
         )) +
         geom_path(stat = "identity",
                   aes(group=as.factor(temperature_2_binom))
                   #aes(linetype=interaction(as.factor(ph):as.factor(temperature_2_binom)))
                   ) +
  geom_errorbar(aes(ymin = value - se, ymax = value + se), width = 0) +
  geom_point(size=2.25) +
  #facet_wrap(~temperature_2_binom) +
  theme_classic() +
  #xlim(3,10) + 
  #ylim(3.25,5) +
  scale_x_discrete(limits=c(3,6,10)) +
  theme(strip.background = element_blank()) +
  ggtitle(NULL) + #, "Average expression of genes with non-linear pattern across temperature\nthat is dependent on pCO2"
  labs(y = NULL, x=NULL)  +
  scale_color_manual(values=c(Increase="darkgreen", Decrease="purple4", 
                              `V-shaped in ambient`=lighten("darkgreen", 0.45), 
                              `A-shaped in ambient`=lighten("purple4", 0.45)), guide="none") #+
  # scale_shape_manual(values=c(19, 17), guide="none") +
  # scale_linetype_manual(values=c("solid", "solid", "dashed", "dashed"), 
  #                       #limits=c("370:Decrease","950:Decrease"),
  #                       # labels=c("370:Decrease"="Ambient pCO2", 
  #                       #          "370:Increase"=NA, "950:Increase"=NA,
  #                       #          "950:Decrease"="High pCO2")
  #                       guide="none") #+
#  guides(colour = guide_legend(order=1))
#dev.off()

# NOTE - this shows the average expression across all genes by group, which was calculated from averages of each gene by group. could consider another way of showing this data that doesn't simplify as much.   

nrow(nonlinear_interaction_degs_Ashaped) #how many A-shaped in ambient pH? 179
nrow(nonlinear_interaction_degs_Vshaped) #how many V-shaped in ambient pH? 71
```


######## LINEAR RESPONSE TO TEMPERATURE & INTERACTIVE EFFECTS

There are 318 genes that respond linearly to temperature, without considering pH

```{r}
summary(DEGs_nl_temperature)
```

There are 1,473 genes that have significant temperature:pH interaction

```{r}
summary(DEGs_nl_interaction)
```
```{r}
# Estimate average logCPM per gene per pH treatment
#library(Rmisc)
tab_exp_avg_interact <- summarySE(
  measurevar = "value",
  groupvars = c("temperature","ph","geneid","temperature_sig", "interact_sig", "logFC_dir"),
  data = tab_exp_df)
tab_exp_avg_interact$temperature_binom_linear <- ifelse(tab_exp_avg$geneid %in% temperature_pos_linear$geneid, "Positive", "Negative") #add whether temperature coef is + or -

# Exploratory plot of linear expression across temperatures grouped by pH (aka pco2) and direction of differential expression
ggplot(data = subset(tab_exp_avg_interact, interact_sig == "Yes" | temperature_sig == "Yes"), #but these can also be significantly non-linear 
       aes(x = temperature, y = value)) +
  geom_path(
    alpha = 0.05,
    size = 0.25,
    stat = "identity",
    aes(group = as.factor(geneid))) +
  scale_x_continuous(breaks=c(3,6,10), labels=c(3,6,10), limits=c(2,11)) +
  facet_grid(temperature_binom_linear ~ ph) +
  theme_classic() +
  theme(strip.background = element_blank()) +
  labs(y = "Avg logCPM", title = "Linear changes in GE output by edgeR")
```

Determine whether it is more probable that gene's expression is linear or non-linear relative to a continuous predictor. We can calculate this relative probability using a likelihood ratio test (LRT). In a previous the code chunk below, we fit linear and 2nd order non-linear models to the expression of each gene before applying LRTs to each gene 

```{r}
## I previously used dlply to fit linear and non-linear models to each gene, multiple-comparison correct, 

# Filter LRT results for genes that don't meet the sig FDR coeff (i.e. those that are not considered nonlinear)
lrt_linear_filt <- filter(lrt_coeff_df, FDR > 0.01)

## Plot sig linear genes according to LRT, grouped by ph and direction of beta 1 coefficient

# Filter for how many gene id's with significant likelihood of just linear effect in LRT
LRT_linear_filt_df <- filter(tab_exp_avg, geneid %in% lrt_linear_filt$geneid)

# Plot
ggplot(data = LRT_linear_filt_df,
       aes(x = temperature, y = value)) +
  geom_path(
    alpha = .025,
    size = 0.25,
    stat = "identity",
    aes(group = as.factor(geneid))
    ) +
  facet_grid(temperature_binom_linear ~ ph) +
  #geom_smooth(method = "loess", se = TRUE, span = 1) +
  theme_classic() +
  theme(strip.background = element_blank()) +
  labs(y = "Avg logCPM", title = "Non-linear changes in GE output by LRTs")
```

### How many genes are identified as not having non-linear expression patterns (i.e. those that are linear) according to the LRT tests? 
That's a ton. But we want to look for overlaps between the LTR-identified linear expression genes, and those that have significant temperature or temperature:ph effects as per edgeR 

```{r}
nrow(as.data.frame(unique(LRT_linear_filt_df$geneid)))
```

```{r, warning=F, message=FALSE}
# Filter down df for gene id's exhibit significant temperature or interaction effect in edgeR and significant likelihood of linear effect in LRT 
edgeR_LRT_linear_df <- tab_exp_avg_interact %>% filter(temperature_sig == "Yes"| interact_sig == "Yes") %>% 
  filter(geneid %in% lrt_linear_filt$geneid)

# Plot linear effects of temperature, by ph
ggplot(data = edgeR_LRT_linear_df, 
       aes(x = temperature, y = value)) +
  geom_path(
    alpha = .25,
    size = 0.25,
    stat = "identity",
    aes(group = as.factor(geneid))
    ) +
  facet_grid(temperature_binom_linear~ph) +
    #geom_smooth(method = "loess", se = TRUE, span = 1) +
  theme_classic() +
  theme(strip.background = element_blank()) +
  labs(y = "Avg logCPM", title = "Linear changes in GE output by edgeR & LRTs")
```

### How many genes are identified as having linear expression with temperature according to both edgeR and LRT? 707 

```{r}
edgeR_LRT_linear_df %>% select(geneid) %>% distinct() %>% 
  filter(geneid %in% unique(unlist(c(lapply(degs, rownames))))) %>% #filter for only DEGs from DESeq2
  nrow()  
```

## How many genes have positive linear expression with temperature and significant interaction between temperature:ph and are DEGs from DESeq2


```{r}
edgeR_LRT_linear_df %>% filter(interact_sig=="Yes") %>% select(geneid) %>% distinct() %>% 
  filter(geneid %in% unique(unlist(c(lapply(degs, rownames))))) %>%  #filter for only DEGs from DESeq2
  nrow() #quite a few! 685

# How many increase with temperature?
edgeR_LRT_linear_df %>% 
  filter(interact_sig=="Yes") %>% 
  filter(temperature_binom_linear=="Positive") %>% 
  filter(geneid %in% unique(unlist(c(lapply(degs, rownames))))) %>%
  select(geneid) %>% distinct() %>% nrow() #282

# How many decrease with temperature?
edgeR_LRT_linear_df %>% 
  filter(interact_sig=="Yes") %>% 
  filter(temperature_binom_linear=="Negative") %>% 
  filter(geneid %in% unique(unlist(c(lapply(degs, rownames))))) %>%
  select(geneid) %>% distinct() %>% nrow() #403
```

```{r}
# Filter down df containing sign. linear genes (from edger AND LRT) for only those with sign. interaction between temp and pH and are DEGs

linear_interaction_df <- filter(edgeR_LRT_linear_df, interact_sig == "Yes") %>% 
      filter(geneid %in% unique(unlist(c(lapply(degs, rownames))))) #filter for only DEGs from DESeq2
linear_interaction_df$gene_id_ph <- paste(linear_interaction_df$geneid,
                                           linear_interaction_df$ph,
                                           sep = "_")

# Add annotation annotation information 
linear_interaction_df <- linear_interaction_df %>%
               left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names", "spid")],
                         by=c("geneid"="gene_gadmor"))
```

Exploring through gene-level plots - genes that respond linearly to temperature and for which pH affects that expression:

```{r}
# Generate list of genes with interaction for a) those that increase with temperature, and b) those that decrease

# a) Increased expression across temperature in ambient pH 
linear_interaction_increase <- 
  linear_interaction_sig %>% as.data.frame() %>% rownames_to_column("geneid") %>% #convert edgeR model results to dataframe to access stats
  filter(geneid %in% unique(linear_interaction_df$geneid)) %>% # select only genes that are linear+interaction 
  filter(geneid %in% tab_exp_avg[tab_exp_avg$temperature_binom_linear == "Positive", "geneid"]) %>% #select genes that increase with temperature as per linear models 
  left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names", "spid")], #annotate with blast results
                         by=c("geneid"="gene_gadmor"))

# b) Decreased expression across temperature in ambient pH 
linear_interaction_decrease <- 
  linear_interaction_sig %>% as.data.frame() %>% rownames_to_column("geneid") %>% #convert edgeR model results to dataframe to access stats
  filter(geneid %in% unique(linear_interaction_df$geneid)) %>% # select only genes that are linear+interaction 
  filter(geneid %in% tab_exp_avg[tab_exp_avg$temperature_binom_linear == "Negative", "geneid"]) %>% #select genes that increase with temperature as per linear models 
  left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names", "spid")], #annotate with blast results
                         by=c("geneid"="gene_gadmor"))
```

Subset linear interactive gene set for only those that are also DEGs from DESeq2

```{r}
# a) Increase expression with temperature 
linear_interaction_degs_increase <- linear_interaction_increase %>% 
  filter(geneid %in% c(lapply(degs, rownames) %>% unlist() %>% unique()))  #select genes that are DEGs 

# b) Decrease expression with temperature  
linear_interaction_degs_decrease <- linear_interaction_decrease %>% 
  filter(geneid %in% c(lapply(degs, rownames) %>% unlist() %>% unique()))   #select genes that are DEGs 

nrow(linear_interaction_degs_increase) #how many increase with temperature?
nrow(linear_interaction_degs_decrease) #how many decrease with temperature?
```

```{r}
# Here I plot genes that have 
# 1) nonlinear expression pattern across temperature as deemed by edgeR & LRT, which is 
# 2) dependent on pH (significant interaction) as deemed by edger, and 
# 3) is differentially expressed among pH or temperature as deemed by DESeq2 
# 4) are in the top 25 genes - i.e. those with the lowest p-values (since there's too many to plot)

genes <- linear_interaction_degs_increase %>% 
  filter(!is.na(spid)) %>% #remove genes that are not annotated
  arrange(PValue) %>% dplyr::slice(1:25) #arrange by pvalue, select top 25 with lowest pvalue 

genes <- linear_interaction_degs_decrease %>%
  filter(!is.na(spid)) %>% #remove genes that are not annotated
  arrange(PValue) %>% dplyr::slice(1:25) #arrange by pvalue, select top 25 with lowest pvalue 

for (i in 1:nrow(genes)) {
  # Plot individual genes 
print(ggplot(data = linear_interaction_df %>% 
               filter(geneid == genes$geneid[i]), 
       aes(
         x = temperature,
         y = value,
         color = as.factor(ph),
         group = as.factor(ph)
         )) +
         geom_path(stat = "identity") +
  geom_errorbar(aes(ymin = value - se, ymax = value + se), width = 0) +
  geom_point() +
  #facet_wrap(~temperature_2_binom) +
  theme_classic() +
  xlim(3,10) +
  #scale_x_continuous(limits=c(3,6,10)) +
  theme(strip.background = element_blank()) +
  labs(y = "logCPM", color = "pCO2 treatment", 
       title = paste(genes$geneid[i], "\n",
       str_wrap(genes[genes$geneid==genes$geneid[i], 
       "protein_names"] %>% unique()))) +
  scale_color_manual(values=c("blue", "red")))
}
```

### Generate a lineplot showing average expression by temperature, ph, and expression pattern for genes that respond linearly to temp, for which slope of that change depends on pH, and genes are also DEGs as per DESeq2
# NOTE: there are 2 genes that respond linearly to temp. without a pH interaction - would need to filter for "interact_sig=="No" below and remove the "ph" grouping variable from both the `summarySE` and `ggplot(aes())` code. 

```{r}
# Average logCPM across different groups according to temperature estimate and ph
edgeR_LRT_linear_degs_df_avg <- 
  summarySE(measurevar = "value",
            groupvars = c("ph", "temperature", "temperature_binom_linear"),
            data = edgeR_LRT_linear_df %>% filter(interact_sig=="Yes") %>% 
              filter(geneid %in% c(linear_interaction_degs_increase$geneid, 
                                   linear_interaction_degs_decrease$geneid))) %>%
  mutate(temperature_binom_linear = factor(ifelse(temperature_binom_linear == "Negative", "Decrease", "Increase")))

# Linear Lineplot
pdf(file="../results/edgeR/Linear-response-genes-plot.pdf", height = 2.75, width = 3)
ggplot(data = edgeR_LRT_linear_degs_df_avg %>%
         filter(ph==370), 
       aes(
         x = temperature,
         y = value ,
         # color = as.factor(ph),
         color = as.factor(temperature_binom_linear),
         group = as.factor(temperature_binom_linear) #,
         #shape=as.factor(ph)
         )) +
         geom_path(stat = "identity",
#                   aes(linetype=interaction(as.factor(ph):as.factor(temperature_binom_linear)))
                   aes(group=temperature_binom_linear)
) +
  geom_errorbar(aes(ymin = value - se, ymax = value + se), width = 0) +
  geom_point(size=2.25) +
#  facet_wrap(~temperature_binom_linear) +
  theme_classic() +
  xlim(3,10) +  ylim(3.25,5) +
  scale_x_discrete(limits=c(3,6,10)) +
  theme(strip.background = element_blank()) +
  ggtitle(NULL) + #, title = "Linear response to temperature"
  labs(y=NULL, x=NULL) +
  scale_color_manual(values=c(Increase="darkgreen", Decrease="purple4", 
                              Concave=lighten("darkgreen", 0.45), Convex=lighten("purple4", 0.45)), guide="none") #+
  #scale_shape_manual(values=c(19, 17), guide="none") +
  #scale_linetype_manual(values=c("solid", "solid", "dashed", "dashed"), 
                        #limits=c("370:Decrease","950:Decrease"),
                        # labels=c("370:Decrease"="Ambient pCO2", 
                        #          "370:Increase"=NA, "950:Increase"=NA,
                        #          "950:Decrease"="High pCO2")
 #                      guide="none") #+
#  guides(colour = guide_legend(order=1))
dev.off()

#Negative=1,261 genes; Positive=687 genes (edgeR, LRT & DESeq2 consensus sets)
```

### Enrichment Analysis - linear and nonlinear gene sets

Here I extract 4 lists of Uniprot SPIDs for DAVID enrichment analysis. These gene sets represent those that respond ... 
  1) linearly to temperature, increasing with temperature 
  2) linearly to temperature, decreasing with temperature  
  3) non-linearly to temperature, convex shape (highest at 6C)
  4) non-linearly to temperature, concave shape (lowest at 6C)

```{r}
# Background set of genes analyzed here 
counts.annot.gadmor %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# 1. linearly to temperature, increasing with temperature 
edgeR_LRT_linear_df %>% 
  filter(temperature_binom_linear=="Positive") %>% 
  filter(geneid %in% unique(unlist(c(lapply(degs, rownames))))) %>%
  select(geneid) %>% distinct() %>%
  left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names", "spid")], #annotate with blast results
                         by=c("geneid"="gene_gadmor")) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# 2. linearly to temperature, decreasing with temperature  
edgeR_LRT_linear_df %>% 
  filter(temperature_binom_linear=="Negative") %>% 
  filter(geneid %in% unique(unlist(c(lapply(degs, rownames))))) %>%
  select(geneid) %>% distinct() %>%
  left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names", "spid")], #annotate with blast results
                         by=c("geneid"="gene_gadmor")) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# 3. non-linearly to temperature, convex shape
## (note- here I combine those with pCO2 interaction and without)
c(nonlinear_interaction_degs_Ashaped %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),
  
  nonlinear_no.interaction_degs %>% filter(temperature_2_binom=="Convex") %>% select(geneid) %>% distinct() %>%
  left_join(gadMor.blast[,c("gene_gadmor", "spid")], by=c("geneid"="gene_gadmor")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector()) %>% length()
  
  write_clip(allow_non_interactive = TRUE)


# 4. non-linearly to temperature, concave shape 
## (note- here I combine those with pCO2 interaction and without)

c(nonlinear_interaction_degs_Vshaped %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector(),

  nonlinear_no.interaction_degs %>% filter(temperature_2_binom=="Concave") %>% select(geneid) %>% distinct() %>%
  left_join(gadMor.blast[,c("gene_gadmor", "spid")], by=c("geneid"="gene_gadmor")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector()) %>%

  write_clip(allow_non_interactive = TRUE)

```

Enrichment analysis - linear and nonlinear - separating genes with significant interaction effect of high pCO2

Here I extract 6 lists of Uniprot SPIDs for DAVID enrichment analysis. These gene sets represent those that respond ... 
  1) linearly to temperature, increasing with temperature, but slope is dependent on pH    
  2) linearly to temperature, decreasing with temperature, but slope is dependent on pH  
  3) non-linearly to temperature, convex shape in ambient pH but concave in low pH
  4) non-linearly to temperature, concave shape in ambient pH but convex in low pH
  5) non-linearly to temperature, no interaction with pH - shape is X
  6) non-linearly to temperature, no interaction with pH - shape is X

```{r}
# Background set of genes analyzed here 
counts.annot.gadmor %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# 1. linearly to temperature, increasing with temperature, but slope is dependent on pH    
linear_interaction_degs_increase %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# 2. linearly to temperature, decreasing with temperature, but slope is dependent on pH  
linear_interaction_degs_decrease %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# 3. non-linearly to temperature, convex shape in ambient pH but concave in low pH
nonlinear_interaction_degs_Ashaped %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# 4. non-linearly to temperature, concave shape in ambient pH but convex in low pH
nonlinear_interaction_degs_Vshaped %>% dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# 5. non-linearly to temperature, Convex shape with temperature, shape not affected by pH
nonlinear_no.interaction_degs %>% filter(temperature_2_binom=="Convex") %>% select(geneid) %>% distinct() %>% 
  left_join(gadMor.blast[,c("gene_gadmor", "spid")], by=c("geneid"="gene_gadmor")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# 6. non-linearly to temperature, Concave shape with temperature, shape not affected by pH
nonlinear_no.interaction_degs %>% filter(temperature_2_binom=="Concave") %>% select(geneid) %>% distinct() %>%
  left_join(gadMor.blast[,c("gene_gadmor", "spid")], by=c("geneid"="gene_gadmor")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)
```

#### Read in enriched biological functions for all patterns

```{r}
filenames <- read_delim(file="../results/GO-enrichment/GO_filenames_linear-nonlinear.txt", delim = "\t", col_names = c("filename", "gene_set"))
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=paste("../results/GO-enrichment/", filenames[i,"filename"], sep=""), delim = "\t"))}

enriched.bp.linear.nonlinear <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term_id", "term_process"), sep = "~") %>%
    separate(gene_set, sep = "-", into = c("shape", "pattern", "interaction"), remove = F) %>% 
  mutate_at(vars(shape, pattern, interaction), as.factor) %>%
  mutate(group=factor(paste(interaction, pattern, sep="_")))
```

### Bubble plot of enrichment results 

#### Genes that respond linearly to temperature, and for which the slope depends on pH (i.e. sign. interaction)

Biological Process GO terms 

```{r}
pdf(file="../results/GO-enrichment/BP-linear-interaction-enriched-bubble.pdf", height = 7, width = 6)
enriched.bp.linear.nonlinear %>% 
  filter(category=="GOTERM_BP_DIRECT") %>% 
  filter(p_value<0.05) %>%
  group_by(shape, pattern, interaction, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.linear.nonlinear %>% select(group, shape, pattern, interaction, category, term_id, term_process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("shape", "pattern", "interaction", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(shape, group, pattern, interaction, p_value) %>% group_by(shape, pattern, interaction) %>% dplyr::slice(1:25) %>% # select the top 25 terms per group
  filter(shape == "linear") %>%
  ungroup() %>% arrange(group, p_value) %>%
  mutate(term_process = factor(term_process, rev(unique(term_process)), ordered = TRUE)) %>%

ggplot(aes(y = term_process, x=pattern, col=pattern)) + 
  geom_point(alpha=.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Change with temperature", values=c(Decrease="#999999", Increase="#4d4d4d"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) +
  coord_cartesian(clip = "off", ylim = c(0, 55)) +
  ggtitle(NULL) +
  # annotate("text", x = 1, y = -1.5, label = "Decrease with\ntemperature", size=2, col="gray15") +
  # annotate("text", x = 2, y = -1.5, label = "Increase with\ntemperature", size=2, col="gray15") +
  annotate("text", x = 1.5, y = 54, label = "Enriched Biological Processes", size=2.5, col="gray15") +
  annotate("text", x = 1.5, y = 53, label = "Linear response to temperature", size=2.5, col="gray15", fontface = 'bold') +
  annotate("text", x = 1.5, y = 52, label = "and pH interaction effect ", size=2.5, col="gray15", fontface = 'italic')
dev.off()
```

Uniprot Keywords

```{r}
pdf(file="../results/GO-enrichment/KW-linear-interaction-enriched-bubble.pdf", height = 4.25, width = 4.25)
enriched.bp.linear.nonlinear %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
  filter(p_value<0.05) %>%
  group_by(shape, pattern, interaction, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.linear.nonlinear %>% select(group, shape, pattern, interaction, category, term_id, term_process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("shape", "pattern", "interaction", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(shape, group, pattern, interaction, p_value) %>% group_by(shape, pattern, interaction) %>% dplyr::slice(1:25) %>% # select the top 25 terms per group
  filter(shape == "linear") %>%
  ungroup() %>% arrange(group, p_value) %>%
  mutate(term_process = factor(term_process, rev(unique(term_process)), ordered = TRUE)) %>%

ggplot(aes(y = term_process, x=pattern, col=pattern)) + 
  geom_point(alpha=.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Pattern across temperatures", values=c(Decrease="#2c7bb6", Increase="#d7191c"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Uniprot Keywords") + xlab(NULL) +
  coord_cartesian(clip = "off", ylim = c(0.25, 28)) +
  ggtitle(NULL) +
  # annotate("text", x = 1, y = -1.5, label = "Decrease with\ntemperature", size=2, col="gray15") +
  # annotate("text", x = 2, y = -1.5, label = "Increase with\ntemperature", size=2, col="gray15") +
  annotate("text", x = 1.5, y = 28, label = "Enriched Uniprot Keywords", size=2.5, col="gray15") +
  annotate("text", x = 1.5, y = 27.25, label = "Linear response to temperature", size=2.5, col="gray15", fontface = 'bold') +
  annotate("text", x = 1.5, y = 26.5, label = "and pH interaction effect ", size=2.5, col="gray15", fontface = 'italic')
dev.off()
```

#### Genes that respond non-linearly to temperature

Biological Process GO terms 

```{r}
pdf(file="../results/GO-enrichment/BP-nonlinear-enriched-bubble.pdf", height = 7, width = 6.5)
enriched.bp.linear.nonlinear %>% 
  filter(category=="GOTERM_BP_DIRECT") %>% 
  filter(p_value<0.05) %>%
  group_by(shape, pattern, interaction, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.linear.nonlinear %>% select(group, shape, pattern, interaction, category, term_id, term_process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("shape", "pattern", "interaction", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(shape, group, pattern, interaction, p_value) %>% group_by(shape, pattern, interaction) %>% dplyr::slice(1:25) %>% # select the top 25 terms per group
  filter(shape == "nonlinear") %>% ungroup() %>% 
  mutate(group=factor(group, levels=c("interaction_Ashaped","nointeraction_Ashaped", "interaction_Ushaped", "nointeraction_Ushaped")),
         shape=as.character(shape), pattern=as.character(pattern)) %>%
  tidyr::complete(category, pattern, interaction) %>% # add rows for groups that are missing from dataframe
  arrange(pattern, interaction, p_value) %>%
  mutate(term_process = factor(term_process, rev(unique(term_process)), ordered = TRUE)) %>%
  #mutate(pattern=as.character(pattern), interaction=as.character(interaction)) %>%

  # plot
ggplot(aes(y = term_process, x=interaction(interaction, pattern), 
           col=pattern, shape=interaction)) + 
  geom_point(alpha=.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Pattern across temperatures\nin ambient pH", 
                     labels=c(`Ashaped`="A-shaped", `Vshaped`='V-shaped'),
                     values=c(`Ashaped`="#999999", `Ushaped`="#4d4d4d"),
             guide = guide_legend(order = 1, override.aes = list(size=4))) +
  scale_shape_manual(name="Significant pH interaction effect", 
                     labels=c(`interaction`="Yes", `nointeraction`='No'),
                     values=c(`interaction`=19, `nointeraction`=17),
             guide = guide_legend(order = 2, override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) +
  coord_cartesian(clip = "off", ylim = c(.5, 38)) +
  ggtitle(NULL) +
  # annotate("text", x = 1, y = -1.5, label = "Decrease with\ntemperature", size=2, col="gray15") +
  # annotate("text", x = 2, y = -1.5, label = "Increase with\ntemperature", size=2, col="gray15") +
  annotate("text", x = 2.5, y = 38, label = "Enriched Biological Processes", size=2.5, col="gray15") +
  annotate("text", x = 2.5, y = 37.25, label = "Non-linear response to temperature", size=2.5, col="gray15", fontface = 'bold') +
  annotate("text", x = 2.5, y = 36.5, label = "by pH interaction effect ", size=2.5, col="gray15", fontface = 'italic')
dev.off()
```

Uniprot Keywords

```{r}
#pdf(file="../results/GO-enrichment/KW-nonlinear-enriched-bubble.pdf", height = 3, width = 4.5)
pdf(file="../results/GO-enrichment/KW-nonlinear-enriched-bubble.pdf", height = 3, width = 4)
enriched.bp.linear.nonlinear %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
  filter(p_value<0.05) %>%
  group_by(shape, pattern, interaction, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.linear.nonlinear %>% select(group, shape, pattern, interaction, category, term_id, term_process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("shape", "pattern", "interaction", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(shape, group, pattern, interaction, p_value) %>% group_by(shape, pattern, interaction) %>% dplyr::slice(1:25) %>% # select the top 25 terms per group
  filter(shape == "nonlinear") %>% ungroup() %>% 
  filter(interaction == "interaction") %>% ungroup() %>% 
  mutate(group=factor(group, levels=c("interaction_Ashaped","nointeraction_Ashaped", "interaction_Ushaped", "nointeraction_Ushaped")),
         shape=as.character(shape), pattern=as.character(pattern)) %>%
  tidyr::complete(category, pattern, interaction) %>% # add rows for groups that are missing from dataframe
  arrange(pattern, interaction, p_value) %>%
  mutate(term_process = factor(term_process, rev(unique(term_process)), ordered = TRUE)) %>%
  #mutate(pattern=as.character(pattern), interaction=as.character(interaction)) %>%

  # plot
ggplot(aes(y = term_process, #shape=pattern, 
#           x=interaction(interaction, pattern), col=interaction)) + # If I also want to show enriched functions in genes where pH affects shape (doesn't add much - just fatty acid and lipid metabolism)
           x=pattern, col=pattern)) + 
  geom_point(alpha=.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
  # scale_shape_manual(name="Pattern across temperatures\nin ambient pH", 
  #                    labels=c(`Ashaped`="A-shaped", `Ushaped`='V-shaped'),
  #                    values=c(`Ashaped`=19, `Ushaped`=17),
  #            guide = guide_legend(order = 1, override.aes = list(size=3))) +
  # scale_color_manual(name="Significant pH interaction effect", 
  #                    labels=c(`interaction`="Yes", `nointeraction`='No'),
  #                    values=c(`interaction`="gray65", `nointeraction`="gray20"),
  #            guide = guide_legend(order = 2, override.aes = list(size=3))) +
  scale_color_manual(name="Pattern across temperatures", 
                     labels=c(`Ashaped`="Decrease", `Ushaped`='Increase'),
#                     values=c(`Ashaped`="gray65", `Ushaped`="gray20"),
                     values=c(`Ashaped`="#2c7bb6", `Ushaped`="#d7191c"),
             guide = guide_legend(order = 1, override.aes = list(size=4))) +

  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Uniprot Keywords") + xlab(NULL) +
  coord_cartesian(clip = "off", ylim = c(.5, 14.5)) +
  ggtitle(NULL) +
  # annotate("text", x = 1, y = -1.5, label = "Decrease with\ntemperature", size=2, col="gray15") +
  # annotate("text", x = 2, y = -1.5, label = "Increase with\ntemperature", size=2, col="gray15") +
  annotate("text", x = 1.5, y = 14.5, label = "Enriched Uniprot Keywords", size=2.5, col="gray15") +
  annotate("text", x = 1.5, y = 13.75, label = "Non-linear response to temperature", size=2.5, col="gray15", fontface = 'bold') +
  annotate("text", x = 1.5, y = 13, label = "by response pattern & pH interaction effect ", size=2.5, col="gray15", fontface = 'italic')
dev.off()
```

### Re-do bubble plots to include both linear and non-linear response processes, either GO terms or Uniprot Keywords

#### Read in enriched biological functions for all patterns

```{r}
filenames <- read_delim(file="../results/GO-enrichment/GO_filenames_linear-nonlinear_all-genes.txt", delim = "\t", col_names = c("filename", "gene_set"))
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=paste("../results/GO-enrichment/", filenames[i,"filename"], sep=""), delim = "\t"))}

enriched.bp.linear.nonlinear_all <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term_id", "term_process"), sep = "~") %>%
    separate(gene_set, sep = "-", into = c("shape", "pattern"), remove = F) %>% 
  mutate_at(vars(shape, pattern), as.factor)
```
#### Enriched Biological Process GO Terms - genes with linear and non-linear expression patterns with temperature

```{r}
pdf(file="../results/GO-enrichment/Enriched-Uniprot-Keywords_linear-nonlinear_Bubble-plot.pdf", height = 4.25, width = 3.5)
#pdf(file="../results/GO-enrichment/Enriched-GO-terms_linear-nonlinear_Bubble-plot.pdf", height = 7, width = 6)

enriched.bp.linear.nonlinear_all %>%
  filter(category=="GOTERM_BP_DIRECT") %>% 
#  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
#  filter(category=="KEGG_PATHWAY") %>%
#  separate(term_id, into = c("term_id", "term_process"), sep = ":") %>%
  filter(p_value<0.05) %>%
  filter(shape=="nonlinear") %>% 
  group_by(shape, pattern, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.linear.nonlinear_all %>% 
#              separate(term_id, into = c("term_id", "term_process"), sep = ":") %>% 
              select(shape, pattern, category, term_id, term_process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("shape", "pattern", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(shape, pattern, p_value) %>% group_by(shape, pattern) %>% #dplyr::slice(1:20) %>% # select the top n terms per group
  arrange(pattern, p_value) %>%
#  arrange(pattern, term_id) %>%
#  ungroup() %>% arrange(term_id) %>%
  mutate(term_process = factor(term_process, rev(unique(term_process)), ordered = TRUE)) %>%
  mutate(pattern=factor(pattern, levels=c("Increase", "Concave", "Decrease", "Convex"), ordered = TRUE)) %>%

  # plot
ggplot(aes(y = term_process, x=pattern, col=pattern)) + 
  geom_point(alpha=.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Pattern across\ntemperature",
                     values=c(Increase="darkgreen", Decrease="purple4", 
                              Concave=lighten("darkgreen", 0.45), Convex=lighten("purple4", 0.45)),
             guide = guide_legend(order = 1, override.aes = list(size=4))) +
  # scale_shape_manual(name="Significant pH interaction effect", 
  #                    labels=c(`interaction`="Yes", `nointeraction`='No'),
  #                    values=c(`interaction`=19, `nointeraction`=17),
  #            guide = guide_legend(order = 2, override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ylab("Enriched Gene Ontology (GO) Term") + xlab(NULL) #+
  # coord_cartesian(clip = "off", ylim = c(.5, 38)) +
  # ggtitle(NULL) +
  # annotate("text", x = 1, y = -1.5, label = "Decrease with\ntemperature", size=2, col="gray15") +
  # annotate("text", x = 2, y = -1.5, label = "Increase with\ntemperature", size=2, col="gray15") +
  # annotate("text", x = 2.5, y = 38, label = "Enriched Biological Processes", size=2.5, col="gray15") +
  # annotate("text", x = 2.5, y = 37.25, label = "Non-linear response to temperature", size=2.5, col="gray15", fontface = 'bold') +
  # annotate("text", x = 2.5, y = 36.5, label = "by pH interaction effect ", size=2.5, col="gray15", fontface = 'italic')
dev.off()
```

Enriched GO biological processes - Network Diagrams using metacoder

Here I save an R objects from GO enrichment results of the linear/nonlinear response genes to visualize in network plots - the visualization is done in RStudio Cloud, as it requires a specific older version of R and some R packages. 

```{r}
# Save R object to then import to RStudio Cloud for use in metacoder 
save(enriched.bp.linear.nonlinear_all, file="../results/GO-enrichment/enriched.bp.linear.nonlinear_all")
```





## Differential Exon + Gene Expression Analysis 


Import exon count summary. Large percentage of exons (~30%) not counted due to ambiguity 

```{r}
# Extract sample names for each column from file path 
samples.exon <- (read_delim(file = "../results/star/featurecounts/featurecounts_exon.summary", col_names = F, show_col_types = FALSE)[1,-1] %>% as.data.frame() %>% t() %>% as.data.frame() %>%
                            mutate(V1=gsub("/home/lspencer/pcod-2022/aligned/star-gadmor/", "", V1)) %>%
                            mutate(V1=gsub(".Aligned.sortedByCoord.out.bam", "", V1)))$V1 %>% as.vector()

# Import exon count summary from featurecounts
counts.exon.summary <- read_delim(file = "../results/star/featurecounts/featurecounts_exon.summary", col_names = F, skip=1)  %>% as.data.frame()

# Add sample names to exon count summary table 
colnames(counts.exon.summary) <- c("results", samples.exon)
#ggplotly(
counts.exon.summary %>% 
  filter(rowSums(across(where(is.numeric)))!=0) %>%
pivot_longer(-results, names_to = "sample", values_to = "count") %>%
  mutate(results=fct_rev(factor(results, ordered = T))) %>%
  ggplot(aes(x=sample, y=count, fill=results)) + 
  geom_bar(position="fill", stat="identity") +
 scale_fill_brewer(palette="Dark2") +
  theme(axis.text.x = element_text(size=6, angle = 90, vjust = 0.5, hjust=1)) +
  theme(legend.position = "bottom") +
  guides(fill=guide_legend(nrow=2,byrow=TRUE, title = ""))#)
```

## Load counts

```{r}
counts.exon <- data.frame(read.table("../results/star/featurecounts/featurecounts_exon", header = T, stringsAsFactors = F, fill = FALSE)) %>%
  mutate(exonID=paste(Geneid, Start, End, sep=".")) %>% select(exonID, everything())

counts.exon <- counts.exon %>% rename_all(~as.character(str_sub(colnames(counts.exon)) %>%
                            gsub("X.home.lspencer.pcod.2022.aligned.star.gadmor.", "", .) %>%
                            gsub(".Aligned.sortedByCoord.out.bam", "", .))) %>% 
  distinct() %>% #for some reason there are 3 identical rows for each exon. Here I remove duplicated rows (determined across all columns)
  column_to_rownames(var="exonID") 
#save(counts.exon, file = "../results/star/featurecounts/counts.exon")

remove.list <- c("PCG194", "PCG137")
counts.exon <- counts.exon[ , -which(names(counts.exon) %in% remove.list)]
```

### Transpose dataframe so each row = a sample (aka "objects"), and each column = genes (aka "variables") 
```{r}
#str(counts.exon) #columns #1-#5 contain extraneous gene info (chr, start, end, strand, length). 
counts.exon.t <- as.data.frame(t(counts.exon[-1:-6])) #remove extraneous columns, transform data to have each sample a row, each column a gene
```

## Optional 

### Pre-filtering - remove low-frequency genes 

NOTE: Should i do this? DESeq2 throws out low-frequency genes anyway, BUT other folks do pre-filter. For example in https://doi.org/10.1186/s12864-017-4392-0 "Genes with mean count less than ten across all samples were removed."

```{r}
keep1 <- colMeans(counts.exon.t, na.rm=TRUE) >= 10 #identify genes with mean count >= 10 across all samples (excluding NAs = 10)
keep2 <- rowSums( counts.exon >= 30 ) >= 0.1*73 #identify genes with counts>=30 across at minimum 10% of the samples
keep <- unique(c(names(which(keep1 == T)), names(which(keep2 == T)))) # list of genes meeting one of the two above criteria
counts.exon.ts <- counts.exon.t[,keep]

print(paste("# exons before filtering:", prettyNum(ncol(counts.exon.t), big.mark=",")))
print(paste("# exons remaining after pre-filtering:", prettyNum(ncol(counts.exon.ts), big.mark=",")))
print(paste("# of exons dropped:", prettyNum(ncol(counts.exon.t) - ncol(counts.exon.ts), big.mark=","), sep=" "))
print(paste("% of fragments remaining after pre-filtering: ", signif(100*sum(counts.exon.ts)/sum(counts.exon.t), digits = 5), "%", sep=""))
print(paste("Number of unique fragments dropped: ", prettyNum(signif(sum(counts.exon.t)-sum(counts.exon.ts), digits = 5), big.mark=",")))
print(paste("% of fragments dropped: ", signif(100*(sum(counts.exon.t)-sum(counts.exon.ts))/sum(counts.exon.t), digits = 5), "%", sep=""))
print(paste("Number of fragments remaining: ", prettyNum(signif(sum(counts.exon.ts), digits = 5), big.mark=",")))
```

#?filterByExpr()

```{r}
counts.exon.treat <- counts.exon.ts %>% rownames_to_column("sample") %>%
  left_join(sample.info[c("tank", "temperature", "ph", "vial_label")], by=c("sample"="vial_label")) %>%
  mutate(treatment=as.factor(paste(ph, temperature, sep="."))) %>%
  mutate(ph=gsub("amb", "370", ph)) %>% #estimate ambient = pH 8.0 and pco2 370
  mutate(ph=gsub("low", "950", ph)) %>% #estimate low pH of 7.6 as pco2 900
  mutate(ph=as.numeric(as.character(ph)),
         temperature=as.numeric(as.character(temperature)),
         treatment=factor(treatment, 
                          levels=c("amb.6", "low.6", "amb.3", "low.3", "amb.10", "low.10"))) %>% 
  select(sample, tank, temperature, ph, treatment, everything())

counts.exon.edgeR <- counts.exon.treat %>% column_to_rownames("sample") %>%
  select(-tank, -temperature, -ph, -treatment) %>%
  t() %>% as.data.frame()

# double check samples (columns) are in the same order as the samples in the annotated count dataframe. Should = TRUE
all(colnames(counts.exon.edgeR) == counts.exon.treat$sample)

# Extract vector of treatments
treatment.exon <- counts.exon.treat$treatment
```

### MarineOmics tutorial analysis 
#### Non-linear effects, interactive effects 

```{r}
# Create a DGEList object 
y.exon <- DGEList(counts.exon.edgeR, group=treatment.exon, 
                  genes = left_join(counts.exon.edgeR %>% rownames_to_column("exonID") %>% select(exonID),  #add gene info - to be used later in alternative splicing analysis
                                    counts.exon[1:6] %>% rownames_to_column("exonID"), by="exonID"))
                                    
# TMM normalization
y.exon <- calcNormFactors(y.exon)
y.exon$samples #check out the normalization factor column 

# Explore exon data by multi-dimensional scaling plot 
plotMDS(y.exon)

# More interesting plot 
points <- c(0,15,1,16,2,17)
colors <- c("goldenrod1", "goldenrod1", "steelblue1", "steelblue1", "tomato1", "tomato1")

# using top 500 genes (default)
plotMDS(y.exon, col=colors[treatment.exon], pch=points[treatment.exon], cex=1.4, main="MDS plot, PC1 x PC2\ntop 500 exons only")
legend("bottomleft", legend=levels(treatment.exon), pch=points, col=colors, ncol=2)

# using all genes
plotMDS(y.exon, col=colors[treatment.exon], pch=points[treatment.exon], cex=1.4, top=nrow(y.exon$counts), main="MDS plot, PC1 x PC2\nall exons")
legend("bottomleft", legend=levels(treatment.exon), pch=points, col=colors, ncol=2)
```

PCoA using code from multiomics tutorial 

```{r}
# Export pcoa loadings
dds.pcoa.exon = pcoa(vegdist(t(df_log <- cpm(y.exon, log = TRUE, prior.count = 2)),
                          method = "euclidean") / 1000)

# Create df of MDS vector loading
scores.exon <- dds.pcoa.exon$vectors

## Plot pcoa loadings of each sample, groouped by time point and pCO2 treatment

# Calculate % variation explained by each eigenvector
percent.exon <- dds.pcoa.exon$values$Eigenvalues
cumulative_percent_variance.exon <- (percent.exon / sum( percent.exon)) * 100

# Prepare information for pcoa plot, then plot
#colors <- c("goldenrod1", "goldenrod1", "steelblue1", "steelblue1", "tomato1", "tomato1")

par(mfrow = c(1, 1))
plot(
  scores.exon[, 1],
  scores.exon[, 2],
  cex = .5,
  cex.axis = 1,
  cex.lab = 1.25,
  xlab = paste("PC1, ", round(cumulative_percent_variance.exon[1], 2), "%"),
  ylab = paste("PC2, ", round(cumulative_percent_variance.exon[2], 2), "%")
  )

# Add visual groupings to pcoa plot
ordihull(
  scores.exon,
  treatment.exon,
  border = NULL,
  lty = 2,
  lwd = .5,
  label = F,
  col = colors,
  draw = "polygon",
  alpha = 100,
  cex = .5
  )

ordispider(scores.exon, treatment.exon, label = F) # Vectors connecting samples in same ph x treatment group
#ordilabel(scores, cex = 0.4) # Label sample IDs
```

Create design matrix 
```{r}
temperature.exon <- counts.exon.treat$temperature
ph.exon <- counts.exon.treat$ph

design.exon <- model.matrix(~temperature.exon + as.factor(ph.exon) + temperature.exon:as.factor(ph.exon))
colnames(design.exon) <- c("(Intercept)", "temperature", "ph", "temperature:ph") #fix column names 
design.exon %>% head()

# Estimate dispersions 
y.exon <- estimateDisp(y.exon, design.exon, robust=TRUE)
y.exon$common.dispersion
plotBCV(y.exon)
```

Differentially expressed exons among temperature (continuous variable), ph (factor variable) and interaction among the two 

Result: 

- Large linear effect of temperature on exon expression. Of the 71,956 exons examined 10,087 significantly decrease with temperature (negative slope), and 9,327 significantly increase with temperature (positive slope)
- No effect of pH (binary, since used as factors), nor does pH interact with temperature to affect exon expression

```{r}
fit.exon <- glmQLFit(y.exon, design.exon, robust=TRUE)
plotQLDisp(fit.exon)

# Test for effect of temperature as a continuous variable 
qlf.exon.temp <- glmQLFTest(fit.exon, coef=2)
is.de.exon.temp <- decideTests(qlf.exon.temp, adjust.method = "fdr", p.value=0.05)
summary(is.de.exon.temp) #loads of differentially expressed exons among temperature
plotMD(qlf.exon.temp, cex=0.6)

# Check out top 12 differentially expressed exons among temperature
topTags(qlf.exon.temp, n = 16) %>% as.data.frame() %>% 
  left_join(
    counts.exon %>% mutate(exonID=paste(Geneid, Start, End, sep="."))) %>%
  pivot_longer(cols = starts_with("PCG"), names_to = "sample", values_to = "count") %>%
  left_join(sample.info[c("tank", "temperature", "ph", "vial_label")], by=c("sample"="vial_label")) %>%
  # group_by(exonID, temperature, ph) %>% dplyr::summarize(mean=mean(count)) %>% 
  mutate(exonID=as.factor(exonID)) %>%
  ggplot(aes(x=temperature, y=log(count), color=temperature:ph)) + 
  geom_point() + theme_minimal() + facet_wrap(~exonID, scales = "free")

# Test for effect of pH as a factor
qlf.exon.ph <- glmQLFTest(fit.exon, coef=3)
is.de.exon.ph <- decideTests(qlf.exon.ph, adjust.method = "fdr", p.value=0.05)
summary(is.de.exon.ph) #no differentially expressed exon among pH

# Test for interaction effect between temperature (continuous variable) and ph (factor variable)
qlf.exon.interact <- glmQLFTest(fit.exon, coef=4)
is.de.exon.interact <- decideTests(qlf.exon.interact, adjust.method = "fdr", p.value=0.05)
summary(is.de.exon.interact) #No interaction between pH and temp 
```

Effect of temperature on differential exon usage 

```{r}

# Effect of temperature^2 on exon usage
temperature.splice <- diffSpliceDGE(fit.exon, coef=2, geneid="Geneid", exonid="Start")
#(temperature.splice.sign <- topSpliceDGE(temperature.splice, test="Simes", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1)) #Simes' method, 45 genes with FDR < 0.1
(temperature.splice.sign <- topSpliceDGE(temperature.splice, test="gene", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1)) #F-tests, 117 genes with FDR < 0.1 

# Plot top 20 genes with evidence of temperature affecting alternative exon usage 
for (i in 1:20) {
  plotSpliceDGE(temperature.splice, geneid=temperature.splice.sign$Geneid[i], genecol="Geneid")
}

# Gene Enrichment Analysis in DAVID

# List of differentially spliced genes 
temperature.splice.sign %>% 
  left_join(
    gadMor.blast.GO[,c("gene_gadmor", "spid", "evalue", "protein_names",
                       "gene_ontology_biological_process")], by=c("Geneid"="gene_gadmor")) %>% 
  filter(!is.na(spid)) %>% #View()
  dplyr::select(spid) %>%  unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

# background list of genes 
topSpliceDGE(temperature.splice, test="gene", n=nrow(nl_temperature_2.splice$genes)) %>% # all genes 
    left_join(
    gadMor.blast.GO[,c("gene_gadmor", "spid")], by=c("Geneid"="gene_gadmor")) %>% 
  filter(!is.na(spid)) %>%  
  dplyr::select(spid) %>%  unlist() %>% as.vector() %>% write_clip(allow_non_interactive = TRUE)

```

Enriched biological process GO terms
```
	GOTERM_BP_DIRECT	intermediate filament cytoskeleton organization	RT		3	3.0	6.1E-3	24.5	1.0E0
	GOTERM_BP_DIRECT	glutathione metabolic process	RT		3	3.0	3.8E-2	9.6	1.0E0
	GOTERM_BP_DIRECT	response to stimulus	RT		4	4.0	4.3E-2	5.1	1.0E0
	GOTERM_BP_DIRECT	cell-cell signaling	RT		3	3.0	7.6E-2	6.5	1.0E0
	GOTERM_BP_DIRECT	adenylate cyclase-activating adrenergic receptor signaling pathway	RT		2	2.0	9.1E-2	21.0	1.0E0
```

Enriched Uniprot Keywords 
```
	UP_KW_BIOLOGICAL_PROCESS	Lipid transport	RT		4	4.0	5.7E-2	4.4	1.0E0
	UP_KW_BIOLOGICAL_PROCESS	Transport	RT		18	17.8	7.8E-2	1.5	1.0E0
	UP_KW_BIOLOGICAL_PROCESS	Sensory transduction	RT		4	4.0	8.4E-2	3.8	1.0E0
```

NOT SURE WHAT THE DIFFERENTIAL EXON EXPRESSION ANALYSIS MEANS, particularly because I have 3 temperature treatments, which are treated as quantitative variables in the fitted model. Is it a regression analysis? 

# Examine nonlinear effects of temperature on exon expression, and interaction with pH 

```{r}
temperature.exon_2 <- temperature.exon^2 #square temperature variable, for potential non-linear effects of temp. on exon expression

# Fit multifactorial design matrix
design.exon <-
  model.matrix(
    ~ 1 + temperature.exon_2 + temperature.exon + ph.exon + temperature.exon_2:ph.exon + temperature.exon:ph.exon) # Generate multivariate edgeR glm
design.exon %>% head()

# Estimate dispersion coefficients
y.exon <- estimateDisp(y.exon, design.exon, robust=TRUE)
y.exon$common.dispersion
plotBCV(y.exon)

### Fit quasi-likelihood, neg binom linear regression with multifactorial design matrix
fit.exon <- glmQLFit(y.exon, design.exon, robust=TRUE)
head(fit.exon$coefficients)
plotQLDisp(fit.exon)
```

### Test for effects of different treatments 

```{r}
# effect of temperature^2
# estimate sign. degs 
nl_temperature_2.exon <-
    glmQLFTest(fit.exon,
    coef = 2,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs
DEGs_nl_temperature_2.exon <-
  decideTestsDGE(nl_temperature_2.exon, adjust.method = "fdr", p.value = 0.05)
summary(DEGs_nl_temperature_2.exon)
plotMD(nl_temperature_2.exon, cex=0.6)


# effect of temperature
# estimate sign. degs 
nl_temperature.exon <-
    glmQLFTest(fit.exon,
    coef = 3,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs
DEGs_nl_temperature.exon <-
  decideTestsDGE(nl_temperature.exon, adjust.method = "fdr", p.value = 0.05)
summary(DEGs_nl_temperature.exon)
plotMD(nl_temperature.exon, cex=0.6)


# effect of low pH relative to ambient pH
# estimate sign. degs 
nl_ph.exon <-
    glmQLFTest(fit.exon,
    coef = 4,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs
DEGs_nl_ph.exon <-
  decideTestsDGE(nl_ph.exon, adjust.method = "fdr", p.value = 0.05)
summary(DEGs_nl_ph.exon)
plotMD(nl_ph.exon, cex=0.6)


# interaction between temperature_2 & pH
# estimate sign. degs 
nl_interaction_2.exon <-
    glmQLFTest(fit.exon,
    coef = 5,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs
DEGs_nl_interaction_2.exon <-
  decideTestsDGE(nl_interaction_2.exon, adjust.method = "fdr", p.value = 0.05)
summary(DEGs_nl_interaction_2.exon)
plotMD(nl_interaction_2.exon, cex=0.6)


# interaction between temperature & pH
# estimate sign. degs 
nl_interaction.exon <-
    glmQLFTest(fit.exon,
    coef = 6,
    contrast = NULL,
    poisson.bound = FALSE) # Estimate significant DEGs
DEGs_nl_interaction.exon <-
  decideTestsDGE(nl_interaction.exon, adjust.method = "fdr", p.value = 0.05)
summary(DEGs_nl_interaction.exon)
plotMD(nl_interaction.exon, cex=0.6)
```

Alternative splicing analysis - differential exon expression, nonlinear model 

For each coefficient in the model (i.e. temperature_2, temperature, pH_2 ... etc.), identify differential exon usage with two statistical methods 

```{r}
# Effect of temperature^2 on exon usage
nl_temperature_2.splice <- diffSpliceDGE(fit.exon, coef=2, geneid="Geneid", exonid="Start")
topSpliceDGE(nl_temperature_2.splice, test="Simes", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1) #Simes' method, no genes with FDR < 0.1
topSpliceDGE(nl_temperature_2.splice, test="gene", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1) #F-tests, 4 genes with FDR < 0.1 

par(mfrow=c(2,2))
plotSpliceDGE(nl_temperature_2.splice, geneid="LOC115539697", genecol="Geneid")
plotSpliceDGE(nl_temperature_2.splice, geneid="col9a2", genecol="Geneid")
plotSpliceDGE(nl_temperature_2.splice, geneid="phb2", genecol="Geneid")
plotSpliceDGE(nl_temperature_2.splice, geneid="cdc5l", genecol="Geneid")

# Effect of temperature on exon usage
nl_temperature.splice <- diffSpliceDGE(fit.exon, coef=3, geneid="Geneid", exonid="Start")
topSpliceDGE(nl_temperature.splice, test="Simes", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1) # No genes
topSpliceDGE(nl_temperature.splice, test="gene", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1) # 5 genes with FDR < 0.1

# Plot 
plotSpliceDGE(nl_temperature_2.splice, geneid="fbl", genecol="Geneid") # first 3 exons higher expression 

# Effect of ph on exon usage
nl_ph.splice <- diffSpliceDGE(fit.exon, coef=4, geneid="Geneid", exonid="Start")
topSpliceDGE(nl_ph.splice, test="Simes", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1) #none
topSpliceDGE(nl_ph.splice, test="gene", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1) #none

# Interactive effect of temperature^2 and ph on exon usage
nl_interaction_2.splice <- diffSpliceDGE(fit.exon, coef=5, geneid="Geneid", exonid="Start")
topSpliceDGE(nl_interaction_2.splice, test="Simes", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1) #4 
topSpliceDGE(nl_interaction_2.splice, test="gene", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1) #5 

par(mfrow=c(3,2))
plotSpliceDGE(nl_temperature_2.splice, geneid="fbxo32", genecol="Geneid")
plotSpliceDGE(nl_temperature_2.splice, geneid="ddit4", genecol="Geneid")
plotSpliceDGE(nl_temperature_2.splice, geneid="LOC115542473", genecol="Geneid")
plotSpliceDGE(nl_temperature_2.splice, geneid="LOC115529695", genecol="Geneid")
plotSpliceDGE(nl_temperature_2.splice, geneid="LOC115531396", genecol="Geneid")

# Interactive effect of temperature and ph on exon usage
nl_interaction.splice <- diffSpliceDGE(fit.exon, coef=5, geneid="Geneid", exonid="Start")
topSpliceDGE(nl_interaction.splice, test="Simes", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1) #4, same as interaction with temperature_2
topSpliceDGE(nl_interaction.splice, test="gene", n=nrow(nl_temperature_2.splice$genes)) %>% filter(FDR<0.1) #5, same as interaction with temperature_2
```



```{r}
"../results/star/featurecounts/featurecounts_exon"

diffSpliceDGE()
topSpliceDGE()
plotSpliceDGE()
```











# BONEYARD 

## Differential expression analysis using my own code based on edgeR tutorial - 

```{r}
con <- makeContrasts(amb.6 - low.6, levels=design)
qlf <- glmQLFTest(fit, contrast=con)
topTags(qlf)
```

```{r}
summary(decideTests(qlf))
plotMD(qlf, cex=0.6)
```

```{r}
tr <- glmTreat(fit, contrast=con, lfc=log2(1.1))
topTags(tr)
summary(decideTests(tr))
plotMD(tr, cex=0.6)
```

## ANOVA-like testing, contrast includes 3 groups, here temperature

```{r}
## Ambient pH 
con <- makeContrasts(
  amb.3vs6 = amb.3 - amb.6,
  amb.3vs10 = amb.3 - amb.10,
  amb.6vs10 = amb.6 - amb.10, levels=design)
anov <- glmQLFTest(fit, contrast=con)
topTags(anov)
summary(decideTests(anov))
```

```{r}
## Low pH 
con <- makeContrasts(
  low.3vs6 = low.3 - low.6,
  low.3vs10 = low.3 - low.10,
  low.6vs10 = low.6 - low.10, levels=design)
anov <- glmQLFTest(fit, contrast=con)
topTags(anov)
summary(decideTests(anov))
```


### Can I do a blocked design, to try to remove any effects of tank 

```{r}
tank <- counts.treat$tank
treatment <- counts.treat$treatment
design <- model.matrix(~tank+treatment)
View(design)
design %>% as.data.frame()
```



```{r}
## Create 'targets' and 'Group dataframe, expressing experimental variables for DEG analysis
pCO2 <- as.numeric(c( 255, 255, 255, 255, 255, 255,
                      530, 530, 530, 530, 530, 530,
                      918, 918, 918, 918, 918, 918))
# Square pCO2 variable
pCO2_2 <- pCO2 ^ 2

treatment <- c("B","B", "B", "B", "B", "B",
               "R", "R", "R", "R", "R", "R",
               "Y", "Y", "Y", "Y", "Y", "Y")

day <- as.numeric(c(7, 7, 7, .5, .5, .5,
                    7, 7, 7, .5, .5, .5,
                    7, 7, 7, .5, .5, .5))

targets <- data.frame(pCO2, day, treatment)
targets$grouping <- paste(targets$pCO2, targets$day, sep = ".")

# The group factor represents the combined levels of an experimental replicate across all variables
Group <- factor(paste(targets$day, targets$pCO2, sep = "_"))

# Fit multifactorial design matrix
design_nl <-
    model.matrix(~ 1 + pCO2_2 + pCO2 + pCO2_2:day + pCO2:day) # Generate multivariate edgeR glm

View(design_nl)
```

