---
title: "GoSeq-enrichment"
author: "Laura Spencer"
date: "2024-06-13"
output: html_document
---

## Let's test out goseq (enrichment program that Ariana uses)

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("goseq")
require(goseq)
require(tidyverse)
`%!in%` = Negate(`%in%`)

if ("GSEABase" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GSEABase")
if ("GOSim" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GOSim")
if ("stats" %in% rownames(installed.packages()) == 'FALSE') install.packages("stats")
if ("ggdendro" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("ggdendro")
if ("GO.db" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("GO.db")
if ("rrvgo" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("rrvgo")

BiocManager::install("org.Dr.eg.db", force=TRUE) #install if needed 

library("GSEABase")
library("GOSim")
library("stats")
library("ggdendro")
library("GO.db")
library("rrvgo")
```

```{r}
diffex.pH.6.filt %>% as.data.frame() %>% 
#  filter(log2FoldChange>0.5)
  filter(log2FoldChange< -0.5)
```

GOSeq Enrichment Analysis with Differentially Expressed Genes 

```{r}
DEGs.4GoSeq <- list(
  
  "OA_Downregulated" = 
    # LFC>0.5 - LOWER IN LOW PH - Downregulated in OA
    res.pH.6 %>% as.data.frame() %>% rownames_to_column("gene") %>%
    mutate(DEG=case_when((gene %in% rownames(diffex.pH.6.filt) & log2FoldChange>0.5) ~ 1, TRUE~0)) %>%
    dplyr::select(gene, log2FoldChange, padj, DEG) %>%
    left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, keyword_id), by=c("gene"="gene_gadmor")) %>% 
  
    left_join(gadmor.gff %>% filter(type=="CDS") %>% mutate(length=end-start) %>%
                group_by(gene) %>% summarize(median_length=round(median(length), digits = 0)), #add median transcript length for each gene 
              by=c("gene"="gene")) %>% filter(!is.na(gene_ontology_i_ds)),

  "OA_Upregulated" = 
    # LFC<0.5 - HIGHER IN LOW PH - Upregulated in OA
    res.pH.6 %>% as.data.frame() %>% rownames_to_column("gene") %>%
    mutate(DEG=case_when((gene %in% rownames(diffex.pH.6.filt) & log2FoldChange< -0.5) ~ 1, TRUE~0)) %>%
    dplyr::select(gene, log2FoldChange, padj, DEG) %>%
    left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, keyword_id), by=c("gene"="gene_gadmor")) %>% 
  
    left_join(gadmor.gff %>% filter(type=="CDS") %>% mutate(length=end-start) %>%
                group_by(gene) %>% summarize(median_length=round(median(length), digits = 0)), #add median transcript length for each gene 
              by=c("gene"="gene")) %>% filter(!is.na(gene_ontology_i_ds)),
  

    "Cold_Upregulated" = 
    # LFC>0.5 - HIGHER IN 3C - Upregulated in Cold Temperature
    res.temp.3.6.amb %>% as.data.frame() %>% rownames_to_column("gene") %>%
    mutate(DEG=case_when((gene %in% rownames(diffex.temp.3.6.amb.filt) & log2FoldChange>0.5) ~ 1, TRUE~0)) %>% 
    dplyr::select(gene, log2FoldChange, padj, DEG) %>%
    left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, keyword_id), by=c("gene"="gene_gadmor")) %>% 
  
    left_join(gadmor.gff %>% filter(type=="CDS") %>% mutate(length=end-start) %>%
                group_by(gene) %>% summarize(median_length=round(median(length), digits = 0)), #add median transcript length for each gene 
              by=c("gene"="gene")) %>% filter(!is.na(gene_ontology_i_ds)),

  "Cold_Downregulated"  =
    # LFC<0.5 - LOWER IN 3C - Downregulated in Cold Temperature
    res.temp.3.6.amb %>% as.data.frame() %>% rownames_to_column("gene") %>%
    mutate(DEG=case_when((gene %in% rownames(diffex.temp.3.6.amb.filt) & log2FoldChange< -0.5) ~ 1, TRUE~0)) %>% 
    dplyr::select(gene, log2FoldChange, padj, DEG) %>%
    left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, keyword_id), by=c("gene"="gene_gadmor")) %>% 
  
    left_join(gadmor.gff %>% filter(type=="CDS") %>% mutate(length=end-start) %>%
                group_by(gene) %>% summarize(median_length=round(median(length), digits = 0)), #add median transcript length for each gene 
              by=c("gene"="gene")) %>% filter(!is.na(gene_ontology_i_ds)),
  
  "Warm_Downregulated" = 
    # LFC>0.5 - LOWER IN 10C - Downregulated in Warm Temperature
    res.temp.6.10.amb %>% as.data.frame() %>% rownames_to_column("gene") %>%
    mutate(DEG=case_when((gene %in% rownames(diffex.temp.6.10.amb.filt) & log2FoldChange>0.5) ~ 1, TRUE~0)) %>% 
    dplyr::select(gene, log2FoldChange, padj, DEG) %>%
    left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, keyword_id), by=c("gene"="gene_gadmor")) %>% 
  
    left_join(gadmor.gff %>% filter(type=="CDS") %>% mutate(length=end-start) %>%
                group_by(gene) %>% summarize(median_length=round(median(length), digits = 0)), #add median transcript length for each gene 
              by=c("gene"="gene")) %>% filter(!is.na(gene_ontology_i_ds)),

  "Warm_Upregulated" = 
    # LFC<0.5 - HIGHER IN 10C - Upregulated in Warm Temperature
    res.temp.6.10.amb %>% as.data.frame() %>% rownames_to_column("gene") %>%
    mutate(DEG=case_when((gene %in% rownames(diffex.temp.6.10.amb.filt) & log2FoldChange< -0.5) ~ 1, TRUE~0)) %>% 
    dplyr::select(gene, log2FoldChange, padj, DEG) %>%
    left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, keyword_id), by=c("gene"="gene_gadmor")) %>% 
  
    left_join(gadmor.gff %>% filter(type=="CDS") %>% mutate(length=end-start) %>%
                group_by(gene) %>% summarize(median_length=round(median(length), digits = 0)), #add median transcript length for each gene 
              by=c("gene"="gene")) %>% filter(!is.na(gene_ontology_i_ds)),
  

  "Cold-OA_Upregulated" = 
    # LFC>0.5 - HIGHER IN 3C-LOW PH - Upregulated in Cold+OA
    res.both.3low.6amb %>% as.data.frame() %>% rownames_to_column("gene") %>%
    mutate(DEG=case_when((gene %in% rownames(diffex.both.3low.6amb.filt) & log2FoldChange>0.5) ~ 1, TRUE~0)) %>% 
    dplyr::select(gene, log2FoldChange, padj, DEG) %>%
    left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, keyword_id), by=c("gene"="gene_gadmor")) %>% 
  
    left_join(gadmor.gff %>% filter(type=="CDS") %>% mutate(length=end-start) %>%
                group_by(gene) %>% summarize(median_length=round(median(length), digits = 0)), #add median transcript length for each gene 
              by=c("gene"="gene")) %>% filter(!is.na(gene_ontology_i_ds)),

  "Cold-OA_Downregulated" = 
    # LFC<0.5 - LOWER IN 3C-LOW PH - Downregulated in Cold+OA
    res.both.3low.6amb %>% as.data.frame() %>% rownames_to_column("gene") %>%
    mutate(DEG=case_when((gene %in% rownames(diffex.both.3low.6amb.filt) & log2FoldChange< -0.5) ~ 1, TRUE~0)) %>% 
    dplyr::select(gene, log2FoldChange, padj, DEG) %>%
    left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, keyword_id), by=c("gene"="gene_gadmor")) %>% 
  
    left_join(gadmor.gff %>% filter(type=="CDS") %>% mutate(length=end-start) %>%
                group_by(gene) %>% summarize(median_length=round(median(length), digits = 0)), #add median transcript length for each gene 
              by=c("gene"="gene")) %>% filter(!is.na(gene_ontology_i_ds)),
    
  "Warm-OA_Upregulated" = 
    # lfc<0.5 - HIGHER IN 10C-LOW PH - Upregulated in Warm+OA
    res.both.10low.6amb %>% as.data.frame() %>% rownames_to_column("gene") %>%
    mutate(DEG=case_when((gene %in% rownames(diffex.both.10low.6amb.filt) & log2FoldChange< -0.5) ~ 1, TRUE~0)) %>% 
    dplyr::select(gene, log2FoldChange, padj, DEG) %>%
    left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, keyword_id), by=c("gene"="gene_gadmor")) %>% 
  
    left_join(gadmor.gff %>% filter(type=="CDS") %>% mutate(length=end-start) %>%
                group_by(gene) %>% summarize(median_length=round(median(length), digits = 0)), #add median transcript length for each gene 
              by=c("gene"="gene")) %>% filter(!is.na(gene_ontology_i_ds)),

  "Warm-OA_Downregulated" = 
    # lfc>0.5 - LOWER IN 10C-LOW PH - Downregulated in Warm+OA
    res.both.10low.6amb %>% as.data.frame() %>% rownames_to_column("gene") %>%
    mutate(DEG=case_when((gene %in% rownames(diffex.both.10low.6amb.filt) & log2FoldChange> 0.5) ~ 1, TRUE~0)) %>% 
    dplyr::select(gene, log2FoldChange, padj, DEG) %>%
    left_join(gadMor.blast.GO %>% dplyr::select(gene_gadmor, gene_ontology_i_ds, keyword_id), by=c("gene"="gene_gadmor")) %>% 
  
    left_join(gadmor.gff %>% filter(type=="CDS") %>% mutate(length=end-start) %>%
                group_by(gene) %>% summarize(median_length=round(median(length), digits = 0)), #add median transcript length for each gene 
              by=c("gene"="gene")) %>% filter(!is.na(gene_ontology_i_ds))) 

```

Conduct functional enrichment for each contrast using goseq and rrvgo. Compare against all genes detected in the dataset. I am choosing to compare against genes in our dataset because we are looking at one developmental stage (larvae). Therefore, I do not want to conduct enrichment of all genes in the genome because many functions including calcification and gametogenesis do not apply to the biological function of larvae.  Conduct at biological process level. Filter for FDR < 0.1

Prep and run GOSeq analysis for each differentially expressed gene set using loop

```{r, warning=F, message=F, echo=F}
GOSeq.results <- data.frame()

for (i in 1:length(DEGs.4GoSeq)) {

    ##== Prep gene lists and weights
  
    ##Construct named vector of all genes (names=gene IDs) with 1 indicating DEG and 0 not a DEG
    gene.vector=setNames(DEGs.4GoSeq[[i]]$DEG, DEGs.4GoSeq[[i]]$gene)

    ### Generate vector with DEG gene IDs
    ID.vector <- DEGs.4GoSeq[[i]] %>%
      filter(DEG==1)%>%
      pull(gene)

    ### Generate length vector for all genes
    LENGTH.vector <- c(DEGs.4GoSeq[[i]]$median_length)

    #weight gene vector by bias for length of gene 
    pwf<-nullp(gene.vector, ID.vector, bias.data=LENGTH.vector) 
    
    ## === Prep GO information 
    
     ##Get a list of GO Terms for all genes detected in our dataset 
    GO.terms <- DEGs.4GoSeq[[i]] %>%
      dplyr::select(gene, gene_ontology_i_ds) %>% dplyr::rename("GOs"="gene_ontology_i_ds")
    
    ##Format to have one goterm per row with gene ID repeated
    split <- strsplit(as.character(GO.terms$GOs), ";") 
    split2 <- data.frame(v1 = rep.int(GO.terms$gene, sapply(split, length)), v2 = unlist(split)) 
    colnames(split2) <- c("gene", "GOs")
    GO.terms<-split2 %>% mutate(GOs=gsub(" ", "", GOs))

    ### === Run GOSeq analysis
    
    #run goseq using Wallenius method for all categories of GO terms 
    GO.wall<- goseq(pwf, ID.vector, gene2cat=GO.terms, test.cats=c("GO:BP"), method="Wallenius", use_genes_without_cat=TRUE)

    GO <- GO.wall[order(GO.wall$over_represented_pvalue),]
    colnames(GO)[1] <- "GOterm"
    
    #adjust p-values (FDR)
    GO$fdr_over <-  p.adjust(GO$over_represented_pvalue, method="fdr") #add adjusted p-values
    GO$fdr_under <-  p.adjust(GO$under_represented_pvalue, method="fdr") #add adjusted p-values

    #Filtering for p < 0.01
    GO <- GO %>%
        dplyr::filter(fdr_over<0.1) %>%
        dplyr::arrange(., ontology, fdr_over) %>%
      mutate(treatment=names(DEGs.4GoSeq[i]))

    #Write file of results 
    write_csv(GO, file = paste0("../results/goseq/goseq_results_", names(DEGs.4GoSeq[i]), ".csv"))
    
    #Add to GOSeq results list
    GOSeq.results <- rbind(GOSeq.results, GO)
    
    print(paste("Treatment: ", names(DEGs.4GoSeq[i]), ", enriched BP GO Terms (@ FDR<10%) = ", nrow(GO), sep=""))
}
```
Use REVIGO to collapse GO terms into parent GO terms using semantic similarity 

```{r}
# Create dataframe to save collapsed GO enrichment results to. 
# To collapse using Revigo there needs to be 2+ GO terms. Here, I add enrichment results for treatments with less than 2 enriched BP 
GOSeq.revigo <- (GOSeq.results %>% group_by(treatment, ontology) %>% summarise(n=n()) %>%
  filter(ontology=="BP") %>% 
#  filter(ontology=="MF") %>% 
#  filter(ontology=="CC") %>% 
    filter(n>1))$treatment

GOSeq.results.collapsed <- GOSeq.results %>% 
  filter(ontology=="BP") %>% 
#  filter(ontology=="MF") %>% 
#  filter(ontology=="CC") %>% 
  filter(treatment %!in% GOSeq.revigo) %>% 
  separate(treatment, sep = "_", into = c("treatment", "response")) %>% mutate(ParentTerm=NA)

for (i in 1:length(GOSeq.revigo)) {

    #Read relevant file of results from goseq analysis  
    go_results<-read_csv(file = paste0("../results/goseq/goseq_results_", GOSeq.revigo[i], ".csv"))
    
    go_results<-go_results%>%
      filter(ontology=="BP")%>%
#      filter(ontology=="MF")%>%
#      filter(ontology=="CC")%>%
      filter(fdr_under != "NA") %>%
      #filter(numInCat>5)%>%
      arrange(., fdr_under)
    
    print(GOSeq.revigo[i])
    print(length(unique(go_results$GOterm)))
    
    #Reduce/collapse GO term set with the rrvgo package 
    simMatrix <- calculateSimMatrix(go_results$GOterm,
                                orgdb="org.Dr.eg.db", #zebrafish database
                                ont=c("BP"),
#                                ont=c("MF"),
#                                ont=c("CC"),
                                method="Rel")
    #calculate similarity 
    scores <- setNames(-log10(go_results$fdr_under), go_results$GOterm)
    reducedTerms <- reduceSimMatrix(simMatrix,
                                scores,
                                threshold=0.7,
                                orgdb="org.Dr.eg.db")
    
    #keep only the goterms from the reduced list
    go_results<-go_results%>%
      filter(GOterm %in% reducedTerms$go)
    
    #add in parent terms to list of go terms 
    go_results$ParentTerm<-reducedTerms$parentTerm[match(go_results$GOterm, reducedTerms$go)]
    
    #add in contrast and regulation
    go_results <- go_results %>% separate(treatment, sep = "_", into = c("treatment", "response"))
    # go_results$direction<-c("upregulated")
    # go_results$contrast<-temp
    
    print(length(unique(go_results$GOterm)))
    
    #add to data frame 
    temp_go_df <- go_results
    GOSeq.results.collapsed <- rbind(GOSeq.results.collapsed, temp_go_df) 
}

GOSeq.results.collapsed %>% 
  mutate(treatment=as.factor(treatment), response=as.factor(response)) %>%
  
  ungroup() %>% arrange(response, rev(treatment), fdr_over) %>%
  mutate(ParentTerm = factor(ParentTerm, rev(unique(ParentTerm)), ordered = TRUE),
         treatment=factor(treatment, ordered = TRUE, levels=c("OA", "Warm", "Warm-OA", "Cold", "Cold-OA"))) %>%

ggplot(aes(y = str_wrap(ParentTerm), x=treatment:response, col=treatment, shape=response)) + 
  geom_point(alpha=0.75, aes(size=-log10(fdr_over))) + #size=count,  alpha=p_value, 
 facet_wrap(~ontology,scales="free", nrow = 2) +
 #  scale_shape_manual(name="Gene Set", labels=c("Upregulated"="More gene transcripts",
 #                                               "Downregulated"="Fewer gene transcripts"),
 #                     values=c("Upregulated"=17, "Downregulated"=16),
 #                     guide = guide_legend(override.aes = list(size=3))) +
 # scale_color_manual(name="Stressor",
 #                     values=c("OA"=lighten("darkorange2", 0.25), 
 #                              "Warm"=lighten("firebrick4", 0.45), 
 #                              "Warm-OA"="firebrick4", 
 #                              "Cold"=lighten("navyblue", 0.45),
 #                              "Cold-OA"="navyblue"),
 #             guide = guide_legend(override.aes = list(size=4))) +  
  scale_size("-Log10 FDR", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(order = 1, override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=7), 
        legend.title = element_text(size=7),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Enriched Biological Processes (GO Terms)")
#   ggtitle("Enriched Molecular Functions (GO Terms)")
#  ggtitle("Enriched Cellular Components (GO Terms)")
```
```{r}
uniprot.keywords <- read_delim("../references/uniprot_keywords_all_2024_06_17.tsv") %>% clean_names()
View(uniprot.keywords)
```

Uniprot Keyword Enrichment Analysis

```{r, warning=F, message=F, echo=F}
GOSeq.results.KW <- data.frame()

for (i in 1:length(DEGs.4GoSeq)) {

    ##== Prep gene lists and weights
  
    ##Construct named vector of all genes (names=gene IDs) with 1 indicating DEG and 0 not a DEG
    gene.vector=setNames(DEGs.4GoSeq[[i]]$DEG, DEGs.4GoSeq[[i]]$gene)

    ### Generate vector with DEG gene IDs
    ID.vector <- DEGs.4GoSeq[[i]] %>%
      filter(DEG==1)%>%
      pull(gene)

    ### Generate length vector for all genes
    LENGTH.vector <- c(DEGs.4GoSeq[[i]]$median_length)

    #weight gene vector by bias for length of gene 
    pwf<-nullp(gene.vector, ID.vector, bias.data=LENGTH.vector) 
    
    ## === Prep GO information 
    
     ##Get a list of GO Terms for all genes detected in our dataset 
    UP.KWs <- DEGs.4GoSeq[[i]] %>%
      dplyr::select(gene, keyword_id) %>% dplyr::rename("KWs"="keyword_id")
    
    ##Format to have one goterm per row with gene ID repeated
    split <- strsplit(as.character(UP.KWs$KWs), ";") 
    split2 <- data.frame(v1 = rep.int(UP.KWs$gene, sapply(split, length)), v2 = unlist(split)) 
    colnames(split2) <- c("gene", "KWs")
    UP.KWs<-split2 %>% mutate(KWs=gsub(" ", "", KWs))

    # Get the gene lists of "numDFinCat" in GO.wall report to add gene info
    cat2gene <- UP.KWs %>%
      filter(gene %in% (DEGs.4GoSeq[[i]] %>% filter(DEG==1))$gene) %>%
      group_by(KWs) %>%
      summarise(genes = toString(gene)) %>%
      ungroup()

    ### === Run GOSeq analysis
    
    #run goseq using Wallenius method for all categories of GO terms 
    KW.wall<- goseq(pwf, ID.vector, gene2cat=UP.KWs, test.cats=c("GO:BP"), method="Wallenius", use_genes_without_cat=TRUE)

    KW <- KW.wall[order(KW.wall$over_represented_pvalue),]
    colnames(KW)[1] <- "KWterm"
    
    #adjust p-values (FDR)
    KW$fdr_over <-  p.adjust(KW$over_represented_pvalue, method="fdr") #add adjusted p-values
    KW$fdr_under <-  p.adjust(KW$under_represented_pvalue, method="fdr") #add adjusted p-values

    #Filtering for p < 0.01
    KW <- KW %>%
        dplyr::filter(fdr_over<0.1) %>%
        dplyr::arrange(., KWterm, fdr_over) %>%
      mutate(treatment=names(DEGs.4GoSeq[i])) %>%
      left_join(uniprot.keywords %>% dplyr::select(keyword_id, name, category), by=c("KWterm"="keyword_id")) %>%
      left_join(cat2gene, by = c("KWterm"="KWs"))
    
    #Write file of results 
    write_csv(KW, file = paste0("../results/goseq/goseq_results_", names(DEGs.4GoSeq[i]), ".csv"))
    
    #Add to GOSeq results list
    GOSeq.results.KW <- rbind(GOSeq.results.KW, KW)
    
    print(paste("Treatment: ", names(DEGs.4GoSeq[i]), ", enriched BP Keywords (@ FDR<10%) = ", nrow(KW), sep=""))
}
```


```{r}
GOSeq.results.KW <- GOSeq.results.KW %>% 
  separate(treatment, into = c("treatment", "response"), sep = "_") %>%
  filter(category=="Biological process") %>%
  mutate(treatment=as.factor(treatment), response=as.factor(response)) 

GOSeq.results.KW %>%
#    filter(response=="Upregulated") %>%
    filter(response=="Downregulated") %>% 
  filter(name!="Hemostasis") %>% #Hemostasis is the parent term for Blood coagulation, both have same stats and genes contributing, so pick child term! 
  # group_by(treatment, response, numDEInCat, genes) %>%
  # dplyr::summarise(fdr_over=min(fdr_over)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  # left_join(GOSeq.results.KW, by = c("treatment", "response", "numDEInCat", "genes", "fdr_over")) %>%  #re-add data

  ungroup() %>% arrange(treatment, fdr_over) %>%
  mutate(name = factor(name, rev(unique(name)), ordered = TRUE),
         treatment=factor(treatment, ordered = TRUE, levels=c("OA", "Warm", "Warm-OA", "Cold", "Cold-OA"))) %>%

ggplot(aes(y = name, x=treatment, col=treatment)) + 
  geom_point(alpha=0.75, aes(size=-log10(fdr_over+0.000000001))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(drop=F, name="Stressor",
                     values=c("OA"=lighten("darkorange2", 0.25),
                              "Warm"=lighten("firebrick4", 0.45),
                              "Warm-OA"="firebrick4",
                              "Cold"=lighten("navyblue", 0.45),
                              "Cold-OA"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 FDR", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(order = 1, override.aes = list(col="gray50"))) +
  scale_x_discrete() + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=7), 
        legend.title = element_text(size=7),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
#  ggtitle("Genes with higher expression")
  ggtitle("Genes with lower expression")

```
