{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "893c947d",
   "metadata": {},
   "source": [
    "We are interested in the expression and haplotype of a particular ZP3 gene (LOC115551355 in the Atlantic cod genome gadMor3.0), as it has previously been implicated as under selection and there is distinct spatial structure in Alaskan waters. We have expression data for this ZP3 gene, which show higher levels in cold (3C) compared to control (6C) and warm temperature (10C). Expression level also decreases linearly with increasing temperature. \n",
    "\n",
    "![zp3-boxplot](../results/ZP3-boxplot.jpg)\n",
    "![zp3-linear](../results/ZP3-linear-models.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a68dc722",
   "metadata": {},
   "source": [
    "We want to identify the ZP3 haplotype of our experimental P. cod larvae from the RNASeq data. In this notebook I will explore the alignment data for the region containing the ZP3 gene only. \n",
    "\n",
    "First I will see about pulling alignment data for only the ZP3 gene. Where is that gene? \n",
    "```\n",
    "(base) [lspencer@sedna GCF_902167405.1]$ cat gadMor3.0_genes.bed | grep \"115551355\"\n",
    "NC_044056.1     2454600 2458007 gene-LOC115551355;Dbxref=GeneID:115551355;Name=LOC115551355;gbkey=Gene;gene=LOC115551355;gene_biotype=protein_coding\n",
    "```\n",
    "\n",
    "The ZP3 gene is on the RefSeq \"NC_044056.1\", which is chromosome 9 on the [Atlantic cod genome](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_902167405.1/). It spans from  2454600-2458007. Let's try to extract alignment data for just that region: \n",
    "\n",
    "\n",
    "```\n",
    "SAMPLES=$(ls *.Aligned.sortedByCoord.out.bam)\n",
    "\n",
    "for sample in ${SAMPLES}\n",
    "do\n",
    "\n",
    "# First index all .bam files\n",
    "samtools index ${sample}\n",
    "done \n",
    "```\n",
    "\n",
    "The VPN was terminated midway, so I reran the above code on all samples starting from PCG157 using this code:\n",
    "\n",
    "```\n",
    "SAMPLES=$(ls *.Aligned.sortedByCoord.out.bam)\n",
    "startfile=\"PCG353.Aligned.sortedByCoord.out.bam\"\n",
    "for FILE in ${SAMPLES}; do\n",
    "    if ! [ \"$FILE\" '<' \"$startfile\" ] ; then\n",
    "        echo indexing $FILE\n",
    "        samtools index $FILE\n",
    "    else\n",
    "        echo Skipping $FILE\n",
    "    fi\n",
    "done\n",
    "```\n",
    "\n",
    "??NC_044056.1:2,454,259 - 2,458,347??\n",
    "Now extract just ZP3 gene region, and index that new bam file \n",
    "```\n",
    "SAMPLES=$(ls *.Aligned.sortedByCoord.out.bam | sed -e 's/.Aligned.sortedByCoord.out.bam//')\n",
    "\n",
    "for sample in ${SAMPLES}\n",
    "do\n",
    "samtools view -b -h ${sample}.Aligned.sortedByCoord.out.bam \"NC_044056.1:2454600-2458007\" > ${sample}.ZP3.bam\n",
    "samtools sort ${sample}.ZP3.bam > ${sample}.ZP3.sorted.bam\n",
    "samtools index ${sample}.ZP3.sorted.bam\n",
    "done\n",
    "```\n",
    "\n",
    "#### What I actually did: \n",
    "\n",
    "Pull alignment files for the ZP3 plus/minus 2kb\n",
    "gene location on complement strand (i.e. minus strand): 2454600 - 2458007\n",
    "pull this range: 2452600-2460007\n",
    "\n",
    "```\n",
    "SAMPLES=$(ls *.Aligned.sortedByCoord.out.bam | sed -e 's/.Aligned.sortedByCoord.out.bam//')\n",
    "\n",
    "for sample in ${SAMPLES}\n",
    "do\n",
    "samtools view -b -h ${sample}.Aligned.sortedByCoord.out.bam \"NC_044056.1:2452600-2460007\" > ${sample}.zp3.bam\n",
    "samtools sort ${sample}.zp3.bam > ${sample}.zp3.sorted.bam\n",
    "samtools index ${sample}.zp3.sorted.bam\n",
    "done\n",
    "```\n",
    "\n",
    "I could use IGV to pull a consensus sequence for each sample individuall, which would be a pain and not super accurate since samples vary considerably in their coverage. So, instead I'll merge alignment data from all samples into one file, then view that in IGV and pull a consensus sequence for the ZP3 gene from all samples. \n",
    "```\n",
    "samtools merge merged.zp3.bam *.zp3.sorted.bam\n",
    "samtools sort merged.zp3.bam > merged.zp3.sorted.bam\n",
    "samtools index merged.zp3.sorted.bam\n",
    "```\n",
    "\n",
    "\n",
    "Nice code to show the last modification date/time: \n",
    "```\n",
    "bams=$(ls *.chr9.sorted.bam)\n",
    "for file in ${bams}\n",
    "do\n",
    "echo ${file}\n",
    "date -r ${file}\n",
    "done\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
