---
title: "More-code-deseq2"
author: "Laura Spencer"
date: '2022-10-25'
output: html_document
---

### Generate pairwise comparison heatmaps 

```{r}
# Create count matrices containing DEGs for each pairwise treatment comparison, rows (genes) ordered by pvalue,

diffex.AM.counts <- subset(assay(vsd.pH), rownames(dds.DESeq.pH) %in% rownames(diffex.AM)) %>% as.data.frame() %>%
  dplyr::select((sample.info %>% arrange(Treatment) %>%
                   filter(Treatment=="Ambient" | Treatment== "Moderate"))$Sample) %>%
  rownames_to_column(var = "gene") %>% arrange(match(gene, c(as.data.frame(diffex.AM) %>% arrange(padj) %>% rownames()))) %>%
  column_to_rownames(var = "gene") %>% as.matrix()

diffex.AL.counts <- subset(assay(vsd.pH), rownames(dds.DESeq.pH) %in% rownames(diffex.AL)) %>% as.data.frame() %>%
  dplyr::select((sample.info %>% arrange(Treatment) %>%
                   filter(Treatment=="Ambient" | Treatment== "Low"))$Sample) %>%
  rownames_to_column(var = "gene") %>% arrange(match(gene, c(as.data.frame(diffex.AL) %>% arrange(padj) %>% rownames()))) %>%
  column_to_rownames(var = "gene") %>% as.matrix()

diffex.ML.counts <- subset(assay(vsd.pH), rownames(dds.DESeq.pH) %in% rownames(diffex.ML)) %>% as.data.frame() %>%
  dplyr::select((sample.info %>% arrange(desc(Treatment)) %>%
                   filter(Treatment=="Moderate" | Treatment== "Low"))$Sample) %>%
  rownames_to_column(var = "gene") %>%
  arrange(match(gene, c(as.data.frame(diffex.ML) %>% arrange(padj) %>% rownames()))) %>% map_df(rev) %>%
  column_to_rownames(var = "gene") %>% as.matrix()

# Define gradient range, breaks
my_colours <- rev(c("springgreen4", "gray95", "slateblue4"))
my_gradient <- colorRampPalette(my_colours)(length(breaksList))

# NOTE, the row-wise scale produces Z-scores (x-mean/sd)

# Draw heatmaps
pdf(file = "../results/heatmap-AM.pdf", width = 6.15, height = 3.15)
(heat.am <- pheatmap(diffex.AM.counts, cluster_cols = FALSE, cluster_rows=TRUE,
         show_rownames=FALSE, show_colnames = FALSE, na.rm=TRUE, scale="row",
         #main = "DEGs among pH 8.0 (ambient) & pH 7.8 (moderate), sorted by p-value",
         treeheight_row = 0, treeheight_col = 0,
         fontsize = 13, cutree_rows = 2,  border_color=F,
         annotation_colors = list(Treatment=c(Ambient="#2c7bb6", Moderate="#fdae61")),
         annotation_col=dds.pH.df[colnames(diffex.AM.counts),"Treatment", drop=FALSE],
         annotation_legend = FALSE, annotation_names_col= FALSE,
         #color = colorRampPalette(rev(brewer.pal(n = 5, name = "BrBG")))(length(breaksList)),
         color = my_gradient,
         breaks = breaksList))
dev.off()

pdf(file = "../results/heatmap-AL.pdf", width = 6.15, height = 3.15)
(heat.al <- pheatmap(diffex.AL.counts, cluster_cols = FALSE, cluster_rows=TRUE,
         show_rownames=FALSE, show_colnames = FALSE, na.rm=TRUE, scale="row",
         #main = "DEGs among pH 8.0 (ambient) & pH 7.5 (low), sorted by p-value",
         treeheight_row = 0, treeheight_col = 0,
         fontsize = 13, cutree_rows = 2, border_color=F,
         annotation_colors = list(Treatment=c(Ambient="#2c7bb6", Low="#d7191c")),
         annotation_col=dds.pH.df[colnames(diffex.AL.counts),"Treatment", drop=FALSE],
         annotation_legend = FALSE, annotation_names_col= FALSE,
         #color = colorRampPalette(rev(brewer.pal(n = 5, name = "BrBG")))(length(breaksList)),
         color = my_gradient,
         breaks = breaksList))
dev.off()

pdf(file = "../results/heatmap-ML.pdf", width = 6.15, height = 3.15)
(heat.ml <- pheatmap(diffex.ML.counts, cluster_cols = FALSE, cluster_rows=TRUE,
         show_rownames=FALSE, show_colnames = FALSE, na.rm=TRUE, scale="row",
         #main = "DEGs among pH 7.8 (moderate) & pH 7.5 (low), sorted by p-value",
         treeheight_row = 0, treeheight_col = 0,
         fontsize = 13, cutree_rows = 2, border_color=F,
         annotation_colors = list(Treatment=c(Moderate="#fdae61", Low="#d7191c")),
         annotation_col=dds.pH.df[colnames(diffex.ML.counts),"Treatment", drop=FALSE],
         annotation_legend = FALSE, annotation_names_col= FALSE,
         #color = colorRampPalette(rev(brewer.pal(n = 5, name = "BrBG")))(length(breaksList)),
         color = my_gradient,
         breaks = breaksList))
dev.off()

save(diffex.AM.counts, file = "../results/deseq2/diffex.AM.counts")
save(diffex.AL.counts, file = "../results/deseq2/diffex.AL.counts")
save(diffex.ML.counts, file = "../results/deseq2/diffex.ML.counts")
```

Volcano plots 

```{r}
# Set axis ticks for all volcano plots
xtick=seq(-6,4, by=2)
ytick=seq(0,15, by=5)
```

```{r}
# ----------- AMBIENT VS MODERATE 
topT <- as.data.frame(res.pco2.AM)

par(mar=c(2.5,2.5,1.5,1), cex.main=.9, cex.axis=.8, cex.lab=.8)

#Adjusted P values (FDR Q values)
with(topT, plot(-1*log2FoldChange, -log10(padj), pch=16, cex=0.45, main="", xlab="", ylab="", xlim=c(-7,4), ylim=c(0,15), xaxt="n", yaxt="n"))
#axis(side=1, at=xtick, labels=F)
#text(par("usr")[3], x=xtick, labels = xtick, pos = 1, xpd = TRUE, cex = 1)
axis(side=2, at=ytick, labels=F)
text(par("usr")[1], ytick, labels = ytick, pos = 2, xpd = TRUE, cex = 1)
title(main="(A) Ambient vs. Moderate OA", adj=0.01, line=-1.2, font.main=1, cex.main=1.3)
title(xlab=NA, ylab=bquote(~-log[10]~P-value), line=1.3, cex.lab=1.15)

# plot genes downregulated in OA (i.e. upregulated in ambient)
# with LFC >1
with(subset(topT, padj<0.05 & -1*log2FoldChange< -1), points(-1*log2FoldChange, -log10(padj), pch=16, col="#2c7bb6", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange< 0  & -1*log2FoldChange> -1), points(-1*log2FoldChange, -log10(padj), pch=16, col="#d1e5f0", cex=0.55))

# plot genes upregulated in moderate OA 
with(subset(topT, padj<0.05 & -1*log2FoldChange>1), points(-1*log2FoldChange, -log10(padj), pch=16, col="#fdae61", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange>0 & -1*log2FoldChange<1), points(-1*log2FoldChange, -log10(padj), pch=16, col="#fddbc7", cex=0.55))


#Add lines for absolute FC>2 and P-value cut-off at FDR Q<0.05
abline(v=0, col="black", lty=3, lwd=1.0)
abline(v=-1, col="black", lty=4, lwd=2.0)
abline(v=1, col="black", lty=4, lwd=2.0)
abline(h=-log10(max(topT$padj[topT$padj<0.05], na.rm=TRUE)), col="black", lty=4, lwd=2.0)
volcano.am <- recordPlot()

pdf(file = "../results/volcano-AM.pdf", width = 6.25, height = 3.75)
print(volcano.am)
dev.off()
```

```{r}
# ----------- AMBIENT VS LOW 
topT <- as.data.frame(res.pco2.AL)

par(mar=c(2.5,2.5,1.5,1), cex.main=.9, cex.axis=.8, cex.lab=.8)

#Adjusted P values (FDR Q values)
with(topT, plot(-1*log2FoldChange, -log10(padj), pch=16, cex=0.45, main="", xlab="", ylab="", xlim=c(-7,4), ylim=c(0,15), xaxt="n", yaxt="n"))
#axis(side=1, at=xtick, labels=F)
#text(par("usr")[3], x=xtick, labels = xtick, pos = 1, xpd = TRUE, cex = 1)
axis(side=2, at=ytick, labels=F)
text(par("usr")[1], ytick, labels = ytick, pos = 2, xpd = TRUE, cex = 1)
title(main="(B) Ambient vs. Severe OA", adj=0.01, line=-1.2, font.main=1, cex.main=1.3)
title(xlab=NA, ylab=bquote(~-log[10]~P-value), line=1.3, cex.lab=1.15)

# plot genes upregulated in ambient 
# with LFC >1
with(subset(topT, padj<0.05 & -1*log2FoldChange< -1), points(-1*log2FoldChange, -log10(padj), pch=16, col="#2c7bb6", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange< 0  & -1*log2FoldChange> -1), points(-1*log2FoldChange, -log10(padj), pch=16, col="#d1e5f0", cex=0.55))

# plot genes upregulated in severe OA 
with(subset(topT, padj<0.05 & -1*log2FoldChange>1), points(-1*log2FoldChange, -log10(padj), pch=16, col="#d7191c", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange>0 & -1*log2FoldChange<1), points(-1*log2FoldChange, -log10(padj), pch=16, col="pink2", cex=0.55))

#Add lines for absolute FC>2 and P-value cut-off at FDR Q<0.05
abline(v=0, col="black", lty=3, lwd=1.0)
abline(v=-1, col="black", lty=4, lwd=1.9)
abline(v=1, col="black", lty=4, lwd=1.9)
abline(h=-log10(max(topT$padj[topT$padj<0.05], na.rm=TRUE)), col="black", lty=4, lwd=1.9)
volcano.al <- recordPlot()

pdf(file = "../results/volcano-AL.pdf", width = 6.25, height = 3.75)
print(volcano.al)
dev.off()

```

```{r}
# ----------- MODERATE VS LOW 
topT <- as.data.frame(res.pco2.ML)

par(mar=c(2.5,2.5,1.5,1), cex.main=.9, cex.axis=.8, cex.lab=.8)
#Adjusted P values (FDR Q values)
with(topT, plot(-1*log2FoldChange, -log10(padj), pch=16, cex=0.45, main="", xlab="", ylab="", xlim=c(-7,4), ylim=c(0,15), xaxt="n", yaxt="n"))
axis(side=1, at=xtick, labels=F)
text(par("usr")[3], x=xtick, labels = xtick, pos = 1, xpd = TRUE, cex = 1)
axis(side=2, at=ytick, labels=F)
text(par("usr")[1], ytick, labels = ytick, pos = 2, xpd = TRUE, cex = 1)
title(main="(C) Moderate OA vs. Severe OA", adj=0.01, line=-1.2, font.main=1, cex.main=1.3)
title(xlab=bquote(~Log[2]~fold~change), ylab=bquote(~-log[10]~P-value), line=1.3, cex.lab=1.15)
# plot genes upregulated in moderate OA 
# with LFC >1
with(subset(topT, padj<0.05 & -1*log2FoldChange< -1), points(-1*log2FoldChange, -log10(padj), pch=16, col="#fdae61", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange< 0  & -1*log2FoldChange> -1), points(-1*log2FoldChange, -log10(padj), pch=16, col="#fddbc7", cex=0.55))

# plot genes upregulated in severe OA 
with(subset(topT, padj<0.05 & -1*log2FoldChange>1), points(-1*log2FoldChange, -log10(padj), pch=16, col="#d7191c", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange>0 & -1*log2FoldChange<1), points(-1*log2FoldChange, -log10(padj), pch=16, col="pink2", cex=0.55))


#Add lines for absolute FC>2 and P-value cut-off at FDR Q<0.05
abline(v=0, col="black", lty=3, lwd=1.0)
abline(v=-1, col="black", lty=4, lwd=2.0)
abline(v=1, col="black", lty=4, lwd=2.0)
abline(h=-log10(max(topT$padj[topT$padj<0.05], na.rm=TRUE)), col="black", lty=4, lwd=2.0)
volcano.ml <- recordPlot()

pdf(file = "../results/volcano-ML.pdf", width = 6.25, height = 3.85)
print(volcano.ml)
dev.off()
```

Save lists of annotated genes DEGs 
Here I merge the DEG lists with Uniprot/Swissprot annotations to get UniprotID and GO information. 

```{r}
diffex.AM.annot <- diffex.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  left_join(gadMor.blast.GO %>% 
              dplyr::select(gene_id, start, end, saccver, spid, gene_uni,evalue, protein_names,
                            gene_ontology_biological_process, gene_ontology_i_ds), 
            by = c("gene"="gene_id")) %>% arrange(padj)
#write.csv(diffex.AM.annot, file = "../results/DEGs.Amb-vs-Mod_annotated.csv", row.names = F, quote = T)

diffex.AL.annot <- diffex.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  left_join(gadMor.blast.GO %>%  
              dplyr::select(gene_id, start, end, saccver, spid, gene_uni,evalue, protein_names,
                            gene_ontology_biological_process, gene_ontology_i_ds), 
            by = c("gene"="gene_id")) %>% arrange(padj)
write.csv(diffex.AL.annot, file = "../results/DEGs.Amb-vs-Low_annotated.csv", row.names = F, quote = T)

diffex.ML.annot <- diffex.ML %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  left_join(gadMor.blast.GO %>% 
              dplyr::select(gene_id, start, end, saccver, spid, gene_uni,evalue, protein_names,
                            gene_ontology_biological_process, gene_ontology_i_ds), 
            by = c("gene"="gene_id")) %>% arrange(padj)  
write.csv(diffex.ML.annot, file = "../results/DEGs.Mod-vs-Low_annotated.csv", row.names = F, quote = T)
  
# Octopamine receptor, SPID Q7JQF1 = g114647
```

How many of the differentially expressed genes are transposable elements? 

```{r}
# After inspecting annotated DEGs, I determined that these strings will identify TEs in my dataset 
TE.genes <-  ("transposon|transposable|LINE|retrotransposable element|transposable element|mobile element jockey|Pol polyprotein")

print(paste("Percentage of Amb vs. Mod UPREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.AM.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.AM.annot %>% filter(log2FoldChange<0, !is.na(spid))), digits = 1)))
print(paste("Percentage of Amb vs. Mod DOWNREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.AM.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.AM.annot %>% filter(log2FoldChange>0, !is.na(spid))), digits = 1)))

print(paste("Percentage of Amb vs. Low UPREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.AL.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.AL.annot %>% filter(log2FoldChange<0, !is.na(spid))), digits = 1)))
print(paste("Percentage of Amb vs. Low DOWNREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.AL.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.AL.annot %>% filter(log2FoldChange>0, !is.na(spid))), digits = 1)))

print(paste("Percentage of Moderate vs. Low UPREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.ML.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.ML.annot %>% filter(log2FoldChange<0, !is.na(spid))), digits = 1)))
print(paste("Percentage of Moderate vs. Low DOWNREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.ML.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.ML.annot %>% filter(log2FoldChange>0, !is.na(spid))), digits = 1)))

```

Calculate average CV of TEs in upregulated / downregulated gene sets 

```{r}
TEs.AM.up <- (diffex.AM.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))$gene
TEs.AM.down <- (diffex.AM.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))$gene

TEs.AL.up <- (diffex.AL.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))$gene
TEs.AL.down <- (diffex.AL.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))$gene

TEs.ML.up <- (diffex.ML.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))$gene
TEs.ML.down <- (diffex.ML.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))$gene

genes.stats %>% 
  filter(gene %in% TEs.AM.up) %>%
  filter(Treatment %in% c("Ambient", "Moderate")) %>% 
  group_by(Treatment) %>% summarise(mean.cv = mean(cv))

genes.stats %>% 
  filter(gene %in% TEs.AL.up) %>%
  filter(Treatment %in% c("Ambient", "Low")) %>% 
  group_by(Treatment) %>% summarise(mean.cv = mean(cv))

genes.stats %>% 
  filter(gene %in% TEs.AM.down) %>%
  filter(Treatment %in% c("Ambient", "Moderate")) %>% 
  group_by(Treatment) %>% summarise(mean.cv = mean(cv))

genes.stats %>% 
  filter(gene %in% TEs.AL.down) %>%
  filter(Treatment %in% c("Ambient", "Low")) %>% 
  group_by(Treatment) %>% summarise(mean.cv = mean(cv))

```

## What % of all genes are TEs? 

```{r}
# What % are TEs across all genes? 
nrow(counts.annot.gadmor %>% filter(str_detect(protein_names, TE.genes)))/nrow(counts.annot.gadmor)

# What % are TEs across all annotated genes? 
nrow(counts.annot.gadmor %>% filter(str_detect(protein_names, TE.genes)) %>% filter(!is.na(SPID)))/nrow(counts.annot.gadmor %>% filter(!is.na(SPID)))
```

## What % of genes are TEs by treatment? 

```{r}
# What % are TEs across all genes? 
nrow(counts.annot.gadmor %>% filter(str_detect(protein_names, TE.genes)) %>% filter(!grepl("E3 SUMO-protein ligase ZBED1",protein_names)))/nrow(counts.annot.gadmor)

# What % are TEs across all annotated genes? 
nrow(counts.annot.gadmor %>% filter(str_detect(protein_names, TE.genes)) %>% filter(!is.na(SPID)))/nrow(counts.annot.gadmor %>% filter(!is.na(SPID)))
```
```{r}
counts.annot.gadmor %>% filter(str_detect(protein_names, TE.genes)) %>% filter(!is.na(SPID)) %>% filter(!grepl("E3 SUMO-protein ligase ZBED1",protein_names)) %>% arrange(protein_names) %>% select(protein_names) %>% distinct() %>% View()
```

## Messing around with the DEGs 
```{r}
res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene,log2FoldChange, pvalue, padj) %>% 
  left_join(gadMor.blast.GO, by = c("gene"="gene_id")) %>% 
  #filter(!is.na(gene_ontology_i_ds)) %>% 
  filter(padj<0.05) %>% View()


# Here I check that the genes with LFC < 0 are UPREGULATED in the lower pH treatment in all pairwise comparisons
res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>%
  filter(padj<0.05) %>% dplyr::select(gene,log2FoldChange, padj) %>% 
  left_join(gadMor.blast.GO, by = c("gene"="gene_id")) %>% 
  filter(!is.na(gene_ontology_i_ds)) %>% View()




test <- diffex.AM %>% as.data.frame() %>% filter(log2FoldChange<0) %>% rownames()

for (i in 1:length(test[1:20])) {
a <-  plotCounts(dds.DESeq.pH, gene=test[i], intgroup=c("Treatment"), returnData = TRUE) %>% 
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
  
print(ggplot(a %>% rownames_to_column("sample"),
       aes(x=Treatment, y=count, color=Treatment, label=sample)) +
  #geom_boxplot(outlier.shape = NA) +
  geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=2) +
    theme_bw() +
    ggtitle(test[i]) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none"))
}
```


## Prep for Ingenuity Pathway Analysis (IPA)
Need tab-delimited file with gene ID column, then for each comparison 2 columns: 1: LFC, 2) p-value 

```{r}
IPA.degs <- cbind(
res.pco2.AM %>% as.data.frame() %>% dplyr::select(log2FoldChange, padj) %>% rename_with( ~ paste0("Amb_vs_Mod_", .x)),
res.pco2.AL %>% as.data.frame() %>% dplyr::select(log2FoldChange, padj) %>% rename_with( ~ paste0("Amb_vs_Low_", .x)),
res.pco2.ML %>% as.data.frame() %>% dplyr::select(log2FoldChange, padj) %>% rename_with( ~ paste0("Mod_vs_Low_", .x))) %>% 
  rownames_to_column("geneID")
save(IPA.degs, file = "../results/IPA/IPA.degs")

# write_delim(IPA.degs, file = "../results/deseq2/DESeq2-results-for-IPA.tab", delim = "\t", 
#               col_names = TRUE, quote = "none")
```


### Examine some of the most consistently differentially expressed genes in OA treatments 

```{r}
# Calculate per-treatment mean, sd, and cv for each gene (using variance-stabilizing normalized counts)

# first extract the normalized counts
gene.counts.norm <- assay(vsd.pH) %>% as.matrix() %>% as.data.frame() %>% rownames_to_column(var = "gene") %>% 
  pivot_longer(-gene, values_to = "counts.vsd", names_to = "Sample") %>%
  left_join(sample.info[,c("Sample", "Treatment")]) %>% 
  mutate_at(c("gene", "Sample"), factor) %>%
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))

# calculate mean, sd, cv and annotate 
# THIS CODE NEEDS TO BE UPDATED - need to re-write deleted code that assigns genes upreg / downreg in each treatment, NOT binning mod/low

genes.stats <- gene.counts.norm %>% group_by(Treatment, gene) %>% 
  summarize(mean=mean(counts.vsd), median=median(counts.vsd), sd=sd(counts.vsd), cv=sd(counts.vsd)/mean(counts.vsd)) %>% 
  mutate(DEG=if_else(gene %in% c(rownames(diffex.AM),rownames(diffex.AL), rownames(diffex.ML)), "DEG", "Non-DEG")) %>%
  mutate(change = case_when(
  gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange < 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange < 0))) ~ "Up-regulated",
  gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange > 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange > 0))) ~ "Down-regulated")) %>% 
  
  mutate(change2 = case_when(
    
    
  ))
  
    mutate(change=replace_na(change,"Non-DEG")) %>% 
  mutate(DEG=factor(DEG, levels=c("Non-DEG", "DEG"))) %>% 
  mutate(change=factor(change, levels=c("Non-DEG", "Down-regulated", "Up-regulated"))) %>%
  ungroup() %>% 
  #left_join(P.plat.bestblast %>% select(id, id_gtf, hit_acc, hit_desc, e_value), by=c("gene"="id")) %>%
  left_join(gadMor.blast.GO %>% dplyr::select(gene_id, spid, protein_names, evalue, gene_ontology_biological_process, gene_ontology_i_ds), by=c("gene"="gene_id")) %>%
  
  # Add Log2FoldChange and padj for each pairwise treatment comparison for each gene
  left_join(cbind(res.pco2.AM %>% as.data.frame() %>% dplyr::rename("L2FC.AM"="log2FoldChange", "padj.AM"="padj") %>% 
  rownames_to_column("gene") %>% dplyr::select(gene, L2FC.AM, padj.AM),
res.pco2.AL %>% as.data.frame() %>% dplyr::rename("L2FC.AL"="log2FoldChange", "padj.AL"="padj") %>% 
  rownames_to_column("gene") %>% dplyr::select(L2FC.AL, padj.AL),
res.pco2.ML %>% as.data.frame() %>% dplyr::rename("L2FC.ML"="log2FoldChange", "padj.ML"="padj") %>% 
  rownames_to_column("gene") %>% dplyr::select(L2FC.ML, padj.ML)), by = "gene")
save(genes.stats, file = "../results/bowtie/per-gene-summary-stats")

hist(genes.stats$mean)
hist(genes.stats$cv, breaks = 50)

# What's the range of mean of the normalized counts, what is considered high? I'd say >10
ggplot(genes.stats) + geom_boxplot(aes(x=Treatment, y=mean, colour=Treatment)) + facet_wrap(~change)

# What's the range of CV of the normalized counts, what is considered low-level variation? Q1 extends to 0.02-0.04
ggplotly(ggplot(genes.stats) + geom_boxplot(aes(x=Treatment, y=cv, colour=Treatment)) + facet_wrap(~change))

# Which genes were highly UP-regulated (mean of transformed counts >10) consistently within an OA treatment? (CV<0.04)
genes.stats %>% filter(mean>10 & cv<.02 & change=="Up-regulated" & Treatment!="Ambient") %>% View()

# Which genes were highly DOWN-regulated (mean of transformed counts >10) consistently within an OA treatment? (CV<0.04)
genes.stats %>% filter(mean>10 & cv<.02 & change=="Down-regulated" & Treatment!="Ambient") %>% View()

# To inspect counts for individual genes of interest, use this code: < --- OUT OF DATE NEED TO USE RKC GENOME GENE NAMES (E.G. "g##")
# e.g. calcium-dependent secretion activator\nevm.TU.HiC_scaffold_92.184
# e.g. heat shock protein\nevm.TU.HiC_scaffold_43.188
# e.g. peroxiredoxin\nevm.TU.HiC_scaffold_16.208
# e.g. superoxide dismutase [Mn], mitochondrial-like\nevm.TU.HiC_scaffold_64.12
# e.g. Methylcytosine dioxygenase TET1\nevm.TU.HiC_scaffold_78.1

# Outlier SNPs
# e.g. nephrin-like \nevm.TU.HiC_scaffold_21.266
# e.g. histone deacetylase 4-like\nevm.TU.HiC_scaffold_91.210
# e.g. ATP-dependent RNA helicase-like, partial\nevm.TU.HiC_scaffold_61.127
# a <-  plotCounts(dds.DESeq.pH, gene="evm.TU.HiC_scaffold_92.184", intgroup=c("Treatment"), returnData = TRUE) %>% 
#   mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
#   
# ggplot(a %>% rownames_to_column("sample"),
#        aes(x=Treatment, y=count, color=Treatment, label=sample)) +
#   #geom_boxplot(outlier.shape = NA) +
#   geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) +
#     #geom_text(size=2) +
#     theme_bw() +
#     ggtitle("calcium-dependent secretion activator\nevm.TU.HiC_scaffold_92.184") +
#     theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
#   scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"))
```

What is the mean Coeff. Variation (CV) for DEGs upregulated by treatment? 
theory - those upregulated in OA have lower CV 
```{r}
# Here the rows with change="Non-DEG" and DEG="Non-DEG" are genes that are not a DEG in any comparison
# But the rows with change="Non-DEG" and DEG="DEG" are genes that are a DEG between some comparison, but not for that specific treatment. 

genes.stats %>% 
  group_by(change, Treatment) %>%
  summarize(mean.cv=mean(cv), sd.cv=sd(cv))

genes.stats %>% 
  group_by(change, Treatment) %>%
  summarize(mean.cv=mean(cv), sd.cv=sd(cv)) %>%
  mutate(pco2=Treatment) %>%
  mutate_at("pco2", function(x) {
    case_when(
      x == "Ambient" ~ 371,
      x == "Moderate" ~ 704,
      x == "Low" ~ 1424)}) %>%
  ggplot(aes(x=pco2, y=mean.cv, group=change, color=change)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=mean.cv-sd.cv, ymax=mean.cv+sd.cv), width=.2,
                 position=position_dodge(0.05))

  


genes.stats %>% filter(DEG=="DEG" & change!="Non-DEG") %>% 
ggplot(aes(x=Treatment, y=log(cv), colour=Treatment)) + geom_boxplot() + facet_wrap(~change) + theme_minimal()

```

## Significant differences in CV among treatment by DEG group? 
- Can't use anova, too many data points. 

```{r}
genes.stats %>% 
  group_by(change, Treatment) %>%
  summarize(mean.cv=mean(cv), sd.cv=sd(cv))

x <- genes.stats %>% filter(change=="Up-regulated")
summary(aov(cv^2 ~ Treatment, data=x))
TukeyHSD(aov(cv^2 ~ Treatment, data=x))

x <- genes.stats %>% filter(change=="Down-regulated")
summary(aov(cv ~ Treatment, data=x))
TukeyHSD(aov(cv ~ Treatment, data=x))

x <- genes.stats %>% filter(DEG=="Non-DEG")
summary(aov(cv ~ Treatment, data=x))
TukeyHSD(aov(cv ~ Treatment, data=x))
```

# Exploring differences in within-group variances 

## Broad differences using multivariate approaches
- Identify centroid (mean of points on PC axes), calculate euclidean distances, compare those  
- Perform a multivariate homogeneity of variance test - betadisper from vegan package

### All genes in matrix 

```{r}
(a <- plotPCA(vsd.pH, intgroup="Treatment", ntop=nrow(assay(vsd.pH))))

# calculate centroids for each 
b <- a$data %>% group_by(Treatment) %>% summarize(PC1.mean=mean(PC1), 
                                             PC1.sd=sd(PC1),
                                             PC2.mean=mean(PC2),
                                             PC2.sd=sd(PC2))
# calculate euclidean distance from treatment centroid for each sample 
c <- a$data %>% left_join(b, by = "Treatment") %>% 
  mutate(dist.cent=sqrt((PC1-PC1.mean)^2+(PC2-PC2.mean)^2)) %>%
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
  
# examine euclidean distances from centroid 
ggplot(c, aes(x=Treatment, y=dist.cent, colour=Treatment)) + 
  geom_boxplot() + geom_jitter(width = .2) +
  ggtitle("Euclidean distances to centroid of PCA by treatment\nAll genes") +
  theme_minimal() + theme(legend.position = "none", axis.title.x = element_blank()) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"))

ggplot(data=c, aes(x=PC1, y=PC2, group=Treatment, colour=Treatment)) + 
  geom_point() + 
    theme_minimal()+ stat_ellipse() +
  geom_segment(aes(x=PC1.mean, y=PC2.mean, xend=PC1, yend=PC2)) +
      geom_point(aes(x=PC1.mean, y=PC2.mean,fill=Treatment), col="black", shape=22, size=4) +
   ggtitle("PCA by Treatment, All genes\n(var-stabilizing transformed)") + theme(legend.position = c(0.08, 0.85)) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) +
    scale_fill_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) 

# Examine homogeneity of multivariate dispersions using all data 

# By examining distance to centroid on PCA
hist(c$dist.cent) #normal distribution? 
shapiro.test(c$dist.cent) #not really 

hist(log(c$dist.cent)) #log transformed = yes  
shapiro.test(log(c$dist.cent)) #log transformed = yes  
summary(aov(log(dist.cent)~Treatment, data=c)) # no diff
TukeyHSD(aov(log(dist.cent)~Treatment, data=c)) # no diff 


# By using the vegan package's `betadisper` function 
# This uses the sample-sample distance matrix, "sampleDistss", generated previously 
all(rownames(dds.pH.df) == attributes(sampleDistss)$Labels) # should === TRUE
print(disper <- betadisper(sampleDistss, dds.pH.df$Treatment)) 
anova(disper) #yes a difference, but how?
TukeyHSD(disper) #Dispersion varies between Ambient & Low pH treatments only 
permutest(disper, pairwise = TRUE) #another way of testing, same result
boxplot(disper)
plot(disper)
```

### DEGs only 

```{r}
degs.vst <- vst(
  DESeqDataSetFromMatrix(
    countData = counts.tk[,unique(c(rownames(diffex.AM),rownames(diffex.AL),rownames(diffex.ML)))] %>% t(),
                              colData = counts.tk[,"Treatment", drop=FALSE] ,
                              design = ~ Treatment), blind=TRUE)

(a <- plotPCA(degs.vst, intgroup="Treatment", ntop=nrow(assay(vsd.pH))))

# calculate centroids for each 
b <- a$data %>% group_by(Treatment) %>% summarize(PC1.mean=mean(PC1), 
                                             PC1.sd=sd(PC1),
                                             PC2.mean=mean(PC2),
                                             PC2.sd=sd(PC2))
# calculate euclidean distance from treatment centroid for each sample 
c <- a$data %>% left_join(b, by = "Treatment") %>% 
  mutate(dist.cent=sqrt((PC1-PC1.mean)^2+(PC2-PC2.mean)^2)) %>%
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
  
# examine euclidean distances from centroid 
ggplot(c, aes(x=Treatment, y=dist.cent, colour=Treatment)) + 
  geom_boxplot() + geom_jitter(width = .2) +
  ggtitle("Euclidean distances to centroid of PCA by treatment\nDEGs only") +
  theme_minimal() + theme(legend.position = "none", axis.title.x = element_blank()) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"))

ggplot(data=c, aes(x=PC1, y=PC2, group=Treatment, colour=Treatment)) + 
  geom_point() + 
    theme_minimal()+ stat_ellipse() +
  geom_segment(aes(x=PC1.mean, y=PC2.mean, xend=PC1, yend=PC2)) +
      geom_point(aes(x=PC1.mean, y=PC2.mean,fill=Treatment), col="black", shape=22, size=4) +
   ggtitle("PCA by Treatment, DEGs only\n(var-stabilizing transformed)") + theme(legend.position = c(0.08, 0.85)) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) +
    scale_fill_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) 

# Examine homogeneity of multivariate dispersions using all data 

# By examining distance to centroid on PCA
hist(c$dist.cent) #normal distribution? 
shapiro.test(c$dist.cent) #not really 

hist(c$dist.cent^.5) #log transformed = yes  
shapiro.test(c$dist.cent^.5) #log transformed = yes  
summary(aov(dist.cent^.5~Treatment, data=c)) # very different 
TukeyHSD(aov(dist.cent^.5~Treatment, data=c)) # ambient diff from other treatments - more spread 


# By using the vegan package's `betadisper` function 
# Need to generate new distance matrix after selecting only DEGs 
sampleDistss.degs <- dist(t(assay(degs.vst)))

all(rownames(dds.pH.df) == attributes(sampleDistss.degs)$Labels) # should === TRUE
print(disper <- betadisper(sampleDistss.degs, dds.pH.df$Treatment)) 
anova(disper) #yes a difference, but how?
TukeyHSD(disper) #Dispersion varies between Ambient & other two factors  
permutest(disper, pairwise = TRUE) #another way of testing, same result
boxplot(disper)
plot(disper)
```

### Excluding DEGs 

```{r}
notdegs.vst <- vst(
  DESeqDataSetFromMatrix(countData = counts.tk[,-which(names(counts.tk) %in% c("Tank", "Treatment", "Treatment_Tank", rownames(diffex.AM),rownames(diffex.AL),rownames(diffex.ML)))] %>% t(),
                              colData = counts.tk[,"Treatment", drop=FALSE] ,
                              design = ~ Treatment), blind=TRUE)

(a <- plotPCA(notdegs.vst, intgroup="Treatment", ntop=nrow(assay(vsd.pH))))

# calculate centroids for each 
b <- a$data %>% group_by(Treatment) %>% summarize(PC1.mean=mean(PC1), 
                                             PC1.sd=sd(PC1),
                                             PC2.mean=mean(PC2),
                                             PC2.sd=sd(PC2))
# calculate euclidean distance from treatment centroid for each sample 
c <- a$data %>% left_join(b, by = "Treatment") %>% 
  mutate(dist.cent=sqrt((PC1-PC1.mean)^2+(PC2-PC2.mean)^2)) %>%
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
  
# examine euclidean distances from centroid 
ggplot(c, aes(x=Treatment, y=dist.cent, colour=Treatment)) + 
  geom_boxplot() + geom_jitter(width = .2) +
  ggtitle("Euclidean distances to centroid of PCA by treatment\nAll genes except DEGs") +
  theme_minimal() + theme(legend.position = "none", axis.title.x = element_blank()) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"))

ggplot(data=c, aes(x=PC1, y=PC2, group=Treatment, colour=Treatment)) + 
  geom_point() + 
    theme_minimal()+ stat_ellipse() +
  geom_segment(aes(x=PC1.mean, y=PC2.mean, xend=PC1, yend=PC2)) +
      geom_point(aes(x=PC1.mean, y=PC2.mean,fill=Treatment), col="black", shape=22, size=4) +
     ggtitle("PCA by Treatment, excluding DEGs\n(var-stabilizing transformed)") + theme(legend.position = c(0.08, 0.85)) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) +
    scale_fill_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) 

# Examine homogeneity of multivariate dispersions using all data 

# By examining distance to centroid on PCA
hist(c$dist.cent) #normal distribution? 
shapiro.test(c$dist.cent) #not really 

hist(c$dist.cent^.5) #sqrt-transformed = yes  
shapiro.test(c$dist.cent^.5) #sqrt transformed = yes  
summary(aov(dist.cent^.5~Treatment, data=c)) # not different at all  
TukeyHSD(aov(dist.cent^.5~Treatment, data=c)) # no differences at all 

# By using the vegan package's `betadisper` function 
# Need to generate new distance matrix after selecting only DEGs 
sampleDistss.notdegs <- dist(t(assay(notdegs.vst)))

all(rownames(dds.pH.df) == attributes(sampleDistss.notdegs)$Labels) # should === TRUE
print(disper <- betadisper(sampleDistss.notdegs, dds.pH.df$Treatment)) 
anova(disper) #yes a difference, but how?
TukeyHSD(disper) #Dispersion varies between Ambient & other two factors  
permutest(disper, pairwise = TRUE) #another way of testing, same result
boxplot(disper)
plot(disper)
```


## ANOTHER Gene-wise variance 

Explore coefficient of variation by treatment, and gene groups 

```{r}
# assay(vsd.pH) #variance-stabilizing transformed count matrix
# mcols(mcols(dds.DESeq.pH))$description #Description of DESeq summary stats columns
# mcols(dds.DESeq.pH) #DESeq summary stats 

# Create dataframe in long format with gene, sample number, transformed counts, and treatment 
# use assay(vsd.pH) for variance-stabilizing transformed data 
# use assay(normTransform(dds.pH)) for log2(n+1) transformation 
d <- assay(vsd.pH) %>% as.matrix() %>% as.data.frame() %>% rownames_to_column(var = "gene") %>% 
  pivot_longer(-gene, values_to = "counts.vsd", names_to = "Sample") %>%
  left_join(sample.info[,c("Sample", "Treatment")]) %>% 
  mutate_at(c("gene", "Sample"), factor) %>%
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))

# Test for homogeneity of variances among treatments  
hist(d$counts.vsd) # first inspect transformed counts for normality - not super convincing 
shapiro.test(d$counts.vsd %>% sample(5000, replace=FALSE)) #not normal 
ks.test(x=d$counts.vsd,y='pnorm',alternative='two.sided')
leveneTest(counts.vsd ~ Treatment, data=d) #variances differ, but how? 
fligner.test(counts.vsd ~ Treatment, data=d) #variances differ, but how? 

ggplot(d %>% group_by(Treatment, gene) %>% summarise(cv=100*sd(counts.vsd)/mean(counts.vsd)), 
       aes(x=Treatment, y=log(cv), fill=Treatment)) + 
  geom_violin() + ggtitle("Normalized counts by treatment\nAll genes")# low looks lower than other treatments

cvs <- d %>% group_by(Treatment, gene) %>% summarise(cv=100*sd(counts.vsd)/mean(counts.vsd))
hist(log(test$cv))

cvs$cv %>% summary()
summary(aov(log(cv+0.001) ~ Treatment, data=cvs))
TukeyHSD(aov(log(cv+0.001) ~ Treatment, data=cvs))

cvs %>% group_by(Treatment) %>% summarise(mean=mean(cv), sd=sd(cv))
```


```{r}
# Calculate & inspect coefficient of variation for each gene within treatments 
e <- d %>% group_by(Treatment, gene) %>% 
  summarize(mean=mean(counts.vsd), sd=sd(counts.vsd), cv=sd(counts.vsd)/mean(counts.vsd))

ggplot(e, aes(x=Treatment, y=sd, colour=Treatment)) + geom_violin() +geom_boxplot() + ggtitle("Standard deviation by treatment\nAll genes")
ggplot(e, aes(x=Treatment, y=cv, colour=Treatment)) + geom_violin() +geom_boxplot() + ggtitle("CV by treatment\nAll genes")
ggplot(e, aes(x=Treatment, y=log(cv), colour=Treatment)) + geom_violin() +geom_boxplot() + ggtitle("log(CV) by treatment\nAll genes")
ggplot(e, aes(x=mean, y=cv, colour=Treatment)) + geom_point(size=1) + facet_wrap(~Treatment) + theme_minimal() + ggtitle("Coefficient of variation of counts (normalized) by treatment\nAll genes")

# inspect cv by treatment - Genes that are differentially expressed between Ambient and the two pH treatments only. 
# NOTE: this does not include genes that are diff. expressed between moderate and low pH (only a handful, though)

f <- e %>% mutate(DEG=if_else(gene %in% c(rownames(diffex.AM),rownames(diffex.AL)), "DEG", "Non-DEG")) %>%
  
  # # This column might have issues - it might erroneously assigns the term "upregulated" to both moderate and low pH treatments even if it was just upregulated in one of those (same for down regulated)
  # mutate(change = case_when(
  # gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange < 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange < 0))) ~ "Up-regulated",
  # gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange > 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange > 0))) ~ "Down-regulated")) %>% 
  # mutate(change=replace_na(change,"Non-DEG")) %>% 

  # Generate new column that indicates, RELATIVE TO THAT TREATMENT, whether the gene is upregulated, downregulated, or not a DEG 
    mutate(change_in_treatment = case_when(
    (gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange < 0))) & Treatment=="Moderate") ~ "Up-regulated",
    (gene %in% c(rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange < 0))) & Treatment=="Low") ~ "Up-regulated",
    (gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange > 0))) & Treatment=="Moderate") ~ "Down-regulated",
    (gene %in% c(rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange > 0))) & Treatment=="Low") ~ "Down-regulated",
    (gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange > 0))) & Treatment=="Ambient") ~ "Up-regulated",
    (gene %in% c(rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange > 0))) & Treatment=="Ambient") ~ "Up-regulated",
    (gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange < 0))) & Treatment=="Ambient") ~ "Down-regulated",
    (gene %in% c(rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange < 0))) & Treatment=="Ambient") ~ "Down-regulated",
    TRUE ~ "Non-DEG")) %>%
  mutate(DEG=factor(DEG, levels=c("Non-DEG", "DEG"))) %>% 
  #mutate(change=factor(change, levels=c("Non-DEG", "Down-regulated", "Up-regulated"))) %>%
  mutate(change_in_treatment=factor(change_in_treatment, levels=c("Non-DEG", "Down-regulated", "Up-regulated"))) %>%
  ungroup() %>%
  #left_join(P.plat.bestblast %>% select(id, id_gtf, hit_acc, hit_desc, e_value), by=c("gene"="id")) %>%
  left_join(gadMor.blast.GO %>% select(gene_id, spid, protein_names, evalue), by=c("gene"="gene_id")) 

# # create vector of gene names that are diff. expressed among moderate and low pH - not included in this variance comparison
# remove <- (f %>% ungroup() %>% filter(DEG=="DEG") %>% filter(change=="Non-DEG"))$gene %>% as.vector()
# removed this from the below ggplot: f %>% filter(!c(gene %in% remove))

# Plot CV by treatment for genes that are upregulated, downregulated, and not degs 
ggplot(f, aes(x=Treatment, y=log(cv), colour=Treatment)) + 
  geom_boxplot() + facet_wrap(~change_in_treatment) + theme_cleveland() + 
  theme(axis.title.x = element_blank()) + ylab("Log(Coeff. variation of transformed counts)") + 
  ggtitle("Gene-wise variance within treatments\nNon-DEGs, and DEGs that are down- and up-regulated in OA") +
  theme(axis.text.x = element_blank(), plot.title = element_text(size=10), legend.position = "bottom")

#ggplotly(
  ggplot(f, 
       aes(x=mean, y=sd, colour=Treatment)) + geom_point(size=1, aes(text=spid)) + facet_wrap(~Treatment+change_in_treatment) + 
  theme_minimal() + theme(plot.title = element_text(size=11), legend.position = "bottom") + 
  ylab("Standard deviation of transformed gene counts") + xlab("Mean of transformed counts") +
  ggtitle("Per-gene variation within treatments (SD) ~ mean of transformed")#, tooltip = "text")

# What are those highly variable genes that are upregulated in moderate and low pH? 
f %>% filter(Treatment!="Ambient") %>% filter(DEG=="DEG") %>% filter(change_in_treatment=="Up-regulated") %>% filter(sd>=1) %>% View()

# What are those genes that are upregulated in moderate and low pH at very consistent levels (very low variability)? These would be critical for function.  

# First define a "low variability" threshold 
ggplotly(ggplot(f, aes(x=Treatment, y=cv, colour=Treatment)) + 
  geom_boxplot() + facet_wrap(~change_in_treatment) + theme_cleveland() + 
  theme(axis.title.x = element_blank()) + ylab("Log(Coeff. variation of transformed counts)") + 
  ggtitle("Gene-wise variance within treatments\nNon-DEGs, and DEGs that are down- and up-regulated in OA") +
  theme(axis.text.x = element_blank(), plot.title = element_text(size=10), legend.position = "bottom"))

# Q1 = 0.02 - 0.03, use .03 
# Check out their functions
f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% filter(!is.na(spid)) %>% View()
f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% filter(!is.na(spid)) %>% View()
f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% filter(!is.na(spid)) %>% View()

# How many upregulated low CV genes in each treatment?
f %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
  group_by(change_in_treatment, Treatment) %>% tally()


# How many genes were upregulated in severe OA treatment relative to ambient? 
rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange < 0)) %>% unique() %>% length()

# How many genes were upregulated in moderate OA treatment relative to ambient? 
rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange < 0)) %>% unique() %>% length()

# How many genes were upregulated in ambient treatment relative to either OA treatment? 
c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange > 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange > 0))) %>%
  unique() %>% length()


# Here I save a list of upregulated DEGs that are also expressed at highly consistent levels (very low variability, CV <0.02).
f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
    write.csv(file = "../results/DEGs.up-low-variability-Ambient_annotated.csv", row.names = F, quote = T)

f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
    write.csv(file = "../results/DEGs.up-low-variability-Moderate_annotated.csv", row.names = F, quote = T)

f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
    write.csv(file = "../results/DEGs.up-low-variability-Severe_annotated.csv", row.names = F, quote = T)

# Save the low CV upregulated genes (i.e. those that are super critical) for other analyses
critical.amb <- f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% dplyr::select(gene) %>% unlist() %>% as.vector()
critical.mod <- f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% dplyr::select(gene) %>% unlist() %>% as.vector()
critical.low <- f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% dplyr::select(gene) %>% unlist() %>% as.vector()

```



```{r}
# 
# # Upregulated DEGs in Moderate (compared to ambient) that are also expressed very consistently 
# res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
#   mutate(critical= case_when(
#     gene %in% critical.mod ~ 1,
#     TRUE ~ 0)) %>%
#   left_join(gadMor.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>% 
#   filter(!is.na(gene_ontology_i_ds)) %>%
#   dplyr::select(gene, critical) %>% 
#   write.csv("../results/GO_MWU/DEGs-upreg-lowCV_amb-mod.csv", quote = F,row.names = F)
# 
# # Upregulated DEGs in Low (compared to ambient) that are also expressed very consistently 
# res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
#   mutate(critical= case_when(
#     gene %in% critical.low ~ 1,
#     TRUE ~ 0)) %>%
#    left_join(gadMor.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>% 
#   filter(!is.na(gene_ontology_i_ds)) %>%
#   dplyr::select(gene, critical) %>% 
#   write.csv("../results/GO_MWU/DEGs-upreg-lowCV_amb-low.csv", quote = F,row.names = F)
# 
# # In comparison, what functions are upregulated in consistent levels in ambient (compared to low)?
# # Note: the filter change=="Down-regulated" refers to the fact that the gene is downregulated in OA treatment, and thus upregualted in ambient treatment 
# 
# res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
#   mutate(critical= case_when(
#     gene %in% critical.amb ~ 1,
#     TRUE ~ 0)) %>% 
#    left_join(gadMor.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>% 
#   filter(!is.na(gene_ontology_i_ds)) %>%
#   dplyr::select(gene, critical) %>% 
#   write.csv("../results/GO_MWU/DEGs-upreg-lowCV_amb.csv", quote = F,row.names = F)
# 
# 
# critical.mod[! critical.mod %in% TE.genes] %>% length()
# critical.mod %>% length()
#     
# # What if we remove Transposable Elements from the list? 
# 
# res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
#   mutate(critical= case_when(
#     gene %in% critical.mod[! critical.mod %in% TE.genes] ~ 1,
#     TRUE ~ 0)) %>%
#   left_join(gadMor.blast.GO, by = c("gene"="gene_id")) %>% 
#   filter(!is.na(gene_ontology_i_ds)) %>%
# 
#   dplyr::select(gene, critical) %>% 
#   write.csv("../results/GO_MWU/DEGs-upreg-lowCV_amb-mod.csv", quote = F,row.names = F)
# 
# # Upregulated DEGs in Low (compared to ambient) that are also expressed very consistently 
# res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
#   mutate(critical= case_when(
#     gene %in% critical.low ~ 1,
#     TRUE ~ 0)) %>%
#    left_join(gadMor.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>% 
#   filter(!is.na(gene_ontology_i_ds)) %>%
#   dplyr::select(gene, critical) %>% 
#   write.csv("../results/GO_MWU/DEGs-upreg-lowCV_amb-low.csv", quote = F,row.names = F)

```

DAVID enrichment with critical genes 

```{r}
# Upregulated DEGs in Ambient (compared to either OA treatment) that are also expressed very consistently 
f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
   filter(!is.na(spid)) %>% select(spid) %>% unlist() %>% as.vector() %>% write_clip()

# Upregulated DEGs in Moderate OA (compared to ambient) that are also expressed very consistently 
f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>%
   filter(!is.na(spid)) %>% select(spid) %>% unlist() %>% as.vector() %>% write_clip()

# Upregulated DEGs in Severe OA (compared to ambient) that are also expressed very consistently 
f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>%
   filter(!is.na(spid)) %>% select(spid) %>% unlist() %>% as.vector() %>% write_clip()

# Background - all genes in data set 
res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  left_join(gadMor.blast.GO %>% dplyr::select(gene_id, spid, protein_names), by = c("gene"="gene_id")) %>% 
  filter(!is.na(spid)) %>% select(spid) %>% unlist() %>% as.vector() %>% write_clip()

# Read in results, create one table with all low CV genes 
enriched.bp.lowcv <- vector(mode = "list", length = 3)
names(enriched.bp.lowcv) <- c("ambient", "moderate", "low")

for (i in 1:3) {
  enriched.bp.lowcv[[i]] <- read_delim(file = paste("../results/GO_MWU/DAVID_lowCV-upreg-", names(enriched.bp.lowcv)[i], ".txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05)
}

enriched.bp.lowcv <- bind_rows(enriched.bp.lowcv, .id = "treatment") %>% 
  separate(Term, into = c("GO", "process"), sep = "~") %>%
  clean_names() %>% 
  select(treatment, go, process, count, percent, p_value, fold_enrichment, fdr) 

enriched.bp.lowcv %>% write_clip() #copy to clipboard to paste into google spreadsheet 

enriched.bp.lowcv %>% filter(fdr<0.1) %>% group_by(treatment) %>% tally()


# what % of low CV genes are TEs?
# Note, the vector "TE.genes" is defined elsewhere 

# In Ambient
(f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>%
  filter(str_detect(protein_names, TE.genes)) %>% nrow()) /
  (f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
  filter(!is.na(spid)) %>% nrow())

# In Moderate OA
(f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>%
  filter(str_detect(protein_names, TE.genes)) %>% nrow()) /
  (f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
  filter(!is.na(spid)) %>% nrow())

# In Severe OA
(f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>%
  filter(str_detect(protein_names, TE.genes)) %>% nrow()) /
  (f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
  filter(!is.na(spid)) %>% nrow())

```

How many upregulated genes with CV < 0.03 in each treatment? What % of upregulated genes are low-CV? 

```{r}
critical.amb %>% length()
critical.mod %>% length()
critical.low %>% length()

critical.amb %>% length() / f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% nrow()
critical.mod %>% length() / f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% nrow()
critical.low %>% length() / f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% nrow()
```

What's the average +/- SD CV? 

```{r}
f %>% group_by(Treatment, change_in_treatment) %>%
  summarise(mean.cv=round(100*mean(cv), digits = 1), sd.cv=round(100*sd(cv), digits = 1))

f %>% filter(change_in_treatment=="Up-regulated")

```

```{r}
f %>% filter(!is.na(spid)) %>% filter(DEG == "DEG") %>% View()
```


<!-- How many don't have GO terms? -->
<!-- ```{r} -->
<!-- res.pco2.AL %>% as.data.frame() %>% dplyr::select(log2FoldChange, padj) %>% -->
<!--   rownames_to_column("gene") %>%  -->
<!--   filter(gene %in% critical.low) %>% -->
<!--   left_join(gadMor.blast.GO %>%  -->
<!--               dplyr::select(gene_id, gene_ontology_i_ds, protein_names, spid, -->
<!--                             gene_names, organism, gene_ontology_biological_process),  -->
<!--             by = c("gene"="gene_id")) %>% filter(!is.na(protein_names)) %>% -->
<!--   mutate(my_category=case_when( -->
<!--     str_detect(protein_names,  -->
<!--     "retrotransposable|LINE|transposable element|transposon|mobile element jockey|Pol polyprotein") ~ "Transposable Elements", -->
<!--     str_detect(protein_names,  -->
<!--                "Exportin-5|methyltransferase|demethylase|Histone deacetylase") ~ "Epigenetic Factor", -->
<!--     str_detect(protein_names,  -->
<!--                     "Zinc finger protein|LIM domain-binding protein 2") ~ "Transcription Regulation",  -->
<!--     str_detect(protein_names,  -->
<!--                "Toll-like receptor Tollo|Serine incorporator 3") ~ "Immune Function", -->
<!--     str_detect(protein_names,  -->
<!--                     "Neurocalcin homolog|Puratrophin-1|Gamma-aminobutyric acid receptor|Orcokinin|Amyloid-beta|Teneurin|Multiple PDZ domain protein|Carboxypeptidase|Innexin|Cytoplasmic protein NCK2|cytokine signaling|Wnt|cAMP|Jerky|Adhesion G protein-coupled receptor") ~ "Signaling",  -->
<!--     str_detect(protein_names, "Cold shock|Bloom syndrome protein") ~ "Stress response", -->
<!--     TRUE ~ "Other")) %>%  -->
<!--   filter(my_category=="Transposable Elements") %>% View() -->
<!--   filter(is.na(gene_ontology_i_ds)) -->
<!-- ``` -->


<!-- # Plot genes from interesting categories  -->

<!-- ```{r} -->
<!-- x <- res.pco2.AL %>% as.data.frame() %>% dplyr::select(log2FoldChange, padj) %>% -->
<!--   rownames_to_column("gene") %>%  -->
<!--   filter(gene %in% critical.low) %>% -->
<!--   left_join(gadMor.blast.GO %>%  -->
<!--               dplyr::select(gene_id, gene_ontology_i_ds, protein_names, spid, -->
<!--                             gene_names, organism, gene_ontology_biological_process),  -->
<!--             by = c("gene"="gene_id")) %>% filter(!is.na(protein_names)) %>% -->
<!--   mutate(my_category=case_when( -->
<!--     str_detect(protein_names,  -->
<!--     "retrotransposable|LINE|transposable element|transposon|mobile element jockey|Pol polyprotein") ~ "Transposable Elements", -->
<!--     str_detect(protein_names,  -->
<!--                "Exportin-5|methyltransferase|demethylase|Histone deacetylase") ~ "Epigenetic Factor", -->
<!--     str_detect(protein_names,  -->
<!--                     "Zinc finger protein|LIM domain-binding protein 2") ~ "Transcription Regulation",  -->
<!--     str_detect(protein_names,  -->
<!--                "Toll-like receptor Tollo|Serine incorporator 3") ~ "Immune Function", -->
<!--     str_detect(protein_names,  -->
<!--                     "Neurocalcin homolog|Puratrophin-1|Gamma-aminobutyric acid receptor|Orcokinin|Amyloid-beta|Teneurin|Multiple PDZ domain protein|Carboxypeptidase|Innexin|Cytoplasmic protein NCK2|cytokine signaling|Wnt|cAMP|Jerky|Adhesion G protein-coupled receptor") ~ "Signaling",  -->
<!--     str_detect(protein_names, "Cold shock|Bloom syndrome protein") ~ "Stress response", -->
<!--     TRUE ~ "Other"))  -->

<!-- y <- x %>% filter(my_category=="Transposable Elements") %>% select(gene) %>% unlist() %>% as.vector() -->

<!-- diff_plots = list() -->
<!-- for (i in 1:length(y)) { -->
<!-- a <-  plotCounts(dds.DESeq.pH, gene=y[i], intgroup=c("Treatment"), returnData = TRUE) %>%  -->
<!--   mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low"))) -->

<!-- b <- ggplot(a %>% rownames_to_column("sample"), -->
<!--        aes(x=Treatment, y=count, color=Treatment, label=sample)) + -->
<!--   #geom_boxplot(outlier.shape = NA) + -->
<!--   geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) + -->
<!--     #geom_text(size=2) + -->
<!--     theme_bw() + -->
<!--     ggtitle(y[i]) + -->
<!--     theme(axis.title.x = element_blank(),  -->
<!--           plot.title = element_text(hjust = 0.5, size=10),  -->
<!--           legend.position = "none") -->
<!-- diff_plots[[i]] <- b -->
<!-- } -->
<!-- diff_plots -->
<!-- ``` -->


<!-- # Here is the average+/-SD for each gene  -->
<!-- # THIS ISN'T INTERESTING - I SHOULD PLOT TES ~ EPIGENETIC STUFF  -->

<!-- ```{r} -->
<!-- # plot mean counts for ambient, moderate, and low pH  -->
<!-- genes.stats %>% pivot_longer(cols = contains("L2FC"), names_to = "contrast", values_to = "L2FC") %>% -->
<!--   filter(gene %in% c(critical.low)) %>%  -->
<!--   #filter(abs(L2FC)>1) %>%  -->
<!--   dplyr::select(-contrast, -L2FC, -contains("padj")) %>% unique() %>%  -->
<!--     mutate(pco2=Treatment) %>% -->
<!--   mutate_at("pco2", function(x) { -->
<!--     case_when( -->
<!--       x == "Ambient" ~ 371, -->
<!--       x == "Moderate" ~ 704, -->
<!--       x == "Low" ~ 1424)}) %>% -->
<!--   filter(str_detect(protein_names,  -->
<!--     "retrotransposable|LINE|transposable element|transposon|mobile element jockey|Pol polyprotein")) %>% -->
<!--   #mutate(treat.num=as.numeric(Treatment)) %>%  #need to change treatment factor to numeric for plotting  -->
<!--   ggplot(aes(x=pco2, y=mean, group=gene)) + geom_point() + geom_line(size=.01) + -->
<!--   theme_minimal() + #facet_wrap(~change) + -->
<!--   #scale_x_continuous(breaks=c(1,2,3), labels=c("Ambient", "Moderate", "Low")) + -->
<!--   theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) + -->
<!--   xlab("Per treatment mean of normalized counts") + ggtitle("Abundances of transposable elements by treatment & relative expression") -->
<!-- ``` -->
<!-- ```{r} -->
<!-- assay(vsd.pH) %>% as.data.frame() %>% rownames_to_column("gene") %>% -->
<!--   left_join(gadMor.blast.GO %>%  -->
<!--               dplyr::select(gene_id, gene_ontology_i_ds, protein_names, spid, -->
<!--                             gene_names, organism, gene_ontology_biological_process),  -->
<!--             by = c("gene"="gene_id")) %>% filter(!is.na(protein_names)) %>% -->
<!--   mutate(my_category=case_when( -->
<!--     str_detect(protein_names,  -->
<!--     "retrotransposable|LINE|transposable element|transposon|mobile element jockey|Pol polyprotein") ~ "Transposable Elements", -->
<!--     str_detect(protein_names,  -->
<!--                "Exportin-5|methyltransferase|demethylase|Histone deacetylase") ~ "Epigenetic Factor", TRUE ~ "Other")) %>% -->
<!--   #filter(gene %in% critical.low) %>% -->
<!--   filter(my_category!="Other") %>% -->
<!--   pivot_longer(cols = starts_with("Tank"), names_to = "Sample", values_to = "counts.vsd") %>% -->
<!--     left_join(sample.info %>% dplyr::select(Sample, Treatment), by="Sample") %>% -->
<!--   ggplot(aes(x=my_category, y=counts.vsd)) + geom_violin() + facet_wrap(~Treatment) -->
<!-- ``` -->

## For all DEGs - Reaction Norm Plot

```{r}
# plot mean counts for ambient, moderate, and low pH 

genes.stats %>% pivot_longer(cols = contains("L2FC"), names_to = "contrast", values_to = "L2FC") %>%
  filter(change!="Non-DEG") %>% 
  filter(abs(L2FC)>1) %>% 
  dplyr::select(-contrast, -L2FC, -contains("padj")) %>% unique() %>% 
    mutate(pco2=Treatment) %>%
  mutate_at("pco2", function(x) {
    case_when(
      x == "Ambient" ~ 371,
      x == "Moderate" ~ 704,
      x == "Low" ~ 1424)}) %>%
  #mutate(treat.num=as.numeric(Treatment)) %>%  #need to change treatment factor to numeric for plotting 
  ggplot(aes(x=pco2, y=mean, group=gene)) + geom_line(size=.01) +
  theme_minimal() + facet_wrap(~change) +
  #scale_x_continuous(breaks=c(1,2,3), labels=c("Ambient", "Moderate", "Low")) +
  theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  xlab("Per treatment mean of normalized counts") + ggtitle("Abundances of DEGs by treatment & relative expression")
```


```{r}
# # NOTE: When I do NOT filter out low-frequency genes prior to the DESeq2 analysis using the mean=10 threshold, some cv=0 for some genes/treatments
# # So I will remove the CV=0 genes before testing for differences 
# f %>% filter(cv!=0)
# hist(log((f %>% filter(cv!=0))$cv))
# summary(aov(log(cv) ~ DEG*Treatment, data=f %>% filter(cv!=0)))
# TukeyHSD(aov(log2(cv) ~ DEG+Treatment+DEG:Treatment, data=f %>% filter(cv!=0)))
# 
# # # Here you can plot the normalized gene counts for those genes. 
# # cv.zero <- (f %>% filter(cv==0))$gene %>% as.vector()
# # diff_plots = list()
# # for (i in 1:length(cv.zero)) {
# # a <-  plotCounts(dds.DESeq.pH, gene=cv.zero[i], intgroup=c("Treatment"), returnData = TRUE)
# # b <- ggplot(a %>% filter(Treatment==c("Ambient", "Moderate", "Low")) 
# #                          %>% rownames_to_column("sample"),
# #        aes(x=Treatment, y=count, color=Treatment, label=sample)) +
# #   geom_point(position=position_jitter(w = 0.1,h = 0)) +
# #     geom_text() +
# #     theme_bw() +
# #     ggtitle(cv.zero[i]) +
# #     theme(plot.title = element_text(hjust = 0.5))
# # diff_plots[[i]] <- b
# # }
# # diff_plots
# 
# # Octopamine receptor, SPID Q7JQF1 = g114647
# 
# # evm.TU.HiC_scaffold_78.1 - Methylcytosine dioxygenase TET1 (EC 1.14.11.n2) (CXXC-type zinc finger protein 6) 
# a <-  plotCounts(dds.DESeq.pH, gene="g114647", intgroup=c("Treatment"), returnData = TRUE) %>% 
#   mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
#   
# ggplot(a %>% rownames_to_column("sample"),
#        aes(x=Treatment, y=count, color=Treatment, label=sample)) +
#   #geom_boxplot(outlier.shape = NA) +
#   geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) +
#     #geom_text(size=2) +
#     theme_bw() +
#     #ggtitle("Octopamine") +
#     theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none")

```


```{r}
# plot mean counts for ambient, moderate, and low pH 
genes.stats %>% pivot_longer(cols = contains("L2FC"), names_to = "contrast", values_to = "L2FC") %>%
  #filter(gene %in% c(critical.low)) %>% 
  #filter(abs(L2FC)>1) %>% 
  filter(change == "Up-regulated") %>% 
  dplyr::select(-contrast, -L2FC, -contains("padj")) %>% unique() %>% 
    mutate(pco2=Treatment) %>%
  mutate_at("pco2", function(x) {
    case_when(
      x == "Ambient" ~ 371,
      x == "Moderate" ~ 704,
      x == "Low" ~ 1424)}) %>%
  filter(str_detect(protein_names, 
    "retrotransposable|LINE|transposable element|transposon|mobile element jockey|Pol polyprotein")) %>%
  #mutate(treat.num=as.numeric(Treatment)) %>%  #need to change treatment factor to numeric for plotting 
  ggplot(aes(x=Treatment, y=cv)) + geom_boxplot() + 
  theme_minimal() + #facet_wrap(~change) +
  #scale_x_continuous(breaks=c(1,2,3), labels=c("Ambient", "Moderate", "Low")) +
  theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  xlab("Per treatment mean of normalized counts") + ggtitle("Abundances of transposable elements by treatment & relative expression")

```

