---
title: "Results and Figures"
author: "Laura Spencer"
date: "2023-01-04"
output: html_document
---

```{r}
require(tidyverse)
require(eulerr)
require(wordcloud2)
require(pheatmap)
require(DESeq2)
require(colorspace)
require(RColorBrewer)
require(patchwork)
require(grid)
require(cowplot)
`%!in%` = Negate(`%in%`)
require(ggplotify)
require(googlesheets4)
require(janitor)
require(lme4)
```

Summary statistics from Emily - growth, condition, mortality 

```{r}
phen_summ <- read_sheet("https://docs.google.com/spreadsheets/d/1RVk14dedsX_OtJhFAdd931b_h6L9cDV8ZJP9rW69F-A/edit?usp=sharing") %>% clean_names() %>% filter(days_post_hatch<36)

phen_summ %>% summarize(mean=mean(n), se=std.error(n))
# Read in overall survival data 
surv <- read_csv("../data/survival_for laura.csv") %>% clean_names() %>% 
  select(tank, surv_percent) %>% distinct() %>% 
    left_join(phen_summ %>% select(tank, temperature, co2) %>% distinct(), by="tank")

install.packages("plotrix")
require(plotrix)

# Overall survivla rate 
surv %>% 
  group_by(temperature, co2) %>% 
  summarize(mean.surv=mean(surv_percent), se=std.error(surv_percent), sd=sd(surv_percent)) #%>% 

# Instantaneous daily mortality rate by treatment. Here we average each tank mortality rate to get treatment averages. 
phen_summ %>% 
  filter(measurement=="MORT.RATE") %>%
  group_by(temperature, co2) %>% 
  summarize(mean.mean=mean(mean), sd=sd(mean), se=std.error(mean)) #%>% 
  # ungroup() %>% 
  # t() %>% write_clip()
  
# Length-based growth rates, averages of each tank-level measurements from 0-11 and 11-28
# Delta length over time 
phen_summ %>% 
  filter(measurement=="LEN") %>% select(-n, -se) %>%  
  pivot_wider(names_from = days_post_hatch, values_from = mean) %>% 
  mutate(gl_0_11=(`11`-`0`)/(11-0), gl_11_28=(`28`-`11`)/(28-11),
         gl_warm=(`11`-`0`)/(11-0), gl_control=(`21`-`0`)/(21-0), gl_cold=(`28`-`0`)/(28-0)) %>% 
  group_by(temperature, co2) %>% 
  summarize(mean.p1=mean(gl_0_11, na.rm=T), se.p1=std.error(gl_0_11, na.rm=T),
            mean.p2=mean(gl_11_28, na.rm=T), se.p2=std.error(gl_11_28, na.rm=T),
            mean.warm=mean(gl_warm, na.rm=T), se.warm=std.error(gl_warm, na.rm=T),
            mean.control=mean(gl_control, na.rm=T), se.control=std.error(gl_control, na.rm=T),
            mean.cold=mean(gl_cold, na.rm=T), se.cold=std.error(gl_cold, na.rm=T)) %>% 
  select(temperature:co2, mean.warm:se.cold) #%>% 
#  ungroup() %>%  t() %>% write_clip()

# Dry weight-based growth rates, averages of each tank-level measurements from 0-11 and 11-28
# difference in the natural log of DW across a specific time period and multiplied by 100
phen_summ %>% 
  filter(measurement=="DW") %>% select(-n, -se) %>% 
  pivot_wider(names_from = days_post_hatch, values_from = mean) %>% 
  mutate(gw_0_11=100*((log(`11`)-log(`0`))/(11-0)), 
         gw_11_28=100*((log(`28`)-log(`11`))/(28-11)),
         gw_warm=100*((log(`11`)-log(`0`))/(11-0)), 
         gw_control=100*((log(`21`)-log(`0`))/(21-0)),
         gw_cold=100*((log(`28`)-log(`0`))/(28-0))) %>% 
  group_by(temperature, co2) %>%  arrange(temperature, co2) %>% #select(-days_post_hatch, -n) %>% distinct() 
  summarize(mean.p1=mean(gw_0_11, na.rm=T), se.p1=std.error(gw_0_11, na.rm=T),
            mean.p2=mean(gw_11_28, na.rm=T), se.p2=std.error(gw_11_28, na.rm=T),
            mean.warm=mean(gw_warm, na.rm=T), se.warm=std.error(gw_warm, na.rm=T),
            mean.control=mean(gw_control, na.rm=T), se.control=std.error(gw_control, na.rm=T),
            mean.cold=mean(gw_cold, na.rm=T), se.cold=std.error(gw_cold, na.rm=T))  %>% 
  select(temperature:co2, mean.warm:se.cold) #%>% 
#  ungroup() %>%  t() %>% write_clip()

# Myotome height based condition factor 
phen_summ %>% #filter(measurement=="dev.MH") %>% arrange(days_post_hatch, temperature, co2)
  filter(measurement=="dev.MH", temperature==10, days_post_hatch==11) %>% # 10C treatment 
#  filter(measurement=="dev.MH", temperature==6, days_post_hatch==21) %>% # 6C treatment 
#  filter(measurement=="dev.MH", temperature==3, days_post_hatch==28) %>% # 3C treatment 
  group_by(temperature, co2) %>% 
  summarize(mean.mean=mean(mean, na.rm=T), se=std.error(mean, na.rm=T)) #%>% 
#  ungroup() %>%  t() %>% write_clip()

# Dry weight based condition factor 
phen_summ %>% 
  filter(measurement=="dev.DW", temperature==10, days_post_hatch==11) %>% # 10C treatment 
#  filter(measurement=="dev.DW", temperature==6, days_post_hatch==21) %>% # 6C treatment 
#  filter(measurement=="dev.DW", temperature==3, days_post_hatch==28) %>% # 3C treatment 
  group_by(temperature, co2) %>% 
  summarize(mean.mean=mean(mean, na.rm=T), se=std.error(mean, na.rm=T)) #%>% 
#  ungroup() %>%  t() %>% write_clip()
```

```{r}
phen_summ %>% filter(measurement=="LEN") %>% 
  ggplot(aes(x=days_post_hatch, y=mean, color=interaction(temperature,co2))) + geom_jitter(width=0.25) + geom_line() + facet_wrap(~interaction(temperature,co2))

summary(lm(mean~temperature:co2 + days_post_hatch, data=phen_summ %>% filter(measurement=="LEN")))

GL.models <- phen_summ %>% filter(measurement=="LEN") %>%
  group_by(temperature,co2) %>%
  nest() %>%
  mutate(model = map(data, ~ lm(mean ~ days_post_hatch, data = .x)))
names(GL.models$model) <- paste0(GL.models$temperature, "_", GL.models$co2)

GL.models$model

(summary(GL.models$model$`3_amb`)[[4]] %>% as.data.frame())[2,1:2]
(summary(GL.models$model$`3_high`)[[4]] %>% as.data.frame())[2,1:2]
(summary(GL.models$model$`6_amb`)[[4]] %>% as.data.frame())[2,1:2]
(summary(GL.models$model$`6_high`)[[4]] %>% as.data.frame())[2,1:2]
(summary(GL.models$model$`10_amb`)[[4]] %>% as.data.frame())[2,1:2]
(summary(GL.models$model$`10_high`)[[4]] %>% as.data.frame())[2,1:2]

#PLot mortality over time?
```

Stats, using tank as replicate 

```{r}
# CAN'T DO STATS ON MORTALITY RATE, SAME RATE FOR EACH TANK 
require(emmeans)
require(car)

mort.df <- phen_summ %>% filter(measurement=="MORT.RATE") %>% 
  select(-days_post_hatch, -n, -se, -measurement) %>% distinct() %>% 
  mutate(temperature=as.factor(temperature), co2=as.factor(co2))

mort.df  %>% mutate(treatment=paste0(temperature,"_",co2),
                 point=case_when(tank==19~"red", TRUE~"black")) %>% 
  ggplot(aes(x=treatment, y=mean, label=tank)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color=point), size=2.5, width=.25) + 
  scale_color_manual(values=c("black","red")) + 
  ggtitle("Daily mortality rates")

mort.mod1 <- glm(mean ~ temperature*co2, family = gaussian, data=mort.df, contrasts = list(temperature="contr.sum", co2="contr.sum"))
summary(mort.mod1)
Anova(mort.mod1, test="F", type = "III")

# LENGTH-BASED GROWTH RATE 
gl <- phen_summ %>% 
  filter(measurement=="LEN") %>% select(-n, -se) %>%  
  pivot_wider(names_from = days_post_hatch, values_from = mean) %>% 
  mutate(gl_warm=(`11`-`0`)/(11-0), gl_control=(`21`-`0`)/(21-0), gl_cold=(`28`-`0`)/(28-0))

# make new dataframe with growth rates calculated from 0 DPH to sampling date for each temperature treatment 
gl.df <- bind_rows(
  gl %>% select(temperature:tank, gl_cold) %>% filter(temperature==3) %>% filter(!is.na(gl_cold)) %>% rename("gl_cold"="gl"),
  gl %>% select(temperature:tank, gl_control) %>% filter(temperature==6) %>% rename("gl_control"="gl"),
  gl %>% select(temperature:tank, gl_warm) %>% filter(temperature==10) %>% rename("gl_warm"="gl")) %>% 
  mutate(temperature=as.factor(temperature), co2=as.factor(co2))
gl.df  %>% mutate(treatment=paste0(temperature,"_",co2),
                 point=case_when(tank==19~"red", TRUE~"black")) %>% 
  ggplot(aes(x=treatment, y=gl, label=tank)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color=point), size=2.5, width=.25) + 
  scale_color_manual(values=c("black","red")) + 
  ggtitle("Growth rates (length)")

gl.mod1 <- glm(gl ~ temperature*co2, family = gaussian, data=gl.df, contrasts = list(temperature="contr.sum", co2="contr.sum")) 
Anova(gl.mod1, test="F", type = "III")
gl.mod2 <- glm(gl ~ temperature, family = gaussian, data=gl.df, contrasts = list(temperature="contr.sum", co2="contr.sum")) 
gl.mod3 <- glm(gl ~ temperature + co2, family = gaussian, data=gl.df, contrasts = list(temperature="contr.sum", co2="contr.sum")) 
AIC(gl.mod1, gl.mod2, gl.mod3)
Anova(gl.mod3, type = "III")
gl_emmeans <- emmeans(gl.mod1, ~ temperature * co2)  # Get marginal means for interaction
pairwise_gl <- pairs(gl_emmeans)  # Pairwise comparisons
summary(pairwise_gl) %>% as.data.frame() %>% filter(p.value<0.05)

# DRY WEIGHT-BASED GROWTH RATE 
gm <- phen_summ %>% 
  filter(measurement=="DW") %>% select(-n, -se) %>%  
  pivot_wider(names_from = days_post_hatch, values_from = mean) %>% 
  mutate(gm_warm=(`11`-`0`)/(11-0), gm_control=(`21`-`0`)/(21-0), gm_cold=(`28`-`0`)/(28-0))

# make new dataframe with growth rates calculated from 0 DPH to sampling date for each temperature treatment 
gm.df <- bind_rows(
  gm %>% select(temperature:tank, gm_cold) %>% filter(temperature==3) %>% filter(!is.na(gm_cold)) %>% rename("gm_cold"="gm"),
  gm %>% select(temperature:tank, gm_control) %>% filter(temperature==6) %>% rename("gm_control"="gm"),
  gm %>% select(temperature:tank, gm_warm) %>% filter(temperature==10) %>% rename("gm_warm"="gm")) %>% 
  mutate(temperature=as.factor(temperature), co2=as.factor(co2))
gm.df %>% mutate(treatment=paste0(temperature,"_",co2),
                 point=case_when(tank==19~"red", TRUE~"black")) %>% 
  ggplot(aes(x=treatment, y=gm, label=tank)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color=point), size=2.5, width=.25) + 
  scale_color_manual(values=c("black","red")) + 
  ggtitle("Growth rates (mass)")

gm.mod1 <- glm(gm ~ temperature*co2, family = gaussian, data=gm.df, contrasts = list(temperature="contr.sum", co2="contr.sum")) #Analysis of deviance table with F statistic - temperature significant effect 
summary(gm.mod1)
Anova(gm.mod1, test="F", type = "III")
gm_emmeans <- emmeans(gm.mod1, ~ temperature)  # Get marginal means for interaction
pairwise_gm <- pairs(gm_emmeans)  # Pairwise comparisons
summary(pairwise_gm) %>% as.data.frame() #%>% filter(p.value<0.05)

# MYOTOME HEIGHT CONDITION FACTOR 
km <- bind_rows(
  phen_summ %>% filter(measurement=="dev.MH", temperature==10, days_post_hatch==11), # 10C treatment 
  phen_summ %>% filter(measurement=="dev.MH", temperature==6, days_post_hatch==21), # 6C treatment 
  phen_summ %>% filter(measurement=="dev.MH", temperature==3, days_post_hatch==28)) %>% 
  mutate(temperature=as.factor(temperature), co2=as.factor(co2))
km %>% mutate(treatment=paste0(temperature,"_",co2),
                 point=case_when(tank==19~"red", TRUE~"black")) %>% 
  ggplot(aes(x=treatment, y=mean, label=tank)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color=point), size=2.5, width=.25) + 
  scale_color_manual(values=c("black","red")) + 
  ggtitle("Condition index (morphometric)")

km.mod1 <- glm(mean ~ temperature*co2, family = gaussian, data=km, contrasts = list(temperature="contr.sum", co2="contr.sum")) #Analysis of deviance table with F statistic - temperature significant effect 
summary(km.mod1)
Anova(km.mod1, test="F", type = "III")
km_emmeans <- emmeans(km.mod1, ~ temperature)  # Get marginal means 
pairwise_km <- pairs(km_emmeans)  # Pairwise comparisons
summary(pairwise_km) %>% as.data.frame() %>% filter(p.value<0.05)

# WEIGHT CONDITION FACTOR 
kw <- bind_rows(
  phen_summ %>% filter(measurement=="dev.DW", temperature==10, days_post_hatch==11), # 10C treatment 
  phen_summ %>% filter(measurement=="dev.DW", temperature==6, days_post_hatch==21), # 6C treatment 
  phen_summ %>% filter(measurement=="dev.DW", temperature==3, days_post_hatch==28)) %>% 
  mutate(temperature=as.factor(temperature), co2=as.factor(co2))
kw %>% mutate(treatment=paste0(temperature,"_",co2),
                 point=case_when(tank==19~"red", TRUE~"black")) %>% 
  ggplot(aes(x=treatment, y=mean, label=tank)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color=point), size=2.5, width=.25) + 
  scale_color_manual(values=c("black","red")) + 
  ggtitle("Condition index (mass)")

kw.mod1 <- glm(mean ~ temperature*co2, family = gaussian, data=kw, contrasts = list(temperature="contr.sum", co2="contr.sum")) 
Anova(kw.mod1, test="F", type = "III")
kw_emmeans <- emmeans(kw.mod1, ~ temperature:co2)  # Get marginal means 
pairwise_kw <- pairs(kw_emmeans)  # Pairwise comparisons
summary(pairwise_kw) %>% as.data.frame() %>% filter(p.value<0.05)

# ALL ANOVAS IN ONE SPOT 

# MORTALITY Rate (daily)
Anova(mort.mod1, test="F", type = "III")
summary(pairs(emmeans(mort.mod1, ~ temperature * co2)))  %>% 
  as.data.frame() %>% filter(p.value<0.05)

# GROWTH LENGTH 
Anova(gl.mod1, test="F", type = "III") #MORTALITY 
summary(pairs(emmeans(gl.mod1, ~ temperature * co2)))  %>% 
  as.data.frame() %>% filter(p.value<0.05)

# GROWTH WEIGHT
Anova(gm.mod1, test="F", type = "III") #MORTALITY 
summary(pairs(emmeans(gm.mod1, ~ temperature * co2)))  %>% 
  as.data.frame() %>% filter(p.value<0.05)

# KMH
Anova(km.mod1, test="F", type = "III") #KMH
summary(pairs(emmeans(km.mod1, ~ temperature * co2)))  %>% 
  as.data.frame() %>% filter(p.value<0.05)

# KDW
Anova(kw.mod1, test="F", type = "III") #KDW
summary(pairs(emmeans(kw.mod1, ~ temperature * co2)))  %>% 
  as.data.frame() %>% filter(p.value<0.05)
```

## Stats reported in paper! Table 3 asterisks, Figure 3 effect sizes. 
```{r}
library(effsize)

# Do pairwise t-tests and cohen d effect sizes 
surv <- surv %>% mutate(treatment=paste0(temperature, ".", co2))
pairwise.t.test(surv$surv_percent, surv$treatment, p.adjust.method = "bonferroni")
surv.test <- surv %>% mutate(successes=round(600*(surv_percent/100)), totals=600) %>% 
  select(treatment:totals)
pairwise.t.test(surv.test$successes, surv.test$treatment, p.adjust.method = "bonferroni")
pairwise.wilcox.test(surv.test$successes, surv.test$treatment) #nonparametric

surv.test.sum <- surv.test %>% group_by(treatment) %>% summarize(surv_sum=sum(successes), total_sum=sum(totals))
pairwise.prop.test(surv.test.sum$surv_sum, surv.test.sum$total_sum, p.adjust.method = "bonferroni")

mort.df <- mort.df %>% mutate(treatment=paste0(temperature, ".", co2))
pairwise.t.test(mort.df$mean, mort.df$treatment, p.adjust.method = "bonferroni")
cohen.d((mort.df%>%filter(treatment=="10.amb"))$mean, (mort.df%>%filter(treatment=="6.amb"))$mean) #warm
cohen.d((mort.df%>%filter(treatment=="10.high"))$mean, (mort.df%>%filter(treatment=="6.amb"))$mean) #warm+OA

km <- km %>% mutate(treatment=paste0(temperature, ".", co2))
pairwise.t.test(km$mean, km$treatment, p.adjust.method = "bonferroni")
cohen.d((km%>%filter(treatment=="10.amb"))$mean, (km%>%filter(treatment=="6.amb"))$mean) #warm
cohen.d((km%>%filter(treatment=="10.high"))$mean, (km%>%filter(treatment=="6.amb"))$mean) #warm+OA
cohen.d((km%>%filter(treatment=="3.amb"))$mean, (km%>%filter(treatment=="6.amb"))$mean) #cold
cohen.d((km%>%filter(treatment=="3.high"))$mean, (km%>%filter(treatment=="6.amb"))$mean) #cold+OA

kw <- kw %>% mutate(treatment=paste0(temperature, ".", co2))
pairwise.t.test(kw$mean, kw.df$treatment, p.adjust.method = "bonferroni")
cohen.d((kw%>%filter(treatment=="3.amb"))$mean, (kw%>%filter(treatment=="6.amb"))$mean) #cold

gl.df <- gl.df %>% mutate(treatment=paste0(temperature, ".", co2))
pairwise.t.test(gl.df$gl, gl.df$treatment, p.adjust.method = "bonferroni")
cohen.d((gl.df%>%filter(treatment=="3.amb"))$gl, (gl.df%>%filter(treatment=="6.amb"))$gl) #cold
cohen.d((gl.df%>%filter(treatment=="3.high"))$gl, (gl.df%>%filter(treatment=="6.amb"))$gl) #cold+OA=

gm.df <- gm.df %>% mutate(treatment=paste0(temperature, ".", co2))
pairwise.t.test(gm.df$gm, gm.df$treatment, p.adjust.method = "bonferroni")
cohen.d((gm.df%>%filter(treatment=="3.amb"))$gm, (gm.df%>%filter(treatment=="6.amb"))$gm) #cold
cohen.d((gm.df%>%filter(treatment=="3.high"))$gm, (gm.df%>%filter(treatment=="6.amb"))$gm) #cold+OA
cohen.d((gm.df%>%filter(treatment=="10.amb"))$gm, (gm.df%>%filter(treatment=="6.amb"))$gm) #warm
cohen.d((gm.df%>%filter(treatment=="10.high"))$gm, (gm.df%>%filter(treatment=="6.amb"))$gm) #warm+OA

?cohen.d()
```

```{r}
phen_summ %>% filter(measurement=="MORT.RATE") %>% select(-days_post_hatch, -n) %>% distinct()
```


```{r, message=F, error=F}
phen_plot <- read_sheet("https://docs.google.com/spreadsheets/d/1Qu1RPkY3Uoq5Dz1_wTHah74SLoVea5wSFmhefrD15HU/edit?usp=sharing", sheet = "Phenotypes for plotting") %>% clean_names()

pheno_plot_df <- phen_plot %>% 
  pivot_longer(cols = control:warm_acid, names_to = "treat", values_to = "value") %>% 
  pivot_wider(names_from = stat, values_from = value) %>% 
  mutate(treat=factor(treat, ordered = T, 
                      levels=c("control", "acid", "warm", "warm_acid", "cool", "cool_acid"))) %>% 
  mutate(pheno=case_when(
    pheno=="mort.inst"~"Instantaneous Mortality",
    pheno=="kmh"~"Condition, morphometric",
    pheno=="kdw"~"Condition, weight",
    pheno=="gl"~"Growth Rate, length",
    pheno=="gm"~"Growth Rate, weight")) %>% 
  mutate(pheno=factor(pheno, ordered = T, 
          levels=c("Instantaneous Mortality", "Condition, morphometric", 
                   "Condition, weight", "Growth Rate, length", "Growth Rate, weight"))) %>% 
  group_by(pheno) %>%
  mutate(
    ymax_adjusted = max(mean + se, na.rm = TRUE) * 1.2,
    ymin_adjusted = min(mean - se, na.rm = TRUE)) %>% ungroup() #%>% 
#  filter(pheno %in% c("mort.inst", "kmh", "gl1")) 

pheno_plot_df %>% ggplot(aes(x=treat, y=mean, group=pheno)) + 
  geom_point(aes(color=treat), size=1.25) + 
  #geom_line() + 
  geom_errorbar(aes(color=treat, ymin = mean-se, ymax = mean+se), width = 0.2) + # Error bars
  theme_minimal() + 
  scale_color_manual(
    labels=c("cool"="Cool", "cool_acid"="Cool+Acidified",
             "control"="Control", "acid"="Acidified",
             "warm"="Warm", "warm_acid"="Warm+Acidified"),
    values=c("control"="gray65",
            "acid"=lighten("darkorange2", 0.25),
            "warm"=lighten("firebrick4", 0.45),
            "warm_acid"="firebrick4",
            "cool"=lighten("navyblue", 0.45),
            "cool_acid"="navyblue")) + 
  facet_wrap(~pheno, scales = "free", ncol = 1)  +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.5))) + # Extend y-axis by 20% above
  theme(
#    axis.text.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    #strip.text = element_blank(),
    strip.background = element_blank(),
    legend.position = "none") +
  labs(x = NULL, y=NULL)

pheno_plot_fig3 <- pheno_plot_df %>% filter(pheno %in% c("Instantaneous Mortality", "Condition, weight", "Growth Rate, length")) %>%
  ggplot(aes(x=treat, y=mean, group=pheno)) + 
  #geom_point(aes(color=treat), size=1.25) + 
  #geom_line() + 
  geom_errorbar(aes(color=treat, ymin = mean-se, ymax = mean+se), width = 0.2) + # Error bars
  scale_color_manual(
    labels=c("cool"="Cool", "cool_acid"="Cool+Acidified",
             "control"="Control", "acid"="Acidified",
             "warm"="Warm", "warm_acid"="Warm+Acidified"),
    values=c("control"="gray65",
            "acid"=lighten("darkorange2", 0.25),
            "warm"=lighten("firebrick4", 0.45),
            "warm_acid"="firebrick4",
            "cool"=lighten("navyblue", 0.45),
            "cool_acid"="navyblue")) + 
  facet_wrap(~pheno, scales = "free", ncol = 1)  +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.5))) + # Extend y-axis by 20% above
#  theme_cleveland() + 
  theme_minimal() + 
  theme(
#    axis.text.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    strip.text = element_blank(),
    axis.text.y=element_text(size=7.25), 
    strip.background = element_blank(),
    legend.position = "none") +
  labs(x = NULL, y=NULL)
  
require(patchwork)
require(clipr)
require(ggpubr)
```


```{r}
fig4.pheno <- 
  read_sheet("https://docs.google.com/spreadsheets/d/1Qu1RPkY3Uoq5Dz1_wTHah74SLoVea5wSFmhefrD15HU/edit?usp=sharing", 
           sheet = "Phenotype effects for plotting") %>% clean_names() %>% 

  mutate(stress=factor(stress, levels=c("OA", "Warm", "Warm-OA", "Cold", "Cold-OA"), ordered = T)) %>%
  #ungroup() %>% arrange(stress, desc(phenotype)) %>%
  mutate(phenotype = factor(phenotype, rev(unique(phenotype)), ordered = TRUE)) %>%

ggplot(aes(y = phenotype, x=stress, col=stress, fill=stress)) + 
  geom_point(alpha=0.65, aes(size=effect, shape=response)) +  #aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 #facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor", 
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4", 
                              "Cold"=lighten("navyblue", 0.45),
                              "Cold-OA"="navyblue"),
                  guide="none") +
 scale_fill_manual(name="Stressor", 
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4", 
                              "Cold"=lighten("navyblue", 0.45),
                              "Cold-OA"="navyblue"),
                  guide="none") +
#                  guide = guide_legend(override.aes = list(size=4))) +
  scale_size("Effect Size\n(phenotypes)", range = c(0.5,4.5), limits=c(1,7), breaks = c(2, 6),
             guide = guide_legend(override.aes = list(col="gray50", shape=17))) +
  scale_shape_manual(name=NULL, values=c("Lower"=25, "Higher"=24), 
                     guide = guide_legend(override.aes = list(col="gray50", size=3))) +
  scale_x_discrete(drop=F) + #Do drop empty factors for temperature contrasts
#  theme_cleveland() + 
  theme_minimal() + 
  theme(
        legend.position = "right",  #plot.margin = margin(5, 5, 5, 60),
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        legend.spacing.y = unit(0.5, "mm"),
        strip.text.x = element_blank()) +
  ggtitle("Phenotypic Effects")
```

Create a dataframe with all examined genes, identifying those of interest 

```{r}
results.df <- counts.annot.gadmor %>%
  dplyr::select(-contains("PCG")) %>% # remove counts from dataframe 
  mutate(
    deg.ph.3=case_when(gene_gadmor %in% rownames(diffex.pH.3.filt) ~ "yes", TRUE ~ "no"),
    deg.ph.6=case_when(gene_gadmor %in% rownames(diffex.pH.6.filt) ~ "yes", TRUE ~ "no"),
    deg.ph.10=case_when(gene_gadmor %in% rownames(diffex.pH.10.filt) ~ "yes", TRUE ~ "no"),
    deg.temp.3.6.amb=case_when(gene_gadmor %in% rownames(diffex.temp.3.6.amb.filt) ~ "yes", TRUE ~ "no"),
    #deg.temp.3.10.amb=case_when(gene_gadmor %in% rownames(diffex.temp.3.10.amb.filt) ~ "yes", TRUE ~ "no"),
    deg.temp.6.10.amb=case_when(gene_gadmor %in% rownames(diffex.temp.6.10.amb.filt) ~ "yes", TRUE ~ "no"),
    deg.temp.3.6.low=case_when(gene_gadmor %in% rownames(diffex.temp.3.6.low.filt) ~ "yes", TRUE ~ "no"),
    #deg.temp.3.10.low=case_when(gene_gadmor %in% rownames(diffex.temp.3.10.low.filt) ~ "yes", TRUE ~ "no"),
    deg.temp.6.10.low=case_when(gene_gadmor %in% rownames(diffex.temp.6.10.low.filt) ~ "yes", TRUE ~ "no"),
    deg.both.3low.6amb=case_when(gene_gadmor %in% rownames(diffex.both.3low.6amb.filt) ~ "yes", TRUE ~ "no"),
    deg.both.10low.6amb=case_when(gene_gadmor %in% rownames(diffex.both.10low.6amb.filt) ~ "yes", TRUE ~ "no")) #,
    #nonlinear.concave=case_when(gene_gadmor %in% unique((filter(nonlinear_df, temperature_2_binom=="Concave"))$geneid) ~ "yes", TRUE ~ "no"),
    #nonlinear.convex=case_when(gene_gadmor %in% unique((filter(nonlinear_df, temperature_2_binom=="Convex"))$geneid) ~ "yes", TRUE ~ "no"))    
    
# Options to identify genes from other analyses in edger, i.e. more linear/nonlinear and alternative splicing 
#    linear.interact.increase=case_when(gene_gadmor %in% linear_interaction_degs_increase$geneid ~ "yes", TRUE ~ "no"),
#    linear.interact.decrease=case_when(gene_gadmor %in% linear_interaction_degs_decrease$geneid ~ "yes", TRUE ~ "no"),
#    nonlinear.interact.Ashaped=case_when(gene_gadmor %in% nonlinear_interaction_degs_Ashaped$geneid ~ "yes", TRUE ~ "no"),
#    nonlinear.interact.Vshaped=case_when(gene_gadmor %in% nonlinear_interaction_degs_Vshaped$geneid ~ "yes", TRUE ~ "no"),
#    nonlinear.no.interact.Ashaped=case_when(gene_gadmor %in% (nonlinear_no.interaction_degs %>% filter(temperature_2_binom=="Convex"))$geneid ~ "yes", TRUE ~ "no"),
#    nonlinear.no.interact.Vshaped=case_when(gene_gadmor %in% (nonlinear_no.interaction_degs %>% filter(temperature_2_binom=="Concave"))$geneid ~ "yes", TRUE ~ "no"))#,
    #temperature.splice=case_when(gene_gadmor %in% temperature.splice.sign$Geneid ~ "yes", TRUE ~ "no"))

View(results.df)
```

Load DAVID enrichment results from DESeq2

```{r}
load(file="../results/GO-enrichment/Final-DAVID/david")
```

Plot any gene of interest using protein name 

### Digestive enzymes that change with larval development: 
- Trypsin (Trypsin-1,-3, -10) - decreases after ~17 DPH in Atlantic cod 
- Amylase (Alpha-amylase) - decreases after ~10 DPH in Atlantic cod 
- phospholipase A2 - increases after ~20 DPH in Atlantic cod 
- Cholecystokinin (CCK), neuropeptide Y (NPY) and Orexin (OX) - decrease steadily in Atlantic cod 

```{r, warning=F, message=F}
protein.name <- "2-iminobutanoate/2-iminopropanoate deaminase"

genes.interest <- gadMor.blast.GO %>% filter(str_detect(protein_names, protein.name)) %>% filter(gene_gadmor %in% rownames(res.pH.3)) 


#enzymes %>% select(gene_gadmor, spid, gene_uni, protein_names) %>% write_clip()

# Are enzymes of interest to RNASeq analyses?
results.df %>% filter(gene_gadmor %in% genes.interest$gene_gadmor)

for (i in 1:nrow(genes.interest)) {
a <-  plotCounts(dds.DESeq.treatment, gene=genes.interest$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #mutate(treatment=factor(treatment, ordered = T, levels = c("3_amb", "3_low", "6_amb", "6_low", "")))
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))

# Plot and facet wrap by ph
print(ggplot(a,
       aes(x=temperature, y=count, color=temperature, label=sample)) +
  geom_boxplot(outlier.shape = NA) + facet_wrap(~ph) +
  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle(str_wrap(genes.interest$protein_names[i])) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=14), legend.position = "none") +
    scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"))
  # scale_color_manual(values=c("3:amb"= "navyblue", "3:low"="royalblue1","6:amb"="darkgreen",
  #                             "6:low"="yellow3","10:amb"="firebrick4", "10:low"="sienna2"))
  )
}

```

Look at lipid metabolism genes that responded to different treatments - how much overlap?

```{r}
# Uniprot KW
OA <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS", process=="Lipid metabolism", stress=="OA"))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector()
Warm <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS", process=="Lipid metabolism", stress=="Warm"))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector()
`Warm+OA` <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS", process=="Lipid metabolism", stress=="Warm-OA"))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector()
Cold <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS", process=="Lipid metabolism", stress=="Cold"))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector()
```

```{r}
venn.lipid <- data.frame(dat = 
             unique(c(`OA`, `Warm`, `Warm+OA`, `Cold`))) %>%
  mutate(`OA` = dat %in% `OA`,
         `Warm` = dat %in% `Warm`,
         `Warm+OA` = dat %in% `Warm+OA`,
         `Cold` = dat %in% `Cold`) %>%
  select(`OA`, `Warm`, `Warm+OA`, `Cold`) %>%
  euler()

#pdf(file="../results/deseq2/venn_single-stressors_no-values.pdf", height = 5.2, width = 6)
eulerr:::plot.euler(venn.lipid, counts = T, quantities = T, 
                      legend = T, edges=F, fontzie=8,
                      font=1, cex=1, alpha=0.5,
                      fills=list(fill=c(lighten("darkorange2", 0.25),
                                        lighten("firebrick4", 0.45), 
                                        "firebrick4", 
                                        lighten("navyblue", 0.45))))
#dev.off()
```

```{r}
intersect(OA, Warm) #6
intersect(OA, `Warm+OA`) #4
intersect(OA, Cold) #2
intersect(Warm, `Warm+OA`) #12
intersect(`Warm`, Cold) #0
intersect(`Warm+OA`, Cold) #0

Reduce(intersect, list(OA, Warm, `Warm+OA`, Cold)) #0
Reduce(intersect, list(OA, Warm, `Warm+OA`)) #4
Reduce(intersect, list(OA, Warm)) #6

setdiff(Cold, c(OA, Warm, `Warm+OA`)) #9, only in cold
setdiff(OA, c(Cold, Warm, `Warm+OA`)) #6, only in OA
setdiff(Warm, c(OA, `Warm+OA`, Cold)) #10, only in Warm
setdiff(`Warm+OA`, c(OA, Warm, Cold)) #3, only in Warm+OA

# These are all the lipid-metabolism related genes diff. expressed in warm and/or warm+oa
c(setdiff(Warm, c(OA, `Warm+OA`, Cold)), #warm only #10
setdiff(`Warm+OA`, c(OA, Warm, Cold)), #warm/oa only #3
setdiff(intersect(Warm, `Warm+OA`), OA)) #warm and warm/oa only #8

# fatty acid catabolism - 1234
# triglyceride catabolism - 123456
# ether lipid / plasmalogen - 12
# phospholipid - 12
# vitamin d metabolism (cholesterol based molecule) - 1
# steroid biosynthesis - 1
# oligosaccharide/glycoprotein - 1
# gangliiosides (neuronal sphingolipids), membrane componenets nerve/brain - 12
# immune response - 1
# detox metabolic intermediate metabolies (enamine/imine)
# Patterns in genes diff. expressed in warm/warm+oa, but not oa: triglyceride and fatty acid metabolism related genes, some possible membrane related genes (e.g. ether lipids / plasmalogens such as "far1"), and a couple misc. genes involved in vitamin d, steroid bio., immune response, and metabolic detox. Only a couple loosely-related to cholesterols. 

setdiff(OA, c(Warm, `Warm+OA`)) #8, in OA not warm/warm+oa
# cholesterol excretion - 1
# vision/vitamin a metabolism - 1
# phospholipids (membrane) - 1
# small-intestine lipid digestion - 12
# neuronal activity - 12 
# fatty acid uptake - 1
# Patterns: a variety of lipid related processes, but not triglyceride/fatty acid catabolism! 

Reduce(intersect, list(OA, `Warm`)) #6 genes diff. expr. in both OA and warm and/or warm+oa
# fatty acid biosynthesis - 1
# triglyceride synthesis, maybe dietary fat absorption?
# prostaglandin biosynthesis (inflammation?)
# vitamin d
# sphingolipids (signaling) - 1

```
Heat map of lipid metabolism related genes 

```{r}
load(file = "../results/deseq2/degs")
degs.unique <- unique(c(rownames(degs$`Amb vs. Low pH @6C`),
rownames(degs$`3 vs. 6C @Amb pH`),
rownames(degs$`6 vs. 10C @Amb pH`), 
rownames(degs$`6 amb vs. 3C low`), 
rownames(degs$`6 amb vs. 10 low`)))

View(counts.annot.gadmor)
lipid.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% (unique(c(OA, Warm, `Warm+OA`)) %>% gsub(" ", "", .))))$gene_gadmor #and only those related to lipid metabolism
#Cold, 

load(file = "../results/deseq2/counts.tk")
load(file = "../results/deseq2/vsd.treatment")

sample.order <- (counts.tk %>% select(treatment) %>% rownames_to_column("sample") %>%
  mutate(treatment=factor(treatment, ordered = T, levels = c("6_amb", "3_low", "3_amb", "6_low", "10_low", "10_amb"))) %>%
  arrange(treatment))$sample

# Generate heat map / cluster lipid related genes
#pdf(file="../results/deseq2/heatmap-lipids.pdf", height = 5, width = 7.25)
cluster.most.variable <- pheatmap(t(assay(vsd.treatment))[sample.order,lipid.genes], 
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
                     show_colnames =TRUE, show_rownames = FALSE,
         border_color = NA,
         fontsize_col=6.5, fontsize = 8,
         cluster_cols = TRUE, cluster_rows = F,
         annotation_colors = list(treatment=
          c("Control"="gray65",
            "Acidification"=lighten("darkorange2", 0.25),
            "Warming"=lighten("firebrick4", 0.45),
            "Warming + Acidification"="firebrick4",
            "Cooling"=lighten("navyblue", 0.45),
            "Cooling + Acidification"="navyblue"
            )),
         annotation_row=as.data.frame(colData(vsd.treatment)[sample.order,c("treatment"), drop=FALSE]) %>%
           mutate(treatment=case_when(
              treatment == "6_amb" ~ "Control",
              treatment == "6_low" ~ "Acidification",
              treatment == "10_amb" ~ "Warming",
              treatment == "10_low" ~ "Warming + Acidification", 
              treatment == "3_amb" ~ "Cooling",
              treatment == "3_low" ~ "Cooling + Acidification"
              )),
         main = "Lipid metabolism gene counts")
#dev.off()
```

Figure out which genes cluster together 

```{r}
lipid.tree <- cluster.most.variable <- pheatmap(t(assay(vsd.treatment))[,lipid.genes], 
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
         cluster_cols = TRUE, cluster_rows = TRUE, silent = TRUE)

plot(lipid.tree$tree_col)
abline(h=10.5, col="red", lty=2, lwd=2)

rbind(t(assay(vsd.treatment))[,lipid.genes],
      cluster = cutree(lipid.tree$tree_col, k = 11)) %>%
  as.data.frame() %>% tail() %>%
  t() %>% as.data.frame() %>% select(cluster) %>%
  rownames_to_column("gene_gadmor") %>%
  left_join(counts.annot.gadmor[c("gene_gadmor", "gene_uni", "spid", "evalue", "protein_names")], 
            by = "gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% 
              select(spid, gene_ontology_biological_process, 
                     gene_ontology_cellular_component, gene_ontology_molecular_function, gene_ontology_i_ds)) %>%
  mutate(cluster=as.factor(cluster)) %>%

  # Filter for clusters in tree 
  filter(cluster %in% c(1)) %>% 

  # # To save Uniprot SPID to clipboard
  # select(spid) %>% na.omit() %>% unlist() %>% as.vector() %>% 
  # write_clip()

  # Generate wordcloud from GO terms 
  select(gene_ontology_biological_process) %>% 
  #select(gene_ontology_cellular_component) %>% 
  na.omit() %>% unlist() %>% as.vector() %>% 
  gsub("\\[[^\\]]*\\]", "", ., perl=TRUE) %>%
  gsub("process", "", ., perl=TRUE) %>% #get rid of word "process" since it's in everything
  wordcloud(min.freq = 2, random.order = F, random.color = F, colors = brewer.pal(2, "Dark2"))
```


Generate heat maps of genes in enriched Uniprot keywords
- Immune System  
- Hemostasis / blood coagulation 
- Electron transport
- Ribosomal Biogenesis / rRNA processing  

```{r}
lipid.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(stress %in% c("Warm", "Warm-OA")) %>%
             filter(grepl('Lipid', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)

immune.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
             filter(grepl('Immun', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)

heme.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
             filter(grepl('Hemostasis|coagulation', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)

electron.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
             filter(grepl('Electron transport', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)

ribosom.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
             filter(grepl('Ribosome|rRNA', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)

stress.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
             filter(grepl('Stress', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)

vision.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
             filter(grepl('Vision|Sensory', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)

adhesion.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
             filter(grepl('adhesion', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)

digestion.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
             filter(grepl('Digestion', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)

steroid.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
             filter(grepl('Steroid', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)
```


```{r}
#load(file = "../results/deseq2/degs")

lipid.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% lipid.spid))$gene_gadmor

immune.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% immune.spid))$gene_gadmor 

heme.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% heme.spid))$gene_gadmor 

electron.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% electron.spid))$gene_gadmor 

ribosom.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% ribosom.spid))$gene_gadmor 

stress.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% stress.spid))$gene_gadmor 

vision.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% vision.spid))$gene_gadmor 

adhesion.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% adhesion.spid))$gene_gadmor

digestion.genes <- c((counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% digestion.spid))$gene_gadmor, "LOC115532524", "LOC115532138", "LOC115543832", "LOC115557385") #ADD LIPID-RELATED DIGESTION

steroid.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% steroid.spid))$gene_gadmor

# Generate heat map / cluster lipid related genes
#pdf(file="../results/deseq2/heatmap-lipids.pdf", height = 5, width = 7.25)
cluster.most.variable <- 
#  pheatmap(t(assay(vsd.treatment))[sample.order,immune.genes], 
#  pheatmap(t(assay(vsd.treatment))[sample.order,heme.genes], 
#  pheatmap(t(assay(vsd.treatment))[sample.order,electron.genes], 
#  pheatmap(t(assay(vsd.treatment))[sample.order,ribosom.genes], 
#  pheatmap(t(assay(vsd.treatment))[sample.order,vision.genes], 
  pheatmap(t(assay(vsd.treatment))[sample.order,vision.genes],   
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
                     show_colnames =TRUE, show_rownames = FALSE,
         border_color = NA,
         fontsize_col=6.5, fontsize = 8,
         cluster_cols = TRUE, cluster_rows = F,
         annotation_colors = list(treatment=
          c("Control"="gray65",
            "Acidification"=lighten("darkorange2", 0.25),
            "Warming"=lighten("firebrick4", 0.45),
            "Warming + Acidification"="firebrick4",
            "Cooling"=lighten("navyblue", 0.45),
            "Cooling + Acidification"="navyblue")),
         annotation_row=as.data.frame(colData(vsd.treatment)[sample.order,c("treatment"), drop=FALSE]) %>%
           mutate(treatment=case_when(
              treatment == "6_amb" ~ "Control",
              treatment == "6_low" ~ "Acidification",
              treatment == "10_amb" ~ "Warming",
              treatment == "10_low" ~ "Warming + Acidification", 
              treatment == "3_amb" ~ "Cooling", 
              treatment == "3_low" ~ "Cooling + Acidification")),
#         main = "Immune system gene counts")
         main = "Hemoglobin & coagulation gene counts")
#         main = "Electron transport gene counts")
#         main = "Ribosome Biogenesis & rRNA Processing gene counts")
#dev.off()
```

MAKE Heatmaps for each process - mean expression shown for each treatment 

```{r}

z <- list(
  "Lipid metabolism genes" = lipid.genes, 
  "Immune genes" = immune.genes, 
  "Hemostasis genes" = heme.genes, 
  "Electron transport genes" = electron.genes, 
  "Ribosomal biogensis genes" = ribosom.genes, 
  "Stress-response genes" = stress.genes, 
  "Vision genes" = vision.genes, 
  "Cell adhesion genes" = adhesion.genes,
  "Digestion genes" = digestion.genes,
  "Steroid metabolism genes" = steroid.genes)

#z <- immune.genes
# z <- heme.genes
# z <- electron.genes
# z <- ribosom.genes
# z <- vision.genes
# z <- adhesion.genes

breaksList<-seq(-1.5, 1.5, by = 0.1)
draw.legend=c(TRUE, FALSE, FALSE,FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE)
legend.position=c("right", "none", "none", "none", "none", "none", "none", "none", "none", "none")
heatmaps_list <- vector(mode = "list", length = length(z))
names(heatmaps_list) <- names(z)

for (i in 1:length(z)) {
  z.lines <- vector("list", length(z[[i]]))
  names(z.lines) <- z[[i]]

  for (j in 1:length(z.lines)) {
  z.lines[[j]] <-  plotCounts(dds.DESeq.treatment, gene=z[[i]][j], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
    rownames_to_column("sample") %>%
    left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
    mutate(treatment=fct_relevel(treatment, c("6_amb", "6_low", "3_amb", "3_low", "10_amb",  "10_low")))
}

  df1 <- bind_rows(z.lines, .id = "gene_gadmor") %>% 
    mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
    group_by(gene_gadmor, treatment) %>%
    dplyr::summarize(mean=mean(count), median=median(count), sd=sd(count), cv=sd(count)/mean(count)) %>% ungroup()
    
  #df2 <- df1 %>% filter(treatment=="6_amb") %>% select(gene_gadmor, mean) %>% rename("mean"="control")
  
  df1.heat <- df1 %>%  remove_rownames() %>%
  dplyr::select(treatment, gene_gadmor, mean) %>%
  pivot_wider(names_from = gene_gadmor, values_from = mean) %>%
  column_to_rownames("treatment")

heatmaps_list[[i]] <- pheatmap(df1.heat %>% t(), 
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="row",
                     show_colnames =FALSE, show_rownames = TRUE,
         border_color = NA,
         fontsize_col=6.5, fontsize = 7,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(length(breaksList)),
         breaks =  breaksList,
         cluster_cols = F, cluster_rows = T, treeheight_row = 0,
         legend=draw.legend[i],  annotation_legend = draw.legend[i],
         # annotation_col = (df1.heat %>% mutate(treatment=rownames(.)) %>%
         #   mutate(treatment=case_when(
         #      treatment == "6_amb" ~ "Control",
         #      treatment == "6_low" ~ "Acidification",
         #      treatment == "10_amb" ~ "Warming",
         #      treatment == "10_low" ~ "Warming + Acidification", 
         #      treatment == "3_amb" ~ "Cooling", 
         #      treatment == "3_low" ~ "Cooling + Acidification")))[,"treatment",drop=FALSE],
         #  annotation_colors = list(treatment=
         #  c("Control"="gray65",
         #    "Acidification"=lighten("darkorange2", 0.25),
         #    "Warming"=lighten("firebrick4", 0.45),
         #    "Warming + Acidification"="firebrick4",
         #    "Cooling"=lighten("navyblue", 0.45),
         #    "Cooling + Acidification"="navyblue")),
         main = names(z[i]), 
         #main="",
         silent=F)
}

```

FIGURE 5 

```{r}
#margin(top,right,bottom,left)
fig5a <- as.ggplot(heatmaps_list$`Lipid metabolism genes`$gtable) + theme(plot.margin = margin(0, 1, 0, 70))
fig5b <- as.ggplot(heatmaps_list$`Immune genes`$gtable) + theme(plot.margin = margin(0, 0, 0, 70)) 
fig5c <- as.ggplot(heatmaps_list$`Vision genes`$gtable) + theme(plot.margin = margin(0, 0, 0, 70))
fig5.spacer <- ggplot() + theme_void() + theme(plot.margin = margin(0,0,-1,-1))

fig5.annot2 <- ggplot() + xlim(c(0,1)) + ylim(c(0,1)) +
  annotate("text", x = 0.05, y = 0.83, size = 4, angle=-90, label = "Control", color="gray35") +
  annotate("text", x = 0.20, y = 0.80, size = 4, angle=-90, label = "Acidifed", color="darkorange2") +
  annotate("text", x = 0.34, y = 0.91, size = 4, angle=-90, label = "Cold", color=lighten("navyblue", 0.5)) +
  annotate("text", x = 0.48, y = 0.56, size = 4, angle=-90, label = "Cold + Acidified", color="navyblue") +
  annotate("text", x = 0.62, y = 0.87, size = 4, angle=-90, label = "Warm", color=lighten("firebrick4", 0.5)) +
  annotate("text", x = 0.76, y = 0.52, size = 4, angle=-90, label = "Warm + Acidified", color="firebrick4") +
  theme_void() + theme(plot.margin = margin(-1, -1, 1, 70))  # Adjust margins around the text

fig5 <- (fig5a) /
        (fig5b + fig5.spacer + plot_layout(widths=c(5.25,1))) /
        (fig5c + fig5.spacer + plot_layout(widths=c(5.25,1))) /
        (fig5.annot2 + fig5.spacer + plot_layout(widths=c(2,1))) +
        plot_layout(heights = c(1.5, 1, 1, .55)) 

pdf(file = "../manuscript/Fig5.pdf", height = 9, width=3.8)
fig5
#grid.locator(unit = "native")# using grid.locator to identify the coordinate for brackets
# add bracket to plot area
grid.brackets(77, 100, 77, 232, type = 4, ticks=NA, col="gray25", h=0.008, lwd=0.8)
grid.brackets(77, 259, 77, 387, type = 4, ticks=NA, col="gray25", h=0.008, lwd=0.8)
grid.brackets(77, 415, 77, 623, type = 4, ticks=NA, col="gray25", h=0.008, lwd=0.8)
dev.off()
# then hand annotate! 
```

FIGURE 6 

```{r}
#margin(top,right,bottom,left)
fig6a <- as.ggplot(heatmaps_list$`Digestion genes`$gtable) + theme(plot.margin = margin(-3, 1, -3, 60))
fig6b <- as.ggplot(heatmaps_list$`Hemostasis genes`$gtable) + theme(plot.margin = margin(-3, 0, -3, 60)) 
fig6c <- as.ggplot(heatmaps_list$`Electron transport genes`$gtable) + theme(plot.margin = margin(-3, 0, -3, 60))
fig6.spacer <- ggplot() + theme_void() + theme(plot.margin = margin(-3,0,-3,-3))

fig6.annot <- ggplot() + xlim(c(0,1)) + ylim(c(0,1)) +
  annotate("text", x = 0.05, y = 0.83, size = 3, angle=-90, label = "Control", color="gray35") +
  annotate("text", x = 0.20, y = 0.80, size = 3, angle=-90, label = "Acidifed", color="darkorange2") +
  annotate("text", x = 0.34, y = 0.91, size = 3, angle=-90, label = "Cold", color=lighten("navyblue", 0.5)) +
  annotate("text", x = 0.48, y = 0.56, size = 3, angle=-90, label = "Cold + Acidified", color="navyblue") +
  annotate("text", x = 0.62, y = 0.87, size = 3, angle=-90, label = "Warm", color=lighten("firebrick4", 0.5)) +
  annotate("text", x = 0.76, y = 0.52, size = 3, angle=-90, label = "Warm + Acidified", color="firebrick4") +
  theme_void() + theme(plot.margin = margin(-1, -1, 1, 70))  # Adjust margins around the text

fig6 <- (fig6a) /
        (fig6b + fig6.spacer + plot_layout(widths=c(5.25,1))) /
        (fig6c + fig6.spacer + plot_layout(widths=c(5.25,1))) /
        (fig5.annot2 + fig6.spacer + plot_layout(widths=c(2,1))) +
        plot_layout(heights = c(0.85, 0.75, 1.1, 1)) 

pdf(file = "../manuscript/Fig6.pdf", height = 4, width=3.8)
fig6
grid.brackets(77, 40, 77, 90, type = 4, ticks=NA, col="gray25", h=0.008, lwd=0.8)
grid.brackets(77, 100, 77, 150, type = 4, ticks=NA, col="gray25", h=0.008, lwd=0.8)
grid.brackets(77, 185, 77, 250, type = 4, ticks=NA, col="gray25", h=0.008, lwd=0.8)
dev.off()
# then hand annotate! 
```

FIGURE 7 - cold heat map 

```{r}
#margin(top,right,bottom,left)
fig7a <- as.ggplot(heatmaps_list$`Ribosomal biogensis genes`$gtable) + theme(plot.margin = margin(-3, 20, -3, 40))
fig7b <- as.ggplot(heatmaps_list$`Stress-response genes`$gtable) + theme(plot.margin = margin(-3, 20, -3, 40)) 
#fig7.spacer <- ggplot() + theme_void() + theme(plot.margin = margin(-3,30,-3,-3))

fig7.annot <- ggplot() + xlim(c(0,1)) + ylim(c(0,1)) +
  annotate("text", x = 0.04, y = 0.83, size = 3, angle=-90, label = "Control", color="gray35") +
  annotate("text", x = 0.17, y = 0.80, size = 3, angle=-90, label = "Acidifed", color="darkorange2") +
  annotate("text", x = 0.29, y = 0.91, size = 3, angle=-90, label = "Cold", color=lighten("navyblue", 0.5)) +
  annotate("text", x = 0.41, y = 0.56, size = 3, angle=-90, label = "Cold + Acidified", color="navyblue") +
  annotate("text", x = 0.54, y = 0.87, size = 3, angle=-90, label = "Warm", color=lighten("firebrick4", 0.5)) +
  annotate("text", x = 0.66, y = 0.52, size = 3, angle=-90, label = "Warm + Acidified", color="firebrick4") +
  theme_void() + theme(plot.margin = margin(-1, 20, 1, 40))  # Adjust margins around the text

fig7 <- fig7a /
        fig7b /
        fig7.annot +
        plot_layout(heights = c(1.25, 1, 0.85)) 

pdf(file = "../manuscript/Fig7.pdf", height = 3.5, width=3.6)
fig7
#grid.brackets(77, 40, 77, 90, type = 4, ticks=NA, col="gray25", h=0.008, lwd=0.8)
#grid.brackets(77, 100, 77, 150, type = 4, ticks=NA, col="gray25", h=0.008, lwd=0.8)
#grid.brackets(77, 185, 77, 250, type = 4, ticks=NA, col="gray25", h=0.008, lwd=0.8)
dev.off()
# then hand annotate! 
```



Lipid metabolism - explore gene clusters and their functions 

```{r}
z <- lipid.genes

z.lines <- vector("list", length(z))
names(z.lines) <- z

for (j in 1:length(z.lines)) {
  z.lines[[j]] <-  plotCounts(dds.DESeq.treatment, gene=z[j], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
    rownames_to_column("sample") %>%
    left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
    mutate(treatment=fct_relevel(treatment, c("6_amb", "6_low", "3_amb", "3_low", "10_amb",  "10_low")))
}

df1 <- bind_rows(z.lines, .id = "gene_gadmor") %>% 
    mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
    group_by(gene_gadmor, treatment) %>%
    dplyr::summarize(mean=mean(count), median=median(count), sd=sd(count), cv=sd(count)/mean(count)) %>% ungroup()
    
df1.heat <- df1 %>%  remove_rownames() %>%
  dplyr::select(treatment, gene_gadmor, mean) %>%
  pivot_wider(names_from = gene_gadmor, values_from = mean) %>%
  column_to_rownames("treatment")


pheatmap(df1.heat %>% t(), 
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="row",
                     show_colnames =FALSE, show_rownames = TRUE,
         border_color = NA,
         fontsize_col=6.5, fontsize = 8,
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(length(breaksList)),
         breaks =  seq(-1.5, 1.5, by = 0.1),
         cluster_cols = F, cluster_rows = T, 
         annotation_col = (df1.heat %>% mutate(treatment=rownames(.)) %>%
           mutate(treatment=case_when(
              treatment == "6_amb" ~ "Control",
              treatment == "6_low" ~ "Acidification",
              treatment == "10_amb" ~ "Warming",
              treatment == "10_low" ~ "Warming + Acidification", 
              treatment == "3_amb" ~ "Cooling", 
              treatment == "3_low" ~ "Cooling + Acidification")))[,"treatment",drop=FALSE],
          annotation_colors = list(treatment=
          c("Control"="gray65",
            "Acidification"=lighten("darkorange2", 0.25),
            "Warming"=lighten("firebrick4", 0.45),
            "Warming + Acidification"="firebrick4",
            "Cooling"=lighten("navyblue", 0.45),
            "Cooling + Acidification"="navyblue")),
         main = "Lipid genes")



lipid.tree <- pheatmap(df1.heat, 
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
         cluster_cols = TRUE, cluster_rows = TRUE, silent = F)

plot(lipid.tree$tree_col)
abline(h=1.85, col="red", lty=2, lwd=2)

rbind(df1.heat,
      cluster = cutree(lipid.tree$tree_col, k = 9)) %>%
  as.data.frame() %>% tail() %>%
  t() %>% as.data.frame() %>% select(cluster) %>%
  rownames_to_column("gene_gadmor") %>%
  left_join(counts.annot.gadmor[c("gene_gadmor", "gene_uni", "spid", "evalue", "protein_names")], 
            by = "gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% 
              select(spid, gene_ontology_biological_process, 
                     gene_ontology_cellular_component, gene_ontology_molecular_function, gene_ontology_i_ds)) %>%
  mutate(cluster=as.factor(cluster)) %>%

  # Filter for clusters in tree 
  filter(cluster %in% c(3)) %>% #dplyr::select(gene_gadmor, cluster) %>% distinct()

  # # To save Uniprot SPID to clipboard
  # select(spid) %>% na.omit() %>% unlist() %>% as.vector() %>% 
  # write_clip()

  # Generate wordcloud from GO terms 
  select(gene_ontology_biological_process) %>% 
  #select(gene_ontology_cellular_component) %>% 
  na.omit() %>% unlist() %>% as.vector() %>% 
  gsub("\\[[^\\]]*\\]", "", ., perl=TRUE) %>%
  gsub("process|metabolic|regulation|lipid", "", ., perl=TRUE) %>% #get rid of word "process" since it's in everything
  wordcloud(min.freq = 2, random.order = F, random.color = F, colors = brewer.pal(2, "Dark2"))

require(wordcloud)
```





Figure out which genes cluster together 

```{r}
immune.tree <- pheatmap(t(assay(vsd.treatment))[,immune.genes], 
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
         cluster_cols = TRUE, cluster_rows = TRUE, silent = TRUE)

plot(immune.tree$tree_col)
abline(h=11.8, col="red", lty=2, lwd=2)

rbind(t(assay(vsd.treatment))[,immune.genes],
      cluster = cutree(immune.tree$tree_col, k = 2)) %>%
  as.data.frame() %>% tail() %>%
  t() %>% as.data.frame() %>% select(cluster) %>%
  rownames_to_column("gene_gadmor") %>%
  left_join(counts.annot.gadmor[c("gene_gadmor", "gene_uni", "spid", "evalue", "protein_names")], 
            by = "gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% 
              select(spid, gene_ontology_biological_process, 
                     gene_ontology_cellular_component, gene_ontology_molecular_function, gene_ontology_i_ds)) %>%
  mutate(cluster=as.factor(cluster)) %>%

  # Filter for clusters in tree 
  #filter(cluster %in% c(2)) %>% 

  # # To save Uniprot SPID to clipboard
  # select(spid) %>% na.omit() %>% unlist() %>% as.vector() %>% 
  # write_clip()

  # Generate wordcloud from GO terms 
  select(gene_ontology_biological_process) %>% 
  #select(gene_ontology_cellular_component) %>%
  #select(protein_names) %>% 
  
  
  na.omit() %>% unlist() %>% as.vector() %>% 
  gsub("\\[[^\\]]*\\]", "", ., perl=TRUE) %>%
  gsub("interleukin-..|interleukin-.|interleukin.", "interleukin", ., perl=TRUE) %>%  #gsub("process|regulation|negative|positive|response|pathway|production", "", ., perl=TRUE) %>% #get rid of word "process" since it's in everything
gsub("process|transposon", "", ., perl=TRUE) %>% #get rid of word "process" since it's in everything
  wordcloud(min.freq = 5, random.order = F, random.color = F, colors = brewer.pal(2, "Dark2"))

```
Generate PCA using differentially expressed genes found to be associated with immune system 

```{r}
pca.princomp.var <- prcomp(t(assay(vsd.treatment)[immune.genes,])) # using the same code that's under the hood of PlotPCA 
pca.eigenval(pca.princomp.var) #The Proporation of Variance = %variance 
pc.percent.var <- pca.eigenval(pca.princomp.var)[2,1:5]*100
pc.percent.var[1:2] %>% sum()
screeplot(pca.princomp.var, bstick=TRUE) #looks like PC 1-3 are significant 

tab.expr.var <- data.frame(sample.id = colnames(assay(vsd.treatment[immune.genes,])),
    EV1.all = pca.princomp.var$x[,1],    # the first eigenvector
    EV2.all = pca.princomp.var$x[,2],    # the second eigenvector
    EV3.all = pca.princomp.var$x[,3],    # the third eigenvector
    stringsAsFactors = FALSE)
#shapiro.test(pca.princomp$x) #sample size too large for shapiro test which is weird 
#hist(pca.princomp$x) #normal? hard to say, maybe
tab.expr.annot.var <- left_join(tab.expr.var, sample.info, by=c("sample.id"="vial_label")) %>% droplevels()

# GGSCATTER with ellipses and stars - PC1 PC2
ggscatter(tab.expr.annot.var, group=c("temperature", "ph"),
          x="EV1.all", y="EV2.all", col="treatment", shape="ph", size=3.5, alpha=0.85, 
          ellipse = FALSE, star.plot = FALSE) +
  theme_minimal() + ggtitle("Gene Expression PC1xPC2\nImmune system DEGs") + 
  xlab(paste("PC1 (", round(pc.percent.var[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.var[2], digits = 1), "%)", sep="")) + 
  labs(color="OA Treatment") + 
  theme(legend.position = "right") + 
  scale_color_manual(name="Treatment", 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4", 
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="High pCO2", 
                              "10_amb"="Warm Temperature", 
                              "10_low"="Warm Temperature + High pCO2", 
                              "3_amb"="Cold Temperature",
                              "3_low"="Cold Temperature + High pCO2")) +
  scale_shape_manual(guide="none", values=c("amb"=19, "low"=17)) + 
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"), order=1))
```


REMAKE IMMUNE HEAT MAP WITH ONLY WARM AND ACIDIFICATION TREATMENTS

```{r}
sample.order.presentation <- (counts.tk %>% select(treatment) %>% rownames_to_column("sample") %>%
                                filter(treatment %!in% c("3_low", "3_amb")) %>%
  mutate(treatment=factor(treatment, ordered = T, levels = c("6_amb", "6_low", "10_low", "10_amb"))) %>%
  arrange(treatment))$sample

cluster.most.variable <- pheatmap(t(assay(vsd.treatment))[sample.order.presentation,immune.genes], 
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
                     show_colnames =TRUE, show_rownames = FALSE,
         border_color = NA,
         fontsize_col=6.5, fontsize = 8,
         cluster_cols = TRUE, cluster_rows = F,
         annotation_colors = list(treatment=
          c("Control"="gray65",
            "Acidification"=lighten("darkorange2", 0.25),
            "Warming"=lighten("firebrick4", 0.45),
            "Warming + Acidification"="firebrick4")),
         annotation_row=as.data.frame(colData(vsd.treatment)[sample.order,c("treatment"), drop=FALSE]) %>%
           mutate(treatment=case_when(
              treatment == "6_amb" ~ "Control",
              treatment == "6_low" ~ "Acidification",
              treatment == "10_amb" ~ "Warming",
              treatment == "10_low" ~ "Warming + Acidification")),
         main = "Immune function gene counts")
```

MAKE BOXPLOTs for each process - one boxplot with all genes 

```{r}
immune.genes.lines <- vector("list", length(immune.genes))
names(immune.genes.lines) <- immune.genes

for (i in 1:length(immune.genes)) {
immune.genes.lines[[i]] <-  plotCounts(dds.DESeq.treatment, gene=immune.genes[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  filter(temperature!=3) %>%
  mutate(treatment=fct_relevel(treatment, c("6_amb", "10_amb",  "10_low","6_low")))
}

df1 <- bind_rows(immune.genes.lines, .id = "gene_gadmor") %>% 
  mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
  group_by(gene_gadmor, treatment) %>%
  dplyr::summarize(mean=mean(count), median=median(count), sd=sd(count), cv=sd(count)/mean(count)) %>% ungroup()
    
df2 <- df1 %>% filter(treatment=="6_amb") %>% select(gene_gadmor, mean) %>% rename("mean"="control")

imm.df <- left_join(df1, df2) %>%
mutate(mean_diff=mean-control) %>% filter(treatment!="6_amb") %>% 
  mutate(log.diff=log(mean_diff))

  # One boxplot, points for all genes
imm.df %>%  ggplot(aes(x=treatment, y=log(mean_diff), color=treatment)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() + ylab("Δ from control, log(counts)") +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  scale_color_manual(name="Treatment", values=c("6_amb"="gray65", "6_low"=lighten("darkorange2", 0.25),
                                                  "10_amb"=lighten("firebrick4", 0.45),  "10_low"="firebrick4")) #+


  # # Boxplots warm and control 
# left_join(df1, df2) %>%
# #mutate(mean_diff=mean-control) %>% 
#   filter(treatment %in% c("6_amb", "10_amb")) %>%
#   mutate(temperature=case_when(treatment=="6_amb" ~ 6, treatment=="10_amb"~10)) %>% 
# #  group_by(gene_gadmor, treatment) %>% 
# 
#   # One boxplot, points for all genes
#   ggplot(aes(x=temperature, y=log(mean), group=gene_gadmor, color=treatment)) +
#   geom_line() + 
#   # geom_boxplot(outlier.shape = NA) +
#   # geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
#   ggtitle("Immune System") + 
#     theme_bw() + ylab("Transcript counts per gene (log-trans)") +
#     theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
#   scale_color_manual(name="Treatment", values=c("6_amb"="gray65", "6_low"=lighten("darkorange2", 0.25),
#                                                   "10_amb"=lighten("firebrick4", 0.45),  "10_low"="firebrick4")) #+
```

BOXPLOT OF LIPID GENES

```{r}
lipid.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% #
             filter(grepl('Lipid', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>%
  gsub(" ", "", .)
lipid.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% lipid.spid))$gene_gadmor #and only those related to immune functions
lipid.genes.lines <- vector("list", length(lipid.genes))
names(lipid.genes.lines) <- lipid.genes

for (i in 1:length(lipid.genes)) {
lipid.genes.lines[[i]] <-  plotCounts(dds.DESeq.treatment, gene=lipid.genes[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  filter(temperature!=3) %>%
  mutate(treatment=fct_relevel(treatment, c("6_amb", "10_amb",  "10_low","6_low")))
}

df1 <- bind_rows(lipid.genes.lines, .id = "gene_gadmor") %>% 
  mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
  group_by(gene_gadmor, treatment) %>%
  dplyr::summarize(mean=mean(count), median=median(count), sd=sd(count), cv=sd(count)/mean(count)) %>% ungroup()
    
df2 <- df1 %>% filter(treatment=="6_amb") %>% select(gene_gadmor, mean) %>% rename("mean"="control")

lip.df <- left_join(df1, df2) %>%
mutate(mean_diff=mean-control) %>% filter(treatment!="6_amb") %>% 
  mutate(log.diff=-1*log(-1*mean_diff))

  # One boxplot, points for all genes
lip.df %>%   ggplot(aes(x=treatment, y=-1*log(-1*mean_diff), color=treatment)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() + ylab("Δ from control, -log(-counts)") +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  scale_color_manual(name="Treatment", values=c("6_amb"="gray65", "6_low"=lighten("darkorange2", 0.25),
                                                  "10_amb"=lighten("firebrick4", 0.45),  "10_low"="firebrick4")) #+
  
#   # Line plots
# lip.df %>% ggplot(aes(x=treatment, y=log(mean), group=gene_gadmor)) + geom_line(size=.01) +
#   theme_minimal() +
#   #scale_x_continuous(breaks=c(1,2,3), labels=c("Ambient", "Moderate", "Low")) +
#   theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
#   xlab("Per treatment mean of normalized counts") + ggtitle("Abundances of DEGs by\ntreatment & relative expression")


# # Boxplots warm and control 
# left_join(df1, df2) %>%
# #mutate(mean_diff=mean-control) %>% 
#   filter(treatment %in% c("6_amb", "10_amb")) %>%
#   group_by(gene_gadmor, treatment) %>% 
# 
#   # One boxplot, points for all genes
#   ggplot(aes(x=treatment, y=log(mean), color=treatment)) +
#   geom_boxplot(outlier.shape = NA) +
#   geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
#   ggtitle("Lipid") + 
#     theme_bw() + ylab("Transcript counts per gene (log-trans)") +
#     theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
#   scale_color_manual(name="Treatment", values=c("6_amb"="gray65", "6_low"=lighten("darkorange2", 0.25),
#                                                   "10_amb"=lighten("firebrick4", 0.45),  "10_low"="firebrick4")) #+

```
Other processes affected by warming 

```{r}
# adhesion genes
adhesion.genes.lines <- vector("list", length(adhesion.genes))
names(adhesion.genes.lines) <- adhesion.genes

for (i in 1:length(adhesion.genes)) {
adhesion.genes.lines[[i]] <-  plotCounts(dds.DESeq.treatment, gene=adhesion.genes[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  filter(temperature!=3) %>%
  mutate(treatment=fct_relevel(treatment, c("6_amb", "10_amb",  "10_low","6_low")))
}

df1 <- bind_rows(adhesion.genes.lines, .id = "gene_gadmor") %>% 
  mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
  group_by(gene_gadmor, treatment) %>%
  dplyr::summarize(mean=mean(count), median=median(count), sd=sd(count), cv=sd(count)/mean(count)) %>% ungroup()
    
df2 <- df1 %>% filter(treatment=="6_amb") %>% select(gene_gadmor, mean) %>% rename("mean"="control")

adhes.df <- left_join(df1, df2) %>%
mutate(mean_diff=mean-control) %>% filter(treatment!="6_amb")  %>% 
  mutate(log.diff=log(mean_diff))


# vision genes
vision.genes.lines <- vector("list", length(vision.genes))
names(vision.genes.lines) <- vision.genes

for (i in 1:length(vision.genes)) {
vision.genes.lines[[i]] <-  plotCounts(dds.DESeq.treatment, gene=vision.genes[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  filter(temperature!=3) %>%
  mutate(treatment=fct_relevel(treatment, c("6_amb", "10_amb",  "10_low","6_low")))
}

df1 <- bind_rows(vision.genes.lines, .id = "gene_gadmor") %>% 
  mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
  group_by(gene_gadmor, treatment) %>%
  dplyr::summarize(mean=mean(count), median=median(count), sd=sd(count), cv=sd(count)/mean(count)) %>% ungroup()
    
df2 <- df1 %>% filter(treatment=="6_amb") %>% select(gene_gadmor, mean) %>% rename("mean"="control")

vision.df <- left_join(df1, df2) %>%
mutate(mean_diff=mean-control) %>% filter(treatment!="6_amb") %>% 
  mutate(log.diff=log(mean_diff))

```

```{r}
bind_rows(list("vision"=vision.df, "adhesion"=adhes.df, "lipid"=lip.df, "immune"=imm.df), .id = "process") %>% 
#  filter(treatment=="10_amb") %>% 
  filter(treatment %in% c("10_amb", "10_low")) %>% 
  mutate(process=factor(process, ordered=T, levels=c("lipid", "immune", "adhesion", "vision"))) %>% 
  filter(process!="vision") %>% 
  ggplot(aes(x=process:treatment, y=log.diff, shape=treatment)) +
  geom_boxplot(outlier.shape = NA) +
  scale_shape_manual(values=c(15,17)) + 
  geom_point(size=1.25, color="gray25", position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() + ylab("Δ from control, log(counts)") + 
  geom_hline(yintercept = 0, linetype="dashed", color="gray40") #+ 
#    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") #+
  # scale_color_manual(name="Treatment", values=c("6_amb"="gray65", "6_low"=lighten("darkorange2", 0.25),
  #                                                 "10_amb"=lighten("firebrick4", 0.45),  "10_low"="firebrick4")) #+
```



BOXPLOT OF DIGESTION GENES

```{r}
digestion.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% #
             filter(grepl('Digestion', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>%
  gsub(" ", "", .)

#digestion.spid <- "P30122" #bile salt lipase 
digestion.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% digestion.spid))$gene_gadmor #and only those related to immune functions
digestion.genes.lines <- vector("list", length(digestion.genes))
names(digestion.genes.lines) <- digestion.genes

for (i in 1:length(digestion.genes)) {
digestion.genes.lines[[i]] <-  plotCounts(dds.DESeq.treatment, gene=digestion.genes[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #filter(temperature!=3) %>%
  mutate(treatment=fct_relevel(treatment, c("6_amb", "6_low", "3_amb", "3_low", "10_amb",  "10_low")))

print(digestion.genes.lines[[i]] %>%
  ggplot(aes(x=treatment, y=count, color=treatment)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() + 
    #ylab("Δ from control, -log(-counts)") +
    ylab("Gene counts") +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), 
          axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "right") +
  scale_color_manual(name="Treatment", 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4", 
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="High pCO2", 
                              "10_amb"="Warm Temperature", 
                              "10_low"="Warm Temperature + High pCO2", 
                              "3_amb"="Cold Temperature",
                              "3_low"="Cold Temperature + High pCO2")))

}
```


BOXPLOT OF HEMOSTASIS GENES

```{r}
hemostasis.spid <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS") %>% #
             filter(grepl('Hemostasis|coagulation', process)))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>%
  gsub(" ", "", .)

#digestion.spid <- "P30122"
hemostasis.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% hemostasis.spid))$gene_gadmor #and only those related to henme functions
hemostasis.genes.lines <- vector("list", length(hemostasis.genes))
names(hemostasis.genes.lines) <- hemostasis.genes
hemostasis.proteins <- hemostasis.genes %>% as.data.frame() %>% set_names("gene_gadmor") %>% left_join(gadMor.blast.GO[c("gene_gadmor", "protein_names")]) %>%
  mutate(protein_names=gsub(" \\(.*", "", protein_names))

for (i in 1:length(hemostasis.genes)) {
hemostasis.genes.lines[[i]] <-  plotCounts(dds.DESeq.treatment, gene=hemostasis.genes[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #filter(temperature!=3) %>%
  mutate(treatment=fct_relevel(treatment, c("6_amb", "6_low", "3_amb", "3_low", "10_amb",  "10_low")))

print(hemostasis.genes.lines[[i]] %>%
  ggplot(aes(x=treatment, y=count, color=treatment)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() + 
    #ylab("Δ from control, -log(-counts)") +
    ggtitle(paste(names(hemostasis.genes.lines[i]), "\n", hemostasis.proteins$protein_names[i], sep="")) +
    ylab("Gene counts") +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), 
          axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "right") +
  scale_color_manual(name="Treatment", 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4", 
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="High pCO2", 
                              "10_amb"="Warm Temperature", 
                              "10_low"="Warm Temperature + High pCO2", 
                              "3_amb"="Cold Temperature",
                              "3_low"="Cold Temperature + High pCO2")))

}
```

### Figure 3: Bubble plots of david enrichment results, but using the median fold change of genes contributing to those enriched processes for bubble size. 


```{r}
# Create lists of DEGs that were expressed higher and lower compared to control in each treatment 
DEGs.4.bubble.plot <- list(
  
  "OA_Downregulated" = 
    diffex.pH.6.filt %>% as.data.frame() %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    rownames_to_column("gene"),
  
  "OA_Upregulated" =  
    diffex.pH.6.filt %>% as.data.frame() %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    rownames_to_column("gene"),

  "Cold_Upregulated" = 
    diffex.temp.3.6.amb.filt %>% as.data.frame() %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    rownames_to_column("gene"),
  
  "Cold_Downregulated"  =
    diffex.temp.3.6.amb.filt %>% as.data.frame() %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    rownames_to_column("gene"),
  
  "Warm_Downregulated" = 
    diffex.temp.6.10.amb.filt %>% as.data.frame() %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    rownames_to_column("gene"),
  
  "Warm_Upregulated" = 
    diffex.temp.6.10.amb.filt %>% as.data.frame() %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    rownames_to_column("gene"),
  
  "Cold-OA_Upregulated" = 
    diffex.both.3low.6amb.filt %>% as.data.frame() %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    rownames_to_column("gene"),
  
  "Cold-OA_Downregulated" = 
    diffex.both.3low.6amb.filt %>% as.data.frame() %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    rownames_to_column("gene"),
  
  "Warm-OA_Upregulated" = 
    diffex.both.10low.6amb.filt %>% as.data.frame() %>%
    filter(dplyr::select(., contains("log2FoldChange"))<(-0.5)) %>% 
    rownames_to_column("gene"),
  
  "Warm-OA_Downregulated" = 
    diffex.both.10low.6amb.filt %>% as.data.frame() %>%
    filter(dplyr::select(., contains("log2FoldChange"))>0.5) %>%
    rownames_to_column("gene")
  ) 

# Select KW biological processes to plot 
processes.kw <- 
  david %>% 
  filter(p_value<0.05) %>%
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(stress, response, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david %>% 
              dplyr::select(stress, response, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("stress", "response", "genes", "p_value", "count")) %>%  #re-add data
  filter(process%!in%c("DNA integration", "DNA recombination")) %>%
  mutate(stress_response=paste(stress, response, sep="_"))

# Extract mean & median log2fc for all genes contributing to enriched processes
processes.kw.fc <- data.frame(stress = character(), response = character(), process = character(), mean.fc=numeric(), median.fc=numeric()) 
for (i in 1:nrow(processes.kw)) {
  process <- processes.kw[i,]$process
  spids <- processes.kw[i,]$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)
  genes <- (counts.annot.gadmor %>% filter(spid %in% spids))$gene_gadmor
  group <- processes.kw[i,]$stress_response
  dat <- DEGs.4.bubble.plot[[group]] %>% filter(gene %in% genes)
  processes.kw.fc[i,] <- c(str_split(group, "_")[[1]][1], str_split(group, "_")[[1]][2], process, abs(mean(dat$log2FoldChange)), abs(median(dat$log2FoldChange)))
}

# Make bubble plots using effect size, i.e. fold change (2^log2fc)
#pdf(file="../results/GO-enrichment/Final-DAVID/Upregulated-KW-effect-size-bubble.pdf", height = 3.05, width = 3.25)
fig4a <- processes.kw.fc %>% mutate(mean.fc=as.numeric(mean.fc), median.fc=as.numeric(median.fc)) %>%
  mutate(stress=factor(stress, levels=c("OA", "Warm", "Warm-OA", "Cold", "Cold-OA"), ordered = T)) %>%
  filter(response=="Upregulated") %>%
  ungroup() %>% arrange(stress, desc(mean.fc)) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=2^mean.fc)) +  #aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 #facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4", 
                              "Cold"=lighten("navyblue", 0.45),
                              "Cold-OA"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("Mean Effect Size\n(Expression)", range = c(1.5,7.5), breaks = c(1.5, 1.75, 2.0),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
#  theme_cleveland() + 
  theme_minimal() + 
  theme(
        legend.position = "right",  #plot.margin = margin(5, 5, 5, 60),
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Higher gene expression")
#dev.off()

#pdf(file="../results/GO-enrichment/Final-DAVID/Downregulated-KW-effect-size-bubble.pdf", height = 1.7, width = 2.15)
fig4b <- processes.kw.fc %>% mutate(mean.fc=as.numeric(mean.fc), median.fc=as.numeric(median.fc)) %>%
  mutate(stress=factor(stress, levels=c("OA", "Warm", "Warm-OA", "Cold", "Cold-OA"), ordered = T)) %>%
  filter(response=="Downregulated") %>%
  ungroup() %>% arrange(stress, desc(mean.fc)) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=2^mean.fc)) +  #aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 #facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4", 
                              "Cold"=lighten("navyblue", 0.45),
                              "Cold-OA"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("Mean Effect Size", range = c(1.5,7.5), breaks = c(1.5, 1.75, 2.0),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=F) + #Do drop empty factors for temperature contrasts
#  theme_cleveland() + 
  theme_minimal() + 
  theme(
        legend.position = "none",  #plot.margin = margin(5, 5, 5, 60),
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.title.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Lower gene expression")
#dev.off()

# Make figure 4 - combined plots of higher and lower expression
# Combine the plots vertically
# (fig4 <- (fig4a / fig4b / pheno_plot_fig3) + plot_layout(heights = c(2, 0.85, 1))) #only enrichment results
library(cowplot)

 (fig4 <- (fig4a / fig4b / fig4.pheno) + plot_layout(heights = c(1.75, 0.8, 0.8))) + #only enrichment results
  theme(plot.margin = margin(t = 0, r = 0, b = 5, l = 0)) + 
  expand_limits(y = -1) + 
  annotate("text", x=1.1, y=-0.1, label="OA", size=2.5, angle = -90) +
  annotate("text", x=2.1, y=-0.1, label="Warm", size=2.5, angle = -90) +
  annotate("text", x=3.1, y=-0.28, label="Warm+OA", size=2.5, angle = -90) +
  annotate("text", x=4.1, y=-0.1, label="Cold", size=2.5, angle = -90) +
  annotate("text", x=5.1, y=-0.26, label="Cold+OA", size=2.5, angle = -90)

# Wrap the plot to respect the extended margins
fig4_wrapped <- ggdraw(fig4) + 
  theme(plot.margin = margin(t = 0, r = 0, b = 35, l = 0))  # Extend bottom margin

# # Combine spacer and plot explicitly for each row
# row1 <- plot_spacer() + fig4a + plot_layout(widths = c(0, 1))  # Spacer and fig4a
# row2 <- plot_spacer() + fig4b + plot_layout(widths = c(0.13, 1))  # Spacer and fig4b
# row3 <- plot_spacer() + pheno_plot_fig3 + plot_layout(widths = c(0.37, 1))  # Spacer and fig4b
# 
# # Stack the rows vertically
# (fig4 <- row1 / row2 / row3 +
#          plot_layout(heights = c(2, 0.85, 1.5)))  # Adjust row heights 

# Save the combined plot as a PDF
ggsave(filename = "../manuscript/Fig4.pdf", fig4_wrapped, width = 4, height = 6.5)
```

Remake those bubble plots using Gene Ontology terms (for supplemental)

```{r}
# Select KW biological processes to plot 
processes.go <- 
  david %>% 
  filter(p_value<0.05) %>%
  filter(category=="GOTERM_BP_DIRECT") %>%

  group_by(stress, response, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david %>% 
              dplyr::select(stress, response, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("stress", "response", "genes", "p_value", "count")) %>%  #re-add data
  filter(process%!in%c("DNA integration", "DNA recombination")) %>%
  mutate(stress_response=paste(stress, response, sep="_"))

# Extract mean & median log2fc for all genes contributing to enriched processes
processes.go.fc <- data.frame(stress = character(), response = character(), process = character(), mean.fc=numeric(), median.fc=numeric()) 
for (i in 1:nrow(processes.go)) {
  process <- processes.go[i,]$process
  spids <- processes.go[i,]$genes %>% strsplit(., ",") %>% unlist() %>% as.vector() %>% gsub(" ", "", .)
  genes <- (counts.annot.gadmor %>% filter(spid %in% spids))$gene_gadmor
  group <- processes.go[i,]$stress_response
  dat <- DEGs.4.bubble.plot[[group]] %>% filter(gene %in% genes)
  processes.go.fc[i,] <- c(str_split(group, "_")[[1]][1], str_split(group, "_")[[1]][2], process, abs(mean(dat$log2FoldChange)), abs(median(dat$log2FoldChange)))
}

# Make bubble plots using effect size, i.e. fold change (2^log2fc)
pdf(file="../results/GO-enrichment/Final-DAVID/Upregulated-GO-effect-size-bubble.pdf", height = 10, width = 6.75)
processes.go.fc %>% mutate(mean.fc=as.numeric(mean.fc), median.fc=as.numeric(median.fc)) %>%
  mutate(stress=factor(stress, levels=c("OA", "Warm", "Warm-OA", "Cold", "Cold-OA"), ordered = T)) %>%
  filter(response=="Upregulated") %>%
  ungroup() %>% arrange(stress, desc(mean.fc)) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=2^mean.fc)) +  #aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 #facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4", 
                              "Cold"=lighten("navyblue", 0.45),
                              "Cold-OA"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("Mean Effect Size", range = c(1.5,7.5), breaks = c(1.5, 1.75, 2.0),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Higher gene expression")
dev.off()

pdf(file="../results/GO-enrichment/Final-DAVID/Downregulated-GO-effect-size-bubble.pdf", height = 3.5, width = 4.75)
processes.go.fc %>% mutate(mean.fc=as.numeric(mean.fc), median.fc=as.numeric(median.fc)) %>%
  mutate(stress=factor(stress, levels=c("OA", "Warm", "Warm-OA", "Cold", "Cold-OA"), ordered = T)) %>%
  filter(response=="Downregulated") %>%
  ungroup() %>% arrange(stress, desc(mean.fc)) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=2^mean.fc)) +  #aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 #facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Stressor",
                     values=c("OA"=lighten("darkorange2", 0.25), 
                              "Warm"=lighten("firebrick4", 0.45), 
                              "Warm-OA"="firebrick4", 
                              "Cold"=lighten("navyblue", 0.45),
                              "Cold-OA"="navyblue"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("Mean Effect Size", range = c(1.5,7.5), breaks = c(1.5, 1.75, 2.0),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=F) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Lower gene expression")
dev.off()
```















Plot genes of interest relating to enzyme analysis 

```{r, warning=FALSE}

enzymes <- 
#gadMor.blast.GO %>% filter(str_detect(protein_names, "itrate synthase")) %>% filter(gene_gadmor %in% rownames(res.pH.3)) #Citrate Synthase
gadMor.blast.GO %>% filter(str_detect(protein_names, "hydroxyacyl-CoA dehydrogenase")) %>% filter(gene_gadmor %in% rownames(res.pH.3)) #HOAD
#gadMor.blast.GO %>% filter(str_detect(protein_names, "lactate dehydrogenase")) %>% filter(gene_gadmor %in% rownames(res.pH.3)) #Lactate Dehydrogenase
#gadMor.blast.GO %>% filter(str_detect(protein_names, "yruvate kinase")) %>% filter(gene_gadmor %in% rownames(res.pH.3)) #Pyruvate kinase

#enzymes %>% select(gene_gadmor, spid, gene_uni, protein_names) %>% write_clip()

# Are enzymes of interest to RNASeq analyses?
#results.df %>% filter(gene_gadmor %in% enzymes$gene_gadmor) %>% View()

for (i in 1:nrow(enzymes)) {
a <-  plotCounts(dds.DESeq.treatment, gene=enzymes$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #mutate(treatment=factor(treatment, ordered = T, levels = c("3_amb", "3_low", "6_amb", "6_low", "")))
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))

# plot 
print(ggplot(a,
       aes(x=interaction(temperature:ph), y=count, color=interaction(temperature:ph))) +
  geom_boxplot(aes(fill=interaction(temperature:ph)), outlier.shape = NA, alpha=0.35) + 
  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle(str_wrap(enzymes$protein_names[i])) +
    theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
          plot.title = element_text(hjust = 0.5, size=11), legend.position = "none") +
  scale_color_manual(values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="navyblue","6:Ambient pCO2"="darkgreen",
                               "6:High pCO2"="darkgreen","10:Ambient pCO2"="firebrick4", "10:High pCO2"="firebrick4")) +
  scale_fill_manual(name="pCO2", values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="white","6:Ambient pCO2"="darkgreen",
                               "6:High pCO2"="white","10:Ambient pCO2"="firebrick4", "10:High pCO2"="white")))

# # Facet wrap by ph
# print(ggplot(a,
#        aes(x=temperature, y=count, color=temperature, label=sample)) +
#   geom_boxplot(outlier.shape = NA) + facet_wrap(~ph) +
#   geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
#     #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
#     theme_bw() +
#     ggtitle(str_wrap(enzymes$protein_names[i])) +
#     theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=14), legend.position = "none") +
#     scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"))
#   # scale_color_manual(values=c("3:amb"= "navyblue", "3:low"="royalblue1","6:amb"="darkgreen",
#   #                             "6:low"="yellow3","10:amb"="firebrick4", "10:low"="sienna2"))
#   )
}
```
Enzyme ratios 

LDH:CS = anaerobic:aerobic contributions 
HOAD:CS = fatty acid contribution towards aerobic metabolism  

```{r}
enzyme.genes <- c("ldhd", "cs", "hadh") #old list before controlling for fish length
#enzyme.genes <- c("cs", "hsd17b10") #new list, no ldhd and new HOAD

enzyme.genes.counts <- vector("list", length(enzyme.genes))
names(enzyme.genes.counts) <- enzyme.genes

# # Were any of the genes tossed out b/c they correlated with fish length? 
# intersect(genes.length, enzyme.genes) #yes, both hadh and ldhd

# Should try to plot gene expression ~ fish length those genes 

for (i in 1:length(enzyme.genes.counts)) {
enzyme.genes.counts[[i]] <-  plotCounts(
#  dds.DESeq.treatment,  #use this DESeq2 object to use counts for genes included in final analysis
  dds.DESeq.treatment.all,  #use this DESeq2 object to use ALL counts (not filtered to remove length-associated ones)
  gene=enzyme.genes[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #filter(treatment %in% b) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))
}

enzyme.ratios <- bind_rows(enzyme.genes.counts, .id = "gene_gadmor") %>% 
  mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
  pivot_wider(names_from=gene_gadmor, values_from=count) %>%
#  mutate(`LDH:CS`=ldhd/cs, `HOAD:CS`=hadh/cs)
  mutate(`HOAD:CS`=hadh/cs)
#  mutate(`HOAD:CS`=hsd17b10/cs)

# # Examine LDH:CS ratio 
# ggplot(enzyme.ratios,
#        aes(x=interaction(temperature:ph), y=`LDH:CS`, color=interaction(temperature:ph))) +
#   geom_boxplot(aes(fill=interaction(temperature:ph)), outlier.shape = NA, alpha=0.35) + 
#   geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
#     #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
#     theme_bw() +
#     ggtitle("LDH:CS Ratio") +
#     theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
#           plot.title = element_text(hjust = 0.5, size=13), legend.position = "none") +
#   scale_color_manual(values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="navyblue","6:Ambient pCO2"="darkgreen",
#                                "6:High pCO2"="darkgreen","10:Ambient pCO2"="firebrick4", "10:High pCO2"="firebrick4")) +
#   scale_fill_manual(name="pCO2", values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="white","6:Ambient pCO2"="darkgreen",
#                                "6:High pCO2"="white","10:Ambient pCO2"="firebrick4", "10:High pCO2"="white"))
# hist(log(enzyme.ratios$`LDH:CS`))
# shapiro.test(log(enzyme.ratios$`LDH:CS`))
# summary(aov(log(`LDH:CS`)~temperature*ph, data=enzyme.ratios)) # Temperature affects LDH:CS ratio 
# TukeyHSD(aov(log(`LDH:CS`)~temperature*ph, data=enzyme.ratios)) # All temperatures differ from each other 

# Examine HOAD:CS ratio
ggplot(enzyme.ratios,
       aes(x=interaction(temperature:ph), y=`HOAD:CS`, color=interaction(temperature:ph))) +
  geom_boxplot(aes(fill=interaction(temperature:ph)), outlier.shape = NA, alpha=0.35) + 
  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle("HOAD:CS Ratio") +
    theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
          plot.title = element_text(hjust = 0.5, size=13), legend.position = "none") +
  scale_color_manual(values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="navyblue","6:Ambient pCO2"="darkgreen",
                               "6:High pCO2"="darkgreen","10:Ambient pCO2"="firebrick4", "10:High pCO2"="firebrick4")) +
  scale_fill_manual(name="pCO2", values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="white","6:Ambient pCO2"="darkgreen",
                               "6:High pCO2"="white","10:Ambient pCO2"="firebrick4", "10:High pCO2"="white"))
hist(log(enzyme.ratios$`HOAD:CS`))
shapiro.test(enzyme.ratios$`HOAD:CS`)
summary(aov(`HOAD:CS`~temperature*ph, data=enzyme.ratios)) # Temperature affects HOAD:CS ratio 
TukeyHSD(aov(`HOAD:CS`~temperature*ph, data=enzyme.ratios)) # Lower HOAD:CS ratio at 10C compared to other temperatures

# Stats on enzymes using ANOVA

hist(log(enzyme.ratios$cs))
shapiro.test(enzyme.ratios$cs)
summary(aov(cs~temperature*ph, data=enzyme.ratios)) # CS affected by temperature and temperature:ph interaction
TukeyHSD(aov(cs~temperature*ph, data=enzyme.ratios)) # CS differs among all temperatures
```

## Examine correlation between fish length and gene counts for genes of interest

```{r}
#HOAD is length-associated. Let's look at HOAD expression ~ fish length
HOAD <- enzyme.genes.counts$hadh %>% left_join(counts.tk %>% rownames_to_column(var="sample") %>% select(sample, length))
cor.test(x=HOAD$length, y=HOAD$count, method = "pearson")

HOAD %>%
  ggplot(aes(x=length, y=count, colour=treatment)) + 
  geom_point(size=2) + 
  theme_minimal() + 
  scale_color_manual(name="Treatment",
                 values=c("6_amb"="gray65",
                          "6_low"=lighten("darkorange2", 0.25),
                          "10_amb"=lighten("firebrick4", 0.45),
                          "10_low"="firebrick4",
                          "3_amb"=lighten("navyblue", 0.45),
                          "3_low"="navyblue"),
                 labels=c("6_amb"="Control",
                          "6_low"="High pCO2",
                          "10_amb"="Warm Temperature",
                          "10_low"="Warm Temperature + High pCO2",
                          "3_amb"="Cold Temperature",
                          "3_low"="Cold Temperature + High pCO2")) +
    geom_smooth(aes(x=length, y=count, colour=treatment), method="lm", se=F) + 
  ggtitle("HOAD expression ~ larval length")

#CS is not length-associated. Let's look at CS expression ~ fish length
cs <- enzyme.genes.counts$cs %>% left_join(counts.tk %>% rownames_to_column(var="sample") %>% select(sample, length))
cor.test(x=cs$length, y=cs$count, method = "pearson")

cs %>%
  ggplot(aes(x=length, y=count, colour=treatment)) + geom_point(size=2) + 
  theme_minimal() + 
  scale_color_manual(name="Treatment",
                 values=c("6_amb"="gray65",
                          "6_low"=lighten("darkorange2", 0.25),
                          "10_amb"=lighten("firebrick4", 0.45),
                          "10_low"="firebrick4",
                          "3_amb"=lighten("navyblue", 0.45),
                          "3_low"="navyblue"),
                 labels=c("6_amb"="Control",
                          "6_low"="High pCO2",
                          "10_amb"="Warm Temperature",
                          "10_low"="Warm Temperature + High pCO2",
                          "3_amb"="Cold Temperature",
                          "3_low"="Cold Temperature + High pCO2")) +
    geom_smooth(aes(x=length, y=count, colour=treatment), method="lm", se=F) + 
  ggtitle("CS expression ~ larval length")
```


### Alternative enzyme-related gene expression analysis 
- Use GO terms, Uniprot Keywords,  and/or KEGG Pathway to identify genes of interest and look at their expression 
-- Citrate synthase: 
--- citrate synthase activity (GO:0036440, Molecular Function)
--- tricarboxylic acid cycle (GO:0006099, Biological Process)
--- aerobic respiration (GO:0009060, Biological Process)
--- anaerobic respiration (GO:0009061, Biological Process)
--- Lipid metabolic process (GO:0006629, Biological Process) 
--- Lipid biosynthetic process (GO:0008610, Biological Process) 
--- Lipid catabolic process (GO:0016042, Biological Process)

```{r, warning=FALSE, message=FALSE}
go <- 
 "GO:0036440" #citrate synthase activity (GO term molecular function)
# "GO:0006099" #tricarboxylic acid cycle (GO term biological process)
# "GO:0009061" # Anaerobic respiration (GO term Biological Process)
# "GO:0009060" # Aerobic respiration (GO term biological process)
#"GO:0008610" #Lipid biosynthetic process
# "GO:0016042" #Lipid catabolic process

go.genes <- results.df %>% left_join(
  gadMor.blast.GO %>% select(spid, gene_ontology_biological_process, gene_ontology_cellular_component, gene_ontology_molecular_function, gene_ontology_i_ds),
  by="spid") %>% 
  filter(grepl(go, gene_ontology_i_ds)) %>% 
  filter(deg.ph.6=="yes" | deg.temp.3.6.amb=="yes" | deg.temp.6.10.amb=="yes" | #deg.temp.3.10.low=="yes" | deg.temp.3.10.amb=="yes" | 
         deg.both.3low.6amb=="yes" | deg.both.10low.6amb=="yes")

# # to paste into Uniprot 
# test$spid %>% unique() %>% write_clip()

for (i in 1:nrow(go.genes)) {
a <-  plotCounts(dds.DESeq.treatment, gene=go.genes$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #filter(treatment %in% b) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))

print(
  ggplot(a,
       aes(x=temperature, y=count, color=temperature, label=sample)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(aes(shape=ph), size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle(str_wrap(go.genes$protein_names[i])) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) + 
  facet_wrap(~ph))
}

# One plot, points or lines for each gene 

go.genes.lines <- vector("list", nrow(go.genes))
names(go.genes.lines) <- go.genes$gene_gadmor

for (i in 1:nrow(test)) {
go.genes.lines[[i]] <-  plotCounts(dds.DESeq.treatment, gene=go.genes$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))
}

bind_rows(go.genes.lines, .id = "gene_gadmor") %>% 
  mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
  group_by(gene_gadmor, temperature, ph) %>%
  dplyr::summarize(mean=mean(count), median=median(count), sd=sd(count), cv=sd(count)/mean(count)) %>%
  group_by(gene_gadmor, temperature, ph) %>% mutate(z_score=scale(mean)) %>%  # figure out how to standardize value across genes (opposed to z-score)

  # # One boxplot, points for all genes
  # ggplot(aes(x=temperature, y=z_score, color=temperature)) +
  # geom_boxplot(outlier.shape = NA) +
  # geom_point(aes(shape=ph), size=2.5, position=position_jitter(w = 0.15,h = 0)) +
  #   #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
  #   theme_bw() +
  #   theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  # scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) +
  # facet_wrap(~ph)

  # Line plots 
  ggplot(aes(x=temperature, y=z_score, group=gene_gadmor)) + geom_line(size=.01) +
  theme_minimal() + facet_wrap(~ph) +
  #scale_x_continuous(breaks=c(1,2,3), labels=c("Ambient", "Moderate", "Low")) +
  theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  xlab("Per treatment mean of normalized counts") + ggtitle("Abundances of DEGs by treatment & relative expression")


```

Is the zona pellucida gene in data? 
Yes! 

```{r}
counts.tk %>% select(LOC115551355) #is this gene in our dataset? 
intersect(genes.length, "LOC115551355") #is it length-associated? yes, so it was removed from DEG analysis

# Using variance stabilization transformed counts 
zp3.vst <- assay(vsd.treatment.all) %>% t() %>% as.data.frame() %>%
  # rownames_to_column("gene_gadmor") %>%
  # filter(gene_gadmor == "LOC115551355") 
  select(LOC115551355) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "treatment")], by=c("sample"="vial_label"))

View(zp3.vst)

(aov.2way <- summary(aov(LOC115551355 ~ temperature*ph, data=zp3.vst)))
aov.2way[[1]] %>% as.data.frame() %>% rownames_to_column("treatment") %>%
  clean_names() %>%
  mutate(p.adj.bonf=((nrow(assay(vsd.treatment))+1)*pr_f))

(tukey <- TukeyHSD(aov(LOC115551355 ~ temperature, data=zp3.vst)))
tukey$`temperature` %>% as.data.frame() %>% rownames_to_column("contrast") %>% arrange(`p adj`) %>%
  clean_names() %>%
  mutate(p.adj.bonf=((nrow(assay(vsd.treatment))+1)*p_adj))

#(fig1 <- 
zp3.vst %>% mutate(treatment=factor(treatment, ordered=T, 
                     levels=c("3_amb", "3_low", "6_amb", "6_low", "10_amb", "10_low"))) %>%
ggplot(aes(x=treatment, y=LOC115551355, color=treatment)) +
  geom_boxplot(outlier.shape = NA, lwd=0.2, alpha=0.75) +
  geom_point(aes(x=treatment, color=treatment), 
             size=1.5, position=position_jitter(w = 0.2,h = 0)) +
    theme_bw() +
    # ggtitle(paste(str_wrap(zona$protein_names[i]), "\n", 
    #               zona$gene_names[i], " - ", zona$species[i], " - ", zona$seq_id[i], sep="")) +
    #ggtitle("Zona pellucida sperm-binding protein 3\n(ZP3, LOC115551355 in GadMor3.0)") + 
  ylab("Transformed counts") + xlab("Treatment") +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_blank(), 
          axis.ticks.x = element_blank(),
          axis.title.y = element_text(size=9),
          plot.title = element_text(hjust = 0.5, size=10), 
          legend.position = "right",
          legend.title = element_text(size=9)) +
  # scale_fill_manual(name="Treatment",
  #                values=c("6_amb"="gray65",
  #                         "6_low"=lighten("darkorange2", 0.25),
  #                         "10_amb"=lighten("firebrick4", 0.45),
  #                         "10_low"="firebrick4",
  #                         "3_amb"=lighten("navyblue", 0.45),
  #                         "3_low"="navyblue"),
  #                labels=c("6_amb"="Control",
  #                         "6_low"="High pCO2",
  #                         "10_amb"="Warm Temperature",
  #                         "10_low"="Warm Temperature + High pCO2",
  #                         "3_amb"="Cold Temperature",
  #                         "3_low"="Cold Temperature + High pCO2")) +
  scale_color_manual(name="Treatment",
                 values=c("6_amb"="gray65",
                          "6_low"=lighten("darkorange2", 0.25),
                          "10_amb"=lighten("firebrick4", 0.45),
                          "10_low"="firebrick4",
                          "3_amb"=lighten("navyblue", 0.45),
                          "3_low"="navyblue"),
                 labels=c("6_amb"="Control",
                          "6_low"="High pCO2",
                          "10_amb"="Warm Temperature",
                          "10_low"="Warm Temperature + High pCO2",
                          "3_amb"="Cold Temperature",
                          "3_low"="Cold Temperature + High pCO2")) +
  annotate("text", x=1.5, y=7.5, label="A", size=5) +
  annotate("text", x=3.5, y=7, label="B", size=5) +
  annotate("text", x=5.5, y=7, label="B", size=5)

#fig2 <- 
zp3.vst %>% mutate(ph=case_when(ph=="amb" ~ "Control", TRUE~"High pCO2")) %>%
ggplot(aes(x=as.numeric(as.character(temperature)), y=LOC115551355)) +
    geom_point(aes(color=temperature), size=1.5, position=position_jitter(w = 0.1,h = 0)) +
    theme_bw() +
    #ggtitle("Zona pellucida sperm-binding protein 3\n(ZP3, LOC115551355 in GadMor3.0)") + 
  ylab("Transformed counts") + xlab("Temperature") +
    theme(axis.title.y = element_text(size=9),
          plot.title = element_text(hjust = 0.5, size=10), 
          legend.position = "none",
          legend.title = element_text(size=9)) +
    geom_smooth(method="lm", formula=y~x, alpha=0.2, colour="gray25") +
    stat_regline_equation(label.x = 3, label.y = 7.7) +
    stat_cor(label.x = 3, label.y = 7.55) +
  scale_color_manual(name="Temperature", values=c("navyblue", "gray35", "firebrick4")) + 
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"), order=1)) +
  scale_x_continuous(breaks = c(3,6,10), labels=c(3,6,10)) +
  facet_wrap(~ph)

#require(gridExtra)
library(cowplot)
plot_grid(fig1, fig2, nrow=1, rel_widths=c(3/5, 2/5))
```

```{r}
# ZP3 ANALYSIS using raw count data
summary(aov(LOC115551355 ~ temperature*ph, data=counts.tk))
(tukey <- TukeyHSD(aov(LOC115551355 ~ temperature, data=counts.tk)))
tukey$`temperature` %>% as.data.frame() %>% rownames_to_column("contrast") %>% arrange(`p adj`) %>%
  clean_names() %>%
  mutate(p.adj.bonf=((nrow(assay(vsd.treatment))+1)*p_adj))

(3.07e-12)*(nrow(assay(vsd.treatment))+1)
```
Figure 3

```{r}
pdf(file="../manuscript/Fig3.pdf", width = 6, height = 3)
counts.tk %>% mutate(treatment=factor(treatment, ordered=T,
                     levels=c("3_amb", "3_low", "6_amb", "6_low", "10_amb", "10_low"))) %>% #select(LOC115551355) %>%
ggplot(aes(x=treatment, y=LOC115551355, fill=treatment)) +
  geom_boxplot(outlier.shape = NA, lwd=0.2, alpha=0.75) +
  geom_point(aes(x=treatment, color=treatment),
             size=1.5, position=position_jitter(w = 0.2,h = 0)) +
    theme_bw() +
    # ggtitle(paste(str_wrap(zona$protein_names[i]), "\n",
    #               zona$gene_names[i], " - ", zona$species[i], " - ", zona$seq_id[i], sep="")) +
    ggtitle("Zona pellucida sperm-binding protein 3\n(ZP3, LOC115551355 in GadMor3.0)") + ylab("Gene counts (raw)") +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          axis.title.y = element_text(size=9),
          plot.title = element_text(hjust = 0.5, size=8),
          legend.position = "right",
          legend.title = element_text(size=8), legend.text = element_text(size=8)) +
  scale_fill_manual(name="Treatment",
                 values=c("6_amb"="gray65",
                          "6_low"=lighten("darkorange2", 0.15),
                          "10_amb"=lighten("firebrick4", 0.60),
                          "10_low"=lighten("firebrick4", 0.15),
                          "3_amb"=lighten("navyblue", 0.60),
                          "3_low"=lighten("navyblue", 0.15)),
                 labels=c("6_amb"="Control",
                          "6_low"="High pCO2",
                          "10_amb"="Warm Temperature",
                          "10_low"="Warm Temperature + High pCO2",
                          "3_amb"="Cold Temperature",
                          "3_low"="Cold Temperature + High pCO2")) +
  scale_color_manual(name="Treatment",
                 values=c("6_amb"="gray65",
                          "6_low"=lighten("darkorange2", 0.25),
                          "10_amb"=lighten("firebrick4", 0.45),
                          "10_low"="firebrick4",
                          "3_amb"=lighten("navyblue", 0.45),
                          "3_low"="navyblue"),
                 labels=c("6_amb"="Control",
                          "6_low"="High pCO2",
                          "10_amb"="Warm Temperature",
                          "10_low"="Warm Temperature + High pCO2",
                          "3_amb"="Cold Temperature",
                          "3_low"="Cold Temperature + High pCO2")) +
  annotate("text", x=1.5, y=65, label="A", size=4) +
  annotate("text", x=3.5, y=30, label="B", size=4) +
  annotate("text", x=5.5, y=30, label="B", size=4)
dev.off()
```


Plot other genes of interest relating to zona pellucida that were in our analysis

```{r, warning=FALSE}
zona <-
gadMor.blast.GO %>% filter(str_detect(protein_names, "ona pellucida")) %>% filter(gene_gadmor %in% rownames(res.pH.3)) %>% filter(seq_id == "NC_044056.1")

# Are zona pellucida genes of interest to RNASeq analyses?
results.df %>% filter(gene_gadmor %in% zona$gene_gadmor) %>% View()

# Yes- differences are between the following:
# 6-amb vs 10-amb
# 6-low vs 10-low
# 6-amb vs 10-low
# 3-amb vs 10-amb

# # Was the zona pellucida gene tossed out b/c it correlated with fish length? 
intersect(genes.length, zona$gene_gadmor) #ZP3 genes in DEG analysis - no 

for (i in 1:nrow(zona)) {
#a <-  plotCounts(dds.DESeq.treatment, gene=zona$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  a <-  plotCounts(dds.DESeq.treatment, gene=zona$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  mutate(treatment=factor(treatment, ordered = T, levels = c("6_amb", "6_low", "3_amb", "3_low", "10_amb", "10_low")))

summary(aov(count ~ temperature*ph, data=a))
(tukey <- TukeyHSD(aov(count ~ temperature, data=a)))
tukey$`temperature` %>% as.data.frame() %>% rownames_to_column("contrast") %>% arrange(`p adj`)

print(ggplot(a,
       aes(x=treatment, y=count, fill=treatment)) +
  geom_boxplot(outlier.shape = NA, lwd=0.2, alpha=0.75) +
  geom_point(aes(x=treatment, color=treatment), 
             size=1.5, position=position_jitter(w = 0.2,h = 0)) +
    theme_bw() +
    # ggtitle(paste(str_wrap(zona$protein_names[i]), "\n", 
    #               zona$gene_names[i], " - ", zona$species[i], " - ", zona$seq_id[i], sep="")) +
    ggtitle("Zona pellucida sperm-binding protein 3 (ZP3)") + 
    theme(axis.title.x = element_blank(), 
          axis.text.x = element_blank(), 
          axis.ticks.x = element_blank(),
          axis.title.y = element_text(size=9),
          plot.title = element_text(hjust = 0.5, size=10), 
          legend.position = "right",
          legend.title = element_text(size=9)) +
  scale_fill_manual(name="Treatment",
                 values=c("6_amb"="gray65",
                          "6_low"=lighten("darkorange2", 0.25),
                          "10_amb"=lighten("firebrick4", 0.45),
                          "10_low"="firebrick4",
                          "3_amb"=lighten("navyblue", 0.45),
                          "3_low"="navyblue"),
                 labels=c("6_amb"="Control",
                          "6_low"="High pCO2",
                          "10_amb"="Warm Temperature",
                          "10_low"="Warm Temperature + High pCO2",
                          "3_amb"="Cold Temperature",
                          "3_low"="Cold Temperature + High pCO2")) +
  scale_color_manual(name="Treatment",
                 values=c("6_amb"="gray65",
                          "6_low"=lighten("darkorange2", 0.25),
                          "10_amb"=lighten("firebrick4", 0.45),
                          "10_low"="firebrick4",
                          "3_amb"=lighten("navyblue", 0.45),
                          "3_low"="navyblue"),
                 labels=c("6_amb"="Control",
                          "6_low"="High pCO2",
                          "10_amb"="Warm Temperature",
                          "10_low"="Warm Temperature + High pCO2",
                          "3_amb"="Cold Temperature",
                          "3_low"="Cold Temperature + High pCO2")))
}

#Zona pellucida 3 is not length-associated. Let's look at zp3 expression ~ fish length
zona.count <- a %>% left_join(counts.tk %>% rownames_to_column(var="sample") %>% select(sample, length))
cor.test(x=zona.count$length, y=zona.count$count, method = "pearson")

zona.count %>%
  ggplot(aes(x=length, y=count, colour=treatment)) + geom_point(size=2) + 
  theme_minimal() + 
  scale_color_manual(name="Treatment",
                 values=c("6_amb"="gray65",
                          "6_low"=lighten("darkorange2", 0.25),
                          "10_amb"=lighten("firebrick4", 0.4),
                          "10_low"="firebrick4",
                          "3_amb"=lighten("navyblue", 0.4),
                          "3_low"="navyblue"),
                 labels=c("6_amb"="Control",
                          "6_low"="High pCO2",
                          "10_amb"="Warm Temperature",
                          "10_low"="Warm Temperature + High pCO2",
                          "3_amb"="Cold Temperature",
                          "3_low"="Cold Temperature + High pCO2")) +
    geom_smooth(aes(x=length, y=count, colour=treatment), method="lm", se=F) + 
  ggtitle("zp3 expression ~ larval length")
```

## Plot interesting genes using Uniprot IDs - each gene separately

```{r, warning=FALSE, message=F}
keyword <- "Biological rhythms"
keyword <- "Vision"
keyword <- "Stress response"
keyword <- "Immunity"
keyword <- "DNA integration"
keyword <- "Urea cycle"
keyword <- "Tricarboxylic acid cycle"
keyword <- "Lipid metabolism"
keyword <- "Cell adhesion"
keyword <- "Digestion"
keyword <- "Hemostasis"

uniprots <- ((enriched.kw.degs.upregulated %>% filter(kw==keyword))$genes)
#uniprots <- c("Q71U07", "Q2KIG3", "P02679", "P14448")
uniprots.vect <- vector(mode="character")

for (i in 1:length(uniprots)) {
  x <- unlist(strsplit(uniprots[i], ", "))
  uniprots.vect <- append(uniprots.vect, x)
}

# Plot all genes in data that match the Uniprot SPIDs of interest
test <- counts.annot.gadmor %>%
  filter(spid %in% unique(uniprots.vect))

# plot only the genes that were DEGs and match the Uniprot SPIDs of interest
test <- results.df %>% 
  # filter(deg.ph.6=="yes" | deg.temp.3.6.amb=="yes" | deg.temp.6.10.amb=="yes" | #deg.temp.3.10.low=="yes" | deg.temp.3.10.amb=="yes" | 
  #        deg.temp.3.6.low=="yes" | deg.temp.6.10.low=="yes") %>%
  filter(deg.ph.6=="yes" |  deg.both.10low.6amb  =="yes") %>%
  filter(spid %in% unique(uniprots.vect))


# plot only the genes that were DEGs and are hemoglobins
test <- results.df %>% 
  filter(deg.ph.6=="yes" | deg.temp.3.6.amb=="yes" | deg.temp.6.10.amb=="yes" | #deg.temp.3.10.low=="yes" | deg.temp.3.10.amb=="yes" | 
         deg.temp.3.6.low=="yes" | deg.temp.6.10.low=="yes") %>%
  filter(grepl("Hemoglobin subunit", protein_names))

# Plot all opsin genes 
test <- results.df %>% 
  # filter(deg.ph.6=="yes" | deg.temp.3.6.amb=="yes" | deg.temp.6.10.amb=="yes" | #deg.temp.3.10.low=="yes" | deg.temp.3.10.amb=="yes" | 
  #        deg.temp.3.6.low=="yes" | deg.temp.6.10.low=="yes") %>%
  filter(gene_gadmor %in% rownames(res.pH.6)) %>%
  filter(grepl("opsin", protein_names))
#  filter(grepl("NR2E3", gene_uni))


# to paste into Uniprot 
#test$spid %>% unique() %>% write_clip()

# Boxplot of gene expression by treatment 
for (i in 1:nrow(test)) {
a <-  plotCounts(dds.DESeq.treatment, gene=test$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #filter(treatment %in% b) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "3_low", "6_amb", "6_low",  "10_amb", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))

print(
  ggplot(a,
       aes(x=treatment, y=count, color=treatment, label=sample)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(aes(shape=ph), size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle(str_wrap(test$protein_names[i])) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  scale_color_manual(name="Treatment", 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4", 
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="High pCO2", 
                              "10_amb"="Warm Temperature", 
                              "10_low"="Warm Temperature + High pCO2", 
                              "3_amb"="Cold Temperature",
                              "3_low"="Cold Temperature + High pCO2")))
}

# Scatter plot of gene expression by fish length 

for (i in 1:nrow(test)) {
a <-  plotCounts(dds.DESeq.treatment, gene=test$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank", "length")], by=c("sample"="vial_label")) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))

#cor.length <- cor.test(x=filter(a, temperature!=10)$length, y=filter(a, temperature!=10)$count, method = "pearson") 
cor.length <- cor.test(x=a$length, y=a$count, method = "pearson")

print(
  ggplot(a 
         #%>% filter(temperature!=10)
         ) +
  geom_point(aes(x=length, y=log(count), color=treatment), size=2.5) +
    theme_bw() +
    ggtitle(str_wrap(test$protein_names[i])) +
    theme(plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  scale_color_manual(name="Treatment", 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4", 
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="High pCO2", 
                              "10_amb"="Warm Temperature", 
                              "10_low"="Warm Temperature + High pCO2", 
                              "3_amb"="Cold Temperature",
                              "3_low"="Cold Temperature + High pCO2")) +
    geom_smooth(data=filter(a, temperature==3), method="lm", aes(x=length, y=count, color=temperature), 
                se=T, color="navyblue", fill="navyblue", alpha=0.25) + 
    geom_smooth(data=filter(a, temperature==6), method="lm", aes(x=length, y=count, color=temperature), 
                se=T, color="darkorange2", fill="darkorange2", alpha=0.25) + 
    geom_smooth(data=filter(a, temperature==10), method="lm", aes(x=length, y=count, color=temperature),
                    se=T, color="firebrick4", fill="firebrick4", alpha=0.25) +
    annotate("text", x=max(a$length)*.98, y=max(a$count)*.95, 
             label=paste("Corr=", round(as.vector(cor.length$estimate),digits = 2), 
                 "\nP-value=", signif(as.vector(cor.length$p.value), digits=2))))
}

```


## Plot "reaction norms" - i.e. mean 

```{r}
test1 <- vector("list", nrow(test))
names(test1) <- test$gene_gadmor

for (i in 1:nrow(test)) {
test1[[i]] <-  plotCounts(dds.DESeq.treatment, gene=test$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #filter(treatment %in% b) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))
}

bind_rows(test1, .id = "gene_gadmor") %>% 
  mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
  group_by(gene_gadmor, temperature, ph) %>%
  dplyr::summarize(mean=mean(count), median=median(count), sd=sd(count), cv=sd(count)/mean(count)) %>%
  group_by(gene_gadmor, temperature, ph) %>% mutate(z_score=scale(mean)) %>%

  # # Plot points for all genes
  # ggplot(aes(x=temperature, y=mean, color=temperature)) +
  # geom_boxplot(outlier.shape = NA) +
  # geom_point(aes(shape=ph), size=2.5, position=position_jitter(w = 0.15,h = 0)) +
  #   #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
  #   theme_bw() +
  #   theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  # scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) +
  # facet_wrap(~ph)

  #Rection norm - too hard to see
  ggplot(aes(x=temperature, y=z_score, group=gene_gadmor)) + geom_line(size=.01) +
  theme_minimal() + facet_wrap(~ph) +
  #scale_x_continuous(breaks=c(1,2,3), labels=c("Ambient", "Moderate", "Low")) +
  theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  xlab("Per treatment mean of normalized counts") + ggtitle("Abundances of DEGs by treatment & relative expression")
```

### Search for Uniprot Keywords and GO Terms that are commonly affected, or uniquely affected, by temperature & pCO2 treatments

```{r}
temp <- enriched.kw.degs.upregulated %>% filter(contrast %!in% c("3vs10.Low-pH", "3vs10.Amb-pH")) %>% filter(!is.na(kw)) %>% filter(background!="Amb-pH")

# Uniprot Keywords that are commonly affected by temperature at high and ambient pCO2, and by pCO2
temp[temp$kw %in% 
  temp$kw[duplicated(temp$kw)],] %>%
  arrange(kw) %>% View()

# Uniprot Keywords that are uniquely affected by temperature at high and ambient pCO2, or by pCO2
temp %>% filter(kw %!in% 
                  (temp[temp$kw %in% temp$kw[duplicated(temp$kw)],])$kw) %>% 
  arrange(contrast) %>% View()

```

