---
title: "Results and Figures"
author: "Laura Spencer"
date: "2023-01-04"
output: html_document
---

```{r}
require(tidyverse)
require(eulerr)
require(wordcloud2)
```

Create a dataframe with all examined genes, identifying those of interest 

```{r}
results.df <- counts.annot.gadmor %>%
  select(-contains("PCG")) %>% # remove counts from dataframe 
  mutate(
    deg.ph.3=case_when(gene_gadmor %in% rownames(diffex.pH.3.filt) ~ "yes", TRUE ~ "no"),
    deg.ph.6=case_when(gene_gadmor %in% rownames(diffex.pH.6.filt) ~ "yes", TRUE ~ "no"),
    deg.ph.10=case_when(gene_gadmor %in% rownames(diffex.pH.10.filt) ~ "yes", TRUE ~ "no"),
    deg.temp.3.6.amb=case_when(gene_gadmor %in% rownames(diffex.temp.3.6.amb.filt) ~ "yes", TRUE ~ "no"),
    #deg.temp.3.10.amb=case_when(gene_gadmor %in% rownames(diffex.temp.3.10.amb.filt) ~ "yes", TRUE ~ "no"),
    deg.temp.6.10.amb=case_when(gene_gadmor %in% rownames(diffex.temp.6.10.amb.filt) ~ "yes", TRUE ~ "no"),
    deg.temp.3.6.low=case_when(gene_gadmor %in% rownames(diffex.temp.3.6.low.filt) ~ "yes", TRUE ~ "no"),
    #deg.temp.3.10.low=case_when(gene_gadmor %in% rownames(diffex.temp.3.10.low.filt) ~ "yes", TRUE ~ "no"),
    deg.temp.6.10.low=case_when(gene_gadmor %in% rownames(diffex.temp.6.10.low.filt) ~ "yes", TRUE ~ "no"),
    deg.both.3low.6amb=case_when(gene_gadmor %in% rownames(diffex.both.3low.6amb.filt) ~ "yes", TRUE ~ "no"),
    deg.both.10low.6amb=case_when(gene_gadmor %in% rownames(diffex.both.10low.6amb.filt) ~ "yes", TRUE ~ "no"),
    nonlinear.concave=case_when(gene_gadmor %in% unique((filter(nonlinear_df, temperature_2_binom=="Concave"))$geneid) ~ "yes", TRUE ~ "no"),
    nonlinear.convex=case_when(gene_gadmor %in% unique((filter(nonlinear_df, temperature_2_binom=="Convex"))$geneid) ~ "yes", TRUE ~ "no"))    
    
# Options to identify genes from other analyses in edger, i.e. more linear/nonlinear and alternative splicing 
#    linear.interact.increase=case_when(gene_gadmor %in% linear_interaction_degs_increase$geneid ~ "yes", TRUE ~ "no"),
#    linear.interact.decrease=case_when(gene_gadmor %in% linear_interaction_degs_decrease$geneid ~ "yes", TRUE ~ "no"),
#    nonlinear.interact.Ashaped=case_when(gene_gadmor %in% nonlinear_interaction_degs_Ashaped$geneid ~ "yes", TRUE ~ "no"),
#    nonlinear.interact.Vshaped=case_when(gene_gadmor %in% nonlinear_interaction_degs_Vshaped$geneid ~ "yes", TRUE ~ "no"),
#    nonlinear.no.interact.Ashaped=case_when(gene_gadmor %in% (nonlinear_no.interaction_degs %>% filter(temperature_2_binom=="Convex"))$geneid ~ "yes", TRUE ~ "no"),
#    nonlinear.no.interact.Vshaped=case_when(gene_gadmor %in% (nonlinear_no.interaction_degs %>% filter(temperature_2_binom=="Concave"))$geneid ~ "yes", TRUE ~ "no"))#,
    #temperature.splice=case_when(gene_gadmor %in% temperature.splice.sign$Geneid ~ "yes", TRUE ~ "no"))

View(results.df)
```

Load DAVID enrichment results from DESeq2

```{r}
load(file="../results/GO-enrichment/Final-DAVID/david")
```

Plot any gene of interest using protein name 

### Digestive enzymes that change with larval development: 
- Trypsin (Trypsin-1,-3, -10) - decreases after ~17 DPH in Atlantic cod 
- Amylase (Alpha-amylase) - decreases after ~10 DPH in Atlantic cod 
- phospholipase A2 - increases after ~20 DPH in Atlantic cod 
- Cholecystokinin (CCK), neuropeptide Y (NPY) and Orexin (OX) - decrease steadily in Atlantic cod 

```{r, warning=F, message=F}
protein.name <- "2-iminobutanoate/2-iminopropanoate deaminase"

genes.interest <- gadMor.blast.GO %>% filter(str_detect(protein_names, protein.name)) %>% filter(gene_gadmor %in% rownames(res.pH.3)) 


#enzymes %>% select(gene_gadmor, spid, gene_uni, protein_names) %>% write_clip()

# Are enzymes of interest to RNASeq analyses?
results.df %>% filter(gene_gadmor %in% genes.interest$gene_gadmor)

for (i in 1:nrow(genes.interest)) {
a <-  plotCounts(dds.DESeq.treatment, gene=genes.interest$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #mutate(treatment=factor(treatment, ordered = T, levels = c("3_amb", "3_low", "6_amb", "6_low", "")))
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))

# Plot and facet wrap by ph
print(ggplot(a,
       aes(x=temperature, y=count, color=temperature, label=sample)) +
  geom_boxplot(outlier.shape = NA) + facet_wrap(~ph) +
  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle(str_wrap(genes.interest$protein_names[i])) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=14), legend.position = "none") +
    scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"))
  # scale_color_manual(values=c("3:amb"= "navyblue", "3:low"="royalblue1","6:amb"="darkgreen",
  #                             "6:low"="yellow3","10:amb"="firebrick4", "10:low"="sienna2"))
  )
}

```

Look at lipid metabolism genes that responded to different treatments - how much overlap?

```{r}
# Uniprot KW
OA <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS", process=="Lipid metabolism", stress=="OA"))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector()
Warm <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS", process=="Lipid metabolism", stress=="Warm"))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector()
`Warm+OA` <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS", process=="Lipid metabolism", stress=="Warm-OA"))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector()
Cold <- (david %>% filter(p_value<0.05, category=="UP_KW_BIOLOGICAL_PROCESS", process=="Lipid metabolism", stress=="Cold"))$genes %>% strsplit(., ",") %>% unlist() %>% as.vector()
```

```{r}
venn.lipid <- data.frame(dat = 
             unique(c(`OA`, `Warm`, `Warm+OA`, `Cold`))) %>%
  mutate(`OA` = dat %in% `OA`,
         `Warm` = dat %in% `Warm`,
         `Warm+OA` = dat %in% `Warm+OA`,
         `Cold` = dat %in% `Cold`) %>%
  select(`OA`, `Warm`, `Warm+OA`, `Cold`) %>%
  euler()

#pdf(file="../results/deseq2/venn_single-stressors_no-values.pdf", height = 5.2, width = 6)
eulerr:::plot.euler(venn.lipid, counts = T, quantities = T, 
                      legend = T, edges=F, fontzie=8,
                      font=1, cex=1, alpha=0.5,
                      fills=list(fill=c(lighten("darkorange2", 0.25),
                                        lighten("firebrick4", 0.45), 
                                        "firebrick4", 
                                        lighten("navyblue", 0.45))))
#dev.off()
```

```{r}
intersect(OA, Warm) #6
intersect(OA, `Warm+OA`) #4
intersect(OA, Cold) #2
intersect(Warm, `Warm+OA`) #12
intersect(`Warm`, Cold) #0
intersect(`Warm+OA`, Cold) #0

Reduce(intersect, list(OA, Warm, `Warm+OA`, Cold)) #0
Reduce(intersect, list(OA, Warm, `Warm+OA`)) #4
Reduce(intersect, list(OA, Warm)) #6

setdiff(Cold, c(OA, Warm, `Warm+OA`)) #9, only in cold
setdiff(OA, c(Cold, Warm, `Warm+OA`)) #6, only in OA
setdiff(Warm, c(OA, `Warm+OA`, Cold)) #10, only in Warm
setdiff(`Warm+OA`, c(OA, Warm, Cold)) #3, only in Warm+OA

# These are all the lipid-metabolism related genes diff. expressed in warm and/or warm+oa
c(setdiff(Warm, c(OA, `Warm+OA`, Cold)), #warm only #10
setdiff(`Warm+OA`, c(OA, Warm, Cold)), #warm/oa only #3
setdiff(intersect(Warm, `Warm+OA`), OA)) #warm and warm/oa only #8

# fatty acid catabolism - 1234
# triglyceride catabolism - 123456
# ether lipid / plasmalogen - 12
# phospholipid - 12
# vitamin d metabolism (cholesterol based molecule) - 1
# steroid biosynthesis - 1
# oligosaccharide/glycoprotein - 1
# gangliiosides (neuronal sphingolipids), membrane componenets nerve/brain - 12
# immune response - 1
# detox metabolic intermediate metabolies (enamine/imine)
# Patterns in genes diff. expressed in warm/warm+oa, but not oa: triglyceride and fatty acid metabolism related genes, some possible membrane related genes (e.g. ether lipids / plasmalogens such as "far1"), and a couple misc. genes involved in vitamin d, steroid bio., immune response, and metabolic detox. Only a couple loosely-related to cholesterols. 

setdiff(OA, c(Warm, `Warm+OA`)) #8, in OA not warm/warm+oa
# cholesterol excretion - 1
# vision/vitamin a metabolism - 1
# phospholipids (membrane) - 1
# small-intestine lipid digestion - 12
# neuronal activity - 12 
# fatty acid uptake - 1
# Patterns: a variety of lipid related processes, but not triglyceride/fatty acid catabolism! 

Reduce(intersect, list(OA, `Warm`)) #6 genes diff. expr. in both OA and warm and/or warm+oa
# fatty acid biosynthesis - 1
# triglyceride synthesis, maybe dietary fat absorption?
# prostaglandin biosynthesis (inflammation?)
# vitamin d
# sphingolipids (signaling) - 1

```
Heat map of lipid metabolism related genes 

```{r}
load(file = "../results/deseq2/degs")
degs.unique <- unique(c(rownames(degs$`Amb vs. Low pH @6C`),
rownames(degs$`3 vs. 6C @Amb pH`),
rownames(degs$`6 vs. 10C @Amb pH`), 
rownames(degs$`6 amb vs. 3C low`), 
rownames(degs$`6 amb vs. 10 low`)))


lipid.genes <- (counts.annot.gadmor %>% filter(gene_gadmor %in% degs.unique) %>% #use only genes that were DEGs
                  filter(spid %in% (unique(c(OA, Warm, Cold, `Warm+OA`)) %>% gsub(" ", "", .))))$gene_gadmor #and only those related to lipid metabolism

load(file = "../results/deseq2/counts.tk")
load(file = "../results/deseq2/vsd.treatment")

sample.order <- (counts.tk %>% select(treatment) %>% rownames_to_column("sample") %>%
  mutate(treatment=factor(treatment, ordered = T, levels = c("6_amb", "3_low", "3_amb", "6_low", "10_low", "10_amb"))) %>%
  arrange(treatment))$sample

# Generate heat map / cluster lipid related genes
#pdf(file="../results/deseq2/heatmap-lipids.pdf", height = 5, width = 7.25)
cluster.most.variable <- pheatmap(t(assay(vsd.treatment))[sample.order,lipid.genes], 
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
                     show_colnames =TRUE, show_rownames = FALSE,
         border_color = NA,
         fontsize_col=6.5, fontsize = 8,
         cluster_cols = TRUE, cluster_rows = F,
         annotation_colors = list(treatment=
          c("Control"="gray65",
            "Acidification"=lighten("darkorange2", 0.25),
            "Warming"=lighten("firebrick4", 0.45),
            "Warming + Acidification"="firebrick4",
            "Cooling"=lighten("navyblue", 0.45),
            "Cooling + Acidification"="navyblue")),
         annotation_row=as.data.frame(colData(vsd.treatment)[sample.order,c("treatment"), drop=FALSE]) %>%
           mutate(treatment=case_when(
              treatment == "6_amb" ~ "Control",
              treatment == "6_low" ~ "Acidification",
              treatment == "10_amb" ~ "Warming",
              treatment == "10_low" ~ "Warming + Acidification", 
              treatment == "3_amb" ~ "Cooling", 
              treatment == "3_low" ~ "Cooling + Acidification")),
         main = "Lipid metabolism gene counts")
#dev.off()
```

Figure out which genes cluster together 

```{r}
lipid.tree <- cluster.most.variable <- pheatmap(t(assay(vsd.treatment))[,lipid.genes], 
         Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column",
         cluster_cols = TRUE, cluster_rows = TRUE, silent = TRUE)

plot(lipid.tree$tree_col)
abline(h=10.5, col="red", lty=2, lwd=2)

rbind(t(assay(vsd.treatment))[,lipid.genes],
      cluster = cutree(lipid.tree$tree_col, k = 11)) %>%
  as.data.frame() %>% tail() %>%
  t() %>% as.data.frame() %>% select(cluster) %>%
  rownames_to_column("gene_gadmor") %>%
  left_join(counts.annot.gadmor[c("gene_gadmor", "gene_uni", "spid", "evalue", "protein_names")], 
            by = "gene_gadmor") %>%
  left_join(gadMor.blast.GO %>% 
              select(spid, gene_ontology_biological_process, 
                     gene_ontology_cellular_component, gene_ontology_molecular_function, gene_ontology_i_ds)) %>%
  mutate(cluster=as.factor(cluster)) %>%

  # Filter for clusters in tree 
  filter(cluster %in% c(5)) %>% 

  # # To save Uniprot SPID to clipboard
  # select(spid) %>% na.omit() %>% unlist() %>% as.vector() %>% 
  # write_clip()

  # Generate wordcloud from GO terms 
  select(gene_ontology_biological_process) %>% na.omit() %>% unlist() %>% as.vector() %>% 
  gsub("\\[[^\\]]*\\]", "", ., perl=TRUE) %>%
  gsub("process", "", ., perl=TRUE) %>% #get rid of word "process" since it's in everything
  wordcloud(min.freq = 2, random.order = F, random.color = F, colors = brewer.pal(2, "Dark2"))
```

Plot genes of interest relating to enzyme analysis 

```{r, warning=FALSE}

enzymes <- 
#gadMor.blast.GO %>% filter(str_detect(protein_names, "itrate synthase")) %>% filter(gene_gadmor %in% rownames(res.pH.3)) #Citrate Synthase
gadMor.blast.GO %>% filter(str_detect(protein_names, "hydroxyacyl-CoA dehydrogenase")) %>% filter(gene_gadmor %in% rownames(res.pH.3)) #HOAD
#gadMor.blast.GO %>% filter(str_detect(protein_names, "lactate dehydrogenase")) %>% filter(gene_gadmor %in% rownames(res.pH.3)) #Lactate Dehydrogenase
#gadMor.blast.GO %>% filter(str_detect(protein_names, "yruvate kinase")) %>% filter(gene_gadmor %in% rownames(res.pH.3)) #Pyruvate kinase

#enzymes %>% select(gene_gadmor, spid, gene_uni, protein_names) %>% write_clip()

# Are enzymes of interest to RNASeq analyses?
#results.df %>% filter(gene_gadmor %in% enzymes$gene_gadmor) %>% View()

for (i in 1:nrow(enzymes)) {
a <-  plotCounts(dds.DESeq.treatment, gene=enzymes$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #mutate(treatment=factor(treatment, ordered = T, levels = c("3_amb", "3_low", "6_amb", "6_low", "")))
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))

# plot 
print(ggplot(a,
       aes(x=interaction(temperature:ph), y=count, color=interaction(temperature:ph))) +
  geom_boxplot(aes(fill=interaction(temperature:ph)), outlier.shape = NA, alpha=0.35) + 
  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle(str_wrap(enzymes$protein_names[i])) +
    theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
          plot.title = element_text(hjust = 0.5, size=11), legend.position = "none") +
  scale_color_manual(values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="navyblue","6:Ambient pCO2"="darkgreen",
                               "6:High pCO2"="darkgreen","10:Ambient pCO2"="firebrick4", "10:High pCO2"="firebrick4")) +
  scale_fill_manual(name="pCO2", values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="white","6:Ambient pCO2"="darkgreen",
                               "6:High pCO2"="white","10:Ambient pCO2"="firebrick4", "10:High pCO2"="white")))

# # Facet wrap by ph
# print(ggplot(a,
#        aes(x=temperature, y=count, color=temperature, label=sample)) +
#   geom_boxplot(outlier.shape = NA) + facet_wrap(~ph) +
#   geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
#     #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
#     theme_bw() +
#     ggtitle(str_wrap(enzymes$protein_names[i])) +
#     theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=14), legend.position = "none") +
#     scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4"))
#   # scale_color_manual(values=c("3:amb"= "navyblue", "3:low"="royalblue1","6:amb"="darkgreen",
#   #                             "6:low"="yellow3","10:amb"="firebrick4", "10:low"="sienna2"))
#   )
}
```
Enzyme ratios 

LDH:CS = anaerobic:aerobic contributions 
HOAD:CS = fatty acid contribution towards aerobic metabolism  

```{r}
enzyme.genes <- c("ldhd", "cs", "hadh") #old list before controlling for fish length
#enzyme.genes <- c("cs", "hsd17b10") #new list, no ldhd and new HOAD

enzyme.genes.counts <- vector("list", length(enzyme.genes))
names(enzyme.genes.counts) <- enzyme.genes

# # Were any of the genes tossed out b/c they correlated with fish length? 
# intersect(genes.length, enzyme.genes) #yes, both hadh and ldhd

# Should try to plot gene expression ~ fish length those genes 

for (i in 1:length(enzyme.genes.counts)) {
enzyme.genes.counts[[i]] <-  plotCounts(
#  dds.DESeq.treatment,  #use this DESeq2 object to use counts for genes included in final analysis
  dds.DESeq.treatment.all,  #use this DESeq2 object to use ALL counts (not filtered to remove length-associated ones)
  gene=enzyme.genes[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #filter(treatment %in% b) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))
}

enzyme.ratios <- bind_rows(enzyme.genes.counts, .id = "gene_gadmor") %>% 
  mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
  pivot_wider(names_from=gene_gadmor, values_from=count) %>%
#  mutate(`LDH:CS`=ldhd/cs, `HOAD:CS`=hadh/cs)
  mutate(`HOAD:CS`=hadh/cs)
#  mutate(`HOAD:CS`=hsd17b10/cs)

# # Examine LDH:CS ratio 
# ggplot(enzyme.ratios,
#        aes(x=interaction(temperature:ph), y=`LDH:CS`, color=interaction(temperature:ph))) +
#   geom_boxplot(aes(fill=interaction(temperature:ph)), outlier.shape = NA, alpha=0.35) + 
#   geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
#     #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
#     theme_bw() +
#     ggtitle("LDH:CS Ratio") +
#     theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
#           plot.title = element_text(hjust = 0.5, size=13), legend.position = "none") +
#   scale_color_manual(values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="navyblue","6:Ambient pCO2"="darkgreen",
#                                "6:High pCO2"="darkgreen","10:Ambient pCO2"="firebrick4", "10:High pCO2"="firebrick4")) +
#   scale_fill_manual(name="pCO2", values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="white","6:Ambient pCO2"="darkgreen",
#                                "6:High pCO2"="white","10:Ambient pCO2"="firebrick4", "10:High pCO2"="white"))
# hist(log(enzyme.ratios$`LDH:CS`))
# shapiro.test(log(enzyme.ratios$`LDH:CS`))
# summary(aov(log(`LDH:CS`)~temperature*ph, data=enzyme.ratios)) # Temperature affects LDH:CS ratio 
# TukeyHSD(aov(log(`LDH:CS`)~temperature*ph, data=enzyme.ratios)) # All temperatures differ from each other 

# Examine HOAD:CS ratio
ggplot(enzyme.ratios,
       aes(x=interaction(temperature:ph), y=`HOAD:CS`, color=interaction(temperature:ph))) +
  geom_boxplot(aes(fill=interaction(temperature:ph)), outlier.shape = NA, alpha=0.35) + 
  geom_point(size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle("HOAD:CS Ratio") +
    theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
          plot.title = element_text(hjust = 0.5, size=13), legend.position = "none") +
  scale_color_manual(values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="navyblue","6:Ambient pCO2"="darkgreen",
                               "6:High pCO2"="darkgreen","10:Ambient pCO2"="firebrick4", "10:High pCO2"="firebrick4")) +
  scale_fill_manual(name="pCO2", values=c("3:Ambient pCO2"= "navyblue", "3:High pCO2"="white","6:Ambient pCO2"="darkgreen",
                               "6:High pCO2"="white","10:Ambient pCO2"="firebrick4", "10:High pCO2"="white"))
hist(log(enzyme.ratios$`HOAD:CS`))
shapiro.test(enzyme.ratios$`HOAD:CS`)
summary(aov(`HOAD:CS`~temperature*ph, data=enzyme.ratios)) # Temperature affects HOAD:CS ratio 
TukeyHSD(aov(`HOAD:CS`~temperature*ph, data=enzyme.ratios)) # Lower HOAD:CS ratio at 10C compared to other temperatures

# Stats on enzymes using ANOVA

hist(log(enzyme.ratios$cs))
shapiro.test(enzyme.ratios$cs)
summary(aov(cs~temperature*ph, data=enzyme.ratios)) # CS affected by temperature and temperature:ph interaction
TukeyHSD(aov(cs~temperature*ph, data=enzyme.ratios)) # CS differs among all temperatures
```

## Examine correlation between fish length and gene counts for genes of interest

```{r}
#HOAD is length-associated. Let's look at HOAD expression ~ fish length
HOAD <- enzyme.genes.counts$hadh %>% left_join(counts.tk %>% rownames_to_column(var="sample") %>% select(sample, length))
cor.test(x=HOAD$length, y=HOAD$count, method = "pearson")

HOAD %>%
  ggplot(aes(x=length, y=count, colour=treatment)) + 
  geom_point(size=2) + 
  theme_minimal() + 
  scale_color_manual(name="Treatment",
                 values=c("6_amb"="gray65",
                          "6_low"=lighten("darkorange2", 0.25),
                          "10_amb"=lighten("firebrick4", 0.45),
                          "10_low"="firebrick4",
                          "3_amb"=lighten("navyblue", 0.45),
                          "3_low"="navyblue"),
                 labels=c("6_amb"="Control",
                          "6_low"="High pCO2",
                          "10_amb"="Warm Temperature",
                          "10_low"="Warm Temperature + High pCO2",
                          "3_amb"="Cold Temperature",
                          "3_low"="Cold Temperature + High pCO2")) +
    geom_smooth(aes(x=length, y=count, colour=treatment), method="lm", se=F) + 
  ggtitle("HOAD expression ~ larval length")

#CS is not length-associated. Let's look at CS expression ~ fish length
cs <- enzyme.genes.counts$cs %>% left_join(counts.tk %>% rownames_to_column(var="sample") %>% select(sample, length))
cor.test(x=cs$length, y=cs$count, method = "pearson")

cs %>%
  ggplot(aes(x=length, y=count, colour=treatment)) + geom_point(size=2) + 
  theme_minimal() + 
  scale_color_manual(name="Treatment",
                 values=c("6_amb"="gray65",
                          "6_low"=lighten("darkorange2", 0.25),
                          "10_amb"=lighten("firebrick4", 0.45),
                          "10_low"="firebrick4",
                          "3_amb"=lighten("navyblue", 0.45),
                          "3_low"="navyblue"),
                 labels=c("6_amb"="Control",
                          "6_low"="High pCO2",
                          "10_amb"="Warm Temperature",
                          "10_low"="Warm Temperature + High pCO2",
                          "3_amb"="Cold Temperature",
                          "3_low"="Cold Temperature + High pCO2")) +
    geom_smooth(aes(x=length, y=count, colour=treatment), method="lm", se=F) + 
  ggtitle("CS expression ~ larval length")
```


### Alternative enzyme-related gene expression analysis 
- Use GO terms, Uniprot Keywords,  and/or KEGG Pathway to identify genes of interest and look at their expression 
-- Citrate synthase: 
--- citrate synthase activity (GO:0036440, Molecular Function)
--- tricarboxylic acid cycle (GO:0006099, Biological Process)
--- aerobic respiration (GO:0009060, Biological Process)
--- anaerobic respiration (GO:0009061, Biological Process)
--- Lipid metabolic process (GO:0006629, Biological Process) 
--- Lipid biosynthetic process (GO:0008610, Biological Process) 
--- Lipid catabolic process (GO:0016042, Biological Process)

```{r, warning=FALSE, message=FALSE}
go <- 
 "GO:0036440" #citrate synthase activity (GO term molecular function)
# "GO:0006099" #tricarboxylic acid cycle (GO term biological process)
# "GO:0009061" # Anaerobic respiration (GO term Biological Process)
# "GO:0009060" # Aerobic respiration (GO term biological process)
#"GO:0008610" #Lipid biosynthetic process
# "GO:0016042" #Lipid catabolic process

go.genes <- results.df %>% left_join(
  gadMor.blast.GO %>% select(spid, gene_ontology_biological_process, gene_ontology_cellular_component, gene_ontology_molecular_function, gene_ontology_i_ds),
  by="spid") %>% 
  filter(grepl(go, gene_ontology_i_ds)) %>% 
  filter(deg.ph.6=="yes" | deg.temp.3.6.amb=="yes" | deg.temp.6.10.amb=="yes" | #deg.temp.3.10.low=="yes" | deg.temp.3.10.amb=="yes" | 
         deg.both.3low.6amb=="yes" | deg.both.10low.6amb=="yes")

# # to paste into Uniprot 
# test$spid %>% unique() %>% write_clip()

for (i in 1:nrow(go.genes)) {
a <-  plotCounts(dds.DESeq.treatment, gene=go.genes$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #filter(treatment %in% b) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))

print(
  ggplot(a,
       aes(x=temperature, y=count, color=temperature, label=sample)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(aes(shape=ph), size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle(str_wrap(go.genes$protein_names[i])) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) + 
  facet_wrap(~ph))
}

# One plot, points or lines for each gene 

go.genes.lines <- vector("list", nrow(go.genes))
names(go.genes.lines) <- go.genes$gene_gadmor

for (i in 1:nrow(test)) {
go.genes.lines[[i]] <-  plotCounts(dds.DESeq.treatment, gene=go.genes$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))
}

bind_rows(go.genes.lines, .id = "gene_gadmor") %>% 
  mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
  group_by(gene_gadmor, temperature, ph) %>%
  dplyr::summarize(mean=mean(count), median=median(count), sd=sd(count), cv=sd(count)/mean(count)) %>%
  group_by(gene_gadmor, temperature, ph) %>% mutate(z_score=scale(mean)) %>%  # figure out how to standardize value across genes (opposed to z-score)

  # # One boxplot, points for all genes
  # ggplot(aes(x=temperature, y=z_score, color=temperature)) +
  # geom_boxplot(outlier.shape = NA) +
  # geom_point(aes(shape=ph), size=2.5, position=position_jitter(w = 0.15,h = 0)) +
  #   #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
  #   theme_bw() +
  #   theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  # scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) +
  # facet_wrap(~ph)

  # Line plots 
  ggplot(aes(x=temperature, y=z_score, group=gene_gadmor)) + geom_line(size=.01) +
  theme_minimal() + facet_wrap(~ph) +
  #scale_x_continuous(breaks=c(1,2,3), labels=c("Ambient", "Moderate", "Low")) +
  theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  xlab("Per treatment mean of normalized counts") + ggtitle("Abundances of DEGs by treatment & relative expression")


```

Plot genes of interest relating to zona pellucida 

```{r, warning=FALSE}
zona <- 
gadMor.blast.GO %>% filter(str_detect(protein_names, "ona pellucida")) %>% filter(gene_gadmor %in% rownames(res.pH.3)) %>% filter(seq_id == "NC_044056.1")

# Are zona pellucida genes of interest to RNASeq analyses?
results.df %>% filter(gene_gadmor %in% zona$gene_gadmor) %>% View()

# Yes- differences are between the following:
# 6-amb vs 10-amb
# 6-low vs 10-low 
# 6-amb vs 10-low
# 3-amb vs 10-amb

# # Was the zona pellucida gene tossed out b/c it correlated with fish length? 
intersect(genes.length, zona$gene_gadmor) #no

for (i in 1:nrow(zona)) {
a <-  plotCounts(dds.DESeq.treatment, gene=zona$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  mutate(treatment=factor(treatment, ordered = T, levels = c("6_amb", "6_low", "3_amb", "3_low", "10_amb", "10_low")))

summary(aov(count ~ temperature*ph, data=a))
(tukey <- TukeyHSD(aov(count ~ temperature*ph, data=a)))
tukey$`temperature:ph` %>% as.data.frame() %>% rownames_to_column("contrast") %>% arrange(`p adj`)

print(ggplot(a,
       aes(x=treatment, y=count, fill=treatment, alpha=treatment)) +
  geom_boxplot(outlier.shape = NA, lwd=0.2) +
  geom_point(aes(x=treatment, color=treatment), 
             size=1.5, position=position_jitter(w = 0.2,h = 0)) +
    theme_bw() +
    # ggtitle(paste(str_wrap(zona$protein_names[i]), "\n", 
    #               zona$gene_names[i], " - ", zona$species[i], " - ", zona$seq_id[i], sep="")) +
    ggtitle("Zona pellucida sperm-binding protein 3 (ZP3)") + 
    theme(axis.title.x = element_blank(), 
          axis.text.x = element_blank(), 
          axis.ticks.x = element_blank(),
          axis.title.y = element_text(size=9),
          plot.title = element_text(hjust = 0.5, size=10), 
          legend.position = "right",
          legend.title = element_text(size=9)) +
  scale_fill_manual(guide="none", values=c("6_amb"="gray25", "6_low"="gray25", 
                                               "3_amb"="navyblue", "10_amb"="firebrick4", 
                                               "3_low"="navyblue", "10_low"="firebrick4"),
                    labels=c("6_amb"="Control", "6_low"="OA", 
                                               "3_amb"="Cold", "10_amb"="Warm", 
                                               "3_low"="Cold+OA", "10_low"="Warm+OA")) +
  scale_color_manual(guide="none", values=c("6_amb"="gray25", "6_low"="gray25", 
                                               "3_amb"="navyblue", "10_amb"="firebrick4", 
                                               "3_low"="navyblue", "10_low"="firebrick4")) +
    scale_alpha_manual(guide="none", values=c(.7, .4, .7, .4, .7, .4)))
}

#Zona pellucida 3 is not length-associated. Let's look at zp3 expression ~ fish length
zona.count <- a %>% left_join(counts.tk %>% rownames_to_column(var="sample") %>% select(sample, length))
cor.test(x=zona.count$length, y=zona.count$count, method = "pearson")

zona.count %>%
  ggplot(aes(x=length, y=count, colour=treatment)) + geom_point(size=2) + 
  theme_minimal() + 
  scale_color_manual(name="Treatment",
                 values=c("6_amb"="gray65",
                          "6_low"=lighten("darkorange2", 0.25),
                          "10_amb"=lighten("firebrick4", 0.45),
                          "10_low"="firebrick4",
                          "3_amb"=lighten("navyblue", 0.45),
                          "3_low"="navyblue"),
                 labels=c("6_amb"="Control",
                          "6_low"="High pCO2",
                          "10_amb"="Warm Temperature",
                          "10_low"="Warm Temperature + High pCO2",
                          "3_amb"="Cold Temperature",
                          "3_low"="Cold Temperature + High pCO2")) +
    geom_smooth(aes(x=length, y=count, colour=treatment), method="lm", se=F) + 
  ggtitle("zp3 expression ~ larval length")
```

## Plot interesting genes using Uniprot IDs - each gene separately

```{r, warning=FALSE, message=F}
keyword <- "Biological rhythms"
keyword <- "Vision"
keyword <- "Stress response"
keyword <- "Immunity"
keyword <- "DNA integration"
keyword <- "Urea cycle"
keyword <- "Tricarboxylic acid cycle"
keyword <- "Cell adhesion"

uniprots <- ((enriched.kw.degs.upregulated %>% filter(kw==keyword))$genes)
uniprots.vect <- vector(mode="character")

for (i in 1:length(uniprots)) {
  x <- unlist(strsplit(uniprots[i], ", "))
  uniprots.vect <- append(uniprots.vect, x)
}

# # Plot all genes in data that match the Uniprot SPIDs of interest
# test <- counts.annot.gadmor %>% 
#   filter(spid %in% unique(uniprots.vect))

# plot only the genes that were DEGs and match the Uniprot SPIDs of interest
test <- results.df %>% 
  filter(deg.ph.6=="yes" | deg.temp.3.6.amb=="yes" | deg.temp.6.10.amb=="yes" | #deg.temp.3.10.low=="yes" | deg.temp.3.10.amb=="yes" | 
         deg.temp.3.6.low=="yes" | deg.temp.6.10.low=="yes") %>%
  filter(spid %in% unique(uniprots.vect))

# to paste into Uniprot 
test$spid %>% unique() %>% write_clip()

# Boxplot of gene expression by treatment 
for (i in 1:nrow(test)) {
a <-  plotCounts(dds.DESeq.treatment, gene=test$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #filter(treatment %in% b) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))

print(
  ggplot(a,
       aes(x=temperature, y=count, color=temperature, label=sample)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(aes(shape=ph), size=2.5, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
    theme_bw() +
    ggtitle(str_wrap(test$protein_names[i])) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) + 
  facet_wrap(~ph))
}


# Scatter plot of gene expression by fish length 

for (i in 1:nrow(test)) {
a <-  plotCounts(dds.DESeq.treatment, gene=test$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank", "length")], by=c("sample"="vial_label")) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))

#cor.length <- cor.test(x=filter(a, temperature!=10)$length, y=filter(a, temperature!=10)$count, method = "pearson") 
cor.length <- cor.test(x=a$length, y=a$count, method = "pearson") 

print(
  ggplot(a 
         #%>% filter(temperature!=10)
         ) +
  geom_point(aes(x=length, y=count, color=treatment), size=2.5) +
    theme_bw() +
    ggtitle(str_wrap(test$protein_names[i])) +
    theme(plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  scale_color_manual(name="Treatment", 
                     values=c("6_amb"="gray65",
                              "6_low"=lighten("darkorange2", 0.25), 
                              "10_amb"=lighten("firebrick4", 0.45), 
                              "10_low"="firebrick4", 
                              "3_amb"=lighten("navyblue", 0.45),
                              "3_low"="navyblue"),
                     labels=c("6_amb"="Control",
                              "6_low"="High pCO2", 
                              "10_amb"="Warm Temperature", 
                              "10_low"="Warm Temperature + High pCO2", 
                              "3_amb"="Cold Temperature",
                              "3_low"="Cold Temperature + High pCO2")) +
    geom_smooth(data=filter(a, temperature==3), method="lm", aes(x=length, y=count, color=temperature), 
                se=T, color="navyblue", fill="navyblue", alpha=0.25) + 
    geom_smooth(data=filter(a, temperature==6), method="lm", aes(x=length, y=count, color=temperature), 
                se=T, color="darkorange2", fill="darkorange2", alpha=0.25) + 
    geom_smooth(data=filter(a, temperature==10), method="lm", aes(x=length, y=count, color=temperature),
                    se=T, color="firebrick4", fill="firebrick4", alpha=0.25) +
    annotate("text", x=max(a$length)*.98, y=max(a$count)*.95, 
             label=paste("Corr=", round(as.vector(cor.length$estimate),digits = 2), 
                 "\nP-value=", signif(as.vector(cor.length$p.value), digits=2))))
}
```


## Plot "reaction norms" - i.e. mean 

```{r}
test1 <- vector("list", nrow(test))
names(test1) <- test$gene_gadmor

for (i in 1:nrow(test)) {
test1[[i]] <-  plotCounts(dds.DESeq.treatment, gene=test$gene_gadmor[i], intgroup=c("treatment"), replaced = TRUE, returnData = TRUE) %>%
  rownames_to_column("sample") %>%
  left_join(sample.info[,c("vial_label", "temperature", "ph", "tank", "treatment_tank")], by=c("sample"="vial_label")) %>%
  #filter(treatment %in% b) %>%
  mutate(treatment=fct_relevel(treatment, c("3_amb", "6_amb", "10_amb", "3_low", "6_low", "10_low"))) %>%
  mutate(ph=as.factor(case_when(ph=="amb" ~ "Ambient pCO2", TRUE ~ "High pCO2")))
}

bind_rows(test1, .id = "gene_gadmor") %>% 
  mutate(gene_gadmor=as.factor(gene_gadmor)) %>%
  group_by(gene_gadmor, temperature, ph) %>%
  dplyr::summarize(mean=mean(count), median=median(count), sd=sd(count), cv=sd(count)/mean(count)) %>%
  group_by(gene_gadmor, temperature, ph) %>% mutate(z_score=scale(mean)) %>%

  # # Plot points for all genes
  # ggplot(aes(x=temperature, y=mean, color=temperature)) +
  # geom_boxplot(outlier.shape = NA) +
  # geom_point(aes(shape=ph), size=2.5, position=position_jitter(w = 0.15,h = 0)) +
  #   #geom_text(size=3.5, position=position_jitter(w = 0.15,h = 0)) +
  #   theme_bw() +
  #   theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
  # scale_color_manual(name="Temperature", values=c("3"= "navyblue", "6"="darkgreen", "10"="firebrick4")) +
  # facet_wrap(~ph)

  #Rection norm - too hard to see
  ggplot(aes(x=temperature, y=z_score, group=gene_gadmor)) + geom_line(size=.01) +
  theme_minimal() + facet_wrap(~ph) +
  #scale_x_continuous(breaks=c(1,2,3), labels=c("Ambient", "Moderate", "Low")) +
  theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  xlab("Per treatment mean of normalized counts") + ggtitle("Abundances of DEGs by treatment & relative expression")
```

### Search for Uniprot Keywords and GO Terms that are commonly affected, or uniquely affected, by temperature & pCO2 treatments

```{r}
temp <- enriched.kw.degs.upregulated %>% filter(contrast %!in% c("3vs10.Low-pH", "3vs10.Amb-pH")) %>% filter(!is.na(kw)) %>% filter(background!="Amb-pH")

# Uniprot Keywords that are commonly affected by temperature at high and ambient pCO2, and by pCO2
temp[temp$kw %in% 
  temp$kw[duplicated(temp$kw)],] %>%
  arrange(kw) %>% View()

# Uniprot Keywords that are uniquely affected by temperature at high and ambient pCO2, or by pCO2
temp %>% filter(kw %!in% 
                  (temp[temp$kw %in% temp$kw[duplicated(temp$kw)],])$kw) %>% 
  arrange(contrast) %>% View()

```

